

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Core components &#8212; EggLib&#39;s documentation</title>
    <link rel="stylesheet" href="../_static/egglibdoc.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '3.0.0b21',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/icon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Data file formats" href="datafiles.html" />
    <link rel="prev" title="C++ Library" href="index.html" />
    <script type="text/javascript">
      // intelligent scrolling of the sidebar content
      $(window).scroll(function() {
        var sb = $('.sphinxsidebarwrapper');
        var win = $(window);
        var sbh = sb.height();
        var offset = $('.sphinxsidebar').position()['top'];
        var wintop = win.scrollTop();
        var winbot = wintop + win.innerHeight();
        var curtop = sb.position()['top'];
        var curbot = curtop + sbh;
        // does sidebar fit in window?
        if (sbh < win.innerHeight()) {
          // yes: easy case -- always keep at the top
          sb.css('top', $u.min([$u.max([0, wintop - offset - 10]),
                                $(document).height() - sbh - 200]));
        } else {
          // no: only scroll if top/bottom edge of sidebar is at
          // top/bottom edge of window
          if (curtop > wintop && curbot > winbot) {
            sb.css('top', $u.max([wintop - offset - 10, 0]));
          } else if (curtop < wintop && curbot < winbot) {
            sb.css('top', $u.min([winbot - sbh - offset - 20,
                                  $(document).height() - sbh - 200]));
          }
        }
      });
    </script>
  </head>
  <body>
<div style="background-color: white; text-align: left; padding: 0px 0px 0px 0px">
<a href="../index.html"><img src="../_static/banner.png" border="0" alt="EggLib"/></a>
</div>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="datafiles.html" title="Data file formats"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="C++ Library"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../contents.html">EggLib&#39;s documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../toc.html" >Contents</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="index.html" accesskey="U">C++ Library</a> &#187;</li> 
      </ul>
    </div>

      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../contents.html">
              <img class="logo" src="../_static/egglib_small.bmp" alt="Logo"/>
            </a></p>
  <h3><a href="../contents.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Core components</a><ul>
<li><a class="reference internal" href="#constants">Constants</a></li>
<li><a class="reference internal" href="#basic-classes">Basic classes</a><ul>
<li><a class="reference internal" href="#dataholder">DataHolder</a></li>
<li><a class="reference internal" href="#geneticcode">GeneticCode</a></li>
<li><a class="reference internal" href="#random">Random</a></li>
</ul>
</li>
<li><a class="reference internal" href="#model-fitting-utilities">Model fitting utilities</a><ul>
<li><a class="reference internal" href="#abc">ABC</a></li>
<li><a class="reference internal" href="#neutral-networks">Neutral networks</a></li>
</ul>
</li>
<li><a class="reference internal" href="#utilities">Utilities</a><ul>
<li><a class="reference internal" href="#interspersealign">IntersperseAlign</a></li>
<li><a class="reference internal" href="#vectorint">VectorInt</a></li>
</ul>
</li>
<li><a class="reference internal" href="#exceptions">Exceptions</a><ul>
<li><a class="reference internal" href="#eggexception">EggException</a></li>
<li><a class="reference internal" href="#eggargumentvalueerror">EggArgumentValueError</a></li>
<li><a class="reference internal" href="#eggformaterror">EggFormatError</a></li>
<li><a class="reference internal" href="#egginvalidalleleerror">EggInvalidAlleleError</a></li>
<li><a class="reference internal" href="#eggmemoryerror">EggMemoryError</a></li>
<li><a class="reference internal" href="#eggopenfileerror">EggOpenFileError</a></li>
<li><a class="reference internal" href="#eggploidyerror">EggPloidyError</a></li>
<li><a class="reference internal" href="#eggruntimeerror">EggRuntimeError</a></li>
<li><a class="reference internal" href="#eggunalignederror">EggUnalignedError</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">C++ Library</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="datafiles.html"
                        title="next chapter">Data file formats</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/cpp/core.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="core-components">
<span id="core"></span><h1>Core components<a class="headerlink" href="#core-components" title="Permalink to this headline">¶</a></h1>
<div class="section" id="constants">
<h2>Constants<a class="headerlink" href="#constants" title="Permalink to this headline">¶</a></h2>
<p>These constant values have use to give special meaning to input
parameters or return values of functions. Usually they mean “data not
available” or “irrelevant value”. Different names are given to the same
value in order to make them more relevant to their context. Read the
documentation carefully.</p>
<dl class="member">
<dt id="_CPPv211MISSINGDATA">
<span id="MISSINGDATA__iC"></span><span class="target" id="project0namespaceegglib_1a5fd22552c475e545c4221175b5eb2f7c"></span><em class="property">const</em> int <code class="descclassname"></code><code class="descname">MISSINGDATA</code><a class="headerlink" href="#_CPPv211MISSINGDATA" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Missing data (large value) </p>
</dd></dl>

<dl class="member">
<dt id="_CPPv23MAX">
<span id="MAX__unsigned-iC"></span><span class="target" id="project0namespaceegglib_1ab498891cdf034be8678af906644e6dc1"></span><em class="property">const</em> unsigned int <code class="descclassname"></code><code class="descname">MAX</code><a class="headerlink" href="#_CPPv23MAX" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Unknown/undefined (large value) </p>
</dd></dl>

<dl class="member">
<dt id="_CPPv27UNKNOWN">
<span id="UNKNOWN__unsigned-iC"></span><span class="target" id="project0namespaceegglib_1a6b564311a87de05e515ec65480ebc236"></span><em class="property">const</em> unsigned int <code class="descclassname"></code><code class="descname">UNKNOWN</code><a class="headerlink" href="#_CPPv27UNKNOWN" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Unknown value (large value) </p>
</dd></dl>

<dl class="member">
<dt id="_CPPv27MISSING">
<span id="MISSING__unsigned-iC"></span><span class="target" id="project0namespaceegglib_1a9a16c0aae203e1ae160f1b5143ffdb6d"></span><em class="property">const</em> unsigned int <code class="descclassname"></code><code class="descname">MISSING</code><a class="headerlink" href="#_CPPv27MISSING" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Missing data (large value) </p>
</dd></dl>

<dl class="member">
<dt id="_CPPv28OUTGROUP">
<span id="OUTGROUP__unsigned-iC"></span><span class="target" id="project0namespaceegglib_1a8ba88f48158e2df67e90bd060d772cd1"></span><em class="property">const</em> unsigned int <code class="descclassname"></code><code class="descname">OUTGROUP</code><a class="headerlink" href="#_CPPv28OUTGROUP" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Outgroup (large value) </p>
</dd></dl>

<dl class="member">
<dt id="_CPPv26BEFORE">
<span id="BEFORE__unsigned-l-iC"></span><span class="target" id="project0namespaceegglib_1ad567f4288226a4ed7a0374b460f2cb27"></span><em class="property">const</em> unsigned long int <code class="descclassname"></code><code class="descname">BEFORE</code><a class="headerlink" href="#_CPPv26BEFORE" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Value before the first (large value) </p>
</dd></dl>

<dl class="member">
<dt id="_CPPv27MAXCHAR">
<span id="MAXCHAR__cC"></span><span class="target" id="project0namespaceegglib_1ae17ab3120fa47f4166f9622fb1a6a9f7"></span><em class="property">const</em> char <code class="descclassname"></code><code class="descname">MAXCHAR</code><a class="headerlink" href="#_CPPv27MAXCHAR" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Unknown value (large value) </p>
</dd></dl>

<dl class="member">
<dt id="_CPPv25UNDEF">
<span id="UNDEF__doubleC"></span><span class="target" id="project0namespaceegglib_1af63e6c601f26236998fb4c3e1181aeec"></span><em class="property">const</em> double <code class="descclassname"></code><code class="descname">UNDEF</code><a class="headerlink" href="#_CPPv25UNDEF" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Unknown value (small / very negative value) </p>
</dd></dl>

</div>
<div class="section" id="basic-classes">
<h2>Basic classes<a class="headerlink" href="#basic-classes" title="Permalink to this headline">¶</a></h2>
<div class="section" id="dataholder">
<h3>DataHolder<a class="headerlink" href="#dataholder" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib10DataHolderE">
<span id="egglib::DataHolder"></span><span class="target" id="project0classegglib_1_1DataHolder"></span><em class="property">class </em></dt>
<dd><p>Integer data set. </p>
<p>Holds a data set with associated sample names and group information. The data consists of given numbers of ingroup and outgroup samples, which can all have a different number of sites, unless the object is configured to be a matrix. In that cases, it is assumed that all samples have the same number of sites as the first loaded sample. There can be any number of group levels (but this number must be the same for all samples), meaning that samples can be described by several group labels in addition to their name. Group labels are not group indices (they do not need to be consecutive). There is a separate data set for sample belonging to the outgroup. There can be any number of outgroup samples. The outgroup has always one level of group labels, but the labels are not initialized. All data are represented by signed integers. Note that none of the accessors performs out-of-bound checking. The user is responsible to provide valid indices. This class follows a memory caching system: allocated memory is never freed with the aim of efficiently reusing the same object.</p>
<p>Header: &lt;egglib-cpp/DataHolder.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder10DataHolderEb">
<span id="egglib::DataHolder::DataHolder__b"></span><span class="target" id="project0classegglib_1_1DataHolder_1a661ee068cc1ea1e1010875f68b8c1e73"></span><code class="descclassname">egglib::DataHolder::</code><code class="descname">DataHolder</code><span class="sig-paren">(</span>bool <em>is_matrix</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder10DataHolderEb" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Default constructor. </p>
<p>Create an empty matrix. The object will be usable only when resizers will have been called.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">is_matrix</span></code> - <p>determines if the object is configured to a matrix (that is, where all ingroup and outgroup samples have the same number of sites). If so, the user is responsible to ensure that all loaded samples are consistent. This value can be changed using the method is_matrix(bool). </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder10DataHolderERK10DataHolder">
<span id="egglib::DataHolder::DataHolder__DataHolderCR"></span><span class="target" id="project0classegglib_1_1DataHolder_1a1260ecb52c864cbfebee2c34b5a2f3ff"></span><code class="descclassname">egglib::DataHolder::</code><code class="descname">DataHolder</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10DataHolder10DataHolderEb" title="egglib::DataHolder::DataHolder">DataHolder</a> &amp;<em>src</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder10DataHolderERK10DataHolder" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Copy constructor. </p>
<p>The reserved memory of the source is not copied. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolderD0Ev">
<span id="egglib::DataHolder::~DataHolder"></span>virtual <span class="target" id="project0classegglib_1_1DataHolder_1a4a6be32dacbe9e4d887efaa9464cf694"></span><code class="descclassname">egglib::DataHolder::</code><code class="descname">~DataHolder</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolderD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder5clearEb">
<span id="egglib::DataHolder::clear__b"></span><span class="target" id="project0classegglib_1_1DataHolder_1a1362bddf3a13884762bba09e48917171"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">clear</code><span class="sig-paren">(</span>bool <em>is_matrix</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder5clearEb" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Clear object. </p>
<p>Actually clears all memory stored by the object (including cache). All memory vector data are effectively lost and memory is released. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder12del_sample_iEj">
<span id="egglib::DataHolder::del_sample_i__unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1a9b217ad620a54c44770744bbdf9e8267"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">del_sample_i</code><span class="sig-paren">(</span>unsigned int <em>sam</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder12del_sample_iEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Delete a sample. </p>
<p>Delete the specified sample and decrease index of all subsequent samples by one. If there is only one sequence in the instance, set the number of sites (the maximal number of sites for non-matrix objects) to 0. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder12del_sample_oEj">
<span id="egglib::DataHolder::del_sample_o__unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1a7799df5f5d3dc29a0f8c90c9d91dedd9"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">del_sample_o</code><span class="sig-paren">(</span>unsigned int <em>sam</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder12del_sample_oEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Delete an outgroup sample. </p>
<p>Delete the specified sample and decrease index of all subsequent samples by one. If there is only one sequence in the instance, set the number of sites (the maximal number of sites for non-matrix objects) to 0. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder9del_sitesEjj">
<span id="egglib::DataHolder::del_sites__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1a33f139e2fd441e7ac02a795dfc28b8ed"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">del_sites</code><span class="sig-paren">(</span>unsigned int <em>start</em>, unsigned int <em>stop</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder9del_sitesEjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Delete a given range of sites. </p>
<p>The sites are removed for all ingroup and outgroup samples. The index must be valid. This method may be used for both matrix and non-matrix objects.</p>
<p>
If the stop argument is larger than the number of sites (the number of sites for this sample in the case of a non-matrix object), then sites are removed until the end of the sequence. If the start argument is larger or equal to the number of sites (the number of sites for this sample in the case of a non-matrix object), then nothing is done. <dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">start</span></code> - <p>start position of the range to remove. </p>
</li>
<li><code class="first docutils literal"><span class="pre">stop</span></code> - <p>stop position of the range to remove (this site IS NOT removed).</p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder11del_sites_iEjjj">
<span id="egglib::DataHolder::del_sites_i__unsigned-i.unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1af29bb9e7fed3537cde5abcb9d6286fc6"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">del_sites_i</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, unsigned int <em>start</em>, unsigned int <em>stop</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder11del_sites_iEjjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Delete a given range of sites for an ingroup sample. </p>
<p>As <a class="reference internal" href="#project0classegglib_1_1DataHolder_1a33f139e2fd441e7ac02a795dfc28b8ed"><span class="std std-ref">del_sites(unsigned int, unsigned int)</span></a> but for a single sample. This method <strong>may not</strong> be called on a matrix object. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder11del_sites_oEjjj">
<span id="egglib::DataHolder::del_sites_o__unsigned-i.unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1af00a1b517df67c60a8eaba524fb50b3d"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">del_sites_o</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, unsigned int <em>start</em>, unsigned int <em>stop</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder11del_sites_oEjjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Delete a given range of sites for an outgroup sample. </p>
<p>As <a class="reference internal" href="#project0classegglib_1_1DataHolder_1a33f139e2fd441e7ac02a795dfc28b8ed"><span class="std std-ref">del_sites(unsigned int, unsigned int)</span></a> but for a single sample. This method <strong>may not</strong> be called on a matrix object. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder4findEjbR9VectorIntjj">
<span id="egglib::DataHolder::find__unsigned-i.b.VectorIntR.unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1a0ad4417480195d2365981734ee8c85a1"></span>unsigned int <code class="descclassname">egglib::DataHolder::</code><code class="descname">find</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, bool <em>of_outgroup</em>, <a class="reference internal" href="#_CPPv2N6egglib9VectorIntE" title="egglib::VectorInt">VectorInt</a> &amp;<em>motif</em>, unsigned int <em>start</em>, unsigned int <em>stop</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder4findEjbR9VectorIntjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Find the start position of the first match of a motif. </p>
<p><dl class="docutils">
<dt><strong>Return</strong></dt>
<dd>The index of the start positon of the first exact match for the passed set of values, or <a class="reference internal" href="#project0namespaceegglib_1ab498891cdf034be8678af906644e6dc1"><span class="std std-ref">egglib::MAX</span></a> if no match was found (within the specified region). </dd>
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">sam</span></code> - <p>sample index. </p>
</li>
<li><code class="first docutils literal"><span class="pre">of_outgroup</span></code> - <p>specifies whether the sample is in the outgroup. </p>
</li>
<li><code class="first docutils literal"><span class="pre">motif</span></code> - <p>the list of integers representing the motif to be found. </p>
</li>
<li><code class="first docutils literal"><span class="pre">start</span></code> - <p>at which to start search. No returned value will be smaller than this value. </p>
</li>
<li><code class="first docutils literal"><span class="pre">stop</span></code> - <p>position at which to stop search (the motif cannot overlap this position). No returned value will be larger than <em>stop</em> - <em>n</em>. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder11get_group_iEjj">
<span id="egglib::DataHolder::get_group_i__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1af2a25df3133ab79cfbf08cc95d6a3325"></span>unsigned int <code class="descclassname">egglib::DataHolder::</code><code class="descname">get_group_i</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, unsigned int <em>lvl</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder11get_group_iEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get a group label. </p>
<p>
The indices must be valid, otherwise a segmentation fault or aberrant behaviour will occur. <dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">sam</span></code> - <p>sample index. </p>
</li>
<li><code class="first docutils literal"><span class="pre">lvl</span></code> - <p>group level index.</p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder11get_group_oEj">
<span id="egglib::DataHolder::get_group_o__unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1a034220b5965926e72c3b67930858b7c8"></span>unsigned int <code class="descclassname">egglib::DataHolder::</code><code class="descname">get_group_o</code><span class="sig-paren">(</span>unsigned int <em>sam</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder11get_group_oEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the group label for an outgroup sample. </p>
<p>
The indexe must be valid, otherwise a segmentation fault or aberrant behaviour will occur. There is necessarily one group level for outgroups, and the default value is 0. <dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">sam</span></code> - <p>sample index.</p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder5get_iEjj">
<span id="egglib::DataHolder::get_i__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1a92566bec9f6fa279cff91e7fcae291e9"></span>int <code class="descclassname">egglib::DataHolder::</code><code class="descname">get_i</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, unsigned int <em>sit</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder5get_iEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an ingroup data entry. </p>
<p>
The indices must be valid, otherwise a segmentation fault or aberrant behaviour will occur. <dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">sam</span></code> - <p>sample index. </p>
</li>
<li><code class="first docutils literal"><span class="pre">sit</span></code> - <p>site index.</p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder13get_is_matrixEv">
<span id="egglib::DataHolder::get_is_matrix"></span><span class="target" id="project0classegglib_1_1DataHolder_1ac707df18ebe194358dfeb6ac2748e1c9"></span>bool <code class="descclassname">egglib::DataHolder::</code><code class="descname">get_is_matrix</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder13get_is_matrixEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Check if the object is configured to be a matrix. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder10get_name_iEj">
<span id="egglib::DataHolder::get_name_i__unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1a3b68a25cc2a7a748bb987971d21e78a6"></span><em class="property">const</em> char *<code class="descclassname">egglib::DataHolder::</code><code class="descname">get_name_i</code><span class="sig-paren">(</span>unsigned int <em>sam</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder10get_name_iEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an ingroup name. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder10get_name_oEj">
<span id="egglib::DataHolder::get_name_o__unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1af9e3a96c137e8e0f7acbfd86aa420c83"></span><em class="property">const</em> char *<code class="descclassname">egglib::DataHolder::</code><code class="descname">get_name_o</code><span class="sig-paren">(</span>unsigned int <em>sam</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder10get_name_oEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an outgroup name. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder11get_ngroupsEv">
<span id="egglib::DataHolder::get_ngroups"></span><span class="target" id="project0classegglib_1_1DataHolder_1ac0655db80d562d684068892bc6b4646a"></span>unsigned int <code class="descclassname">egglib::DataHolder::</code><code class="descname">get_ngroups</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder11get_ngroupsEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the number of group levels. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder10get_nsam_iEv">
<span id="egglib::DataHolder::get_nsam_i"></span><span class="target" id="project0classegglib_1_1DataHolder_1a9edf800cbbecdeb844794d9c7628bd60"></span>unsigned int <code class="descclassname">egglib::DataHolder::</code><code class="descname">get_nsam_i</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder10get_nsam_iEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the number of ingroup samples. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder10get_nsam_oEv">
<span id="egglib::DataHolder::get_nsam_o"></span><span class="target" id="project0classegglib_1_1DataHolder_1a40be949e683b4ce6422f55196050959d"></span>unsigned int <code class="descclassname">egglib::DataHolder::</code><code class="descname">get_nsam_o</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder10get_nsam_oEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the number of outgroup samples. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder8get_nsitEv">
<span id="egglib::DataHolder::get_nsit"></span><span class="target" id="project0classegglib_1_1DataHolder_1a7d1f80a655a96d3d1559a33f95d2e9bf"></span>unsigned int <code class="descclassname">egglib::DataHolder::</code><code class="descname">get_nsit</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder8get_nsitEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the number of sites. </p>
<p>This method <strong>may not</strong> be called on a non-matrix object. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder10get_nsit_iEj">
<span id="egglib::DataHolder::get_nsit_i__unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1afdb5c47eb7509f3e4504beca0718df9c"></span>unsigned int <code class="descclassname">egglib::DataHolder::</code><code class="descname">get_nsit_i</code><span class="sig-paren">(</span>unsigned int <em>sam</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder10get_nsit_iEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the number of sites for an ingroup sample. </p>
<p>This method <strong>may not</strong> be called on a matrix object. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder10get_nsit_oEj">
<span id="egglib::DataHolder::get_nsit_o__unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1abb59bf499bdc96efee991bfedab5dbf6"></span>unsigned int <code class="descclassname">egglib::DataHolder::</code><code class="descname">get_nsit_o</code><span class="sig-paren">(</span>unsigned int <em>sam</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder10get_nsit_oEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the number of sites for an outgroup sample. </p>
<p>This method <strong>may not</strong> be called on a matrix object. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder5get_oEjj">
<span id="egglib::DataHolder::get_o__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1a935c7e78a55e632aca0885a6f15c2c48"></span>int <code class="descclassname">egglib::DataHolder::</code><code class="descname">get_o</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, unsigned int <em>sit</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder5get_oEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an outgroup data entry. </p>
<p>
The indices must be valid, otherwise a segmentation fault or aberrant behaviour will occur. <dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">sam</span></code> - <p>sample index. </p>
</li>
<li><code class="first docutils literal"><span class="pre">sit</span></code> - <p>site index.</p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder12insert_sitesEjj">
<span id="egglib::DataHolder::insert_sites__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1a483f62cdd190b5e2f90b95712b17c131"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">insert_sites</code><span class="sig-paren">(</span>unsigned int <em>pos</em>, unsigned int <em>num</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder12insert_sitesEjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Insert sites at a given position. </p>
<p>Increase the number of sites for all samples. This method may be used for matrix or non-matrix objects. Note that the inserted sites are not initialized.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">pos</span></code> - <p>the position at which to insert sites. The new sites are inserted before the specified index. Use 0 to add sites at the beginning of the sequence, and the current number of sites to add sites at the end. If the value is larger than the current number of sites, sites are added at the end of the sequence. Therefore it is possible to use <a class="reference internal" href="#project0namespaceegglib_1ab498891cdf034be8678af906644e6dc1"><span class="std std-ref">egglib::MAX</span></a> as the position to specify that new sites must be inserted at the end. </p>
</li>
<li><code class="first docutils literal"><span class="pre">num</span></code> - <p>number of sites at add. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder14insert_sites_iEjjj">
<span id="egglib::DataHolder::insert_sites_i__unsigned-i.unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1af3df67a229de317c5cf1c250f1346271"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">insert_sites_i</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, unsigned int <em>pos</em>, unsigned int <em>num</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder14insert_sites_iEjjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Insert sites at a given position for an ingroup sample. </p>
<p>As insert_sites(unsigned int, unsigned int, int) but for only one sample of the ingroup. Available only for non-matrix objects. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder14insert_sites_oEjjj">
<span id="egglib::DataHolder::insert_sites_o__unsigned-i.unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1a7752b7404a54452a155a72ed05dfc6cd"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">insert_sites_o</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, unsigned int <em>pos</em>, unsigned int <em>num</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder14insert_sites_oEjjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Insert sites at a given position for an outgroup sample. </p>
<p>As insert_sites(unsigned int, unsigned int, int) but for only one sample of the outgroup. Available only for non-matrix objects. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder8is_equalEv">
<span id="egglib::DataHolder::is_equal"></span><span class="target" id="project0classegglib_1_1DataHolder_1aaeb592dd50fefcba55123c0a8a05183d"></span>bool <code class="descclassname">egglib::DataHolder::</code><code class="descname">is_equal</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder8is_equalEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Test if all sequences have the same length. </p>
<p>True if no sequences at all. Only valid for containers. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder13name_append_iEjPKc">
<span id="egglib::DataHolder::name_append_i__unsigned-i.cCP"></span><span class="target" id="project0classegglib_1_1DataHolder_1a7670afa9fc096d9d7f7d9a394a035f19"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">name_append_i</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, <em class="property">const</em> char *<em>ch</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder13name_append_iEjPKc" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add character at the end of specified ingroup name. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder13name_append_oEjPKc">
<span id="egglib::DataHolder::name_append_o__unsigned-i.cCP"></span><span class="target" id="project0classegglib_1_1DataHolder_1a1b4de7d1d7cb043bb9fbba24980726ca"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">name_append_o</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, <em class="property">const</em> char *<em>ch</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder13name_append_oEjPKc" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add characters at the end of specified ingroup name. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder15name_appendch_iEjc">
<span id="egglib::DataHolder::name_appendch_i__unsigned-i.c"></span><span class="target" id="project0classegglib_1_1DataHolder_1a036330d202690811d921225944c91c9f"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">name_appendch_i</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, char <em>ch</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder15name_appendch_iEjc" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add a character to the specified ingroup name. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder15name_appendch_oEjc">
<span id="egglib::DataHolder::name_appendch_o__unsigned-i.c"></span><span class="target" id="project0classegglib_1_1DataHolder_1affe6a6503a1a5844af559c0df1a40600"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">name_appendch_o</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, char <em>ch</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder15name_appendch_oEjc" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add a character to the specified outgroup name. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolderaSERK10DataHolder">
<span id="egglib::DataHolder::assign-operator__DataHolderCR"></span><span class="target" id="project0classegglib_1_1DataHolder_1acee1c9216315cfef18a47ef44412366d"></span><a class="reference internal" href="#_CPPv2N6egglib10DataHolderE" title="egglib::DataHolder">DataHolder</a> &amp;<code class="descclassname">egglib::DataHolder::</code><code class="descname">operator=</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10DataHolderE" title="egglib::DataHolder">DataHolder</a> &amp;<em>src</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolderaSERK10DataHolder" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Assignment operator. </p>
<p>The reserved memory of the source is not copied. The reserved memory of the current object is retained. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder7reserveEjjjjj">
<span id="egglib::DataHolder::reserve__unsigned-i.unsigned-i.unsigned-i.unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1a0e93fb0afb1bc08176ffc73e7d6708a7"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">reserve</code><span class="sig-paren">(</span>unsigned int <em>nsi</em>, unsigned int <em>nso</em>, unsigned int <em>ln</em>, unsigned int <em>ng</em>, unsigned int <em>ls</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder7reserveEjjjjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Reserve memory to speed up data loading. </p>
<p>This method does not change the size of the data set contained in the instance, but reserves memory in order to speed up incremental loading of data. The passed values are not required to be accurate. In case the instance has allocated more memory than what is requested, nothing is done (this applies to both dimensions independently). It is always valid to use 0 for any values (in that case, nothing is done). Note that one character is always pre-allocated for all names.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">nsi</span></code> - <p>expected number of ingroup samples. </p>
</li>
<li><code class="first docutils literal"><span class="pre">nso</span></code> - <p>expected number of outgroup samples. </p>
</li>
<li><code class="first docutils literal"><span class="pre">ln</span></code> - <p>expected length of names. </p>
</li>
<li><code class="first docutils literal"><span class="pre">ng</span></code> - <p>expected number of groups. </p>
</li>
<li><code class="first docutils literal"><span class="pre">ls</span></code> - <p>expected number of sites (the same for all ingroup and outgroup samples, whichever the object is a matrix or not). </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder5resetEb">
<span id="egglib::DataHolder::reset__b"></span><span class="target" id="project0classegglib_1_1DataHolder_1ad9cf618002111375704d635f1b063ccc"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">reset</code><span class="sig-paren">(</span>bool <em>is_matrix</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder5resetEb" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Restore object to initial state. </p>
<p>This method is designed to allow reusing the object and reusing previously allocated memory. All data contained in the instance is considered to be lost, but allocated memory is actually retained to speed up later resizing operations. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder11set_group_iEjjj">
<span id="egglib::DataHolder::set_group_i__unsigned-i.unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1ae3fae222450df66f044697aa569cadef"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">set_group_i</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, unsigned int <em>lvl</em>, unsigned int <em>label</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder11set_group_iEjjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set a group label. </p>
<p>
The indices must be valid, otherwise a segmentation fault or aberrant behaviour will occur. <dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">sam</span></code> - <p>sample index. </p>
</li>
<li><code class="first docutils literal"><span class="pre">lvl</span></code> - <p>group level index. </p>
</li>
<li><code class="first docutils literal"><span class="pre">label</span></code> - <p>group label.</p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder11set_group_oEjj">
<span id="egglib::DataHolder::set_group_o__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1a86e66a0e4a0eeddb6f1213421a390fd3"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">set_group_o</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, unsigned int <em>label</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder11set_group_oEjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set the group label for an outgroup sample. </p>
<p>
The indexe must be valid, otherwise a segmentation fault or aberrant behaviour will occur. There is necessarily one group level for outgroups. <dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">sam</span></code> - <p>sample index. </p>
</li>
<li><code class="first docutils literal"><span class="pre">label</span></code> - <p>group label.</p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder5set_iEjji">
<span id="egglib::DataHolder::set_i__unsigned-i.unsigned-i.i"></span><span class="target" id="project0classegglib_1_1DataHolder_1aabee69ce2eeeebab65d24e09d22b35e9"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">set_i</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, unsigned int <em>sit</em>, int <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder5set_iEjji" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set an ingroup data entry. </p>
<p>
The indices must be valid, otherwise a segmentation fault or aberrant behaviour will occur. <dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">sam</span></code> - <p>sample index. </p>
</li>
<li><code class="first docutils literal"><span class="pre">sit</span></code> - <p>site index. </p>
</li>
<li><code class="first docutils literal"><span class="pre">value</span></code> - <p>allele value.</p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder13set_is_matrixEb">
<span id="egglib::DataHolder::set_is_matrix__b"></span><span class="target" id="project0classegglib_1_1DataHolder_1add554c3f63096d57bbb384374fa6f0e4"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">set_is_matrix</code><span class="sig-paren">(</span>bool <em>flag</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder13set_is_matrixEb" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Configure the object (not) to be a matrix. </p>
<p>If a non-matrix object is converted to a matrix, the user is responsible of ensuring that all samples (including outgroup) have the same number of samples. The method will assume that all samples have the same number of sites as the first sample among the ingroup and outgroup. There is no requirement for converting a matrix to a non-matrix. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder10set_name_iEjPKc">
<span id="egglib::DataHolder::set_name_i__unsigned-i.cCP"></span><span class="target" id="project0classegglib_1_1DataHolder_1ae197cb2f5937f23bb30d94b8ee3e98a5"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">set_name_i</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, <em class="property">const</em> char *<em>name</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder10set_name_iEjPKc" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set an ingroup name. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder10set_name_oEjPKc">
<span id="egglib::DataHolder::set_name_o__unsigned-i.cCP"></span><span class="target" id="project0classegglib_1_1DataHolder_1ad7550bc2af4336648526fac60eb4edec"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">set_name_o</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, <em class="property">const</em> char *<em>name</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder10set_name_oEjPKc" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set an outgroup name. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder11set_ngroupsEj">
<span id="egglib::DataHolder::set_ngroups__unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1a41efb8a232782547ced2a1f91f511645"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">set_ngroups</code><span class="sig-paren">(</span>unsigned int <em>ngrp</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder11set_ngroupsEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set the number of group levels. </p>
<p>Perform memory allocation as needed but does not initialize new values. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder10set_nsam_iEj">
<span id="egglib::DataHolder::set_nsam_i__unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1a8955b30541a50b642a124d614f5af00b"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">set_nsam_i</code><span class="sig-paren">(</span>unsigned int <em>nsam</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder10set_nsam_iEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set the number of ingroup samples. </p>
<p>Perform memory allocation as needed but does not initialize new values (except names). If the object is a matrix, new samples are set to have the current number of sites. Otherwise, new samples have no sites. Set the number of samples to a smaller value equals to remove the last samples. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder10set_nsam_oEj">
<span id="egglib::DataHolder::set_nsam_o__unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1a661af771de40c7a627cf217bbbd2043b"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">set_nsam_o</code><span class="sig-paren">(</span>unsigned int <em>nsam</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder10set_nsam_oEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set the number of outgroup samples. </p>
<p>As nsam(unsigned int) but for the outgroup data table. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder8set_nsitEj">
<span id="egglib::DataHolder::set_nsit__unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1a45f7145fad1e627db62d8d507b845bd6"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">set_nsit</code><span class="sig-paren">(</span>unsigned int <em>val</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder8set_nsitEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set the number of sites. </p>
<p>Perform memory allocation as needed but does not initialize new values. It is possible to use this method for both matrix and non-matrix objects. In both cases, the effective result is that all ingroup and outgroup samples are resized to the specified number of sites. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder10set_nsit_iEjj">
<span id="egglib::DataHolder::set_nsit_i__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1a847e55d5153361626e84cae7a05912d2"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">set_nsit_i</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, unsigned int <em>val</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder10set_nsit_iEjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set the number of sites for an ingroup sample. </p>
<p>Similar to nsit(unsigned int) but for only one sample of the ingroup. Available only for non-matrix objects. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder10set_nsit_oEjj">
<span id="egglib::DataHolder::set_nsit_o__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1abea0db11f99b977f61cd50099bc63b06"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">set_nsit_o</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, unsigned int <em>val</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder10set_nsit_oEjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set the number of sites for an outgroup sample. </p>
<p>Similar to nsit(unsigned int) but for only one sample of the outgroup. Available only for non-matrix objects. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder5set_oEjji">
<span id="egglib::DataHolder::set_o__unsigned-i.unsigned-i.i"></span><span class="target" id="project0classegglib_1_1DataHolder_1ad43661404818b7b32ead8362323f7dce"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">set_o</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, unsigned int <em>sit</em>, int <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder5set_oEjji" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set an outgroup data entry. </p>
<p>
The indices must be valid, otherwise a segmentation fault or aberrant behaviour will occur. <dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">sam</span></code> - <p>sample index. </p>
</li>
<li><code class="first docutils literal"><span class="pre">sit</span></code> - <p>site index. </p>
</li>
<li><code class="first docutils literal"><span class="pre">value</span></code> - <p>allele value.</p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder11to_outgroupEjj">
<span id="egglib::DataHolder::to_outgroup__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1DataHolder_1afb1802a061830557c62bf1b963cc3e30"></span>void <code class="descclassname">egglib::DataHolder::</code><code class="descname">to_outgroup</code><span class="sig-paren">(</span>unsigned int <em>sam</em>, unsigned int <em>label</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder11to_outgroupEjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Move a sample to the outgroup. </p>
<p>The specified sample is moved to the outgroup and its group labels are discarded. Obviously, this decreases the ingroup size by 1, and increases the outgroup size accordingly.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">sam</span></code> - <p>ingroup sample index. </p>
</li>
<li><code class="first docutils literal"><span class="pre">label</span></code> - <p>group label to assign to the sample after it it is moved to the outgroup (use 0 if not relevant). </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder14valid_phyml_aaEv">
<span id="egglib::DataHolder::valid_phyml_aa"></span><span class="target" id="project0classegglib_1_1DataHolder_1a72e994247ff2167f09b03f3583434b05"></span>bool <code class="descclassname">egglib::DataHolder::</code><code class="descname">valid_phyml_aa</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder14valid_phyml_aaEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>True if all data are amino acids expected by PhyML (for alignment) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder17valid_phyml_namesEv">
<span id="egglib::DataHolder::valid_phyml_names"></span><span class="target" id="project0classegglib_1_1DataHolder_1a0a81dc621bcdd226ed3214fb5930c7f1"></span>bool <code class="descclassname">egglib::DataHolder::</code><code class="descname">valid_phyml_names</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder17valid_phyml_namesEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>True if all names are got for PhyMl. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10DataHolder14valid_phyml_ntEv">
<span id="egglib::DataHolder::valid_phyml_nt"></span><span class="target" id="project0classegglib_1_1DataHolder_1aad648c0c4a7dc0d6cdde55003e4e0cb2"></span>bool <code class="descclassname">egglib::DataHolder::</code><code class="descname">valid_phyml_nt</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10DataHolder14valid_phyml_ntEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>True if all data are nucleotides expected by PhyML (for alignment) </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="geneticcode">
<h3>GeneticCode<a class="headerlink" href="#geneticcode" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib11GeneticCodeE">
<span id="egglib::GeneticCode"></span><span class="target" id="project0classegglib_1_1GeneticCode"></span><em class="property">class </em></dt>
<dd><p>Hold genetic code tables. </p>
<p>Handle genetic codes. All genetic codes defined by the National Centor for Biotechnology Information are supported.</p>
<p>Header: &lt;egglib-cpp/GeneticCode.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode11GeneticCodeEj">
<span id="egglib::GeneticCode::GeneticCode__unsigned-i"></span><span class="target" id="project0classegglib_1_1GeneticCode_1adf6989c7fe5e3ff6c103fc033b689283"></span><code class="descclassname">egglib::GeneticCode::</code><code class="descname">GeneticCode</code><span class="sig-paren">(</span>unsigned int <em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode11GeneticCodeEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor. </p>
<p>Build an instance of the genetic code. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode11GeneticCodeEv">
<span id="egglib::GeneticCode::GeneticCode"></span><span class="target" id="project0classegglib_1_1GeneticCode_1a24e9c320e5456b72026ed2a01ef094b3"></span><code class="descclassname">egglib::GeneticCode::</code><code class="descname">GeneticCode</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode11GeneticCodeEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Default constructor. </p>
<p>Like GeneticCode(int), except that the code used is 1 (standard). </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode9aminoacidEj">
<span id="egglib::GeneticCode::aminoacid__unsigned-i"></span><span class="target" id="project0classegglib_1_1GeneticCode_1a4f626748658501ba732967a9d42c91a4"></span>char <code class="descclassname">egglib::GeneticCode::</code><code class="descname">aminoacid</code><span class="sig-paren">(</span>unsigned int <em>codon</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode9aminoacidEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Returns the translation of a codon. </p>
<p>The codon should be passed as an integer code (see <a class="reference internal" href="#project0classegglib_1_1GeneticCode_1aea1d701310aa3b0a43bbb9a81169821e"><span class="std std-ref">codon()</span></a>). This methods returns the single amino acid code for valid codons (represented by integers in the range 0-63), and ‘X’ for any other integer. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode8get_codeEv">
<span id="egglib::GeneticCode::get_code"></span><span class="target" id="project0classegglib_1_1GeneticCode_1a9dd406e851ac67e92841f9608d91e7cc"></span>unsigned int <code class="descclassname">egglib::GeneticCode::</code><code class="descname">get_code</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode8get_codeEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the current genetic code. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode4nameEv">
<span id="egglib::GeneticCode::name"></span><span class="target" id="project0classegglib_1_1GeneticCode_1a860a900733b4494b77121df7259a064f"></span><em class="property">const</em> char *<code class="descclassname">egglib::GeneticCode::</code><code class="descname">name</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode4nameEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the name of the genetic code. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode7NSsitesEjb">
<span id="egglib::GeneticCode::NSsites__unsigned-i.b"></span><span class="target" id="project0classegglib_1_1GeneticCode_1ab81ed61ba51a1fec17b407a4bd2d5913"></span>double <code class="descclassname">egglib::GeneticCode::</code><code class="descname">NSsites</code><span class="sig-paren">(</span>unsigned int <em>codon</em>, bool <em>ignorestop</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode7NSsitesEjb" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Give the number of non-synonymous sites of a codon. </p>
<p>The number is in the range 0-3 (3 is all changes at all of the three positions would lead to a non-synonymous change).</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">codon</span></code> - <p>codon integer code (see <a class="reference internal" href="#project0classegglib_1_1GeneticCode_1aea1d701310aa3b0a43bbb9a81169821e"><span class="std std-ref">codon()</span></a>).</p>
</li>
<li><code class="first docutils literal"><span class="pre">ignorestop</span></code> - <p>if true, potential changes to stop codons are excluded and all stop codons return 0; if false, changes to stop codons are considered to be non-synonymous. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode7NSsitesERK10SiteHolderRK10SiteHolderRK10SiteHolderRjbiiii">
<span id="egglib::GeneticCode::NSsites__SiteHolderCR.SiteHolderCR.SiteHolderCR.unsigned-iR.b.i.i.i.i"></span><span class="target" id="project0classegglib_1_1GeneticCode_1a21880cd71260f502a54ebe83a7a005a1"></span>double <code class="descclassname">egglib::GeneticCode::</code><code class="descname">NSsites</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="diversity.html#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site1</em>, <em class="property">const</em> <a class="reference internal" href="diversity.html#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site2</em>, <em class="property">const</em> <a class="reference internal" href="diversity.html#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site3</em>, unsigned int &amp;<em>num_samples</em>, bool <em>ignorestop</em>, int <em>A</em>, int <em>C</em>, int <em>G</em>, int <em>T</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode7NSsitesERK10SiteHolderRK10SiteHolderRK10SiteHolderRjbiiii" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Give the number of non-synonymous sites of a codon site. </p>
<p>The number is in the range 0-3 (3 is all changes at all of the three positions would lead to a non-synonymous change). This is the same as NSistes(unsigned int, bool), but average over all samples based on provided Site instances.</p>
<p>
All three codon positions must have the same number of samples such as the ith nucleotides at the three sites give the codon for the ith sample. (Same ploidy as well.)<dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">site1</span></code> - <p>first position of the codon site. </p>
</li>
<li><code class="first docutils literal"><span class="pre">site2</span></code> - <p>second position of the codon site. </p>
</li>
<li><code class="first docutils literal"><span class="pre">site3</span></code> - <p>third position of the codon site. </p>
</li>
<li><code class="first docutils literal"><span class="pre">num_samples</span></code> - <p>variable used to provide the number of samples analyzed by the method (that is, number of samples minus number of samples containing at least one missing data). The original value of the variable is ignored and is modified by the instance. If 0, the return value should be ignored. </p>
</li>
<li><code class="first docutils literal"><span class="pre">ignorestop</span></code> - <p>if true, potential changes to stop codons are excluded and all stop codons are treated as missing data; if false, changes to stop codons are considered to be non-synonymous. </p>
</li>
<li><code class="first docutils literal"><span class="pre">A</span></code> - <p>integer value representing bases A. </p>
</li>
<li><code class="first docutils literal"><span class="pre">C</span></code> - <p>integer value representing bases C. </p>
</li>
<li><code class="first docutils literal"><span class="pre">G</span></code> - <p>integer value representing bases G. </p>
</li>
<li><code class="first docutils literal"><span class="pre">T</span></code> - <p>integer value representing bases T.</p>
</li>
</ul>
</dd>
</dl>
</p>
<p>All nucleotides represented by an integer allele values not matching either of the four values passed as the A, C, G and T arguments are considered as missing data.</p>
<p><dl class="docutils">
<dt><strong>Warning</strong></dt>
<dd>This method will not accept coding sequences mixing upper and lower case characters. It is however possible to configure how the four nucleotides are represented. </dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode7NSsitesERK10SiteHolderRjbiiii">
<span id="egglib::GeneticCode::NSsites__SiteHolderCR.unsigned-iR.b.i.i.i.i"></span><span class="target" id="project0classegglib_1_1GeneticCode_1a63b29806b3d47658e684657efdd7fa4d"></span>double <code class="descclassname">egglib::GeneticCode::</code><code class="descname">NSsites</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="diversity.html#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>codons</em>, unsigned int &amp;<em>num_samples</em>, bool <em>ignorestop</em>, int <em>A</em>, int <em>C</em>, int <em>G</em>, int <em>T</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode7NSsitesERK10SiteHolderRjbiiii" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Give the number of non-synonymous sites of a codon site. </p>
<p>See NSsites(const SiteHolder&amp;, const SiteHolder&amp;, const SiteHolder&amp;, unsigned int&amp;, bool, int, int C, int, int). Do the same, except that this version takes a single <a class="reference internal" href="diversity.html#project0classegglib_1_1SiteHolder"><span class="std std-ref">SiteHolder</span></a> reference instance of three, and the <a class="reference internal" href="diversity.html#project0classegglib_1_1SiteHolder"><span class="std std-ref">SiteHolder</span></a> reference passed to this function contains integer alleles representing codons, that is in the range 0-63. Other values are considered to be missing data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode8set_codeEj">
<span id="egglib::GeneticCode::set_code__unsigned-i"></span><span class="target" id="project0classegglib_1_1GeneticCode_1a8aa0397d9e2f0abd9933129c946132a5"></span>void <code class="descclassname">egglib::GeneticCode::</code><code class="descname">set_code</code><span class="sig-paren">(</span>unsigned int <em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode8set_codeEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set the genetic code. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode6SsitesEjb">
<span id="egglib::GeneticCode::Ssites__unsigned-i.b"></span><span class="target" id="project0classegglib_1_1GeneticCode_1a408974b20ec1f3abb15525c4768b4366"></span>double <code class="descclassname">egglib::GeneticCode::</code><code class="descname">Ssites</code><span class="sig-paren">(</span>unsigned int <em>codon</em>, bool <em>ignorestop</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode6SsitesEjb" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Give the number of synonymous sites of a codon. </p>
<p>The number is in the range 0-3 (3 is all changes at all of the three positions would lead to a non-synonymous change).</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">codon</span></code> - <p>codon integer code (see <a class="reference internal" href="#project0classegglib_1_1GeneticCode_1aea1d701310aa3b0a43bbb9a81169821e"><span class="std std-ref">codon()</span></a>).</p>
</li>
<li><code class="first docutils literal"><span class="pre">ignorestop</span></code> - <p>if true, potential changes to stop codons are excluded and all stop codons return 0; if false, changes to stop codons are considered to be non-synonymous. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode6SsitesERK10SiteHolderRK10SiteHolderRK10SiteHolderRjbiiii">
<span id="egglib::GeneticCode::Ssites__SiteHolderCR.SiteHolderCR.SiteHolderCR.unsigned-iR.b.i.i.i.i"></span><span class="target" id="project0classegglib_1_1GeneticCode_1a9d67b25aad6f2c7badbd66bd009fb429"></span>double <code class="descclassname">egglib::GeneticCode::</code><code class="descname">Ssites</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="diversity.html#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site1</em>, <em class="property">const</em> <a class="reference internal" href="diversity.html#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site2</em>, <em class="property">const</em> <a class="reference internal" href="diversity.html#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site3</em>, unsigned int &amp;<em>num_samples</em>, bool <em>ignorestop</em>, int <em>A</em>, int <em>C</em>, int <em>G</em>, int <em>T</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode6SsitesERK10SiteHolderRK10SiteHolderRK10SiteHolderRjbiiii" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Like NSsites, for synonmous sites. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode6SsitesERK10SiteHolderRjbiiii">
<span id="egglib::GeneticCode::Ssites__SiteHolderCR.unsigned-iR.b.i.i.i.i"></span><span class="target" id="project0classegglib_1_1GeneticCode_1a65358f761ed72c707f4d113a44d27af8"></span>double <code class="descclassname">egglib::GeneticCode::</code><code class="descname">Ssites</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="diversity.html#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>codons</em>, unsigned int &amp;<em>num_samples</em>, bool <em>ignorestop</em>, int <em>A</em>, int <em>C</em>, int <em>G</em>, int <em>T</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode6SsitesERK10SiteHolderRjbiiii" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Like NSsites, but for synonymous sites. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode5startEj">
<span id="egglib::GeneticCode::start__unsigned-i"></span><span class="target" id="project0classegglib_1_1GeneticCode_1a42915d5ec4151260df9b32d17f8168cf"></span>bool <code class="descclassname">egglib::GeneticCode::</code><code class="descname">start</code><span class="sig-paren">(</span>unsigned int <em>codon</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode5startEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Tells if a codon is an initiation codon. </p>
<p>The codon should be passed as an integer code (see <a class="reference internal" href="#project0classegglib_1_1GeneticCode_1aea1d701310aa3b0a43bbb9a81169821e"><span class="std std-ref">codon()</span></a>). This methods returns <code class="docutils literal"><span class="pre">true</span></code> if the codon in an initiation codon (including any of the alternative initiation codons known for the genetic code of the set for the current object), and <code class="docutils literal"><span class="pre">false</span></code> otherwise (including for invalid codon codes). </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode9translateEiiib">
<span id="egglib::GeneticCode::translate__i.i.i.b"></span><span class="target" id="project0classegglib_1_1GeneticCode_1ae8fe58e6dd6d476a15a9aee08dee2d7b"></span>int <code class="descclassname">egglib::GeneticCode::</code><code class="descname">translate</code><span class="sig-paren">(</span>int <em>first</em>, int <em>second</em>, int <em>third</em>, bool <em>smart</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode9translateEiiib" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Translate a codon directly. </p>
<p>
Codon positions should be ASCII-coded. Return ‘X’ if missing data or invalid nucleotides. For fourfold degenerate positions. Codons including non-ambiguity characters always return ‘X’ (even ‘?’ at a fourfold degenerate position), except if the codon is ‘’ (in that case, ‘-‘ is returned). <dl class="docutils">
<dt><strong>Return</strong></dt>
<dd>ASCII-coded aminoacid.</dd>
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">first</span></code> - <p>first codon position. </p>
</li>
<li><code class="first docutils literal"><span class="pre">second</span></code> - <p>second codon position. </p>
</li>
<li><code class="first docutils literal"><span class="pre">third</span></code> - <p>third codon position. </p>
</li>
<li><code class="first docutils literal"><span class="pre">smart</span></code> - <p>smart translation. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

</div>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Static Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode4baseEjj">
<span id="egglib::GeneticCode::base__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1GeneticCode_1a41c2c7a637452ce1f2690ec4acbf9a29"></span><em class="property">static</em> char <code class="descclassname">egglib::GeneticCode::</code><code class="descname">base</code><span class="sig-paren">(</span>unsigned int <em>codon</em>, unsigned int <em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode4baseEjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Returns one of the base of a codon. </p>
<p>This method can be called on the class directly (as in <code class="docutils literal"><span class="pre">GeneticCode::base(0,</span> <span class="pre">0)</span></code> and it is not dependent on the specification of a genetic code).</p>
<p><dl class="docutils">
<dt><strong>Return</strong></dt>
<dd>Returns the character at the specified position of the codon (as an upper-case character).</dd>
<dt><strong>Warning</strong></dt>
<dd>The methods returns ‘?’ if it cannot perform base extraction, but it is not guaranteed that all invalid arguments will be detected properly. </dd>
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">codon</span></code> - <p>the codon should be passed as an integer code (see <a class="reference internal" href="#project0classegglib_1_1GeneticCode_1aea1d701310aa3b0a43bbb9a81169821e"><span class="std std-ref">codon()</span></a>). Only values in the range 0-63 are supported. </p>
</li>
<li><code class="first docutils literal"><span class="pre">index</span></code> - <p>index of the base to extract (only 0, 1 and 2 are accepted; other values will result in aberrant outcome). </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode5codonEccc">
<span id="egglib::GeneticCode::codon__c.c.c"></span><span class="target" id="project0classegglib_1_1GeneticCode_1aea1d701310aa3b0a43bbb9a81169821e"></span>unsigned int <code class="descclassname">egglib::GeneticCode::</code><code class="descname">codon</code><span class="sig-paren">(</span>char <em>first</em>, char <em>second</em>, char <em>third</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode5codonEccc" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Return the integer code for a codon. </p>
<p>The first, second and third bases of the codon must be passed as character arguments. The case of characters is ignored. Returns and integer in the range [0, 63] for the 64 codons (see the table below).</p>
<p>
The codons are identified by single integers as given by the table below:<dl class="docutils">
<dt><strong>Warning</strong></dt>
<dd>The base ‘U’, although biologically relevant, is treated as an invalid base.</dd>
<dt><strong>Note</strong></dt>
<dd>As a static method, this method can be called as <code class="docutils literal"><span class="pre">GeneticCode::codon(base1,</span> <span class="pre">base2,</span> <span class="pre">base3)</span></code> directly (it does not require instanciation of an object) and it is not dependent on any genetic code specification.</dd>
</dl>
</p>
<p>
All other triplets: <a class="reference internal" href="#project0namespaceegglib_1a6b564311a87de05e515ec65480ebc236"><span class="std std-ref">egglib.UNKNOWN</span></a>. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode5diff1Ejj">
<span id="egglib::GeneticCode::diff1__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1GeneticCode_1a12c4a31bc23ee9fcb67c45dec6f22be9"></span><em class="property">static</em> bool <code class="descclassname">egglib::GeneticCode::</code><code class="descname">diff1</code><span class="sig-paren">(</span>unsigned int <em>codon1</em>, unsigned int <em>codon2</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode5diff1Ejj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Check if the first position of two codons is identical. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode5diff2Ejj">
<span id="egglib::GeneticCode::diff2__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1GeneticCode_1a1099dc0b7395e59340531a1334bb1f82"></span><em class="property">static</em> bool <code class="descclassname">egglib::GeneticCode::</code><code class="descname">diff2</code><span class="sig-paren">(</span>unsigned int <em>codon1</em>, unsigned int <em>codon2</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode5diff2Ejj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Check if the second position of two codons is identical. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode5diff3Ejj">
<span id="egglib::GeneticCode::diff3__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1GeneticCode_1a4a966e5cab3479222d04e1039a737eda"></span><em class="property">static</em> bool <code class="descclassname">egglib::GeneticCode::</code><code class="descname">diff3</code><span class="sig-paren">(</span>unsigned int <em>codon1</em>, unsigned int <em>codon2</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode5diff3Ejj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Check if the third position of two codons is identical. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode9int2codonEjjj">
<span id="egglib::GeneticCode::int2codon__unsigned-i.unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1GeneticCode_1a3b60daac7b0d5b754140322cd5052b1c"></span><em class="property">static</em> unsigned int <code class="descclassname">egglib::GeneticCode::</code><code class="descname">int2codon</code><span class="sig-paren">(</span>unsigned int <em>base1</em>, unsigned int <em>base2</em>, unsigned int <em>base3</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode9int2codonEjjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Return the integer code for a codon. </p>
<p>The first, second and third bases of the codon must be passed as integer code, according to the following mapping: 0 for T, 1 for C, 2 for G and 3 for T. This code must absolutely be following and no other value may be passed. Returns and integer in the range [0, 63] for the 64 codons (see the documentation of the method <a class="reference internal" href="#project0classegglib_1_1GeneticCode_1aea1d701310aa3b0a43bbb9a81169821e"><span class="std std-ref">codon(char, char, char)</span></a>).</p>
<p><dl class="docutils">
<dt><strong>Note</strong></dt>
<dd>As a static method, this method can be called as <code class="docutils literal"><span class="pre">GeneticCode::int2codon(base1,</span> <span class="pre">base2,</span> <span class="pre">base3)</span></code> directly (it does not require instanciation of an object) and it is not dependent on any genetic code specification. </dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode5ndiffEjj">
<span id="egglib::GeneticCode::ndiff__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1GeneticCode_1ae2e4194b15eaa091de7d53fca0ab5066"></span><em class="property">static</em> unsigned int <code class="descclassname">egglib::GeneticCode::</code><code class="descname">ndiff</code><span class="sig-paren">(</span>unsigned int <em>codon1</em>, unsigned int <em>codon2</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode5ndiffEjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Returns the number of nucleotide differences between two codons. </p>
<p>This method can be called on the class directly (as in <code class="docutils literal"><span class="pre">GeneticCode::base(0,</span> <span class="pre">0)</span></code> and it is not dependent on the specification of a genetic code).</p>
<p>This method is only valid if both arguments are less than<ol class="arabic simple">
<li>Returns only 0, 1, 2 or 3. </li>
</ol>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib11GeneticCode9num_codesEv">
<span id="egglib::GeneticCode::num_codes"></span><span class="target" id="project0classegglib_1_1GeneticCode_1afcfc908e798676be5691c800b4474109"></span>unsigned int <code class="descclassname">egglib::GeneticCode::</code><code class="descname">num_codes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib11GeneticCode9num_codesEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Get the number of available codes. </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="random">
<h3>Random<a class="headerlink" href="#random" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib6RandomE">
<span id="egglib::Random"></span><span class="target" id="project0classegglib_1_1Random"></span><em class="property">class </em></dt>
<dd><p>Pseudo-random number generator. </p>
<p>This class implements the Mersenne Twister algorithm for pseudo-random number generation. It is based on work by Makoto Matsumoto and Takuji Nishimura (see <a class="reference external" href="http://www.math.sci.hiroshima-u.ac.jp/~m-MAT/MT/emt.html">http://www.math.sci.hiroshima-u.ac.jp/~m-MAT/MT/emt.html</a>) and Jasper Bedaux (see <a class="reference external" href="http://www.bedaux.net/mtrand/">http://www.bedaux.net/mtrand/</a>) for the core generator, and the <a class="reference internal" href="#project0classegglib_1_1Random"><span class="std std-ref">Random</span></a> class of Egglib up to 2.2 for conversion to other laws than uniform.</p>
<p>Note that different instances of the class have independent chain of pseudo-random numbers. If several instances have the same seed, they will generate the exact same chain of pseudo-random numbers. Note that this applies if the default constructor is used and that instances are created within the same second.</p>
<p>All non-uniform distribution laws generators are based either on the <a class="reference internal" href="#project0classegglib_1_1Random_1a65efeea7718e4d622103f2d5e35f8e99"><span class="std std-ref">rand_int32()</span></a> or the standard (half-open, 32 bit) <a class="reference internal" href="#project0classegglib_1_1Random_1a30d1e43990eafb316df89c96cd213e6b"><span class="std std-ref">uniform()</span></a> methods.</p>
<p>Header: &lt;egglib-cpp/Random.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib6Random6RandomEv">
<span id="egglib::Random::Random"></span><span class="target" id="project0classegglib_1_1Random_1aa0cd23a16025cfeef7b8810eeb2a5d36"></span><code class="descclassname">egglib::Random::</code><code class="descname">Random</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Random6RandomEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor with default seed. </p>
<p>Uses the current system clock second as seed. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Random6RandomEm">
<span id="egglib::Random::Random__unsigned-l"></span><span class="target" id="project0classegglib_1_1Random_1aa22144173c7a0d90af7f9c166618f794"></span><code class="descclassname">egglib::Random::</code><code class="descname">Random</code><span class="sig-paren">(</span>unsigned long <em>s</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Random6RandomEm" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor with custom seed. </p>
<p>Favor large, high-complexity seeds. When using different instances of <a class="reference internal" href="#project0classegglib_1_1Random"><span class="std std-ref">Random</span></a> in a program, or different processes using <a class="reference internal" href="#project0classegglib_1_1Random"><span class="std std-ref">Random</span></a>, ensure they are all seeded using different seeds. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6RandomD0Ev">
<span id="egglib::Random::~Random"></span>virtual <span class="target" id="project0classegglib_1_1Random_1a70c16ca0bcf0c1d2b4fdbc8b0948af9b"></span><code class="descclassname">egglib::Random::</code><code class="descname">~Random</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6RandomD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Random9binomrandEld">
<span id="egglib::Random::binomrand__l.double"></span><span class="target" id="project0classegglib_1_1Random_1a699a8cbb7ca75137a85eb366d987bbe5"></span>unsigned long <code class="descclassname">egglib::Random::</code><code class="descname">binomrand</code><span class="sig-paren">(</span>long <em>n</em>, double <em>p</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Random9binomrandEld" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Draws a number from a binomial law. </p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">n</span></code> - <p>number of tests (must be &gt;=0). </p>
</li>
<li><code class="first docutils literal"><span class="pre">p</span></code> - <p>test probability. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Random5brandEv">
<span id="egglib::Random::brand"></span><span class="target" id="project0classegglib_1_1Random_1a27f2a560ed6911a313f8c179f202e1fb"></span>bool <code class="descclassname">egglib::Random::</code><code class="descname">brand</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Random5brandEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Boolean integer. </p>
<p>Return true with probability 0.5. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Random5erandEd">
<span id="egglib::Random::erand__double"></span><span class="target" id="project0classegglib_1_1Random_1a736ea2abafc876112016cd688de4bf8b"></span>double <code class="descclassname">egglib::Random::</code><code class="descname">erand</code><span class="sig-paren">(</span>double <em>expectation</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Random5erandEd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Draws a number from an exponential distribution. </p>
<p>Beware, the argument is the distribution’s mean (and is also 1/lambda where lambda is the rate parameter). </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Random8get_seedEv">
<span id="egglib::Random::get_seed"></span><span class="target" id="project0classegglib_1_1Random_1a18890ccbb335621c0024da042bfc2265"></span>unsigned long <code class="descclassname">egglib::Random::</code><code class="descname">get_seed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Random8get_seedEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get seed value. </p>
<p>Return the value of the seed that was used to initiate the instance. If the generator was re-seeded, return the seed value passed at that point. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Random5grandEd">
<span id="egglib::Random::grand__double"></span><span class="target" id="project0classegglib_1_1Random_1a85de9534be4059c43c1ef3909644d95d"></span>unsigned int <code class="descclassname">egglib::Random::</code><code class="descname">grand</code><span class="sig-paren">(</span>double <em>param</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Random5grandEd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Draws a number from a geometric law. </p>
<p>The argument is the geometric law parameter. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Random5irandEj">
<span id="egglib::Random::irand__unsigned-i"></span><span class="target" id="project0classegglib_1_1Random_1aa8b8d32761d3b0fdc3b3544a52cfc748"></span>unsigned int <code class="descclassname">egglib::Random::</code><code class="descname">irand</code><span class="sig-paren">(</span>unsigned int <em>ncards</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Random5irandEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Draws a uniform integer. </p>
<p>The argument is the number of values that can be generated. Returns an integer in the range [0, ncards-1]. Therefore, ncards is not included in the range. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Random5nrandEv">
<span id="egglib::Random::nrand"></span><span class="target" id="project0classegglib_1_1Random_1af85a0ea298a7982d05de74119500443a"></span>double <code class="descclassname">egglib::Random::</code><code class="descname">nrand</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Random5nrandEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Draws a number from a normal distribution. </p>
<p>Return a normal variation with expectation 0 and standard deviation 1. The algorithm used is the polar form of the Box-Muller algorithm. A draw is performed every two calls unless the instance is re-seeded. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Random5prandEd">
<span id="egglib::Random::prand__double"></span><span class="target" id="project0classegglib_1_1Random_1ac8b7be570b108ba974d865c179f24914"></span>unsigned int <code class="descclassname">egglib::Random::</code><code class="descname">prand</code><span class="sig-paren">(</span>double <em>p</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Random5prandEd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Draws an integer from a Poisson distribution. </p>
<p>The argument is the Poisson distribution parameters. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Random10rand_int32Ev">
<span id="egglib::Random::rand_int32"></span><span class="target" id="project0classegglib_1_1Random_1a65efeea7718e4d622103f2d5e35f8e99"></span>unsigned long <code class="descclassname">egglib::Random::</code><code class="descname">rand_int32</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Random10rand_int32Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Generate a 32-bit random integer. </p>
<p>Returns an integer in the range [0, 4294967295] (that is in the range [0, 2^32-1]. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Random8set_seedEm">
<span id="egglib::Random::set_seed__unsigned-l"></span><span class="target" id="project0classegglib_1_1Random_1a349820317a7470ff29945a1649bd225a"></span>void <code class="descclassname">egglib::Random::</code><code class="descname">set_seed</code><span class="sig-paren">(</span>unsigned long <em>s</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Random8set_seedEm" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Re-seed an instance. </p>
<p>Favor large, high-complexity seeds. When using different instances of <a class="reference internal" href="#project0classegglib_1_1Random"><span class="std std-ref">Random</span></a> in a program, or different processes using <a class="reference internal" href="#project0classegglib_1_1Random"><span class="std std-ref">Random</span></a>, ensure they are all seeded using different seeds. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Random7uniformEv">
<span id="egglib::Random::uniform"></span><span class="target" id="project0classegglib_1_1Random_1a30d1e43990eafb316df89c96cd213e6b"></span>double <code class="descclassname">egglib::Random::</code><code class="descname">uniform</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Random7uniformEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Generate a real in the half-open interval [0,1) </p>
<p>0 is included but not 1. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Random9uniform53Ev">
<span id="egglib::Random::uniform53"></span><span class="target" id="project0classegglib_1_1Random_1ae59f56cb1eb16ac4dd8bdb7ffa2153a4"></span>double <code class="descclassname">egglib::Random::</code><code class="descname">uniform53</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Random9uniform53Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Generate a 53-bit real. </p>
<p>The value has increased precision: even uniform integer pseudo-random numbers can take a finite number of values (2^32 of them, that is). This method increases the complexity of return values, with a cost as increased computing time. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Random9uniformclEv">
<span id="egglib::Random::uniformcl"></span><span class="target" id="project0classegglib_1_1Random_1a46438c49b90a269f3eb8eb80ac1f1269"></span>double <code class="descclassname">egglib::Random::</code><code class="descname">uniformcl</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Random9uniformclEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Generate a real in the closed interval [0,1]. </p>
<p>Both 0 and 1 are included. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Random9uniformopEv">
<span id="egglib::Random::uniformop"></span><span class="target" id="project0classegglib_1_1Random_1a3bfd32d1301cb4c44276ebae7a8b66a6"></span>double <code class="descclassname">egglib::Random::</code><code class="descname">uniformop</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Random9uniformopEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Generate a real in the open interval (0,1) </p>
<p>Neither 0 nor 1 is included. </p>
</dd></dl>

</div>
</dd></dl>

</div>
</div>
<div class="section" id="model-fitting-utilities">
<h2>Model fitting utilities<a class="headerlink" href="#model-fitting-utilities" title="Permalink to this headline">¶</a></h2>
<div class="section" id="abc">
<h3>ABC<a class="headerlink" href="#abc" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib3ABCE">
<span id="egglib::ABC"></span><span class="target" id="project0classegglib_1_1ABC"></span><em class="property">class </em></dt>
<dd><p>Model estimation by Approximate Bayesian Computation. </p>
<p>It is required to set the number of statistics and at least one input file name before performing analysis. The analysis itself consists in several steps. (1) Computation of the threshold, which requires to read through all files and imports statistics. In the process, the standard deviation of all statistics will be calculated and will be available. (2) Computation of Euclidean distances and weights and generation of a second sample file with weights and (non-standardized) statistics with only non-null weights exported. This step requires that the observed statistics have been set (between steps 1 and 2). (3) Local-linear regression using the fit method. While in the previous steps several models can be mixed, in this step a single model can be processed at a time. The output is a simple file with adjusted parameters only. </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Types</p>
<dl class="type">
<dt id="_CPPv2N6egglib3ABC13TransformModeE">
<span id="egglib::ABC::TransformMode"></span><span class="target" id="project0classegglib_1_1ABC_1aaa6c4a8b099a272e3e4cc2d85b43fd86"></span>enum <code class="descclassname">egglib::ABC::</code><code class="descname">TransformMode</code><a class="headerlink" href="#_CPPv2N6egglib3ABC13TransformModeE" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Modes for parameter transformation. </p>
<p><em>Values:</em></p>
<dl class="member">
<dt id="_CPPv2N6egglib3ABC4NONEE">
<span id="egglib::ABC::NONE"></span><span class="target" id="project0classegglib_1_1ABC_1aaa6c4a8b099a272e3e4cc2d85b43fd86ac157bdf0b85a40d2619cbc8bc1ae5fe2"></span></dt>
<dd></dd></dl>

<dl class="member">
<dt id="_CPPv2N6egglib3ABC3LOGE">
<span id="egglib::ABC::LOG"></span><span class="target" id="project0classegglib_1_1ABC_1aaa6c4a8b099a272e3e4cc2d85b43fd86acd7ffe737ad5fb21fbd7499886934910"></span></dt>
<dd></dd></dl>

<dl class="member">
<dt id="_CPPv2N6egglib3ABC3TANE">
<span id="egglib::ABC::TAN"></span><span class="target" id="project0classegglib_1_1ABC_1aaa6c4a8b099a272e3e4cc2d85b43fd86ac457138bcc3482d817346fc4d6308983"></span></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib3ABC3ABCEv">
<span id="egglib::ABC::ABC"></span><span class="target" id="project0classegglib_1_1ABC_1af55bb81fb3f74a45a2285fa1b954e2e5"></span><code class="descclassname">egglib::ABC::</code><code class="descname">ABC</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3ABC3ABCEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3ABCD0Ev">
<span id="egglib::ABC::~ABC"></span><span class="target" id="project0classegglib_1_1ABC_1a4a9961b9a2cbe211764bb51310fa7b07"></span><code class="descclassname">egglib::ABC::</code><code class="descname">~ABC</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3ABCD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3ABC9add_fnameEPKcj">
<span id="egglib::ABC::add_fname__cCP.unsigned-i"></span><span class="target" id="project0classegglib_1_1ABC_1a2eddc9d1f4f81d5fcb43cc7baed6dd49"></span>void <code class="descclassname">egglib::ABC::</code><code class="descname">add_fname</code><span class="sig-paren">(</span><em class="property">const</em> char *<em>fname</em>, unsigned int <em>number_of_params</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3ABC9add_fnameEPKcj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Adds a file name. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3ABC13get_thresholdEd">
<span id="egglib::ABC::get_threshold__double"></span><span class="target" id="project0classegglib_1_1ABC_1adcf04af15e7e5212c660b739829122c0"></span>void <code class="descclassname">egglib::ABC::</code><code class="descname">get_threshold</code><span class="sig-paren">(</span>double <em>tolerance</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3ABC13get_thresholdEd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Gets the regression threshold. </p>
<p>If several files are loaded, the data will be aggregated (note that they must all contain the same number of statistics). At least one file must have been set, and the number of statistics must have been set as well.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">tolerance</span></code> - <p>rejection threshold (proportion of points in the local region. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3ABC17number_of_samplesEv">
<span id="egglib::ABC::number_of_samples"></span><span class="target" id="project0classegglib_1_1ABC_1aa5e9cf2d74d8b98e5333e5c38313213c"></span>unsigned int <code class="descclassname">egglib::ABC::</code><code class="descname">number_of_samples</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3ABC17number_of_samplesEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Gets number of imported data samples. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3ABC22number_of_samples_partEj">
<span id="egglib::ABC::number_of_samples_part__unsigned-i"></span><span class="target" id="project0classegglib_1_1ABC_1a7324b6c8e71a50a7f3f489f06fc936e0"></span>unsigned int <code class="descclassname">egglib::ABC::</code><code class="descname">number_of_samples_part</code><span class="sig-paren">(</span>unsigned int <em>i</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3ABC22number_of_samples_partEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Gets number of imported data samples for a given file. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3ABC20number_of_statisticsEj">
<span id="egglib::ABC::number_of_statistics__unsigned-i"></span><span class="target" id="project0classegglib_1_1ABC_1abdfea6ccf0bcd7f428235488f5374ea6"></span>void <code class="descclassname">egglib::ABC::</code><code class="descname">number_of_statistics</code><span class="sig-paren">(</span>unsigned int <em>ns</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3ABC20number_of_statisticsEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Sets number of statistics. </p>
<p>If data was already present in the instance, it will all be cleared. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3ABC3obsEjd">
<span id="egglib::ABC::obs__unsigned-i.double"></span><span class="target" id="project0classegglib_1_1ABC_1ad60d64a56d8dbb1eee32dbf2c93d5931"></span>void <code class="descclassname">egglib::ABC::</code><code class="descname">obs</code><span class="sig-paren">(</span>unsigned int <em>index</em>, double <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3ABC3obsEjd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Sets a summary statistics. </p>
<p>The number of statistics must have been set, and the index must not be out of bound. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3ABC10regressionEPKcPKc13TransformModePKc">
<span id="egglib::ABC::regression__cCP.cCP.TransformMode.cCP"></span><span class="target" id="project0classegglib_1_1ABC_1ac3fc5cbf42c77d603bdca48ace3e06c5"></span>unsigned int <code class="descclassname">egglib::ABC::</code><code class="descname">regression</code><span class="sig-paren">(</span><em class="property">const</em> char *<em>infname</em>, <em class="property">const</em> char *<em>outfname</em>, <a class="reference internal" href="#_CPPv2N6egglib3ABC13TransformModeE" title="egglib::ABC::TransformMode">TransformMode</a> <em>mode</em>, <em class="property">const</em> char *<em>header</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3ABC10regressionEPKcPKc13TransformModePKc" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Performs regression step. </p>
<p><dl class="docutils">
<dt><strong>Return</strong></dt>
<dd>Number of data point processed </dd>
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">infname</span></code> - <p>input file name (generated using rejection) </p>
</li>
<li><code class="first docutils literal"><span class="pre">outfname</span></code> - <p>output file name (final posterior) </p>
</li>
<li><code class="first docutils literal"><span class="pre">mode</span></code> - <p>transformation mode </p>
</li>
<li><code class="first docutils literal"><span class="pre">header</span></code> - <p>outfile file header (name of parameters; if empty string, no header is printed)</p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3ABC9rejectionEPKcbb">
<span id="egglib::ABC::rejection__cCP.b.b"></span><span class="target" id="project0classegglib_1_1ABC_1a422326d6dbd62416166b82d1b4e7143d"></span>unsigned int <code class="descclassname">egglib::ABC::</code><code class="descname">rejection</code><span class="sig-paren">(</span><em class="property">const</em> char *<em>outfname</em>, bool <em>exportlabels</em>, bool <em>strip</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3ABC9rejectionEPKcbb" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Performs rejection step. </p>
<p>The observed value must have been entered (otherwise the results will be meaningless), and the threshold must have been computed.</p>
<p><dl class="docutils">
<dt><strong>Return</strong></dt>
<dd>the number of points in the local region. </dd>
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">outfname</span></code> - <p>the name of the intermediary file. </p>
</li>
<li><code class="first docutils literal"><span class="pre">exportlabels</span></code> - <p>if true: exports a tag at the beginning of each line to identify the file or origin of each accepted sample (starting from 1). </p>
</li>
<li><code class="first docutils literal"><span class="pre">strip</span></code> - <p>if true: remove statistics and weights (only export statistics of accepted points; then the file cannot be used for regression). </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3ABC2sdEj">
<span id="egglib::ABC::sd__unsigned-i"></span><span class="target" id="project0classegglib_1_1ABC_1a9ee271962a29e405a015418c46771b02"></span>double <code class="descclassname">egglib::ABC::</code><code class="descname">sd</code><span class="sig-paren">(</span>unsigned int <em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3ABC2sdEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Gets a standard deviation. </p>
<p>The <a class="reference internal" href="#project0classegglib_1_1ABC_1adcf04af15e7e5212c660b739829122c0"><span class="std std-ref">get_threshold()</span></a> method must have been called, and the index must not be out of bound. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3ABC9thresholdEv">
<span id="egglib::ABC::threshold"></span><span class="target" id="project0classegglib_1_1ABC_1a5b9ed2b701daf3902e8c748ae0d36b61"></span>double <code class="descclassname">egglib::ABC::</code><code class="descname">threshold</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3ABC9thresholdEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Gets Euclidean threshold. </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="neutral-networks">
<h3>Neutral networks<a class="headerlink" href="#neutral-networks" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib4nnet7NetworkE">
<span id="egglib::nnet::Network"></span><span class="target" id="project0classegglib_1_1nnet_1_1Network"></span><em class="property">class </em></dt>
<dd><p>Training and prediction with neural networks. </p>
<p>This classes implements the back propagation algorithm for training neural network.</p>
<p>The constructor of Neural generates a bare, unusable instance. It is required set up the network with the <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Network_1a2d40b2a455fe760bff2874dfbb86491e"><span class="std std-ref">setup()</span></a> method and initialize weights (normally using the <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Network_1a22d0e305b2d5f8b57ac4cbf97decf652"><span class="std std-ref">init_weights()</span></a> method, or by setting manually all weights) before calling <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Network_1a83d41d8b0e2f19a14a4fc629b92171c6"><span class="std std-ref">train()</span></a>. More information is available in the documentation of the <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Network_1a2d40b2a455fe760bff2874dfbb86491e"><span class="std std-ref">setup()</span></a> method.</p>
<p>Header: &lt;egglib-cpp/Neural.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib4nnet7Network7NetworkEv">
<span id="egglib::nnet::Network::Network"></span><span class="target" id="project0classegglib_1_1nnet_1_1Network_1a5f56923e4eac55c8c08d72599b0a0d41"></span><code class="descclassname">egglib::nnet::Network::</code><code class="descname">Network</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet7Network7NetworkEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet7NetworkD0Ev">
<span id="egglib::nnet::Network::~Network"></span>virtual <span class="target" id="project0classegglib_1_1nnet_1_1Network_1ac9062f3da8a725626fd7e7bd27449220"></span><code class="descclassname">egglib::nnet::Network::</code><code class="descname">~Network</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet7NetworkD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet7Network12init_weightsEdd">
<span id="egglib::nnet::Network::init_weights__double.double"></span><span class="target" id="project0classegglib_1_1nnet_1_1Network_1a22d0e305b2d5f8b57ac4cbf97decf652"></span>void <code class="descclassname">egglib::nnet::Network::</code><code class="descname">init_weights</code><span class="sig-paren">(</span>double <em>range_input</em>, double <em>range_output</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet7Network12init_weightsEdd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Initialize weights. </p>
<p>It is required to call this method before calling training, otherwise the weights have undefined values. All weights are initialized to random values from a uniform distribution in the range [-X, X] with X is the range_input argument for all weights connecting input variables to neurons of the hidden layer, and the range_output arguments for all weights connecting neurons of the hidden layer to the output neurons.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">range_input</span></code> - <p>bound for first-level weights. </p>
</li>
<li><code class="first docutils literal"><span class="pre">range_output</span></code> - <p>bound for second-level weights. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet7Network8num_iterEv">
<span id="egglib::nnet::Network::num_iter"></span><span class="target" id="project0classegglib_1_1nnet_1_1Network_1ade21d2123080c6f6d6ed4ceb7126c14c"></span>unsigned int <code class="descclassname">egglib::nnet::Network::</code><code class="descname">num_iter</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet7Network8num_iterEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the number of training iterations. </p>
<p>The internal counter is reset by <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Network_1a2d40b2a455fe760bff2874dfbb86491e"><span class="std std-ref">setup()</span></a>. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet7Network7predictERK4Datajb">
<span id="egglib::nnet::Network::predict__DataCR.unsigned-i.b"></span><span class="target" id="project0classegglib_1_1nnet_1_1Network_1ab5d76b7e73e57fa6b21378ea6b0712e2"></span>void <code class="descclassname">egglib::nnet::Network::</code><code class="descname">predict</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib4nnet4DataE" title="egglib::nnet::Data">Data</a> &amp;<em>data</em>, unsigned int <em>pattern</em>, bool <em>compute_error</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet7Network7predictERK4Datajb" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Use the neural network to predict output. </p>
<p>The network must have been trained using a <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Data"><span class="std std-ref">Data</span></a> instance with the same number of input and output variables. This method generates an array of predicted values based on the current values of the weight (after training) that can be accessed using the <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Network_1aa6f9dc59b3249f919faff13bca2c512e"><span class="std std-ref">prediction()</span></a> method. If the compute_error argument is true, the error is accessed using error().</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">data</span></code> - <p>data set with the correct number of input (always) and output variables (unless compute_errors is false). </p>
</li>
<li><code class="first docutils literal"><span class="pre">pattern</span></code> - <p>pattern to process. </p>
</li>
<li><code class="first docutils literal"><span class="pre">compute_error</span></code> - <p>if false, don’t compute errors, and output variables are not considered. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet7Network10predictionEj">
<span id="egglib::nnet::Network::prediction__unsigned-i"></span><span class="target" id="project0classegglib_1_1nnet_1_1Network_1aa6f9dc59b3249f919faff13bca2c512e"></span>double <code class="descclassname">egglib::nnet::Network::</code><code class="descname">prediction</code><span class="sig-paren">(</span>unsigned int <em>output_var</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet7Network10predictionEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get a predicted output. </p>
<p>The <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Network_1ab5d76b7e73e57fa6b21378ea6b0712e2"><span class="std std-ref">predict()</span></a> method must have been called. Get the predicted value for one of the output variables. Warning: training (using the method <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Network_1a83d41d8b0e2f19a14a4fc629b92171c6"><span class="std std-ref">train()</span></a>) modifies the predicted values and invalidates the results of the method <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Network_1ab5d76b7e73e57fa6b21378ea6b0712e2"><span class="std std-ref">predict()</span></a>.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">output_var</span></code> - <p>index of the output variable (must be smalled than the number of output variables defined by both the <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Data"><span class="std std-ref">Data</span></a> instances used for training and for prediction). </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet7Network5setupERK4DatajjRK4Datajjj18ActivationFunction18ActivationFunctionddddP6Random">
<span id="egglib::nnet::Network::setup__DataCR.unsigned-i.unsigned-i.DataCR.unsigned-i.unsigned-i.unsigned-i.ActivationFunction.ActivationFunction.double.double.double.double.RandomP"></span><span class="target" id="project0classegglib_1_1nnet_1_1Network_1a2d40b2a455fe760bff2874dfbb86491e"></span>void <code class="descclassname">egglib::nnet::Network::</code><code class="descname">setup</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib4nnet4DataE" title="egglib::nnet::Data">Data</a> &amp;<em>training_data</em>, unsigned int <em>training_start</em>, unsigned int <em>training_stop</em>, <em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib4nnet4DataE" title="egglib::nnet::Data">Data</a> &amp;<em>testing_data</em>, unsigned int <em>testing_start</em>, unsigned int <em>testing_stop</em>, unsigned int <em>num_hidden</em>, ActivationFunction <em>fun1</em>, ActivationFunction <em>fun2</em>, double <em>rate1</em>, double <em>rate2</em>, double <em>bound</em>, double <em>momentum</em>, <a class="reference internal" href="#_CPPv2N6egglib6RandomE" title="egglib::Random">Random</a> *<em>random</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet7Network5setupERK4DatajjRK4Datajjj18ActivationFunction18ActivationFunctionddddP6Random" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Setup the network. </p>
<p>Upon call to this method, the network is set up based on the number of input and output variables of the provided data set, and the specified number of neurons in the hidden layer. Note that all neurons (from both the hidden and output layers) are automatically connected to an additional neuron generating a constant input of 1.0 (the bias).</p>
<p>Loading a training data set (using the train_data) is required. The first and last-plus-one indexes must be passed to indicate which patterns must be processed for training. Note that the pattern corresponding to the train_stop argument is not included. It is possible to use a train_stop value larger than the number of patterns in train_data. To use all patterns, use train_start=0 and train_stop=<a class="reference internal" href="#project0namespaceegglib_1ab498891cdf034be8678af906644e6dc1"><span class="std std-ref">egglib::MAX</span></a>.</p>
<p>Loading a test data set is not mandatory but advisable. The test data set is not used for training but allows to evaluate the predictive ability of the network. To skip this option, pass any data set as test_data (for example, the same as train_data) and set test_start &gt;= test_stop.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">training_data</span></code> - <p>a training data set with at least one input variable, at least one output variable, at least one pattern and all data loaded. The object passed must not modified until training is finished. </p>
</li>
<li><code class="first docutils literal"><span class="pre">training_start</span></code> - <p>index of the first pattern to process for training. </p>
</li>
<li><code class="first docutils literal"><span class="pre">training_stop</span></code> - <p>index of the pattern immediately after the last pattern to process for training. </p>
</li>
<li><code class="first docutils literal"><span class="pre">testing_data</span></code> - <p>a data set to use for testing the predictive ability of the network (not used for fitting). The data set must have the same number of input and output variables as train_data (preferably it will the same object, this non-overlapping ranges of patterns to process). </p>
</li>
<li><code class="first docutils literal"><span class="pre">testing_start</span></code> - <p>index of the first pattern to process for testing. </p>
</li>
<li><code class="first docutils literal"><span class="pre">testing_stop</span></code> - <p>index of the pattern immediately after the last pattern to process for test. </p>
</li>
<li><code class="first docutils literal"><span class="pre">num_hidden</span></code> - <p>number of neurons in the hidden layer of the network. </p>
</li>
<li><code class="first docutils literal"><span class="pre">fun1</span></code> - <p>activation function to use for neurons of the hidden layer. </p>
</li>
<li><code class="first docutils literal"><span class="pre">fun2</span></code> - <p>activation function to use for output neurons. </p>
</li>
<li><code class="first docutils literal"><span class="pre">rate1</span></code> - <p>rate for weights from input to hidden layer. </p>
</li>
<li><code class="first docutils literal"><span class="pre">rate2</span></code> - <p>rate for weights form hidden layer to output. </p>
</li>
<li><code class="first docutils literal"><span class="pre">bound</span></code> - <p>extreme value (for both signs) for weight values. </p>
</li>
<li><code class="first docutils literal"><span class="pre">momentum</span></code> - <p>proportion of the previous weight change to apply to all weight changes (use 0.0 to skip momentum). </p>
</li>
<li><code class="first docutils literal"><span class="pre">random</span></code> - <p><a class="reference internal" href="#project0classegglib_1_1Random"><span class="std std-ref">Random</span></a> object to use for generating random numbers (only used for initial weights). </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet7Network13testing_errorEv">
<span id="egglib::nnet::Network::testing_error"></span><span class="target" id="project0classegglib_1_1nnet_1_1Network_1a991faff73d9ffaf60854d2ae0700fd32"></span>double <code class="descclassname">egglib::nnet::Network::</code><code class="descname">testing_error</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet7Network13testing_errorEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get total error for testing data. </p>
<p>As <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Network_1a39a62b4f7c310155990d162c6fbb6179"><span class="std std-ref">training_error()</span></a> but for the testing data set. Only valid if testing data has been passed to <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Network_1a2d40b2a455fe760bff2874dfbb86491e"><span class="std std-ref">setup()</span></a>. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet7Network5trainEj">
<span id="egglib::nnet::Network::train__unsigned-i"></span><span class="target" id="project0classegglib_1_1nnet_1_1Network_1a83d41d8b0e2f19a14a4fc629b92171c6"></span>void <code class="descclassname">egglib::nnet::Network::</code><code class="descname">train</code><span class="sig-paren">(</span>unsigned int <em>num_iter</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet7Network5trainEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Train the neural network. </p>
<p>Train the network for a fixed number of iterations. This method must be called iteratively until a given stop criterion is fulfilled. After each call to <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Network_1a83d41d8b0e2f19a14a4fc629b92171c6"><span class="std std-ref">train()</span></a>, it is possible to access the the current values of weights, the error for training data and, if testing data have been loaded, the error for testing data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet7Network14training_errorEv">
<span id="egglib::nnet::Network::training_error"></span><span class="target" id="project0classegglib_1_1nnet_1_1Network_1a39a62b4f7c310155990d162c6fbb6179"></span>double <code class="descclassname">egglib::nnet::Network::</code><code class="descname">training_error</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet7Network14training_errorEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get total error for training data. </p>
<p>The error is computed sqrt(sum((pred[i]-obs[i])^2)) where sqrt is the square root function, sum is the sum over all output nodes, pred[i] is the predicted value provided by output neuron i and obs[i] is the observed value for output variable i. Computed using the training data only. Requires <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Network_1a83d41d8b0e2f19a14a4fc629b92171c6"><span class="std std-ref">train()</span></a> but also modified if <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Network_1ab5d76b7e73e57fa6b21378ea6b0712e2"><span class="std std-ref">predict()</span></a> is called with compute_error=true. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet7Network13weight_hiddenEjj">
<span id="egglib::nnet::Network::weight_hidden__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1nnet_1_1Network_1af3412a7e6b77eacedc4c805167580818"></span>double <code class="descclassname">egglib::nnet::Network::</code><code class="descname">weight_hidden</code><span class="sig-paren">(</span>unsigned int <em>i</em>, unsigned int <em>j</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet7Network13weight_hiddenEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Value of a weight to a hidden neuron. </p>
<p>Access to the current value of the weight of the connexion of a hidden neuron to an input variable or to the bias neuron. The bias weight’s index is equal to the number of input variables.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">i</span></code> - <p>index of the hidden neuron. </p>
</li>
<li><code class="first docutils literal"><span class="pre">j</span></code> - <p>index of the input variable. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet7Network13weight_hiddenEjjd">
<span id="egglib::nnet::Network::weight_hidden__unsigned-i.unsigned-i.double"></span><span class="target" id="project0classegglib_1_1nnet_1_1Network_1a39384138be7c1be7bf57b16c3a84855c"></span>void <code class="descclassname">egglib::nnet::Network::</code><code class="descname">weight_hidden</code><span class="sig-paren">(</span>unsigned int <em>i</em>, unsigned int <em>j</em>, double <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet7Network13weight_hiddenEjjd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set the value of a weight to a hidden neuron. </p>
<p>Set the value of the weight of the connexion of a hidden neuron to an input variable or to the bias neuron. The bias weight’s index is equal to the number of input variables.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">i</span></code> - <p>index of the hidden neuron. </p>
</li>
<li><code class="first docutils literal"><span class="pre">j</span></code> - <p>index of the input variable. </p>
</li>
<li><code class="first docutils literal"><span class="pre">value</span></code> - <p>weight value. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet7Network13weight_outputEjj">
<span id="egglib::nnet::Network::weight_output__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1nnet_1_1Network_1a7eb9523d31726bc15ad6e2c329dfb8df"></span>double <code class="descclassname">egglib::nnet::Network::</code><code class="descname">weight_output</code><span class="sig-paren">(</span>unsigned int <em>i</em>, unsigned int <em>j</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet7Network13weight_outputEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Value of a weight to an output neuron. </p>
<p>Access to the current value of the weight of the connexion of an output neuron to hidden neuron or to the bias neuron. The bias weight’s index is equal to the number of hidden neurons.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">i</span></code> - <p>index of the output neuron. </p>
</li>
<li><code class="first docutils literal"><span class="pre">j</span></code> - <p>index of the hidden neuron. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet7Network13weight_outputEjjd">
<span id="egglib::nnet::Network::weight_output__unsigned-i.unsigned-i.double"></span><span class="target" id="project0classegglib_1_1nnet_1_1Network_1aff38d662fb39ba83d6be0e9a5c45ec03"></span>void <code class="descclassname">egglib::nnet::Network::</code><code class="descname">weight_output</code><span class="sig-paren">(</span>unsigned int <em>i</em>, unsigned int <em>j</em>, double <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet7Network13weight_outputEjjd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set the value of a weight to an output neuron. </p>
<p>Set the value of the weight of the connexion of an output variable to a hidden neuron or to the bias neuron. The bias weight’s index is equal to the number of hidden neurons.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">i</span></code> - <p>index of the output neuron. </p>
</li>
<li><code class="first docutils literal"><span class="pre">j</span></code> - <p>index of the hidden neuron. </p>
</li>
<li><code class="first docutils literal"><span class="pre">value</span></code> - <p>weight value. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

</div>
</dd></dl>

<dl class="class">
<dt id="_CPPv2N6egglib4nnet6NeuronE">
<span id="egglib::nnet::Neuron"></span><span class="target" id="project0classegglib_1_1nnet_1_1Neuron"></span><em class="property">class </em></dt>
<dd><p>Neuron, that is a node of a neural network </p>
<p>Header: &lt;egglib-cpp/Neural.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib4nnet6Neuron6NeuronEv">
<span id="egglib::nnet::Neuron::Neuron"></span><span class="target" id="project0classegglib_1_1nnet_1_1Neuron_1a209bdec294205b220f16df82169f2f2f"></span><code class="descclassname">egglib::nnet::Neuron::</code><code class="descname">Neuron</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet6Neuron6NeuronEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Create a neuron. </p>
<p>The neuron is created naked, empty and unusable. The user must use <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Neuron_1a55d1fcc185bd6401fd0e1b69d780687f"><span class="std std-ref">config()</span></a> before doing anything with it. To reuse a neuron after it has been used (e.g. for training again the same network), it is required to call <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Neuron_1a55d1fcc185bd6401fd0e1b69d780687f"><span class="std std-ref">config()</span></a> again. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet6NeuronD0Ev">
<span id="egglib::nnet::Neuron::~Neuron"></span>virtual <span class="target" id="project0classegglib_1_1nnet_1_1Neuron_1a7e9e9bae49cbb1b6ba9d2c9e8a3bada3"></span><code class="descclassname">egglib::nnet::Neuron::</code><code class="descname">~Neuron</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet6NeuronD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Delete a neuron. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet6Neuron8activateEv">
<span id="egglib::nnet::Neuron::activate"></span><span class="target" id="project0classegglib_1_1nnet_1_1Neuron_1a8011fcbeb5efeda20bd2dbffbbb70165"></span>void <code class="descclassname">egglib::nnet::Neuron::</code><code class="descname">activate</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet6Neuron8activateEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Activate the neuron. </p>
<p>Process all incoming connexions, apply weights and the activation function to generate the output. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet6Neuron6configE18ActivationFunctionj">
<span id="egglib::nnet::Neuron::config__ActivationFunction.unsigned-i"></span><span class="target" id="project0classegglib_1_1nnet_1_1Neuron_1a55d1fcc185bd6401fd0e1b69d780687f"></span>void <code class="descclassname">egglib::nnet::Neuron::</code><code class="descname">config</code><span class="sig-paren">(</span>ActivationFunction <em>fun</em>, unsigned int <em>num_input</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet6Neuron6configE18ActivationFunctionj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set up the neuron. </p>
<p>The object is reset, but previously data may not have been reinitialized.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">fun</span></code> - <p>function used for activation of this neuron. </p>
</li>
<li><code class="first docutils literal"><span class="pre">num_input</span></code> - <p>number of incoming connexions (either neurons of the previous layer or input variables) of this neuron. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet6Neuron5deltaEj">
<span id="egglib::nnet::Neuron::delta__unsigned-i"></span><span class="target" id="project0classegglib_1_1nnet_1_1Neuron_1a8e06213023b53c00409463235e7f9d49"></span>double <code class="descclassname">egglib::nnet::Neuron::</code><code class="descname">delta</code><span class="sig-paren">(</span>unsigned int <em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet6Neuron5deltaEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the delta value for a given weight. </p>
<p>The <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Neuron_1adc648361d3247cb949f952f5fc9e4d09"><span class="std std-ref">propagate()</span></a> method must have been called, which itself requires that the neuron had been previously loaded with all needed data and activated. This method returns the change of the specified weight value based on the propagated error. The delta value is computed as f’(I) * E[index] where f’ is the derivative of the activation function, I is the input value for this neuron and E[index] is the error associated to the weight in question. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet6Neuron10get_outputEv">
<span id="egglib::nnet::Neuron::get_output"></span><span class="target" id="project0classegglib_1_1nnet_1_1Neuron_1a72ee8d390a358dc015f508edbcbc91c0"></span>double <code class="descclassname">egglib::nnet::Neuron::</code><code class="descname">get_output</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet6Neuron10get_outputEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Collect the output. </p>
<p>Get the output value. Calling this method does not update the output if any input data has changed (make sure to call the <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Neuron_1a8011fcbeb5efeda20bd2dbffbbb70165"><span class="std std-ref">activate()</span></a> method for this). </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet6Neuron10get_weightEj">
<span id="egglib::nnet::Neuron::get_weight__unsigned-i"></span><span class="target" id="project0classegglib_1_1nnet_1_1Neuron_1a5141129fc1ca99014e26182c3e0c9c35"></span>double <code class="descclassname">egglib::nnet::Neuron::</code><code class="descname">get_weight</code><span class="sig-paren">(</span>unsigned int <em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet6Neuron10get_weightEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get a weight. </p>
<p>Get the current valueo of the weight applied to a given incoming connexion. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet6Neuron9propagateEjd">
<span id="egglib::nnet::Neuron::propagate__unsigned-i.double"></span><span class="target" id="project0classegglib_1_1nnet_1_1Neuron_1adc648361d3247cb949f952f5fc9e4d09"></span>void <code class="descclassname">egglib::nnet::Neuron::</code><code class="descname">propagate</code><span class="sig-paren">(</span>unsigned int <em>index</em>, double <em>val</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet6Neuron9propagateEjd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Propagate error for a given weight. </p>
<p>This method can only be used with neurons that have loaded input and have been activated. After the error has been propagated, the delta value can be obtained using <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Neuron_1a8e06213023b53c00409463235e7f9d49"><span class="std std-ref">delta()</span></a>, typically for propagating errors to the previous layer. The delta value is the value passed as val to this method multiplied by the value derivative of the activation function at the current input value of this neuron. Note that the neuron’s weight are not modified by this method, but only when <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Neuron_1adfac0170a03ab60215a27106236c0a8c"><span class="std std-ref">update()</span></a> is called.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">index</span></code> - <p>index of one of the weights of this neuron. </p>
</li>
<li><code class="first docutils literal"><span class="pre">val</span></code> - <p>propagated error for this weight (for an output neuron: this neuron’s error times the output value of the corresponding neuron of the hidden layer; for a hidden neuron: the sum of delta values for all output neurons, weighted by the corresponding weights. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet6Neuron9set_inputEjd">
<span id="egglib::nnet::Neuron::set_input__unsigned-i.double"></span><span class="target" id="project0classegglib_1_1nnet_1_1Neuron_1abf28e34dd554d836d748771a5a82897a"></span>void <code class="descclassname">egglib::nnet::Neuron::</code><code class="descname">set_input</code><span class="sig-paren">(</span>unsigned int <em>index</em>, double <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet6Neuron9set_inputEjd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set an input value. </p>
<p>Load a value for a given incoming connexion. Until changed, the value will be remembered. Changing the value does not update the output value of this neuron. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet6Neuron10set_weightEjd">
<span id="egglib::nnet::Neuron::set_weight__unsigned-i.double"></span><span class="target" id="project0classegglib_1_1nnet_1_1Neuron_1aeb15ca7757cd743eb520670986175ed2"></span>void <code class="descclassname">egglib::nnet::Neuron::</code><code class="descname">set_weight</code><span class="sig-paren">(</span>unsigned int <em>index</em>, double <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet6Neuron10set_weightEjd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set a weight. </p>
<p>Load the value of the weight to be applied to a given incoming connexion. Until changed, the value will be remembered. Changing the value does not update the output value of this neuron. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet6Neuron6updateEddd">
<span id="egglib::nnet::Neuron::update__double.double.double"></span><span class="target" id="project0classegglib_1_1nnet_1_1Neuron_1adfac0170a03ab60215a27106236c0a8c"></span>void <code class="descclassname">egglib::nnet::Neuron::</code><code class="descname">update</code><span class="sig-paren">(</span>double <em>rate</em>, double <em>bound</em>, double <em>momentum</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet6Neuron6updateEddd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Update all weights. </p>
<p>The <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Neuron_1adc648361d3247cb949f952f5fc9e4d09"><span class="std std-ref">propagate()</span></a> method must have been called for all weights. This method applies all delta values.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">rate</span></code> - <p>learning rate. </p>
</li>
<li><code class="first docutils literal"><span class="pre">bound</span></code> - <p>absolute limit value: weights are bound to the range [-bound, +bound]. </p>
</li>
<li><code class="first docutils literal"><span class="pre">momentum</span></code> - <p>proportion of the previous weight change to apply to the new change. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

</div>
</dd></dl>

<dl class="class">
<dt id="_CPPv2N6egglib4nnet4DataE">
<span id="egglib::nnet::Data"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data"></span><em class="property">class </em></dt>
<dd><p>Holds input and output data for training neural networks. </p>
<p>When creating a data set, the user must first set the number of patterns with num_patterns(), of input variables with num_input() and of output variables with num_output(). Only then it is possible to load data using input() and output(). Make sure to load every declared slots as data are not initialized.</p>
<p>
Header: &lt;egglib-cpp/Neural.hpp&gt; <dl class="docutils">
<dt><strong>Note</strong></dt>
<dd>In case a same object must be reused will smaller values of num_input and/or num_output more larger num_patterns, it is more efficient to call the methods num_input() and num_output() before num_patterns().</dd>
</dl>
</p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data4DataEv">
<span id="egglib::nnet::Data::Data"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1a51d806137473e32bc8e082f8beb3133a"></span><code class="descclassname">egglib::nnet::Data::</code><code class="descname">Data</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data4DataEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data4DataERK4Data">
<span id="egglib::nnet::Data::Data__DataCR"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1aaa4a67fc2299cc9619a56c817255a283"></span><code class="descclassname">egglib::nnet::Data::</code><code class="descname">Data</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib4nnet4Data4DataEv" title="egglib::nnet::Data::Data">Data</a> &amp;<em>src</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data4DataERK4Data" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Copy constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4DataD0Ev">
<span id="egglib::nnet::Data::~Data"></span>virtual <span class="target" id="project0classegglib_1_1nnet_1_1Data_1a6da8ae30c57cc1732f240fc3496ee7de"></span><code class="descclassname">egglib::nnet::Data::</code><code class="descname">~Data</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4DataD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data9get_inputEjj">
<span id="egglib::nnet::Data::get_input__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1a2d8bb05ef48ee862dae8a6d4a696b5e3"></span>double <code class="descclassname">egglib::nnet::Data::</code><code class="descname">get_input</code><span class="sig-paren">(</span>unsigned int <em>pattern</em>, unsigned int <em>variable</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data9get_inputEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an input data item. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data13get_num_inputEv">
<span id="egglib::nnet::Data::get_num_input"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1ad4b43ee61b0a0beaf3218c47eec07ae5"></span>unsigned int <code class="descclassname">egglib::nnet::Data::</code><code class="descname">get_num_input</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data13get_num_inputEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get number of input variables. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data14get_num_outputEv">
<span id="egglib::nnet::Data::get_num_output"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1a5384a48da5d5531dee9a46cbfe1d062d"></span>unsigned int <code class="descclassname">egglib::nnet::Data::</code><code class="descname">get_num_output</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data14get_num_outputEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get number of output variables. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data16get_num_patternsEv">
<span id="egglib::nnet::Data::get_num_patterns"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1a44b369ff198bf429a099fcc5b60f05b1"></span>unsigned int <code class="descclassname">egglib::nnet::Data::</code><code class="descname">get_num_patterns</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data16get_num_patternsEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get number of patterns. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data10get_outputEjj">
<span id="egglib::nnet::Data::get_output__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1a59d4ae5787a1b27cef9ba981e3fcfb32"></span>double <code class="descclassname">egglib::nnet::Data::</code><code class="descname">get_output</code><span class="sig-paren">(</span>unsigned int <em>pattern</em>, unsigned int <em>variable</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data10get_outputEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an output data item. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data10mean_inputEj">
<span id="egglib::nnet::Data::mean_input__unsigned-i"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1a6e28c6cd7edffd473391d3652a087d1c"></span>double <code class="descclassname">egglib::nnet::Data::</code><code class="descname">mean_input</code><span class="sig-paren">(</span>unsigned int <em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data10mean_inputEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the mean for an input variable. </p>
<p>If data have been normalized using the method <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Data_1acd0de676568888d848beb97dcc53ae47"><span class="std std-ref">normalize()</span></a>, this method returns the mean of a given input variable in the original data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data11mean_outputEj">
<span id="egglib::nnet::Data::mean_output__unsigned-i"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1a5786d368c78e0d30f231d8caf430e974"></span>double <code class="descclassname">egglib::nnet::Data::</code><code class="descname">mean_output</code><span class="sig-paren">(</span>unsigned int <em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data11mean_outputEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the mean for an output variable. </p>
<p>If data have been normalized using the method <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Data_1acd0de676568888d848beb97dcc53ae47"><span class="std std-ref">normalize()</span></a>, this method returns the mean of a given output variable in the original data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data9normalizeEv">
<span id="egglib::nnet::Data::normalize"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1acd0de676568888d848beb97dcc53ae47"></span>void <code class="descclassname">egglib::nnet::Data::</code><code class="descname">normalize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data9normalizeEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Normalize data for all input and output variables. </p>
<p>All data are modified permanently and the average and standard deviation for each input and output variable are saved and remain accessible using the methods <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Data_1a6e28c6cd7edffd473391d3652a087d1c"><span class="std std-ref">mean_input()</span></a>, <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Data_1a5786d368c78e0d30f231d8caf430e974"><span class="std std-ref">mean_output()</span></a>, <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Data_1aa704ba173a20caac8f241fbc8d208346"><span class="std std-ref">std_input()</span></a> and <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Data_1a926df3c27d065e571dd6d279f73b9e9d"><span class="std std-ref">std_output()</span></a>, until <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Data_1acd0de676568888d848beb97dcc53ae47"><span class="std std-ref">normalize()</span></a> is called again or the number of input or output variables or the number of patterns is modified. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data15normalize_inputEjdd">
<span id="egglib::nnet::Data::normalize_input__unsigned-i.double.double"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1a8e66eb97615addac6bc0f411e3ec3571"></span>void <code class="descclassname">egglib::nnet::Data::</code><code class="descname">normalize_input</code><span class="sig-paren">(</span>unsigned int <em>index</em>, double <em>mean</em>, double <em>std</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data15normalize_inputEjdd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Normalize data for an input variable. </p>
<p>All data are modified permanently. The passed mean and standard deviation are not saved. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data16normalize_outputEjdd">
<span id="egglib::nnet::Data::normalize_output__unsigned-i.double.double"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1a38b0cb0265bcbc1b9d8984c3b001b756"></span>void <code class="descclassname">egglib::nnet::Data::</code><code class="descname">normalize_output</code><span class="sig-paren">(</span>unsigned int <em>index</em>, double <em>mean</em>, double <em>std</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data16normalize_outputEjdd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Normalize data for an output variable. </p>
<p>All data are modified permanently. The passed mean and standard deviation are not saved. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4DataaSERK4Data">
<span id="egglib::nnet::Data::assign-operator__DataCR"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1a88f37837774108af8c87c1d7004374cc"></span><a class="reference internal" href="#_CPPv2N6egglib4nnet4DataE" title="egglib::nnet::Data">Data</a> &amp;<code class="descclassname">egglib::nnet::Data::</code><code class="descname">operator=</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib4nnet4DataE" title="egglib::nnet::Data">Data</a> &amp;<em>src</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4DataaSERK4Data" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Copy assignment operator. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data9set_inputEjjd">
<span id="egglib::nnet::Data::set_input__unsigned-i.unsigned-i.double"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1a91485e2aed32c9841412c0ae10bc4beb"></span>void <code class="descclassname">egglib::nnet::Data::</code><code class="descname">set_input</code><span class="sig-paren">(</span>unsigned int <em>pattern</em>, unsigned int <em>variable</em>, double <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data9set_inputEjjd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Load an input data item. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data13set_num_inputEj">
<span id="egglib::nnet::Data::set_num_input__unsigned-i"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1a54760180c9e03cebd010b423c6989fbd"></span>void <code class="descclassname">egglib::nnet::Data::</code><code class="descname">set_num_input</code><span class="sig-paren">(</span>unsigned int <em>num</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data13set_num_inputEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set number of input variables. </p>
<p>Invalidate all means and standard deviations if the data was previously normalized. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data14set_num_outputEj">
<span id="egglib::nnet::Data::set_num_output__unsigned-i"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1a070436562fdc40b21b1fb45ac4200459"></span>void <code class="descclassname">egglib::nnet::Data::</code><code class="descname">set_num_output</code><span class="sig-paren">(</span>unsigned int <em>num</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data14set_num_outputEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set number of output variables. </p>
<p>Invalidate all means and standard deviations if the data was previously normalized. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data16set_num_patternsEj">
<span id="egglib::nnet::Data::set_num_patterns__unsigned-i"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1a43c35e8c50ab03a237471377f45d1118"></span>void <code class="descclassname">egglib::nnet::Data::</code><code class="descname">set_num_patterns</code><span class="sig-paren">(</span>unsigned int <em>num</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data16set_num_patternsEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set number of patterns. </p>
<p>Invalidate all means and standard deviations if the data was previously normalized. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data10set_outputEjjd">
<span id="egglib::nnet::Data::set_output__unsigned-i.unsigned-i.double"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1a5f047b3d327dd304665c6e80e18e2ba6"></span>void <code class="descclassname">egglib::nnet::Data::</code><code class="descname">set_output</code><span class="sig-paren">(</span>unsigned int <em>pattern</em>, unsigned int <em>variable</em>, double <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data10set_outputEjjd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Load an output data item. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data9std_inputEj">
<span id="egglib::nnet::Data::std_input__unsigned-i"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1aa704ba173a20caac8f241fbc8d208346"></span>double <code class="descclassname">egglib::nnet::Data::</code><code class="descname">std_input</code><span class="sig-paren">(</span>unsigned int <em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data9std_inputEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the standard deviation for an input variable. </p>
<p>If data have been normalized using the method <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Data_1acd0de676568888d848beb97dcc53ae47"><span class="std std-ref">normalize()</span></a>, this method returns the standard deviation of a given input variable in the original data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data10std_outputEj">
<span id="egglib::nnet::Data::std_output__unsigned-i"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1a926df3c27d065e571dd6d279f73b9e9d"></span>double <code class="descclassname">egglib::nnet::Data::</code><code class="descname">std_output</code><span class="sig-paren">(</span>unsigned int <em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data10std_outputEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the standard deviation for an output variable. </p>
<p>If data have been normalized using the method <a class="reference internal" href="#project0classegglib_1_1nnet_1_1Data_1acd0de676568888d848beb97dcc53ae47"><span class="std std-ref">normalize()</span></a>, this method returns the standard deviation of a given output variable in the original data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib4nnet4Data18unnormalize_outputEjdd">
<span id="egglib::nnet::Data::unnormalize_output__unsigned-i.double.double"></span><span class="target" id="project0classegglib_1_1nnet_1_1Data_1ae01e370a8e221106e1429b8b82b38856"></span>void <code class="descclassname">egglib::nnet::Data::</code><code class="descname">unnormalize_output</code><span class="sig-paren">(</span>unsigned int <em>index</em>, double <em>mean</em>, double <em>std</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib4nnet4Data18unnormalize_outputEjdd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Unnormalize data for an output variable. </p>
<p>All data are modified permanently. The passed mean and standard deviation are not saved. </p>
</dd></dl>

</div>
</dd></dl>

</div>
</div>
<div class="section" id="utilities">
<h2>Utilities<a class="headerlink" href="#utilities" title="Permalink to this headline">¶</a></h2>
<div class="section" id="interspersealign">
<h3>IntersperseAlign<a class="headerlink" href="#interspersealign" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib16IntersperseAlignE">
<span id="egglib::IntersperseAlign"></span><span class="target" id="project0classegglib_1_1IntersperseAlign"></span><em class="property">class </em></dt>
<dd><p>Insert non-varying sites within alignments. </p>
<p>This class allows to add non-varying sites within an alignment at given positions. The procedure below must be strictly followed:</p>
<p><ul class="simple">
<li>Create an instance. The constructors takes no arguments.</li>
<li>Load a <a class="reference internal" href="#project0classegglib_1_1DataHolder"><span class="std std-ref">DataHolder</span></a> instance using <a class="reference internal" href="#project0classegglib_1_1IntersperseAlign_1a8824c6c24b26d5b406af9307bae6ad41"><span class="std std-ref">load()</span></a>. It is required that it is an alignment (a matrix). The instance will create an array of positions internally.</li>
<li>Specify the desired length of the final alignment using <a class="reference internal" href="#project0classegglib_1_1IntersperseAlign_1a311eb9facd24a8c2fd63874baab657dd"><span class="std std-ref">set_length()</span></a>.</li>
<li>Specify the position of all sites of the original alignment. This can be achieved by three ways:<ul>
<li>Specify manually all positions as real numbers using <a class="reference internal" href="#project0classegglib_1_1IntersperseAlign_1afd419aff81801237ab8bc693d07be959"><span class="std std-ref">set_position()</span></a>.</li>
<li>Specify manually all positions as extant indexes (as integer values) using <a class="reference internal" href="#project0classegglib_1_1IntersperseAlign_1aaa738c98aaacbb9ce3c8d41d972f696c"><span class="std std-ref">set_round_position()</span></a> for all positions. If this approach is used, it is necessary to set the round option of intersperse to false.</li>
<li>Pass the reference to the <a class="reference internal" href="coalesce.html#project0classegglib_1_1Coalesce"><span class="std std-ref">Coalesce</span></a> instance that has generated the alignment (assuming it is a simulation) and let the instance find itself the site positions, with the method <a class="reference internal" href="#project0classegglib_1_1IntersperseAlign_1ad086ec846e043b569437f63c5cd53806"><span class="std std-ref">get_positions()</span></a>.</li>
</ul>
</li>
<li>Specify the list of alleles values used for non-varying positions using <a class="reference internal" href="#project0classegglib_1_1IntersperseAlign_1aaf91e726b94a213d08f1780e93c8bcce"><span class="std std-ref">set_num_alleles()</span></a> and then <a class="reference internal" href="#project0classegglib_1_1IntersperseAlign_1aad5f1bdcb3ec1ecdb0cbe2137341924d"><span class="std std-ref">set_allele()</span></a> as many times as needed. If there is more than one allele, non-varying alleles will be picked randomly. This method can be skipped (by default, the value corresponding to <code class="docutils literal"><span class="pre">A</span></code> will be used).</li>
<li>Provide the address of a random number generator using <a class="reference internal" href="#project0classegglib_1_1IntersperseAlign_1a21829540d0575f00851e8cc91ca682da"><span class="std std-ref">set_random()</span></a> (it is always needed).</li>
<li>Call <a class="reference internal" href="#project0classegglib_1_1IntersperseAlign_1a730d1143d22a4f331018164c6a21aa20"><span class="std std-ref">intersperse()</span></a>. This will change the <a class="reference internal" href="#project0classegglib_1_1DataHolder"><span class="std std-ref">DataHolder</span></a> originally loaded.</li>
</ul>
</p>
<p>Header: &lt;egglib-cpp/DataHolder.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib16IntersperseAlign16IntersperseAlignEv">
<span id="egglib::IntersperseAlign::IntersperseAlign"></span><span class="target" id="project0classegglib_1_1IntersperseAlign_1ac9b92a00f153df6a8534beeb09ecc487"></span><code class="descclassname">egglib::IntersperseAlign::</code><code class="descname">IntersperseAlign</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16IntersperseAlign16IntersperseAlignEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16IntersperseAlignD0Ev">
<span id="egglib::IntersperseAlign::~IntersperseAlign"></span><span class="target" id="project0classegglib_1_1IntersperseAlign_1a05154f9b8d24cd213ae8e87f47fe0139"></span><code class="descclassname">egglib::IntersperseAlign::</code><code class="descname">~IntersperseAlign</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16IntersperseAlignD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16IntersperseAlign13get_positionsERK8Coalesce">
<span id="egglib::IntersperseAlign::get_positions__CoalesceCR"></span><span class="target" id="project0classegglib_1_1IntersperseAlign_1ad086ec846e043b569437f63c5cd53806"></span>void <code class="descclassname">egglib::IntersperseAlign::</code><code class="descname">get_positions</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="coalesce.html#_CPPv2N6egglib8CoalesceE" title="egglib::Coalesce">Coalesce</a> &amp;<em>coalesce</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16IntersperseAlign13get_positionsERK8Coalesce" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Gets automatically the positions of sites of the original alignment. </p>
<p>A <a class="reference internal" href="#project0classegglib_1_1DataHolder"><span class="std std-ref">DataHolder</span></a> reference must have been loaded using <a class="reference internal" href="#project0classegglib_1_1IntersperseAlign_1a8824c6c24b26d5b406af9307bae6ad41"><span class="std std-ref">load()</span></a>, and this <a class="reference internal" href="#project0classegglib_1_1DataHolder"><span class="std std-ref">DataHolder</span></a> object must be the last one simulated using the <a class="reference internal" href="coalesce.html#project0classegglib_1_1Coalesce"><span class="std std-ref">Coalesce</span></a> object whose reference is passed. This method will load the positions of all sites as provided by <a class="reference internal" href="coalesce.html#project0classegglib_1_1Coalesce"><span class="std std-ref">Coalesce</span></a>. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16IntersperseAlign11intersperseEb">
<span id="egglib::IntersperseAlign::intersperse__b"></span><span class="target" id="project0classegglib_1_1IntersperseAlign_1a730d1143d22a4f331018164c6a21aa20"></span>void <code class="descclassname">egglib::IntersperseAlign::</code><code class="descname">intersperse</code><span class="sig-paren">(</span>bool <em>round_positions</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16IntersperseAlign11intersperseEb" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Insert non-varying sites. </p>
<p>This method modifies the <a class="reference internal" href="#project0classegglib_1_1DataHolder"><span class="std std-ref">DataHolder</span></a> instance that has been loaded. It is required to have loaded one, and to have specified the position of each of its sites. It is also logical (but not formally required) to have specified the desired length of the alignment. It is possible to specify more than one alleles for inserted positions. It is required to have passed a random number generator.</p>
<p>After call to this method, the loaded <a class="reference internal" href="#project0classegglib_1_1DataHolder"><span class="std std-ref">DataHolder</span></a> instance will have a length equal to the value specified using <a class="reference internal" href="#project0classegglib_1_1IntersperseAlign_1a311eb9facd24a8c2fd63874baab657dd"><span class="std std-ref">set_length()</span></a>, unless the original <a class="reference internal" href="#project0classegglib_1_1DataHolder"><span class="std std-ref">DataHolder</span></a> was longer (in such case, it is not changed at all).</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">round_positions</span></code> - <p>a boolean indicating if site positions must be rounded. Set it to false if already rounded positions have been provided. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16IntersperseAlign4loadER10DataHolder">
<span id="egglib::IntersperseAlign::load__DataHolderR"></span><span class="target" id="project0classegglib_1_1IntersperseAlign_1a8824c6c24b26d5b406af9307bae6ad41"></span>void <code class="descclassname">egglib::IntersperseAlign::</code><code class="descname">load</code><span class="sig-paren">(</span><a class="reference internal" href="#_CPPv2N6egglib10DataHolderE" title="egglib::DataHolder">DataHolder</a> &amp;<em>data</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16IntersperseAlign4loadER10DataHolder" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Loads a data set. </p>
<p>The loaded data set may contain any number of sites (even zero).</p>
<p>This method does not reset the random number generator, the final alignment length, the position of sites (unless the new alignment has a different number of sites compared with the previous one) or the number and value of non-varying alleles. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16IntersperseAlign10set_alleleEji">
<span id="egglib::IntersperseAlign::set_allele__unsigned-i.i"></span><span class="target" id="project0classegglib_1_1IntersperseAlign_1aad5f1bdcb3ec1ecdb0cbe2137341924d"></span>void <code class="descclassname">egglib::IntersperseAlign::</code><code class="descname">set_allele</code><span class="sig-paren">(</span>unsigned int <em>index</em>, int <em>allele</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16IntersperseAlign10set_alleleEji" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Sets an allele for inserted positions. </p>
<p>The number of alleles must have been fixed using <a class="reference internal" href="#project0classegglib_1_1IntersperseAlign_1aaf91e726b94a213d08f1780e93c8bcce"><span class="std std-ref">set_num_alleles()</span></a>. The default value for the first allele is <code class="docutils literal"><span class="pre">A</span></code>. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16IntersperseAlign10set_lengthEj">
<span id="egglib::IntersperseAlign::set_length__unsigned-i"></span><span class="target" id="project0classegglib_1_1IntersperseAlign_1a311eb9facd24a8c2fd63874baab657dd"></span>void <code class="descclassname">egglib::IntersperseAlign::</code><code class="descname">set_length</code><span class="sig-paren">(</span>unsigned int <em>length</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16IntersperseAlign10set_lengthEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Specifies the desired length of the final alignment. </p>
<p>If this method is skipped, the default value is 0 (the final alignment is identical to the original one), or the previously specified value (if <a class="reference internal" href="#project0classegglib_1_1IntersperseAlign_1a311eb9facd24a8c2fd63874baab657dd"><span class="std std-ref">set_length()</span></a> has been called previously). </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16IntersperseAlign15set_num_allelesEj">
<span id="egglib::IntersperseAlign::set_num_alleles__unsigned-i"></span><span class="target" id="project0classegglib_1_1IntersperseAlign_1aaf91e726b94a213d08f1780e93c8bcce"></span>void <code class="descclassname">egglib::IntersperseAlign::</code><code class="descname">set_num_alleles</code><span class="sig-paren">(</span>unsigned int <em>num</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16IntersperseAlign15set_num_allelesEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Specifies the number of possible alleles at inserted positions. </p>
<p>This method may be called at any time. The value must be at least one. If more than one, alleles at inserted positions will be picked randomly (they will be fixed among samples). All alleles must be specified using <a class="reference internal" href="#project0classegglib_1_1IntersperseAlign_1aad5f1bdcb3ec1ecdb0cbe2137341924d"><span class="std std-ref">set_allele()</span></a>. The default value is one and the default first allele is <code class="docutils literal"><span class="pre">A</span></code>. It is possible not to specify the first allele even if the number of alleles is increased (the <code class="docutils literal"><span class="pre">A</span></code> value will be retained). </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16IntersperseAlign12set_positionEjd">
<span id="egglib::IntersperseAlign::set_position__unsigned-i.double"></span><span class="target" id="project0classegglib_1_1IntersperseAlign_1afd419aff81801237ab8bc693d07be959"></span>void <code class="descclassname">egglib::IntersperseAlign::</code><code class="descname">set_position</code><span class="sig-paren">(</span>unsigned int <em>index</em>, double <em>position</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16IntersperseAlign12set_positionEjd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Sets the position of one of the sites of the original alignment. </p>
<p>A <a class="reference internal" href="#project0classegglib_1_1DataHolder"><span class="std std-ref">DataHolder</span></a> reference must have been loaded using <a class="reference internal" href="#project0classegglib_1_1IntersperseAlign_1a8824c6c24b26d5b406af9307bae6ad41"><span class="std std-ref">load()</span></a>. This method allows to specify the position of each of the sites of the passed <a class="reference internal" href="#project0classegglib_1_1DataHolder"><span class="std std-ref">DataHolder</span></a> instance. Note that the position of all sites must be specified, that positions must always be increasing (consecutive positions might be equal), and all positions must be at least 0 and at most 1. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16IntersperseAlign10set_randomEP6Random">
<span id="egglib::IntersperseAlign::set_random__RandomP"></span><span class="target" id="project0classegglib_1_1IntersperseAlign_1a21829540d0575f00851e8cc91ca682da"></span>void <code class="descclassname">egglib::IntersperseAlign::</code><code class="descname">set_random</code><span class="sig-paren">(</span><a class="reference internal" href="#_CPPv2N6egglib6RandomE" title="egglib::Random">Random</a> *<em>random</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16IntersperseAlign10set_randomEP6Random" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Provides a random number generator. </p>
<p>It is always required to provide a random number generator, even if <a class="reference internal" href="#project0classegglib_1_1IntersperseAlign_1aaf91e726b94a213d08f1780e93c8bcce"><span class="std std-ref">set_num_alleles()</span></a> is one. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16IntersperseAlign18set_round_positionEjj">
<span id="egglib::IntersperseAlign::set_round_position__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1IntersperseAlign_1aaa738c98aaacbb9ce3c8d41d972f696c"></span>void <code class="descclassname">egglib::IntersperseAlign::</code><code class="descname">set_round_position</code><span class="sig-paren">(</span>unsigned int <em>index</em>, unsigned int <em>position</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16IntersperseAlign18set_round_positionEjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Sets the position of one of the sites of the original alignment. </p>
<p>A <a class="reference internal" href="#project0classegglib_1_1DataHolder"><span class="std std-ref">DataHolder</span></a> reference must have been loaded using <a class="reference internal" href="#project0classegglib_1_1IntersperseAlign_1a8824c6c24b26d5b406af9307bae6ad41"><span class="std std-ref">load()</span></a>. This method allows to specify the position of each of the sites of the passed <a class="reference internal" href="#project0classegglib_1_1DataHolder"><span class="std std-ref">DataHolder</span></a> instance. Note that the position of all sites must be specified, that positions must always be increasing (consecutive positions might be equal), and all positions must be at least 0 and at most ls-1 where ls is the length of the final alignment.</p>
<p>If you use this method, you must use it for all sites and then set the argument of <a class="reference internal" href="#project0classegglib_1_1IntersperseAlign_1a730d1143d22a4f331018164c6a21aa20"><span class="std std-ref">intersperse()</span></a> to false. </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="vectorint">
<h3>VectorInt<a class="headerlink" href="#vectorint" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib9VectorIntE">
<span id="egglib::VectorInt"></span><span class="target" id="project0classegglib_1_1VectorInt"></span><em class="property">class </em></dt>
<dd><p>Minimal reimplementation of a vector&lt;int&gt; </p>
<p>Header: &lt;egglib-cpp/DataHolder.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib9VectorInt9VectorIntEv">
<span id="egglib::VectorInt::VectorInt"></span><span class="target" id="project0classegglib_1_1VectorInt_1aa5a771144446d4bbdeff4dab3880e08d"></span><code class="descclassname">egglib::VectorInt::</code><code class="descname">VectorInt</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9VectorInt9VectorIntEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor (default: 0 values) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib9VectorInt9VectorIntERK9VectorInt">
<span id="egglib::VectorInt::VectorInt__VectorIntCR"></span><span class="target" id="project0classegglib_1_1VectorInt_1ae4a83a3e697ab525e6419b77ed70579a"></span><code class="descclassname">egglib::VectorInt::</code><code class="descname">VectorInt</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib9VectorInt9VectorIntEv" title="egglib::VectorInt::VectorInt">VectorInt</a> &amp;<em>src</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9VectorInt9VectorIntERK9VectorInt" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Copy constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib9VectorIntD0Ev">
<span id="egglib::VectorInt::~VectorInt"></span>virtual <span class="target" id="project0classegglib_1_1VectorInt_1abc5e8168b310a058b7f0ea5cd84ab8dc"></span><code class="descclassname">egglib::VectorInt::</code><code class="descname">~VectorInt</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9VectorIntD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib9VectorInt5clearEv">
<span id="egglib::VectorInt::clear"></span><span class="target" id="project0classegglib_1_1VectorInt_1ac8bb3912a3ce86b15842e79d0b421204"></span>void <code class="descclassname">egglib::VectorInt::</code><code class="descname">clear</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9VectorInt5clearEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Release memory. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib9VectorInt8get_itemEj">
<span id="egglib::VectorInt::get_item__unsigned-i"></span><span class="target" id="project0classegglib_1_1VectorInt_1adb54314458a93d2bff006dc129ec2d8e"></span>int <code class="descclassname">egglib::VectorInt::</code><code class="descname">get_item</code><span class="sig-paren">(</span>unsigned int <em>i</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9VectorInt8get_itemEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get a value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib9VectorInt14get_num_valuesEv">
<span id="egglib::VectorInt::get_num_values"></span><span class="target" id="project0classegglib_1_1VectorInt_1aee0535a0cf860c710995244a996077b6"></span>unsigned int <code class="descclassname">egglib::VectorInt::</code><code class="descname">get_num_values</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9VectorInt14get_num_valuesEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the number of values. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib9VectorIntaSERK9VectorInt">
<span id="egglib::VectorInt::assign-operator__VectorIntCR"></span><span class="target" id="project0classegglib_1_1VectorInt_1a8f4ca49ebde091d3126c41dc2b881daf"></span><a class="reference internal" href="#_CPPv2N6egglib9VectorIntE" title="egglib::VectorInt">VectorInt</a> &amp;<code class="descclassname">egglib::VectorInt::</code><code class="descname">operator=</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib9VectorIntE" title="egglib::VectorInt">VectorInt</a> &amp;<em>src</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9VectorIntaSERK9VectorInt" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Copy assignment operator. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib9VectorInt8set_itemEji">
<span id="egglib::VectorInt::set_item__unsigned-i.i"></span><span class="target" id="project0classegglib_1_1VectorInt_1a9d442afca6b4c460b842fd9f3f7c8215"></span>void <code class="descclassname">egglib::VectorInt::</code><code class="descname">set_item</code><span class="sig-paren">(</span>unsigned int <em>i</em>, int <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9VectorInt8set_itemEji" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set a value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib9VectorInt14set_num_valuesEj">
<span id="egglib::VectorInt::set_num_values__unsigned-i"></span><span class="target" id="project0classegglib_1_1VectorInt_1af1cc984d676c993962e865de6fc8eabd"></span>void <code class="descclassname">egglib::VectorInt::</code><code class="descname">set_num_values</code><span class="sig-paren">(</span>unsigned int <em>n</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9VectorInt14set_num_valuesEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set the number of vqlues (values are not initialized) </p>
</dd></dl>

</div>
</dd></dl>

</div>
</div>
<div class="section" id="exceptions">
<h2>Exceptions<a class="headerlink" href="#exceptions" title="Permalink to this headline">¶</a></h2>
<div class="section" id="eggexception">
<h3>EggException<a class="headerlink" href="#eggexception" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib12EggExceptionE">
<span id="egglib::EggException"></span><span class="target" id="project0classegglib_1_1EggException"></span><em class="property">class </em></dt>
<dd><p>Base exception type for errors occurring in this library. </p>
<p>Header: &lt;egglib-cpp/egglib.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib12EggException12EggExceptionEv">
<span id="egglib::EggException::EggException"></span><span class="target" id="project0classegglib_1_1EggException_1ac6440008c5ae192c3b1a56dab4d91290"></span><code class="descclassname">egglib::EggException::</code><code class="descname">EggException</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12EggException12EggExceptionEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor with empty error message. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12EggException12EggExceptionEPKc">
<span id="egglib::EggException::EggException__cCP"></span><span class="target" id="project0classegglib_1_1EggException_1acf6e24c79b073d366de6bcd466befef0"></span><code class="descclassname">egglib::EggException::</code><code class="descname">EggException</code><span class="sig-paren">(</span><em class="property">const</em> char *<em>message</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12EggException12EggExceptionEPKc" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Creates the exception. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12EggExceptionD0Ev">
<span id="egglib::EggException::~EggException"></span><span class="target" id="project0classegglib_1_1EggException_1a40b17157e271cd0bb5ea15f515b7296b"></span><code class="descclassname">egglib::EggException::</code><code class="descname">~EggException</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12EggExceptionD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12EggException4whatEv">
<span id="egglib::EggException::what"></span>virtual <span class="target" id="project0classegglib_1_1EggException_1a8d4b96162e93e11e5816d83702578af8"></span><em class="property">const</em> char *<code class="descclassname">egglib::EggException::</code><code class="descname">what</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12EggException4whatEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Gets error message. </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="eggargumentvalueerror">
<h3>EggArgumentValueError<a class="headerlink" href="#eggargumentvalueerror" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib21EggArgumentValueErrorE">
<span id="egglib::EggArgumentValueError"></span><span class="target" id="project0classegglib_1_1EggArgumentValueError"></span><em class="property">class </em></dt>
<dd><p>Exception type for argument value errors. </p>
<p>Header: &lt;egglib-cpp/egglib.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib21EggArgumentValueError21EggArgumentValueErrorEPKc">
<span id="egglib::EggArgumentValueError::EggArgumentValueError__cCP"></span><span class="target" id="project0classegglib_1_1EggArgumentValueError_1adbcd4872c70dbad55d98af07f67d9c02"></span><code class="descclassname">egglib::EggArgumentValueError::</code><code class="descname">EggArgumentValueError</code><span class="sig-paren">(</span><em class="property">const</em> char *<em>m</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib21EggArgumentValueError21EggArgumentValueErrorEPKc" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Creates the exception. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib21EggArgumentValueErrorD0Ev">
<span id="egglib::EggArgumentValueError::~EggArgumentValueError"></span><span class="target" id="project0classegglib_1_1EggArgumentValueError_1a710d98aa7a085b3ee092051b49b81ea2"></span><code class="descclassname">egglib::EggArgumentValueError::</code><code class="descname">~EggArgumentValueError</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib21EggArgumentValueErrorD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="eggformaterror">
<h3>EggFormatError<a class="headerlink" href="#eggformaterror" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib14EggFormatErrorE">
<span id="egglib::EggFormatError"></span><span class="target" id="project0classegglib_1_1EggFormatError"></span><em class="property">class </em></dt>
<dd><p>Exception type for file/string parsing errors. </p>
<p>Header: &lt;egglib-cpp/egglib.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib14EggFormatError14EggFormatErrorEPKcjPKcPKccPKc">
<span id="egglib::EggFormatError::EggFormatError__cCP.unsigned-i.cCP.cCP.c.cCP"></span><span class="target" id="project0classegglib_1_1EggFormatError_1a11bf77a2ea5f3bd6c30d2196ff370c51"></span><code class="descclassname">egglib::EggFormatError::</code><code class="descname">EggFormatError</code><span class="sig-paren">(</span><em class="property">const</em> char *<em>fileName</em>, unsigned int <em>line</em>, <em class="property">const</em> char *<em>expectedFormat</em>, <em class="property">const</em> char *<em>m</em>, char <em>c</em>, <em class="property">const</em> char *<em>paste_end</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib14EggFormatError14EggFormatErrorEPKcjPKcPKccPKc" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Creates the exception. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib14EggFormatErrorD0Ev">
<span id="egglib::EggFormatError::~EggFormatError"></span><span class="target" id="project0classegglib_1_1EggFormatError_1a4a89a9be9841b8f62c49db42d6dc908d"></span><code class="descclassname">egglib::EggFormatError::</code><code class="descname">~EggFormatError</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib14EggFormatErrorD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib14EggFormatError9characterEv">
<span id="egglib::EggFormatError::character"></span><span class="target" id="project0classegglib_1_1EggFormatError_1ab5f6f377e1e32d194bad304e3e84e43e"></span>char <code class="descclassname">egglib::EggFormatError::</code><code class="descname">character</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib14EggFormatError9characterEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Get character. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib14EggFormatError4infoEv">
<span id="egglib::EggFormatError::info"></span><span class="target" id="project0classegglib_1_1EggFormatError_1a5384565c07d8c05739cd2238a2dd556d"></span><em class="property">const</em> char *<code class="descclassname">egglib::EggFormatError::</code><code class="descname">info</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib14EggFormatError4infoEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Get additional information field. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib14EggFormatError4lineEv">
<span id="egglib::EggFormatError::line"></span><span class="target" id="project0classegglib_1_1EggFormatError_1a3065bb3d6a184bbacd0e748c99345667"></span>unsigned int <code class="descclassname">egglib::EggFormatError::</code><code class="descname">line</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib14EggFormatError4lineEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Get line number. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib14EggFormatError1mEv">
<span id="egglib::EggFormatError::m"></span><span class="target" id="project0classegglib_1_1EggFormatError_1a6e0c22d2a5dc949e40f01950d2b65a37"></span><em class="property">const</em> char *<code class="descclassname">egglib::EggFormatError::</code><code class="descname">m</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib14EggFormatError1mEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Get bare error message (before formatting) </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="egginvalidalleleerror">
<h3>EggInvalidAlleleError<a class="headerlink" href="#egginvalidalleleerror" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib21EggInvalidAlleleErrorE">
<span id="egglib::EggInvalidAlleleError"></span><span class="target" id="project0classegglib_1_1EggInvalidAlleleError"></span><em class="property">class </em></dt>
<dd><p>Exception type for invalid allele. </p>
<p>Header: &lt;egglib-cpp/egglib.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib21EggInvalidAlleleError21EggInvalidAlleleErrorEijj">
<span id="egglib::EggInvalidAlleleError::EggInvalidAlleleError__i.unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1EggInvalidAlleleError_1a7547e8f4e9497d8466ffc5b2911ee8b4"></span><code class="descclassname">egglib::EggInvalidAlleleError::</code><code class="descname">EggInvalidAlleleError</code><span class="sig-paren">(</span>int <em>c</em>, unsigned int <em>seqIndex</em>, unsigned int <em>posIndex</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib21EggInvalidAlleleError21EggInvalidAlleleErrorEijj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Creates the exception. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib21EggInvalidAlleleErrorD0Ev">
<span id="egglib::EggInvalidAlleleError::~EggInvalidAlleleError"></span><span class="target" id="project0classegglib_1_1EggInvalidAlleleError_1a94df7785dfab10471a90c6de0e57bc84"></span><code class="descclassname">egglib::EggInvalidAlleleError::</code><code class="descname">~EggInvalidAlleleError</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib21EggInvalidAlleleErrorD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="eggmemoryerror">
<h3>EggMemoryError<a class="headerlink" href="#eggmemoryerror" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib14EggMemoryErrorE">
<span id="egglib::EggMemoryError"></span><span class="target" id="project0classegglib_1_1EggMemoryError"></span><em class="property">class </em></dt>
<dd><p>Exception type for memory errors. </p>
<p>There is a macro EGGMEM which stands for <a class="reference internal" href="#project0classegglib_1_1EggMemoryError"><span class="std std-ref">EggMemoryError</span></a>(<strong>LINE</strong>, <strong>FILE</strong>).</p>
<p>Header: &lt;egglib-cpp/egglib.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib14EggMemoryError14EggMemoryErrorEjPKc">
<span id="egglib::EggMemoryError::EggMemoryError__unsigned-i.cCP"></span><span class="target" id="project0classegglib_1_1EggMemoryError_1a999c9a9dd1c5fc23747805e10a4a61f0"></span><code class="descclassname">egglib::EggMemoryError::</code><code class="descname">EggMemoryError</code><span class="sig-paren">(</span>unsigned int <em>line</em>, <em class="property">const</em> char *<em>file</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib14EggMemoryError14EggMemoryErrorEjPKc" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Creates the exception. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib14EggMemoryErrorD0Ev">
<span id="egglib::EggMemoryError::~EggMemoryError"></span><span class="target" id="project0classegglib_1_1EggMemoryError_1a60a7a84866d977101306c19b9b1aed45"></span><code class="descclassname">egglib::EggMemoryError::</code><code class="descname">~EggMemoryError</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib14EggMemoryErrorD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="eggopenfileerror">
<h3>EggOpenFileError<a class="headerlink" href="#eggopenfileerror" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib16EggOpenFileErrorE">
<span id="egglib::EggOpenFileError"></span><span class="target" id="project0classegglib_1_1EggOpenFileError"></span><em class="property">class </em></dt>
<dd><p>Exception type for errors while opening a file. </p>
<p>Header: &lt;egglib-cpp/egglib.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib16EggOpenFileError16EggOpenFileErrorEPKc">
<span id="egglib::EggOpenFileError::EggOpenFileError__cCP"></span><span class="target" id="project0classegglib_1_1EggOpenFileError_1aab3d5b2be120f3b1977c021483011e25"></span><code class="descclassname">egglib::EggOpenFileError::</code><code class="descname">EggOpenFileError</code><span class="sig-paren">(</span><em class="property">const</em> char *<em>fileName</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16EggOpenFileError16EggOpenFileErrorEPKc" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Creates the exception. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16EggOpenFileErrorD0Ev">
<span id="egglib::EggOpenFileError::~EggOpenFileError"></span><span class="target" id="project0classegglib_1_1EggOpenFileError_1ab79613471bfa59483c7e4ead8c98b2da"></span><code class="descclassname">egglib::EggOpenFileError::</code><code class="descname">~EggOpenFileError</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16EggOpenFileErrorD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="eggploidyerror">
<h3>EggPloidyError<a class="headerlink" href="#eggploidyerror" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib14EggPloidyErrorE">
<span id="egglib::EggPloidyError"></span><span class="target" id="project0classegglib_1_1EggPloidyError"></span><em class="property">class </em></dt>
<dd><p>Exception type for inconsistent ploidy over individuals EggInvalidChromosomeError</p>
<p>Header: &lt;egglib-cpp/egglib.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib14EggPloidyError14EggPloidyErrorEv">
<span id="egglib::EggPloidyError::EggPloidyError"></span><span class="target" id="project0classegglib_1_1EggPloidyError_1a62285a2f0fe254125b5147692d17f31e"></span><code class="descclassname">egglib::EggPloidyError::</code><code class="descname">EggPloidyError</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib14EggPloidyError14EggPloidyErrorEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Creates the exception. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib14EggPloidyErrorD0Ev">
<span id="egglib::EggPloidyError::~EggPloidyError"></span><span class="target" id="project0classegglib_1_1EggPloidyError_1aeef4379d8c009c772f0ceb18514e6cb5"></span><code class="descclassname">egglib::EggPloidyError::</code><code class="descname">~EggPloidyError</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib14EggPloidyErrorD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="eggruntimeerror">
<h3>EggRuntimeError<a class="headerlink" href="#eggruntimeerror" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib15EggRuntimeErrorE">
<span id="egglib::EggRuntimeError"></span><span class="target" id="project0classegglib_1_1EggRuntimeError"></span><em class="property">class </em></dt>
<dd><p>Exception type for runtime errors. </p>
<p>Runtime error definition is rather large. Includes bugs as well as logical errors.</p>
<p>Header: &lt;egglib-cpp/egglib.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib15EggRuntimeError15EggRuntimeErrorEPKc">
<span id="egglib::EggRuntimeError::EggRuntimeError__cCP"></span><span class="target" id="project0classegglib_1_1EggRuntimeError_1a5abda8922cda1fabf9e2a8236c219804"></span><code class="descclassname">egglib::EggRuntimeError::</code><code class="descname">EggRuntimeError</code><span class="sig-paren">(</span><em class="property">const</em> char *<em>m</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15EggRuntimeError15EggRuntimeErrorEPKc" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Creates the exception. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15EggRuntimeErrorD0Ev">
<span id="egglib::EggRuntimeError::~EggRuntimeError"></span><span class="target" id="project0classegglib_1_1EggRuntimeError_1a849917d7fb2c4b25cefd345ab19d9d72"></span><code class="descclassname">egglib::EggRuntimeError::</code><code class="descname">~EggRuntimeError</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15EggRuntimeErrorD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="eggunalignederror">
<h3>EggUnalignedError<a class="headerlink" href="#eggunalignederror" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib17EggUnalignedErrorE">
<span id="egglib::EggUnalignedError"></span><span class="target" id="project0classegglib_1_1EggUnalignedError"></span><em class="property">class </em></dt>
<dd><p>Exception type for unaligned sequences. </p>
<p>Header: &lt;egglib-cpp/egglib.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib17EggUnalignedError17EggUnalignedErrorEv">
<span id="egglib::EggUnalignedError::EggUnalignedError"></span><span class="target" id="project0classegglib_1_1EggUnalignedError_1a9a068f612f01d1d3a96a00e9c4762a68"></span><code class="descclassname">egglib::EggUnalignedError::</code><code class="descname">EggUnalignedError</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib17EggUnalignedError17EggUnalignedErrorEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Creates the exception. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib17EggUnalignedErrorD0Ev">
<span id="egglib::EggUnalignedError::~EggUnalignedError"></span><span class="target" id="project0classegglib_1_1EggUnalignedError_1a4d1d74c9559471433f3bb0fb289959f3"></span><code class="descclassname">egglib::EggUnalignedError::</code><code class="descname">~EggUnalignedError</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib17EggUnalignedErrorD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

</div>
</dd></dl>

</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="datafiles.html" title="Data file formats"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="C++ Library"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../contents.html">EggLib&#39;s documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../toc.html" >Contents</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="index.html" >C++ Library</a> &#187;</li> 
      </ul>
    </div>
<!--
        <address style="text-align: right;"><small>Hosted by&nbsp;
        <a href="http://sourceforge.net/projects/egglib"><img src="http://sflogo.sourceforge.net/sflogo.php?group_id=232965&amp;type=8" width="80" height="15" alt="Get EggLib at SourceForge.net. Fast, secure and Free Open Source software downloads" / ALIGN=RIGHT BORDER=0></a>
        </address>
-->

  </body>
</html>