

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Diversity analyses &#8212; EggLib&#39;s documentation</title>
    <link rel="stylesheet" href="../_static/egglibdoc.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '3.0.0b21',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/icon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="Coalescent simulator" href="coalesce.html" />
    <script type="text/javascript">
      // intelligent scrolling of the sidebar content
      $(window).scroll(function() {
        var sb = $('.sphinxsidebarwrapper');
        var win = $(window);
        var sbh = sb.height();
        var offset = $('.sphinxsidebar').position()['top'];
        var wintop = win.scrollTop();
        var winbot = wintop + win.innerHeight();
        var curtop = sb.position()['top'];
        var curbot = curtop + sbh;
        // does sidebar fit in window?
        if (sbh < win.innerHeight()) {
          // yes: easy case -- always keep at the top
          sb.css('top', $u.min([$u.max([0, wintop - offset - 10]),
                                $(document).height() - sbh - 200]));
        } else {
          // no: only scroll if top/bottom edge of sidebar is at
          // top/bottom edge of window
          if (curtop > wintop && curbot > winbot) {
            sb.css('top', $u.max([wintop - offset - 10, 0]));
          } else if (curtop < wintop && curbot < winbot) {
            sb.css('top', $u.min([winbot - sbh - offset - 20,
                                  $(document).height() - sbh - 200]));
          }
        }
      });
    </script>
  </head>
  <body>
<div style="background-color: white; text-align: left; padding: 0px 0px 0px 0px">
<a href="../index.html"><img src="../_static/banner.png" border="0" alt="EggLib"/></a>
</div>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="coalesce.html" title="Coalescent simulator"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../contents.html">EggLib&#39;s documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../toc.html" >Contents</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="index.html" accesskey="U">C++ Library</a> &#187;</li> 
      </ul>
    </div>

      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../contents.html">
              <img class="logo" src="../_static/egglib_small.bmp" alt="Logo"/>
            </a></p>
  <h3><a href="../contents.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Diversity analyses</a><ul>
<li><a class="reference internal" href="#elementary-classes">Elementary classes</a><ul>
<li><a class="reference internal" href="#filter">Filter</a></li>
<li><a class="reference internal" href="#structure">Structure</a></li>
</ul>
</li>
<li><a class="reference internal" href="#classes-analysing-frequencies">Classes analysing frequencies</a><ul>
<li><a class="reference internal" href="#freqholder">FreqHolder</a></li>
<li><a class="reference internal" href="#freqset">FreqSet</a></li>
</ul>
</li>
<li><a class="reference internal" href="#site-level-operations">Site-level operations</a><ul>
<li><a class="reference internal" href="#siteholder">SiteHolder</a></li>
<li><a class="reference internal" href="#sitegeno">SiteGeno</a></li>
<li><a class="reference internal" href="#sitediversity">SiteDiversity</a></li>
<li><a class="reference internal" href="#codingsite">CodingSite</a></li>
<li><a class="reference internal" href="#allelestatus">AlleleStatus</a></li>
<li><a class="reference internal" href="#computev">ComputeV</a></li>
</ul>
</li>
<li><a class="reference internal" href="#generic-classes">Generic classes</a><ul>
<li><a class="reference internal" href="#diversity1">Diversity1</a></li>
<li><a class="reference internal" href="#diversity2">Diversity2</a></li>
</ul>
</li>
<li><a class="reference internal" href="#other-statistics">Other statistics</a><ul>
<li><a class="reference internal" href="#haplotypes">Haplotypes</a></li>
<li><a class="reference internal" href="#paralogpi">ParalogPi</a></li>
<li><a class="reference internal" href="#rd">Rd</a></li>
<li><a class="reference internal" href="#fu-s-f">Fu’s F</a></li>
</ul>
</li>
<li><a class="reference internal" href="#linkage-analysis">Linkage analysis</a><ul>
<li><a class="reference internal" href="#pairwiseld">PairwiseLD</a></li>
<li><a class="reference internal" href="#matrixld">MatrixLD</a></li>
</ul>
</li>
<li><a class="reference internal" href="#extended-haplotype-homozygosity">Extended haplotype homozygosity</a></li>
<li><a class="reference internal" href="#helpers">Helpers</a><ul>
<li><a class="reference internal" href="#constants">Constants</a></li>
<li><a class="reference internal" href="#functions">Functions</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="coalesce.html"
                        title="previous chapter">Coalescent simulator</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/cpp/diversity.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="diversity-analyses">
<span id="diversity"></span><h1>Diversity analyses<a class="headerlink" href="#diversity-analyses" title="Permalink to this headline">¶</a></h1>
<div class="section" id="elementary-classes">
<h2>Elementary classes<a class="headerlink" href="#elementary-classes" title="Permalink to this headline">¶</a></h2>
<div class="section" id="filter">
<h3>Filter<a class="headerlink" href="#filter" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib6FilterE">
<span id="egglib::Filter"></span><span class="target" id="project0classegglib_1_1Filter"></span><em class="property">class </em></dt>
<dd><p>Holds lists of valid (exploitable and missing) data codes. </p>
<p>This class holds two list of integer values: one corresponding to data codes that should be treated as valid and exploitable, and one corresponding to data codes that should be treated as valid but missing. If the list of exploitable data is empty, all data are considered to be exploitable (even those who are missing).</p>
<p>Header: &lt;egglib-cpp/Filter.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib6Filter6FilterEv">
<span id="egglib::Filter::Filter"></span><span class="target" id="project0classegglib_1_1Filter_1ad724ba04cbde6118ceddbd6635e96eaa"></span><code class="descclassname">egglib::Filter::</code><code class="descname">Filter</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Filter6FilterEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Filter6FilterERK6Filter">
<span id="egglib::Filter::Filter__FilterCR"></span><span class="target" id="project0classegglib_1_1Filter_1a4931f802b3599a266ea294e223ee5dae"></span><code class="descclassname">egglib::Filter::</code><code class="descname">Filter</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib6Filter6FilterEv" title="egglib::Filter::Filter">Filter</a> &amp;<em>src</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Filter6FilterERK6Filter" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Copy constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6FilterD0Ev">
<span id="egglib::Filter::~Filter"></span>virtual <span class="target" id="project0classegglib_1_1Filter_1afd45f479845303491834388e3ab74cf4"></span><code class="descclassname">egglib::Filter::</code><code class="descname">~Filter</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6FilterD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Filter15add_exploitableEi">
<span id="egglib::Filter::add_exploitable__i"></span><span class="target" id="project0classegglib_1_1Filter_1a884865100fa5893300d4cf4331505f3f"></span>void <code class="descclassname">egglib::Filter::</code><code class="descname">add_exploitable</code><span class="sig-paren">(</span>int <em>code</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Filter15add_exploitableEi" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add a exploitable data code. </p>
<p>By default (that is, if no exploitable data codes have been entered), all data values are considered to be exploitable, even if they appear in missing. If one or more data have been set as exploitable, only those will be considered to be exploitable. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Filter21add_exploitable_rangeEii">
<span id="egglib::Filter::add_exploitable_range__i.i"></span><span class="target" id="project0classegglib_1_1Filter_1a3aeabd0189ed6cecbe529bd6804af2f4"></span>void <code class="descclassname">egglib::Filter::</code><code class="descname">add_exploitable_range</code><span class="sig-paren">(</span>int <em>first</em>, int <em>last</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Filter21add_exploitable_rangeEii" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add a range of exploitable data codes. </p>
<p>Like <a class="reference internal" href="#project0classegglib_1_1Filter_1a884865100fa5893300d4cf4331505f3f"><span class="std std-ref">add_exploitable(int)</span></a>, but add all values included between first and last (both included). It is not required at all that all exploitable codes have a synonym. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Filter26add_exploitable_with_aliasEii">
<span id="egglib::Filter::add_exploitable_with_alias__i.i"></span><span class="target" id="project0classegglib_1_1Filter_1ac5267012bb41f625f3701f8b8cf942c4"></span>void <code class="descclassname">egglib::Filter::</code><code class="descname">add_exploitable_with_alias</code><span class="sig-paren">(</span>int <em>code</em>, int <em>alias</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Filter26add_exploitable_with_aliasEii" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add a exploitable data code with synonym character. </p>
<p>Similar to <a class="reference internal" href="#project0classegglib_1_1Filter_1a884865100fa5893300d4cf4331505f3f"><span class="std std-ref">add_exploitable(int)</span></a>, except that a second parameter is passed to specify a synonym for the main code. If a data item matching <code class="docutils literal"><span class="pre">alias</span></code> is passed to <a class="reference internal" href="#project0classegglib_1_1Filter_1a9d68ce2d52bdb3c05b3c711855762d6a"><span class="std std-ref">is_exploitable()</span></a>, it will be replaced by <code class="docutils literal"><span class="pre">code</span></code> and the function will return <code class="docutils literal"><span class="pre">true</span></code>. For a large range of continguous codes, do not use this method within a loop and use the appropriate and more efficient method <a class="reference internal" href="#project0classegglib_1_1Filter_1a3aeabd0189ed6cecbe529bd6804af2f4"><span class="std std-ref">add_exploitable_range(int, int)</span></a>. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Filter11add_missingEi">
<span id="egglib::Filter::add_missing__i"></span><span class="target" id="project0classegglib_1_1Filter_1afa1674235df886976f47ae9567dd1650"></span>void <code class="descclassname">egglib::Filter::</code><code class="descname">add_missing</code><span class="sig-paren">(</span>int <em>code</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Filter11add_missingEi" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add a missing data code. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Filter22add_missing_with_aliasEii">
<span id="egglib::Filter::add_missing_with_alias__i.i"></span><span class="target" id="project0classegglib_1_1Filter_1a23e84282c09c13c2c73b4f165aec541f"></span>void <code class="descclassname">egglib::Filter::</code><code class="descname">add_missing_with_alias</code><span class="sig-paren">(</span>int <em>code</em>, int <em>alias</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Filter22add_missing_with_aliasEii" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add a missing data code with synonym character. </p>
<p>Similar to <a class="reference internal" href="#project0classegglib_1_1Filter_1afa1674235df886976f47ae9567dd1650"><span class="std std-ref">add_missing(int)</span></a>, except that a second parameter is passed to specify a synonym for the main code. If a data item matching <code class="docutils literal"><span class="pre">alias</span></code> is passed to is_missing(int), it will be replaced by <code class="docutils literal"><span class="pre">code</span></code> and the function will return <code class="docutils literal"><span class="pre">true</span></code>. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Filter5checkEiRb">
<span id="egglib::Filter::check__i.bR"></span><span class="target" id="project0classegglib_1_1Filter_1a64369543459c1295032ebc861501adaa"></span>int <code class="descclassname">egglib::Filter::</code><code class="descname">check</code><span class="sig-paren">(</span>int <em>code</em>, bool &amp;<em>flag</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Filter5checkEiRb" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Check data value. </p>
<p>Return: (i) the allelic value if it is exploitable as is; (ii) the main value if the provided value is a synonym for an exploitable value; (iii) MISSINGDATA if it is one of the missing data codes or synonyms; (iv) MISSINGDATA if the value is invalid (and in addition the flag argument is set to true). </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Filter5clearEv">
<span id="egglib::Filter::clear"></span><span class="target" id="project0classegglib_1_1Filter_1ac8bb3912a3ce86b15842e79d0b421204"></span>void <code class="descclassname">egglib::Filter::</code><code class="descname">clear</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Filter5clearEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Reset instance. </p>
<p>The instance is delivered as newly created, and memory is actually released. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Filter14is_exploitableERi">
<span id="egglib::Filter::is_exploitable__iR"></span><span class="target" id="project0classegglib_1_1Filter_1a9d68ce2d52bdb3c05b3c711855762d6a"></span>bool <code class="descclassname">egglib::Filter::</code><code class="descname">is_exploitable</code><span class="sig-paren">(</span>int &amp;<em>code</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Filter14is_exploitableERi" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Check if a data value is exploitable. </p>
<p>If the list of exploitable data is empty, then this method always returns <code class="docutils literal"><span class="pre">true</span></code>. Otherwise, it checks if the value matches one of the value in the exploitable list. If the value does not match any of the exploitable list, but matches one of the synonyms entered using add_exploitable(int, int), then the method returns <code class="docutils literal"><span class="pre">true</span></code> and modifies the passed value (so that the synonym is modified to the reference exploitable code). Otherwise, it never modified the value </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Filter10is_missingERi">
<span id="egglib::Filter::is_missing__iR"></span><span class="target" id="project0classegglib_1_1Filter_1ab4b09526543689f73f4b816b3ee47117"></span>bool <code class="descclassname">egglib::Filter::</code><code class="descname">is_missing</code><span class="sig-paren">(</span>int &amp;<em>code</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Filter10is_missingERi" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Check if a data value is missing. </p>
<p>Returns <code class="docutils literal"><span class="pre">true</span></code> if the code matches one of the missing data codes. Otherwise, returns <code class="docutils literal"><span class="pre">true</span></code> and modifies the value if the code matches one of the synonyms. If also not, return <code class="docutils literal"><span class="pre">false</span></code>. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6FilteraSERK6Filter">
<span id="egglib::Filter::assign-operator__FilterCR"></span><span class="target" id="project0classegglib_1_1Filter_1a9578e3c57c2b36ab7be3ff61a09be5da"></span><a class="reference internal" href="#_CPPv2N6egglib6FilterE" title="egglib::Filter">Filter</a> &amp;<code class="descclassname">egglib::Filter::</code><code class="descname">operator=</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib6FilterE" title="egglib::Filter">Filter</a> &amp;<em>src</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6FilteraSERK6Filter" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Copy assignment operator. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib6Filter7reserveEjjjjj">
<span id="egglib::Filter::reserve__unsigned-i.unsigned-i.unsigned-i.unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1Filter_1af2f73c6b584be578a9fff66e90b76eba"></span>void <code class="descclassname">egglib::Filter::</code><code class="descname">reserve</code><span class="sig-paren">(</span>unsigned int <em>num_expl</em>, unsigned int <em>num_expl_ranges</em>, unsigned int <em>num_missing</em>, unsigned int <em>num_synonyms</em>, unsigned int <em>num_missing_synonyms</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib6Filter7reserveEjjjjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Reserve memory. </p>
<p>The method pre-allocates data arrays in order to speed up subsequent loading operations (up to the numbers passed). The instance is not formally changed by this method, and it is absolutely not required to call this method prior setting valid or missing data codes.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">num_expl</span></code> - <p>expected number of exploitable data codes. </p>
</li>
<li><code class="first docutils literal"><span class="pre">num_expl_ranges</span></code> - <p>expected number of ranges of exploitable data codes. </p>
</li>
<li><code class="first docutils literal"><span class="pre">num_missing</span></code> - <p>expected number of missing data codes. </p>
</li>
<li><code class="first docutils literal"><span class="pre">num_synonyms</span></code> - <p>expected number of synonyms (it is not required that all exploitable codes have a synonym). </p>
</li>
<li><code class="first docutils literal"><span class="pre">num_missing_synonyms</span></code> - <p>expected number of synonym for missing data (it is not required that all missing codes have a synonym). </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="structure">
<h3>Structure<a class="headerlink" href="#structure" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib15StructureHolderE">
<span id="egglib::StructureHolder"></span><span class="target" id="project0classegglib_1_1StructureHolder"></span><em class="property">class </em></dt>
<dd><p>Manage hierarchical group structure. </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder11add_clusterEj">
<span id="egglib::StructureHolder::add_cluster__unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureHolder_1af385bd7feb275f69acec6c633a43acac"></span><a class="reference internal" href="#_CPPv2N6egglib16StructureClusterE" title="egglib::StructureCluster">StructureCluster</a> *<code class="descclassname">egglib::StructureHolder::</code><code class="descname">add_cluster</code><span class="sig-paren">(</span>unsigned int <em>label</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder11add_clusterEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add a cluster with no samples in it. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder22add_individual_ingroupEjP16StructureClusterP19StructurePopulation">
<span id="egglib::StructureHolder::add_individual_ingroup__unsigned-i.StructureClusterP.StructurePopulationP"></span><span class="target" id="project0classegglib_1_1StructureHolder_1aeef54df008a1ba67030281836280b380"></span><a class="reference internal" href="#_CPPv2N6egglib14StructureIndivE" title="egglib::StructureIndiv">StructureIndiv</a> *<code class="descclassname">egglib::StructureHolder::</code><code class="descname">add_individual_ingroup</code><span class="sig-paren">(</span>unsigned int <em>label</em>, <a class="reference internal" href="#_CPPv2N6egglib16StructureClusterE" title="egglib::StructureCluster">StructureCluster</a> *<em>cluster</em>, <a class="reference internal" href="#_CPPv2N6egglib19StructurePopulationE" title="egglib::StructurePopulation">StructurePopulation</a> *<em>population</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder22add_individual_ingroupEjP16StructureClusterP19StructurePopulation" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add an ingroup individual with no samples in it. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder23add_individual_outgroupEj">
<span id="egglib::StructureHolder::add_individual_outgroup__unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a9a796eb42f695255a7fc8276b5b72f35"></span><a class="reference internal" href="#_CPPv2N6egglib14StructureIndivE" title="egglib::StructureIndiv">StructureIndiv</a> *<code class="descclassname">egglib::StructureHolder::</code><code class="descname">add_individual_outgroup</code><span class="sig-paren">(</span>unsigned int <em>label</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder23add_individual_outgroupEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add an outgroup individual with no samples in it. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder14add_pop_filterEj">
<span id="egglib::StructureHolder::add_pop_filter__unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureHolder_1aa86b10dcd8c65980b0ed1df88f0e0de2"></span>void <code class="descclassname">egglib::StructureHolder::</code><code class="descname">add_pop_filter</code><span class="sig-paren">(</span>unsigned int <em>lbl</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder14add_pop_filterEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add a population label to filter. </p>
<p>If at least one is passed, process only those passed. By default, include all populations. Use <a class="reference internal" href="#project0classegglib_1_1StructureHolder_1a8e76f8a1de2a950c90946a4c88b40b03"><span class="std std-ref">reset_filter()</span></a> to reset to default. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder14add_populationEjP16StructureCluster">
<span id="egglib::StructureHolder::add_population__unsigned-i.StructureClusterP"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a5a3e7ed3e271bac90a0bd888f37ea81b"></span><a class="reference internal" href="#_CPPv2N6egglib19StructurePopulationE" title="egglib::StructurePopulation">StructurePopulation</a> *<code class="descclassname">egglib::StructureHolder::</code><code class="descname">add_population</code><span class="sig-paren">(</span>unsigned int <em>label</em>, <a class="reference internal" href="#_CPPv2N6egglib16StructureClusterE" title="egglib::StructureCluster">StructureCluster</a> *<em>cluster</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder14add_populationEjP16StructureCluster" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add a population with no samples in it. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder18add_sample_ingroupEjP16StructureClusterP19StructurePopulationP14StructureIndiv">
<span id="egglib::StructureHolder::add_sample_ingroup__unsigned-i.StructureClusterP.StructurePopulationP.StructureIndivP"></span><span class="target" id="project0classegglib_1_1StructureHolder_1ae5354dc770dde468354a7614d7c995a5"></span>void <code class="descclassname">egglib::StructureHolder::</code><code class="descname">add_sample_ingroup</code><span class="sig-paren">(</span>unsigned int <em>sam_idx</em>, <a class="reference internal" href="#_CPPv2N6egglib16StructureClusterE" title="egglib::StructureCluster">StructureCluster</a> *<em>cluster</em>, <a class="reference internal" href="#_CPPv2N6egglib19StructurePopulationE" title="egglib::StructurePopulation">StructurePopulation</a> *<em>population</em>, <a class="reference internal" href="#_CPPv2N6egglib14StructureIndivE" title="egglib::StructureIndiv">StructureIndiv</a> *<em>indiv</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder18add_sample_ingroupEjP16StructureClusterP19StructurePopulationP14StructureIndiv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add one ingroup sample. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder19add_sample_outgroupEjP14StructureIndiv">
<span id="egglib::StructureHolder::add_sample_outgroup__unsigned-i.StructureIndivP"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a9dc3af9b932cb6e1316b61763d54a45d"></span>void <code class="descclassname">egglib::StructureHolder::</code><code class="descname">add_sample_outgroup</code><span class="sig-paren">(</span>unsigned int <em>sam_idx</em>, <a class="reference internal" href="#_CPPv2N6egglib14StructureIndivE" title="egglib::StructureIndiv">StructureIndiv</a> *<em>indiv</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder19add_sample_outgroupEjP14StructureIndiv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add one outgroup sample. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder12check_ploidyEj">
<span id="egglib::StructureHolder::check_ploidy__unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a3ef5078aab3f727f70c625315105f4ef"></span>void <code class="descclassname">egglib::StructureHolder::</code><code class="descname">check_ploidy</code><span class="sig-paren">(</span>unsigned int <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder12check_ploidyEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Ensure ploidy is consistent and optionally equal to passed value. </p>
<p>Automatically called by <a class="reference internal" href="#project0classegglib_1_1StructureHolder_1a97131940fc7a8f764aaab1057bcc25c9"><span class="std std-ref">get_structure()</span></a>. Need to be called if <a class="reference internal" href="#project0classegglib_1_1StructureHolder_1a685065741b9079baf591de590606ac4c"><span class="std std-ref">process_ingroup()</span></a> and/or <a class="reference internal" href="#project0classegglib_1_1StructureHolder_1add72d2128ab1ae9f5a5f0942f9616ab3"><span class="std std-ref">process_outgroup()</span></a> is used.</p>
<p>Value must be &gt;0. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder4copyERK15StructureHolder">
<span id="egglib::StructureHolder::copy__StructureHolderCR"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a549101b0ed3ab9febffd64b468f325d0"></span>void <code class="descclassname">egglib::StructureHolder::</code><code class="descname">copy</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib15StructureHolderE" title="egglib::StructureHolder">StructureHolder</a> &amp;<em>source</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder4copyERK15StructureHolder" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Copy data frome source object. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder11get_clusterEj">
<span id="egglib::StructureHolder::get_cluster__unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureHolder_1ae637e3ef05a98b44f8f37bf03d2786d5"></span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib16StructureClusterE" title="egglib::StructureCluster">StructureCluster</a> &amp;<code class="descclassname">egglib::StructureHolder::</code><code class="descname">get_cluster</code><span class="sig-paren">(</span>unsigned int <em>idx</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder11get_clusterEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get a cluster. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder17get_indiv_ingroupEj">
<span id="egglib::StructureHolder::get_indiv_ingroup__unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a593497ec4a7b97973b8324de7de806f3"></span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib14StructureIndivE" title="egglib::StructureIndiv">StructureIndiv</a> &amp;<code class="descclassname">egglib::StructureHolder::</code><code class="descname">get_indiv_ingroup</code><span class="sig-paren">(</span>unsigned int <em>idx</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder17get_indiv_ingroupEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an ingroup individual. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder18get_indiv_outgroupEj">
<span id="egglib::StructureHolder::get_indiv_outgroup__unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a2092279491b9722ba53ad35322b3eead"></span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib14StructureIndivE" title="egglib::StructureIndiv">StructureIndiv</a> &amp;<code class="descclassname">egglib::StructureHolder::</code><code class="descname">get_indiv_outgroup</code><span class="sig-paren">(</span>unsigned int <em>idx</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder18get_indiv_outgroupEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an outgroup individual. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder6get_noEv">
<span id="egglib::StructureHolder::get_no"></span><span class="target" id="project0classegglib_1_1StructureHolder_1aa9848e2e31dc542d47a2865ed0b7493f"></span>unsigned int <code class="descclassname">egglib::StructureHolder::</code><code class="descname">get_no</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder6get_noEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get number of outgroup samples. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder10get_no_reqEv">
<span id="egglib::StructureHolder::get_no_req"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a8fa1d65084027c7bfb1b55b80fff4d96"></span>unsigned int <code class="descclassname">egglib::StructureHolder::</code><code class="descname">get_no_req</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder10get_no_reqEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get required number of outgroup samples. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder6get_nsEv">
<span id="egglib::StructureHolder::get_ns"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a24580f8f2eb415030ad4e25fbc94cf87"></span>unsigned int <code class="descclassname">egglib::StructureHolder::</code><code class="descname">get_ns</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder6get_nsEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get number of ingroup samples. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder10get_ns_reqEv">
<span id="egglib::StructureHolder::get_ns_req"></span><span class="target" id="project0classegglib_1_1StructureHolder_1aa9f5945b55b34d5951440bb24e848696"></span>unsigned int <code class="descclassname">egglib::StructureHolder::</code><code class="descname">get_ns_req</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder10get_ns_reqEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get required number of ingroup samples. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder10get_ploidyEv">
<span id="egglib::StructureHolder::get_ploidy"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a93794a39636914f681ec33ff62e93c3e"></span>unsigned int <code class="descclassname">egglib::StructureHolder::</code><code class="descname">get_ploidy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder10get_ploidyEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get ploidy. </p>
<p>Default is UNKNOWN. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder13get_pop_indexEj">
<span id="egglib::StructureHolder::get_pop_index__unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a3c3591f1d6be78ff5d017dcf3b1eb59a"></span>unsigned int <code class="descclassname">egglib::StructureHolder::</code><code class="descname">get_pop_index</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder13get_pop_indexEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Index of the population containing this sample (default: MISSING). </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder14get_populationEj">
<span id="egglib::StructureHolder::get_population__unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a4e372a90170f514afcacfc4cf684860c"></span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib19StructurePopulationE" title="egglib::StructurePopulation">StructurePopulation</a> &amp;<code class="descclassname">egglib::StructureHolder::</code><code class="descname">get_population</code><span class="sig-paren">(</span>unsigned int <em>idx</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder14get_populationEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get a population. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder13get_structureER10DataHolderjjjjb">
<span id="egglib::StructureHolder::get_structure__DataHolderR.unsigned-i.unsigned-i.unsigned-i.unsigned-i.b"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a97131940fc7a8f764aaab1057bcc25c9"></span>void <code class="descclassname">egglib::StructureHolder::</code><code class="descname">get_structure</code><span class="sig-paren">(</span><a class="reference internal" href="core.html#_CPPv2N6egglib10DataHolderE" title="egglib::DataHolder">DataHolder</a> &amp;<em>data</em>, unsigned int <em>lvl_clust</em>, unsigned int <em>lvl_pop</em>, unsigned int <em>lvl_indiv</em>, unsigned int <em>ploidy</em>, bool <em>skip_outgroup</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder13get_structureER10DataHolderjjjjb" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Process labels from a <a class="reference internal" href="core.html#project0classegglib_1_1DataHolder"><span class="std std-ref">DataHolder</span></a>. </p>
<p>Use UNKNOWN for any level to skip (but skipping individuals is not the same as skipping clusters/pops). You must set pop filter separately. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder9num_clustEv">
<span id="egglib::StructureHolder::num_clust"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a759e6d0b43824e7c845254f54453a89f"></span>unsigned int <code class="descclassname">egglib::StructureHolder::</code><code class="descname">num_clust</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder9num_clustEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of clusters. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder17num_indiv_ingroupEv">
<span id="egglib::StructureHolder::num_indiv_ingroup"></span><span class="target" id="project0classegglib_1_1StructureHolder_1adac7ee577c49ee575dd8d679495dd2b0"></span>unsigned int <code class="descclassname">egglib::StructureHolder::</code><code class="descname">num_indiv_ingroup</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder17num_indiv_ingroupEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of ingroup individuals (total). </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder18num_indiv_outgroupEv">
<span id="egglib::StructureHolder::num_indiv_outgroup"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a7f0acfddf56be666492125d47835481a"></span>unsigned int <code class="descclassname">egglib::StructureHolder::</code><code class="descname">num_indiv_outgroup</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder18num_indiv_outgroupEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of outgroup individuals. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder7num_popEv">
<span id="egglib::StructureHolder::num_pop"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a60bd848ce0d1789c9485a2b5ac6ec2bf"></span>unsigned int <code class="descclassname">egglib::StructureHolder::</code><code class="descname">num_pop</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder7num_popEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of populations (total). </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder15process_ingroupEjjjj">
<span id="egglib::StructureHolder::process_ingroup__unsigned-i.unsigned-i.unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a685065741b9079baf591de590606ac4c"></span>void <code class="descclassname">egglib::StructureHolder::</code><code class="descname">process_ingroup</code><span class="sig-paren">(</span>unsigned int <em>idx</em>, unsigned int <em>lbl_clust</em>, unsigned int <em>lbl_pop</em>, unsigned int <em>lbl_indiv</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder15process_ingroupEjjjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Process one ingroup sample. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder16process_outgroupEjj">
<span id="egglib::StructureHolder::process_outgroup__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureHolder_1add72d2128ab1ae9f5a5f0942f9616ab3"></span>void <code class="descclassname">egglib::StructureHolder::</code><code class="descname">process_outgroup</code><span class="sig-paren">(</span>unsigned int <em>idx</em>, unsigned int <em>lbl_indiv</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder16process_outgroupEjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Process one outgroup sample. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder14reserve_filterEj">
<span id="egglib::StructureHolder::reserve_filter__unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a391a227ea13870869c7cafd476595c4d"></span>void <code class="descclassname">egglib::StructureHolder::</code><code class="descname">reserve_filter</code><span class="sig-paren">(</span>unsigned int <em>howmany</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder14reserve_filterEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Pre-alloc filter table. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder5resetEv">
<span id="egglib::StructureHolder::reset"></span><span class="target" id="project0classegglib_1_1StructureHolder_1ad20897c5c8bd47f5d4005989bead0e55"></span>void <code class="descclassname">egglib::StructureHolder::</code><code class="descname">reset</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder5resetEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Reset to defaults. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib15StructureHolder12reset_filterEv">
<span id="egglib::StructureHolder::reset_filter"></span><span class="target" id="project0classegglib_1_1StructureHolder_1a8e76f8a1de2a950c90946a4c88b40b03"></span>void <code class="descclassname">egglib::StructureHolder::</code><code class="descname">reset_filter</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib15StructureHolder12reset_filterEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Reset pop filter only. </p>
</dd></dl>

</div>
</dd></dl>

<dl class="class">
<dt id="_CPPv2N6egglib16StructureClusterE">
<span id="egglib::StructureCluster"></span><span class="target" id="project0classegglib_1_1StructureCluster"></span><em class="property">class </em></dt>
<dd><p>Manage a cluster. </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib16StructureCluster9add_indivEP19StructurePopulationj">
<span id="egglib::StructureCluster::add_indiv__StructurePopulationP.unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureCluster_1a2658d71b665b6f74719d310c1f44e40d"></span><a class="reference internal" href="#_CPPv2N6egglib14StructureIndivE" title="egglib::StructureIndiv">StructureIndiv</a> *<code class="descclassname">egglib::StructureCluster::</code><code class="descname">add_indiv</code><span class="sig-paren">(</span><a class="reference internal" href="#_CPPv2N6egglib19StructurePopulationE" title="egglib::StructurePopulation">StructurePopulation</a> *<em>pop</em>, unsigned int <em>label</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16StructureCluster9add_indivEP19StructurePopulationj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add and create an individual. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16StructureCluster7add_popEj">
<span id="egglib::StructureCluster::add_pop__unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureCluster_1a9604cbd50a5d16572e3396db0d93ac71"></span><a class="reference internal" href="#_CPPv2N6egglib19StructurePopulationE" title="egglib::StructurePopulation">StructurePopulation</a> *<code class="descclassname">egglib::StructureCluster::</code><code class="descname">add_pop</code><span class="sig-paren">(</span>unsigned int <em>label</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16StructureCluster7add_popEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add and create a population. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16StructureCluster10add_sampleEv">
<span id="egglib::StructureCluster::add_sample"></span><span class="target" id="project0classegglib_1_1StructureCluster_1a6c105b90e5d632d21d69a77e0bbe02a1"></span>void <code class="descclassname">egglib::StructureCluster::</code><code class="descname">add_sample</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16StructureCluster10add_sampleEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add a sample. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16StructureCluster9get_indivEj">
<span id="egglib::StructureCluster::get_indiv__unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureCluster_1a0a443318cf88288844a64fc4b790c089"></span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib14StructureIndivE" title="egglib::StructureIndiv">StructureIndiv</a> &amp;<code class="descclassname">egglib::StructureCluster::</code><code class="descname">get_indiv</code><span class="sig-paren">(</span>unsigned int <em>idx</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16StructureCluster9get_indivEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an individual. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16StructureCluster9get_labelEv">
<span id="egglib::StructureCluster::get_label"></span><span class="target" id="project0classegglib_1_1StructureCluster_1ae9e590b88847ed3a4e714f87ee0976b6"></span>unsigned int <code class="descclassname">egglib::StructureCluster::</code><code class="descname">get_label</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16StructureCluster9get_labelEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get label. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16StructureCluster14get_populationEj">
<span id="egglib::StructureCluster::get_population__unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureCluster_1a4e372a90170f514afcacfc4cf684860c"></span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib19StructurePopulationE" title="egglib::StructurePopulation">StructurePopulation</a> &amp;<code class="descclassname">egglib::StructureCluster::</code><code class="descname">get_population</code><span class="sig-paren">(</span>unsigned int <em>idx</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16StructureCluster14get_populationEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get a population. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16StructureCluster9num_indivEv">
<span id="egglib::StructureCluster::num_indiv"></span><span class="target" id="project0classegglib_1_1StructureCluster_1a69ffd0fc203827dc6a8ab3a2190be461"></span>unsigned int <code class="descclassname">egglib::StructureCluster::</code><code class="descname">num_indiv</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16StructureCluster9num_indivEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of individuals (total for this cluster). </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16StructureCluster7num_popEv">
<span id="egglib::StructureCluster::num_pop"></span><span class="target" id="project0classegglib_1_1StructureCluster_1a60bd848ce0d1789c9485a2b5ac6ec2bf"></span>unsigned int <code class="descclassname">egglib::StructureCluster::</code><code class="descname">num_pop</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16StructureCluster7num_popEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of populations. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib16StructureCluster5resetEP15StructureHolderj">
<span id="egglib::StructureCluster::reset__StructureHolderP.unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureCluster_1ac9f53646ad51324aeb98e11bb42a4202"></span>void <code class="descclassname">egglib::StructureCluster::</code><code class="descname">reset</code><span class="sig-paren">(</span><a class="reference internal" href="#_CPPv2N6egglib15StructureHolderE" title="egglib::StructureHolder">StructureHolder</a> *<em>holder</em>, unsigned int <em>label</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib16StructureCluster5resetEP15StructureHolderj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Restore defaults. </p>
</dd></dl>

</div>
</dd></dl>

<dl class="class">
<dt id="_CPPv2N6egglib19StructurePopulationE">
<span id="egglib::StructurePopulation"></span><span class="target" id="project0classegglib_1_1StructurePopulation"></span><em class="property">class </em></dt>
<dd><p>Manage a population. </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib19StructurePopulation9add_indivEj">
<span id="egglib::StructurePopulation::add_indiv__unsigned-i"></span><span class="target" id="project0classegglib_1_1StructurePopulation_1a36c5d76014f1566bf733d175806c2b27"></span><a class="reference internal" href="#_CPPv2N6egglib14StructureIndivE" title="egglib::StructureIndiv">StructureIndiv</a> *<code class="descclassname">egglib::StructurePopulation::</code><code class="descname">add_indiv</code><span class="sig-paren">(</span>unsigned int <em>label</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib19StructurePopulation9add_indivEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add and create an individual. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib19StructurePopulation10add_sampleEv">
<span id="egglib::StructurePopulation::add_sample"></span><span class="target" id="project0classegglib_1_1StructurePopulation_1a6c105b90e5d632d21d69a77e0bbe02a1"></span>void <code class="descclassname">egglib::StructurePopulation::</code><code class="descname">add_sample</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib19StructurePopulation10add_sampleEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add a sample. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib19StructurePopulation11get_clusterEv">
<span id="egglib::StructurePopulation::get_cluster"></span><span class="target" id="project0classegglib_1_1StructurePopulation_1ad4c28dea84fb3c365c80a4fd26ea5664"></span><a class="reference internal" href="#_CPPv2N6egglib16StructureClusterE" title="egglib::StructureCluster">StructureCluster</a> *<code class="descclassname">egglib::StructurePopulation::</code><code class="descname">get_cluster</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib19StructurePopulation11get_clusterEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Get containing cluster. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib19StructurePopulation9get_indivEj">
<span id="egglib::StructurePopulation::get_indiv__unsigned-i"></span><span class="target" id="project0classegglib_1_1StructurePopulation_1a0a443318cf88288844a64fc4b790c089"></span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib14StructureIndivE" title="egglib::StructureIndiv">StructureIndiv</a> &amp;<code class="descclassname">egglib::StructurePopulation::</code><code class="descname">get_indiv</code><span class="sig-paren">(</span>unsigned int <em>idx</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib19StructurePopulation9get_indivEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an individual. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib19StructurePopulation9get_labelEv">
<span id="egglib::StructurePopulation::get_label"></span><span class="target" id="project0classegglib_1_1StructurePopulation_1ae9e590b88847ed3a4e714f87ee0976b6"></span>unsigned int <code class="descclassname">egglib::StructurePopulation::</code><code class="descname">get_label</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib19StructurePopulation9get_labelEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get label. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib19StructurePopulation9num_indivEv">
<span id="egglib::StructurePopulation::num_indiv"></span><span class="target" id="project0classegglib_1_1StructurePopulation_1a69ffd0fc203827dc6a8ab3a2190be461"></span>unsigned int <code class="descclassname">egglib::StructurePopulation::</code><code class="descname">num_indiv</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib19StructurePopulation9num_indivEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of individuals. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib19StructurePopulation5resetEP15StructureHolderP16StructureClusterj">
<span id="egglib::StructurePopulation::reset__StructureHolderP.StructureClusterP.unsigned-i"></span><span class="target" id="project0classegglib_1_1StructurePopulation_1a83b684c4125e5aab03b6e06aec91e66e"></span>void <code class="descclassname">egglib::StructurePopulation::</code><code class="descname">reset</code><span class="sig-paren">(</span><a class="reference internal" href="#_CPPv2N6egglib15StructureHolderE" title="egglib::StructureHolder">StructureHolder</a> *<em>holder</em>, <a class="reference internal" href="#_CPPv2N6egglib16StructureClusterE" title="egglib::StructureCluster">StructureCluster</a> *<em>cluster</em>, unsigned int <em>label</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib19StructurePopulation5resetEP15StructureHolderP16StructureClusterj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Restore defaults. </p>
</dd></dl>

</div>
</dd></dl>

<dl class="class">
<dt id="_CPPv2N6egglib14StructureIndivE">
<span id="egglib::StructureIndiv"></span><span class="target" id="project0classegglib_1_1StructureIndiv"></span><em class="property">class </em></dt>
<dd><p>Manage an individual. </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib14StructureIndiv10add_sampleEj">
<span id="egglib::StructureIndiv::add_sample__unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureIndiv_1ab1afa788538c1b602c11aa5c07355923"></span>void <code class="descclassname">egglib::StructureIndiv::</code><code class="descname">add_sample</code><span class="sig-paren">(</span>unsigned int <em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib14StructureIndiv10add_sampleEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add a sample. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib14StructureIndiv11get_clusterEv">
<span id="egglib::StructureIndiv::get_cluster"></span><span class="target" id="project0classegglib_1_1StructureIndiv_1ad4c28dea84fb3c365c80a4fd26ea5664"></span><a class="reference internal" href="#_CPPv2N6egglib16StructureClusterE" title="egglib::StructureCluster">StructureCluster</a> *<code class="descclassname">egglib::StructureIndiv::</code><code class="descname">get_cluster</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib14StructureIndiv11get_clusterEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Get containing cluster (NULL if outgroup). </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib14StructureIndiv9get_labelEv">
<span id="egglib::StructureIndiv::get_label"></span><span class="target" id="project0classegglib_1_1StructureIndiv_1ae9e590b88847ed3a4e714f87ee0976b6"></span>unsigned int <code class="descclassname">egglib::StructureIndiv::</code><code class="descname">get_label</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib14StructureIndiv9get_labelEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get label. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib14StructureIndiv14get_populationEv">
<span id="egglib::StructureIndiv::get_population"></span><span class="target" id="project0classegglib_1_1StructureIndiv_1ac98264ef8fc53d5d8ecd479cc54628b3"></span><a class="reference internal" href="#_CPPv2N6egglib19StructurePopulationE" title="egglib::StructurePopulation">StructurePopulation</a> *<code class="descclassname">egglib::StructureIndiv::</code><code class="descname">get_population</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib14StructureIndiv14get_populationEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Get containing population (NULL if outgroup). </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib14StructureIndiv10get_sampleEj">
<span id="egglib::StructureIndiv::get_sample__unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureIndiv_1a7362c06433f669987631d9085a487cb8"></span>unsigned int <code class="descclassname">egglib::StructureIndiv::</code><code class="descname">get_sample</code><span class="sig-paren">(</span>unsigned int <em>idx</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib14StructureIndiv10get_sampleEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get a sample. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib14StructureIndiv11num_samplesEv">
<span id="egglib::StructureIndiv::num_samples"></span><span class="target" id="project0classegglib_1_1StructureIndiv_1aba32ed97991ec044c35d2692148f02d4"></span>unsigned int <code class="descclassname">egglib::StructureIndiv::</code><code class="descname">num_samples</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib14StructureIndiv11num_samplesEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of samples. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib14StructureIndiv5resetEP15StructureHolderP16StructureClusterP19StructurePopulationj">
<span id="egglib::StructureIndiv::reset__StructureHolderP.StructureClusterP.StructurePopulationP.unsigned-i"></span><span class="target" id="project0classegglib_1_1StructureIndiv_1aa4a96cfc9d40641dbe9913c30c1c03c8"></span>void <code class="descclassname">egglib::StructureIndiv::</code><code class="descname">reset</code><span class="sig-paren">(</span><a class="reference internal" href="#_CPPv2N6egglib15StructureHolderE" title="egglib::StructureHolder">StructureHolder</a> *<em>holder</em>, <a class="reference internal" href="#_CPPv2N6egglib16StructureClusterE" title="egglib::StructureCluster">StructureCluster</a> *<em>cluster</em>, <a class="reference internal" href="#_CPPv2N6egglib19StructurePopulationE" title="egglib::StructurePopulation">StructurePopulation</a> *<em>population</em>, unsigned int <em>label</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib14StructureIndiv5resetEP15StructureHolderP16StructureClusterP19StructurePopulationj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Restore defaults. </p>
</dd></dl>

</div>
</dd></dl>

</div>
</div>
<div class="section" id="classes-analysing-frequencies">
<h2>Classes analysing frequencies<a class="headerlink" href="#classes-analysing-frequencies" title="Permalink to this headline">¶</a></h2>
<div class="section" id="freqholder">
<h3>FreqHolder<a class="headerlink" href="#freqholder" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib10FreqHolderE">
<span id="egglib::FreqHolder"></span><span class="target" id="project0classegglib_1_1FreqHolder"></span><em class="property">class </em></dt>
<dd><p>Class holding frequencies for all compartments for a site. </p>
<p>Possible uses of this class:</p>
<p><ul class="simple">
<li>Process a site with structure stored in a <a class="reference internal" href="#project0classegglib_1_1StructureHolder"><span class="std std-ref">StructureHolder</span></a>:<ul>
<li>setup_structure(structure, ploidy, flag) and keep structure available</li>
<li><a class="reference internal" href="#project0classegglib_1_1FreqHolder_1a1388b48adb83f4c15e5081ab021f0a49"><span class="std std-ref">process_site()</span></a></li>
</ul>
</li>
<li>Process a site without structure or with a manual structure:<ul>
<li>setup_raw(nc, np, no, ploidy, flag)</li>
<li>setup_pop(i, cluster, relative, ns) for all populations</li>
<li><a class="reference internal" href="#project0classegglib_1_1FreqHolder_1a1388b48adb83f4c15e5081ab021f0a49"><span class="std std-ref">process_site()</span></a> assuming all individuals are consecutive</li>
</ul>
</li>
<li>Enter frequencies manually:<ul>
<li>setup_raw(nc, np, no, ploidy, flag)</li>
<li>setup_pop(i, cluster, relative, ns) for all populations</li>
<li>set_nall(na, ng)</li>
<li><a class="reference internal" href="#project0classegglib_1_1FreqSet_1a8ae7f1adf9509a43d90e562f25bbd058"><span class="std std-ref">FreqSet.incr_allele()</span></a></li>
<li>[FreqSet.ncr_genotype()]</li>
<li>[<a class="reference internal" href="#project0classegglib_1_1FreqSet_1a2e29029abae0b1e5dad34a6ca7fce695"><span class="std std-ref">FreqSet.tell_het()</span></a>]</li>
</ul>
</li>
<li>Process data from a VCF parser:<ul>
<li><a class="reference internal" href="#project0classegglib_1_1FreqHolder_1a0eb3dd7497097429d027116613f22a19"><span class="std std-ref">process_vcf()</span></a> </li>
</ul>
</li>
</ul>
</p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder10FreqHolderEv">
<span id="egglib::FreqHolder::FreqHolder"></span><span class="target" id="project0classegglib_1_1FreqHolder_1a1d033fedd97d40758e482b0485a4b03b"></span><code class="descclassname">egglib::FreqHolder::</code><code class="descname">FreqHolder</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder10FreqHolderEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolderD0Ev">
<span id="egglib::FreqHolder::~FreqHolder"></span><span class="target" id="project0classegglib_1_1FreqHolder_1add122b43577dad01a6f90615ed004345"></span><code class="descclassname">egglib::FreqHolder::</code><code class="descname">~FreqHolder</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolderD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder6alleleEj">
<span id="egglib::FreqHolder::allele__unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqHolder_1ae5bbe34d3b96745d5e7c681ad2f01459"></span>int <code class="descclassname">egglib::FreqHolder::</code><code class="descname">allele</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder6alleleEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an allele value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder13cluster_indexEj">
<span id="egglib::FreqHolder::cluster_index__unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqHolder_1aa35caa767c423a4c236d583a01ee6766"></span>unsigned int <code class="descclassname">egglib::FreqHolder::</code><code class="descname">cluster_index</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder13cluster_indexEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the index of the cluster of a given population. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder11frq_clusterEj">
<span id="egglib::FreqHolder::frq_cluster__unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqHolder_1aa10a53aba426600f1eb4164ae4f10515"></span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib7FreqSetE" title="egglib::FreqSet">FreqSet</a> &amp;<code class="descclassname">egglib::FreqHolder::</code><code class="descname">frq_cluster</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder11frq_clusterEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get frequencies in a cluster. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder11frq_ingroupEv">
<span id="egglib::FreqHolder::frq_ingroup"></span><span class="target" id="project0classegglib_1_1FreqHolder_1af39c5b1786a582255c38b7af03e9b7ce"></span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib7FreqSetE" title="egglib::FreqSet">FreqSet</a> &amp;<code class="descclassname">egglib::FreqHolder::</code><code class="descname">frq_ingroup</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder11frq_ingroupEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get frequencies in whole ingroup. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder12frq_outgroupEv">
<span id="egglib::FreqHolder::frq_outgroup"></span><span class="target" id="project0classegglib_1_1FreqHolder_1a0860491b6cb6ef5324a57a39cde798a8"></span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib7FreqSetE" title="egglib::FreqSet">FreqSet</a> &amp;<code class="descclassname">egglib::FreqHolder::</code><code class="descname">frq_outgroup</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder12frq_outgroupEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get frequencies in outgroup. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder14frq_populationEj">
<span id="egglib::FreqHolder::frq_population__unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqHolder_1a7ed875c2246c44d057c65de2c65bf86a"></span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib7FreqSetE" title="egglib::FreqSet">FreqSet</a> &amp;<code class="descclassname">egglib::FreqHolder::</code><code class="descname">frq_population</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder14frq_populationEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get frequencies in a population. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder8genotypeEj">
<span id="egglib::FreqHolder::genotype__unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqHolder_1af4efcf3fd9675f892f408e9b70d5b7e0"></span><em class="property">const</em> unsigned int *<code class="descclassname">egglib::FreqHolder::</code><code class="descname">genotype</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder8genotypeEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get a genotype (as array of allele indexes) (none if haploid) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder12genotype_hetEj">
<span id="egglib::FreqHolder::genotype_het__unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqHolder_1a6c805d14cbe003e279df937fbc8bd772"></span>bool <code class="descclassname">egglib::FreqHolder::</code><code class="descname">genotype_het</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder12genotype_hetEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>True if genotype is heterozygote. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder13genotype_itemEjj">
<span id="egglib::FreqHolder::genotype_item__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqHolder_1aad2e722524db07c8614adcf75352d2d2"></span>unsigned int <code class="descclassname">egglib::FreqHolder::</code><code class="descname">genotype_item</code><span class="sig-paren">(</span>unsigned int, unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder13genotype_itemEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get part of a genotype. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder11num_allelesEv">
<span id="egglib::FreqHolder::num_alleles"></span><span class="target" id="project0classegglib_1_1FreqHolder_1a5434000611eea3568aec6698fec6fec7"></span>unsigned int <code class="descclassname">egglib::FreqHolder::</code><code class="descname">num_alleles</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder11num_allelesEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of alleles. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder12num_clustersEv">
<span id="egglib::FreqHolder::num_clusters"></span><span class="target" id="project0classegglib_1_1FreqHolder_1a700f8cae4125577ca237dc669142316e"></span>unsigned int <code class="descclassname">egglib::FreqHolder::</code><code class="descname">num_clusters</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder12num_clustersEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get number of clusters. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder13num_genotypesEv">
<span id="egglib::FreqHolder::num_genotypes"></span><span class="target" id="project0classegglib_1_1FreqHolder_1a17cd0f7d2263c544af4cb353f75584c7"></span>unsigned int <code class="descclassname">egglib::FreqHolder::</code><code class="descname">num_genotypes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder13num_genotypesEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of genotypes with non-null frequency (0 if haploid) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder15num_populationsEv">
<span id="egglib::FreqHolder::num_populations"></span><span class="target" id="project0classegglib_1_1FreqHolder_1a64faf8fea8af3c9d9f0c3490c42daf3b"></span>unsigned int <code class="descclassname">egglib::FreqHolder::</code><code class="descname">num_populations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder15num_populationsEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get number of populations. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder6ploidyEv">
<span id="egglib::FreqHolder::ploidy"></span><span class="target" id="project0classegglib_1_1FreqHolder_1a526257f81406097ffa795e33ec58938b"></span>unsigned int <code class="descclassname">egglib::FreqHolder::</code><code class="descname">ploidy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder6ploidyEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Ploidy. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder12process_siteERK10SiteHolder">
<span id="egglib::FreqHolder::process_site__SiteHolderCR"></span><span class="target" id="project0classegglib_1_1FreqHolder_1a1388b48adb83f4c15e5081ab021f0a49"></span>void <code class="descclassname">egglib::FreqHolder::</code><code class="descname">process_site</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder12process_siteERK10SiteHolder" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Compute frequencies (structure cannot have individual level) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder11process_vcfERK9VcfParser">
<span id="egglib::FreqHolder::process_vcf__VcfParserCR"></span><span class="target" id="project0classegglib_1_1FreqHolder_1a0eb3dd7497097429d027116613f22a19"></span>void <code class="descclassname">egglib::FreqHolder::</code><code class="descname">process_vcf</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="datafiles.html#_CPPv2N6egglib9VcfParserE" title="egglib::VcfParser">VcfParser</a> &amp;<em>vcf</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder11process_vcfERK9VcfParser" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Get frequencies from VCF. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder17set_genotype_itemEjjj">
<span id="egglib::FreqHolder::set_genotype_item__unsigned-i.unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqHolder_1a159c6eaab9dd7f44260befe034abf76b"></span>void <code class="descclassname">egglib::FreqHolder::</code><code class="descname">set_genotype_item</code><span class="sig-paren">(</span>unsigned int <em>i</em>, unsigned int <em>j</em>, unsigned int <em>a</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder17set_genotype_itemEjjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set part of a genotype. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder8set_nallEjj">
<span id="egglib::FreqHolder::set_nall__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqHolder_1a21a8c57944a70f3af5f365465d3278b6"></span>void <code class="descclassname">egglib::FreqHolder::</code><code class="descname">set_nall</code><span class="sig-paren">(</span>unsigned int <em>na</em>, unsigned int <em>ng</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder8set_nallEjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Before loading frequencies manually. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder9setup_popEjjjj">
<span id="egglib::FreqHolder::setup_pop__unsigned-i.unsigned-i.unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqHolder_1a4f6a351a425193e31b10757f5357c50d"></span>void <code class="descclassname">egglib::FreqHolder::</code><code class="descname">setup_pop</code><span class="sig-paren">(</span>unsigned int <em>i</em>, unsigned int <em>clu_idx</em>, unsigned int <em>rel_idx</em>, unsigned int <em>ns</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder9setup_popEjjjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Follows <a class="reference internal" href="#project0classegglib_1_1FreqHolder_1a5cd4adcd4b23755f7d1d00dff0e95422"><span class="std std-ref">setup_raw()</span></a> (for all pops) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder9setup_rawEjjjj">
<span id="egglib::FreqHolder::setup_raw__unsigned-i.unsigned-i.unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqHolder_1a5cd4adcd4b23755f7d1d00dff0e95422"></span>void <code class="descclassname">egglib::FreqHolder::</code><code class="descname">setup_raw</code><span class="sig-paren">(</span>unsigned int <em>nc</em>, unsigned int <em>np</em>, unsigned int <em>no</em>, unsigned int <em>ploidy</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder9setup_rawEjjjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Setup manual structure. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10FreqHolder15setup_structureEPK15StructureHolderj">
<span id="egglib::FreqHolder::setup_structure__StructureHolderCP.unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqHolder_1a1724dcf2cfe7dd637d803dfe9d911c74"></span>void <code class="descclassname">egglib::FreqHolder::</code><code class="descname">setup_structure</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib15StructureHolderE" title="egglib::StructureHolder">StructureHolder</a> *<em>structure</em>, unsigned int <em>ploidy</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10FreqHolder15setup_structureEPK15StructureHolderj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set up based on provided structure (no individual level) </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="freqset">
<h3>FreqSet<a class="headerlink" href="#freqset" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib7FreqSetE">
<span id="egglib::FreqSet"></span><span class="target" id="project0classegglib_1_1FreqSet"></span><em class="property">class </em></dt>
<dd><p>Class holding frequencies in a given compartment. </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib7FreqSet7FreqSetEv">
<span id="egglib::FreqSet::FreqSet"></span><span class="target" id="project0classegglib_1_1FreqSet_1ad8b841630a18a018dd4894394488a67d"></span><code class="descclassname">egglib::FreqSet::</code><code class="descname">FreqSet</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib7FreqSet7FreqSetEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor (all empty) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib7FreqSetD0Ev">
<span id="egglib::FreqSet::~FreqSet"></span><span class="target" id="project0classegglib_1_1FreqSet_1a4b72ac0f4333d5d71aed5f0b9ca5fa80"></span><code class="descclassname">egglib::FreqSet::</code><code class="descname">~FreqSet</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib7FreqSetD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib7FreqSet13add_genotypesEj">
<span id="egglib::FreqSet::add_genotypes__unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqSet_1ac92cc3b811325f5bb4f80b298bc99683"></span>void <code class="descclassname">egglib::FreqSet::</code><code class="descname">add_genotypes</code><span class="sig-paren">(</span>unsigned int <em>num</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib7FreqSet13add_genotypesEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add genotypes. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib7FreqSet7frq_allEj">
<span id="egglib::FreqSet::frq_all__unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqSet_1af89f32553df18e1be44460dd4e552151"></span>unsigned int <code class="descclassname">egglib::FreqSet::</code><code class="descname">frq_all</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib7FreqSet7frq_allEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an allele frequency. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib7FreqSet7frq_genEj">
<span id="egglib::FreqSet::frq_gen__unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqSet_1a2b7cb1cb1aef7379dbb72f27c83a3341"></span>unsigned int <code class="descclassname">egglib::FreqSet::</code><code class="descname">frq_gen</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib7FreqSet7frq_genEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an genotype frequency. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib7FreqSet7frq_hetEj">
<span id="egglib::FreqSet::frq_het__unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqSet_1a56225bf0e70fe1f95dd9ae9df564c4e9"></span>unsigned int <code class="descclassname">egglib::FreqSet::</code><code class="descname">frq_het</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib7FreqSet7frq_hetEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Frequency of heterozygotes have &gt;= 1 copies of allele. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib7FreqSet11incr_alleleEjj">
<span id="egglib::FreqSet::incr_allele__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqSet_1a8ae7f1adf9509a43d90e562f25bbd058"></span>void <code class="descclassname">egglib::FreqSet::</code><code class="descname">incr_allele</code><span class="sig-paren">(</span>unsigned int <em>all_idx</em>, unsigned int <em>num</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib7FreqSet11incr_alleleEjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Increment frequency of a given allele. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib7FreqSet13incr_genotypeEjj">
<span id="egglib::FreqSet::incr_genotype__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqSet_1a6386735a4231b64b0336fcc67df032f2"></span>void <code class="descclassname">egglib::FreqSet::</code><code class="descname">incr_genotype</code><span class="sig-paren">(</span>unsigned int <em>gen_idx</em>, unsigned int <em>num</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib7FreqSet13incr_genotypeEjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Increment frequency of a given genotype. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib7FreqSet5nieffEv">
<span id="egglib::FreqSet::nieff"></span><span class="target" id="project0classegglib_1_1FreqSet_1ada84377bcff2bdc38aaf88c5533f3ac7"></span>unsigned int <code class="descclassname">egglib::FreqSet::</code><code class="descname">nieff</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib7FreqSet5nieffEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Total frequency (number of individuals) (0 if haploid) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib7FreqSet5nseffEv">
<span id="egglib::FreqSet::nseff"></span><span class="target" id="project0classegglib_1_1FreqSet_1a291e81cff28e32c3a25ff41996260784"></span>unsigned int <code class="descclassname">egglib::FreqSet::</code><code class="descname">nseff</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib7FreqSet5nseffEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Total frequency (number of samples) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib7FreqSet11num_allelesEv">
<span id="egglib::FreqSet::num_alleles"></span><span class="target" id="project0classegglib_1_1FreqSet_1a5434000611eea3568aec6698fec6fec7"></span>unsigned int <code class="descclassname">egglib::FreqSet::</code><code class="descname">num_alleles</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib7FreqSet11num_allelesEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of alleles (equal to user-provided value) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib7FreqSet15num_alleles_effEv">
<span id="egglib::FreqSet::num_alleles_eff"></span><span class="target" id="project0classegglib_1_1FreqSet_1a343742ed329d0db14679b5054190b175"></span>unsigned int <code class="descclassname">egglib::FreqSet::</code><code class="descname">num_alleles_eff</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib7FreqSet15num_alleles_effEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of alleles with non-null frequency. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib7FreqSet13num_genotypesEv">
<span id="egglib::FreqSet::num_genotypes"></span><span class="target" id="project0classegglib_1_1FreqSet_1a17cd0f7d2263c544af4cb353f75584c7"></span>unsigned int <code class="descclassname">egglib::FreqSet::</code><code class="descname">num_genotypes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib7FreqSet13num_genotypesEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of genotypes (user-provided) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib7FreqSet17num_genotypes_effEv">
<span id="egglib::FreqSet::num_genotypes_eff"></span><span class="target" id="project0classegglib_1_1FreqSet_1a702a2790e322aa2b07c661a252f9ebda"></span>unsigned int <code class="descclassname">egglib::FreqSet::</code><code class="descname">num_genotypes_eff</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib7FreqSet17num_genotypes_effEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of genotypes with non-null frequency. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib7FreqSet5resetEj">
<span id="egglib::FreqSet::reset__unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqSet_1aab9755515d7cb3abb10b486126e3cad8"></span>void <code class="descclassname">egglib::FreqSet::</code><code class="descname">reset</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib7FreqSet5resetEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set number of alleles (set nsam/ngen to 0) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib7FreqSet5setupEv">
<span id="egglib::FreqSet::setup"></span><span class="target" id="project0classegglib_1_1FreqSet_1a4fc01d736fe50cf5b977f755b675f11d"></span>void <code class="descclassname">egglib::FreqSet::</code><code class="descname">setup</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib7FreqSet5setupEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Setup. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib7FreqSet8tell_hetEjj">
<span id="egglib::FreqSet::tell_het__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1FreqSet_1a2e29029abae0b1e5dad34a6ca7fce695"></span>void <code class="descclassname">egglib::FreqSet::</code><code class="descname">tell_het</code><span class="sig-paren">(</span>unsigned int <em>i</em>, unsigned int <em>a</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib7FreqSet8tell_hetEjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Tell the class that genotype i is heterozygote for allele a call it several times! don’t change frequencies after that! </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib7FreqSet7tot_hetEv">
<span id="egglib::FreqSet::tot_het"></span><span class="target" id="project0classegglib_1_1FreqSet_1ad107c6710ab3bb351fec6c84c636fc70"></span>unsigned int <code class="descclassname">egglib::FreqSet::</code><code class="descname">tot_het</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib7FreqSet7tot_hetEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Total frequency of heterozygotes. </p>
</dd></dl>

</div>
</dd></dl>

</div>
</div>
<div class="section" id="site-level-operations">
<h2>Site-level operations<a class="headerlink" href="#site-level-operations" title="Permalink to this headline">¶</a></h2>
<div class="section" id="siteholder">
<h3>SiteHolder<a class="headerlink" href="#siteholder" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib10SiteHolderE">
<span id="egglib::SiteHolder"></span><span class="target" id="project0classegglib_1_1SiteHolder"></span><em class="property">class </em></dt>
<dd><p>Holds data for a site for diversity analysis. </p>
<p>Usage of this class: first set the ploidy. Then, either load an alignment or data from a VCF, or individuals manually. Before loading individuals manually, it is required to pre-set the number such as the indexes will exist. If you don’t set all samples manually with <a class="reference internal" href="#project0classegglib_1_1SiteHolder_1a037cbc7feeb15fb23a8ca9000ed171c5"><span class="std std-ref">load_ing()</span></a> or <a class="reference internal" href="#project0classegglib_1_1SiteHolder_1abc047e45d1a063ac22c5e0b1a816464f"><span class="std std-ref">load_otg()</span></a>, you must force set the alleles to the default value. Note: the instance is not reset unless you ask it. Data will add up. </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder10SiteHolderEv">
<span id="egglib::SiteHolder::SiteHolder"></span><span class="target" id="project0classegglib_1_1SiteHolder_1a996b7c31e444b9fe1a98fc550d8d903f"></span><code class="descclassname">egglib::SiteHolder::</code><code class="descname">SiteHolder</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder10SiteHolderEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor (ploidy = 1) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder10SiteHolderEj">
<span id="egglib::SiteHolder::SiteHolder__unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteHolder_1ad41836398b66bbfe135be5672b0c0926"></span><code class="descclassname">egglib::SiteHolder::</code><code class="descname">SiteHolder</code><span class="sig-paren">(</span>unsigned int <em>ploidy</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder10SiteHolderEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolderD0Ev">
<span id="egglib::SiteHolder::~SiteHolder"></span>virtual <span class="target" id="project0classegglib_1_1SiteHolder_1a24bae58231ac5af8c8b159cad23b37ca"></span><code class="descclassname">egglib::SiteHolder::</code><code class="descname">~SiteHolder</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolderD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder7add_ingEj">
<span id="egglib::SiteHolder::add_ing__unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteHolder_1a91cd59b921d8dfa488d419b05e26df3a"></span>void <code class="descclassname">egglib::SiteHolder::</code><code class="descname">add_ing</code><span class="sig-paren">(</span>unsigned int <em>num</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder7add_ingEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add ingroup individuals. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder7add_otgEj">
<span id="egglib::SiteHolder::add_otg__unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteHolder_1ae9be95a0785c9f5e9adad96421214d2c"></span>void <code class="descclassname">egglib::SiteHolder::</code><code class="descname">add_otg</code><span class="sig-paren">(</span>unsigned int <em>num</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder7add_otgEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Add outgroup individuals. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder10get_alleleEj">
<span id="egglib::SiteHolder::get_allele__unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteHolder_1a22545649903a465c326bfd8cfe9d6e99"></span>int <code class="descclassname">egglib::SiteHolder::</code><code class="descname">get_allele</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder10get_alleleEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an allele (MISSINGDATA for MISSING) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder5get_iEjj">
<span id="egglib::SiteHolder::get_i__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteHolder_1afa62a9c88968e8a509cb71b0123206f7"></span>unsigned int <code class="descclassname">egglib::SiteHolder::</code><code class="descname">get_i</code><span class="sig-paren">(</span>unsigned int <em>idv</em>, unsigned int <em>chrom</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder5get_iEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get allele index for ingroup (MISSING for missing data) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder11get_missingEv">
<span id="egglib::SiteHolder::get_missing"></span><span class="target" id="project0classegglib_1_1SiteHolder_1a2662bf46d8d35e3ba0d4af771f438d0e"></span>unsigned int <code class="descclassname">egglib::SiteHolder::</code><code class="descname">get_missing</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder11get_missingEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of missing alleles found in the last processed data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder15get_missing_ingEv">
<span id="egglib::SiteHolder::get_missing_ing"></span><span class="target" id="project0classegglib_1_1SiteHolder_1ab9bdc8e79f1ec8d42f52c092f98744fa"></span>unsigned int <code class="descclassname">egglib::SiteHolder::</code><code class="descname">get_missing_ing</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder15get_missing_ingEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Total number of missing alleles in ingroup. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder15get_missing_otgEv">
<span id="egglib::SiteHolder::get_missing_otg"></span><span class="target" id="project0classegglib_1_1SiteHolder_1ac3098dba905c5a5445fedeb1b3ba535a"></span>unsigned int <code class="descclassname">egglib::SiteHolder::</code><code class="descname">get_missing_otg</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder15get_missing_otgEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Total number of missing alleles in outgroup. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder8get_nallEv">
<span id="egglib::SiteHolder::get_nall"></span><span class="target" id="project0classegglib_1_1SiteHolder_1a3f357b7cc92a4f0c44a7519e2b212a48"></span>unsigned int <code class="descclassname">egglib::SiteHolder::</code><code class="descname">get_nall</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder8get_nallEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of alleles. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder12get_nall_ingEv">
<span id="egglib::SiteHolder::get_nall_ing"></span><span class="target" id="project0classegglib_1_1SiteHolder_1a1d7c7d48441e4cdacd31ba45a29e9267"></span>unsigned int <code class="descclassname">egglib::SiteHolder::</code><code class="descname">get_nall_ing</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder12get_nall_ingEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of alleles in ingroup only. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder8get_ningEv">
<span id="egglib::SiteHolder::get_ning"></span><span class="target" id="project0classegglib_1_1SiteHolder_1abcefa4fc6d7fada70a348dfd636d3d09"></span>unsigned int <code class="descclassname">egglib::SiteHolder::</code><code class="descname">get_ning</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder8get_ningEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get number of ingroup individuals. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder8get_noutEv">
<span id="egglib::SiteHolder::get_nout"></span><span class="target" id="project0classegglib_1_1SiteHolder_1a3dfddace305288428e49c6055a53ea4b"></span>unsigned int <code class="descclassname">egglib::SiteHolder::</code><code class="descname">get_nout</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder8get_noutEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get number of outgroup individuals. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder5get_oEjj">
<span id="egglib::SiteHolder::get_o__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteHolder_1a5fe51d36e99b5af125879b2b3f403cd4"></span>unsigned int <code class="descclassname">egglib::SiteHolder::</code><code class="descname">get_o</code><span class="sig-paren">(</span>unsigned int <em>idv</em>, unsigned int <em>chrom</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder5get_oEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get allele index for outgroup (MISSING for missing data) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder6get_piEj">
<span id="egglib::SiteHolder::get_pi__unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteHolder_1ae6bd9af965bcb16e5277f22b01771de9"></span><em class="property">const</em> unsigned int *<code class="descclassname">egglib::SiteHolder::</code><code class="descname">get_pi</code><span class="sig-paren">(</span>unsigned int <em>idv</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder6get_piEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>get_i as pointer </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder10get_ploidyEv">
<span id="egglib::SiteHolder::get_ploidy"></span><span class="target" id="project0classegglib_1_1SiteHolder_1a93794a39636914f681ec33ff62e93c3e"></span>unsigned int <code class="descclassname">egglib::SiteHolder::</code><code class="descname">get_ploidy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder10get_ploidyEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get ploidy. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder6get_poEj">
<span id="egglib::SiteHolder::get_po__unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteHolder_1a7ffcef6dc6e89edfdc96796bd1a2df61"></span><em class="property">const</em> unsigned int *<code class="descclassname">egglib::SiteHolder::</code><code class="descname">get_po</code><span class="sig-paren">(</span>unsigned int <em>idv</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder6get_poEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>get_o as pointer </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder14get_straight_iEj">
<span id="egglib::SiteHolder::get_straight_i__unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteHolder_1aeea2e96a222ef766e1b32f156303954e"></span>unsigned int <code class="descclassname">egglib::SiteHolder::</code><code class="descname">get_straight_i</code><span class="sig-paren">(</span>unsigned int <em>sam</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder14get_straight_iEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get allele index for ingroup (MISSING for missing data) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder14get_straight_oEj">
<span id="egglib::SiteHolder::get_straight_o__unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteHolder_1a6b9c3ee811bbc5f1c6cfd0295c36c793"></span>unsigned int <code class="descclassname">egglib::SiteHolder::</code><code class="descname">get_straight_o</code><span class="sig-paren">(</span>unsigned int <em>sam</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder14get_straight_oEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get allele index for outgroup (MISSING for missing data) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder15get_tot_missingEv">
<span id="egglib::SiteHolder::get_tot_missing"></span><span class="target" id="project0classegglib_1_1SiteHolder_1a88841b47b8d0c7791c1325f14cedd5a9"></span>unsigned int <code class="descclassname">egglib::SiteHolder::</code><code class="descname">get_tot_missing</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder15get_tot_missingEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Total number of missing alleles. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder8load_ingEjji">
<span id="egglib::SiteHolder::load_ing__unsigned-i.unsigned-i.i"></span><span class="target" id="project0classegglib_1_1SiteHolder_1a037cbc7feeb15fb23a8ca9000ed171c5"></span>void <code class="descclassname">egglib::SiteHolder::</code><code class="descname">load_ing</code><span class="sig-paren">(</span>unsigned int <em>idv</em>, unsigned int <em>chrom</em>, int <em>allele</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder8load_ingEjji" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Analyze an ingroup allele (MISSINGDATA for missing data) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder8load_otgEjji">
<span id="egglib::SiteHolder::load_otg__unsigned-i.unsigned-i.i"></span><span class="target" id="project0classegglib_1_1SiteHolder_1abc047e45d1a063ac22c5e0b1a816464f"></span>void <code class="descclassname">egglib::SiteHolder::</code><code class="descname">load_otg</code><span class="sig-paren">(</span>unsigned int <em>idv</em>, unsigned int <em>chrom</em>, int <em>allele</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder8load_otgEjji" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Analyze an outgroup allele (MISSINGDATA for missing data) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder13process_alignERK10DataHolderjPK15StructureHolderRK6Filterjb">
<span id="egglib::SiteHolder::process_align__DataHolderCR.unsigned-i.StructureHolderCP.FilterCR.unsigned-i.b"></span><span class="target" id="project0classegglib_1_1SiteHolder_1a1466ca2a6a7e9e30ca236ad67eb658d5"></span>bool <code class="descclassname">egglib::SiteHolder::</code><code class="descname">process_align</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="core.html#_CPPv2N6egglib10DataHolderE" title="egglib::DataHolder">DataHolder</a> &amp;<em>data</em>, unsigned int <em>idx</em>, <em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib15StructureHolderE" title="egglib::StructureHolder">StructureHolder</a> *<em>struc</em>, <em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib6FilterE" title="egglib::Filter">Filter</a> &amp;<em>filtr</em>, unsigned int <em>max_missing</em>, bool <em>consider_outgroup_missing</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder13process_alignERK10DataHolderjPK15StructureHolderRK6Filterjb" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Process an alignment. </p>
<p>Does not reset instance! Ploidy must be defined before!</p>
<p><dl class="docutils">
<dt><strong>Return</strong></dt>
<dd>true if the number of missing data was not exceeded. </dd>
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">data</span></code> - <p>an alignment. </p>
</li>
<li><code class="first docutils literal"><span class="pre">idx</span></code> - <p>index of the site to process. </p>
</li>
<li><code class="first docutils literal"><span class="pre">struc</span></code> - <p>the structure to use (NULL can be passed to process all samples as haploid individuals). </p>
</li>
<li><code class="first docutils literal"><span class="pre">filtr</span></code> - <p>the allele filter. </p>
</li>
<li><code class="first docutils literal"><span class="pre">max_missing</span></code> - <p>maximum number of missing alleles. If there are more missing data, stop processing and return false. Then, the instance should absolutely not be used further. </p>
</li>
<li><code class="first docutils literal"><span class="pre">consider_outgroup_missing</span></code> - <p>if true, consider missing data for the max_missing argument, otherwise, only count missing data of the ingroup.</p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder11process_vcfER9VcfParserjjj">
<span id="egglib::SiteHolder::process_vcf__VcfParserR.unsigned-i.unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteHolder_1a8b8967456b771db2037853178c57bad7"></span>bool <code class="descclassname">egglib::SiteHolder::</code><code class="descname">process_vcf</code><span class="sig-paren">(</span><a class="reference internal" href="datafiles.html#_CPPv2N6egglib9VcfParserE" title="egglib::VcfParser">VcfParser</a> &amp;<em>data</em>, unsigned int <em>start</em>, unsigned int <em>stop</em>, unsigned int <em>max_missing</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder11process_vcfER9VcfParserjjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Import allelic data and compute frequencies from VCF data. </p>
<p>Beware: this method does not reset the instance.</p>
<p><dl class="docutils">
<dt><strong>Return</strong></dt>
<dd>A boolean specifying whether processing was completed. </dd>
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">data</span></code> - <p>a <a class="reference internal" href="datafiles.html#project0classegglib_1_1VcfParser"><span class="std std-ref">VcfParser</span></a> reference containing data and having the GT format field filled. </p>
</li>
<li><code class="first docutils literal"><span class="pre">start</span></code> - <p>index of the first sample to consider. </p>
</li>
<li><code class="first docutils literal"><span class="pre">stop</span></code> - <p>index of the last sample to consider. </p>
</li>
<li><code class="first docutils literal"><span class="pre">max_missing</span></code> - <p>maximum number of missing alleles. If this proportion is processing is stopped and <a class="reference internal" href="#project0classegglib_1_1SiteHolder_1a2662bf46d8d35e3ba0d4af771f438d0e"><span class="std std-ref">get_missing()</span></a> returns max_missing + 1. Only missing data in this data set, and in the ingroup, are considered.</p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder5resetEj">
<span id="egglib::SiteHolder::reset__unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteHolder_1a5bd301e37ecf2756d5d0e1553ee5c836"></span>void <code class="descclassname">egglib::SiteHolder::</code><code class="descname">reset</code><span class="sig-paren">(</span>unsigned int <em>ploidy</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder5resetEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Reset all to defaults. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder10set_alleleEji">
<span id="egglib::SiteHolder::set_allele__unsigned-i.i"></span><span class="target" id="project0classegglib_1_1SiteHolder_1ad5dcdd8d354a5370e72666e11fa208f7"></span>void <code class="descclassname">egglib::SiteHolder::</code><code class="descname">set_allele</code><span class="sig-paren">(</span>unsigned int, int <em>a</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder10set_alleleEji" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Set an allele value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder5set_iEjjj">
<span id="egglib::SiteHolder::set_i__unsigned-i.unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteHolder_1ac40bbeaf0c0d4eccde6b22959470b9c4"></span>void <code class="descclassname">egglib::SiteHolder::</code><code class="descname">set_i</code><span class="sig-paren">(</span>unsigned int <em>idv</em>, unsigned int <em>chrom</em>, unsigned int <em>all</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder5set_iEjjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set allele index for ingroup (MISSING for missing data) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder8set_nallEjj">
<span id="egglib::SiteHolder::set_nall__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteHolder_1ab6e43448f5921beb657772cf9dcf75d4"></span>void <code class="descclassname">egglib::SiteHolder::</code><code class="descname">set_nall</code><span class="sig-paren">(</span>unsigned int <em>all</em>, unsigned int <em>ing</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder8set_nallEjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set number of alleles (up to you that all is consistent) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10SiteHolder5set_oEjjj">
<span id="egglib::SiteHolder::set_o__unsigned-i.unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteHolder_1ab19868b9691934ba2f948467b2c11160"></span>void <code class="descclassname">egglib::SiteHolder::</code><code class="descname">set_o</code><span class="sig-paren">(</span>unsigned int <em>idv</em>, unsigned int <em>chrom</em>, unsigned int <em>all</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10SiteHolder5set_oEjjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set allele index for outgroup (MISSING for missing data) </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="sitegeno">
<h3>SiteGeno<a class="headerlink" href="#sitegeno" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib8SiteGenoE">
<span id="egglib::SiteGeno"></span><span class="target" id="project0classegglib_1_1SiteGeno"></span><em class="property">class </em></dt>
<dd><p><a class="reference internal" href="#project0classegglib_1_1SiteHolder"><span class="std std-ref">SiteHolder</span></a> subclass to transform regular data to genotypic. </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib8SiteGeno5homozEj">
<span id="egglib::SiteGeno::homoz__unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteGeno_1a1f381a687ace92c0de0529f5c12c8638"></span>bool <code class="descclassname">egglib::SiteGeno::</code><code class="descname">homoz</code><span class="sig-paren">(</span>unsigned int <em>genotype</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8SiteGeno5homozEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>tell if a genotype is homozygote </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8SiteGeno7processERK10SiteHolder">
<span id="egglib::SiteGeno::process__SiteHolderCR"></span><span class="target" id="project0classegglib_1_1SiteGeno_1abb62a94c11a57ef384dd8feb4c6f3e05"></span>void <code class="descclassname">egglib::SiteGeno::</code><code class="descname">process</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>src</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8SiteGeno7processERK10SiteHolder" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>reset and get data from a site </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="sitediversity">
<h3>SiteDiversity<a class="headerlink" href="#sitediversity" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib13SiteDiversityE">
<span id="egglib::SiteDiversity"></span><span class="target" id="project0classegglib_1_1SiteDiversity"></span><em class="property">class </em></dt>
<dd><p>Diversity analyses at the level of a site </p>
<p>Computes standard diversity indexes for a unique site or marker.</p>
<p><a class="reference internal" href="#project0classegglib_1_1SiteDiversity_1a60ce3598ff895b75117337fc20cdca00"><span class="std std-ref">process()</span></a> and <a class="reference internal" href="#project0classegglib_1_1SiteDiversity_1a43523fca1fab0b3b748ff95bc276d1bd"><span class="std std-ref">average()</span></a> return a composite flag.</p>
<p>Statistics:<ul class="simple">
<li>If fstats_diplo is called<ul>
<li>npop_eff2 (pops with &gt;= 1 indiv)</li>
</ul>
</li>
<li>If fstats_haplo is called<ul>
<li>npop_eff3 (pops with &gt;= 1 sample)</li>
</ul>
</li>
<li>If fstats_hier is called<ul>
<li>nclu</li>
<li>nclu_eff (&gt;= 1 pops each with &gt;= 1 indiv)</li>
<li>npop_eff2 (same as for fstats_diplo)</li>
</ul>
</li>
<li>flag&amp;1 (always on for <a class="reference internal" href="#project0classegglib_1_1SiteDiversity_1a60ce3598ff895b75117337fc20cdca00"><span class="std std-ref">process()</span></a>):<ul>
<li>ns</li>
</ul>
</li>
<li>flag&amp;2:<ul>
<li>npop</li>
<li>Aglob</li>
<li>Aing</li>
<li>Stot</li>
<li>pairdiff</li>
<li>He</li>
<li>R</li>
<li>npop_eff1 (pops with &gt;= 2 samples)</li>
<li>He[pop] (for pops with &gt;= 2 samples)</li>
<li>pairdiff_pop[pop1][pop2] (for pops with &gt;= 2 samples and pop2 != pop1)</li>
</ul>
</li>
<li>flag&amp;4:<ul>
<li>thetaIAM</li>
<li>thetaSMM</li>
</ul>
</li>
<li>flag&amp;8:<ul>
<li>Ho</li>
<li>Hi</li>
</ul>
</li>
<li>flag&amp;16:<ul>
<li>Sder</li>
<li>der</li>
</ul>
</li>
<li>flag&amp;2048:<ul>
<li>Aout</li>
</ul>
</li>
<li>flag&amp;32:<ul>
<li>n</li>
<li>d</li>
</ul>
</li>
<li>flag&amp;64:<ul>
<li>a</li>
<li>b</li>
<li>c</li>
</ul>
</li>
<li>flag&amp;128:<ul>
<li>a0</li>
<li>b1</li>
<li>b2</li>
<li>c0</li>
</ul>
</li>
<li>flag&amp;256:<ul>
<li>JostD</li>
</ul>
</li>
<li>flag&amp;2048:<ul>
<li>Hst</li>
</ul>
</li>
<li>flag&amp;4096:<ul>
<li>Gst</li>
</ul>
</li>
<li>flag&amp;8192:<ul>
<li>Gste</li>
</ul>
</li>
<li>flag&amp;512: ns, Aglob, Aing, Aout, Stot, Sder, and der are actually integers</li>
<li>flag&amp;1024: site is polymorphic / there is at least one polymorphic site<ul>
<li>MAF</li>
<li>MAF_pop</li>
</ul>
</li>
</ul>
</p>
<p>Fit = 1 - c/(a+b+c) Fst = a/(a+b+c) Fis = 1 - c/(b+c)</p>
<p>Fst = n/d</p>
<p>Fit = 1 - c0/(a0+b2+b1+c0) Fst = (a0+b2)/(a0+b2+b1+c0) Fct = a0/(a0+b2+b1+c0) Fis = 1 - c0/(b1+c0)</p>
<p>Hst = 1 - Hs / He Gst = 1 - Hs / Httilde Gste = 1 - Hse / Hte</p>
<p>Requires: stats() </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity13SiteDiversityEv">
<span id="egglib::SiteDiversity::SiteDiversity"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a5549e909059999085e046a2da3bf6426"></span><code class="descclassname">egglib::SiteDiversity::</code><code class="descname">SiteDiversity</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity13SiteDiversityEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversityD0Ev">
<span id="egglib::SiteDiversity::~SiteDiversity"></span>virtual <span class="target" id="project0classegglib_1_1SiteDiversity_1aace69c34d3a4411adc5498f1f1c401ed"></span><code class="descclassname">egglib::SiteDiversity::</code><code class="descname">~SiteDiversity</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversityD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity1aEv">
<span id="egglib::SiteDiversity::a"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a7f0799082988106ce030952a5d2fbb2e"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">a</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity1aEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Computed by fstats_diplo() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity2a0Ev">
<span id="egglib::SiteDiversity::a0"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1af6c74f1463142e197e5d0ebcfd34cf2b"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">a0</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity2a0Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Computed by fstats_hier() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity5AglobEv">
<span id="egglib::SiteDiversity::Aglob"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a67106b147d4f8b51438524236f39e765"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">Aglob</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity5AglobEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of alleles (including outgroup-specific alleles) (stats) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity4AingEv">
<span id="egglib::SiteDiversity::Aing"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a032aea63de8af3d44b8d19d60e22679c"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">Aing</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity4AingEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of alleles excluding outgroup-specific alleles (stats) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity4AoutEv">
<span id="egglib::SiteDiversity::Aout"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a467ae02eba32838a60c58d7870c24228"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">Aout</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity4AoutEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of different alleles in the outgroup (stats) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity7averageEv">
<span id="egglib::SiteDiversity::average"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a43523fca1fab0b3b748ff95bc276d1bd"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">average</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity7averageEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Compute the average of all stats (except those per pop) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity1bEv">
<span id="egglib::SiteDiversity::b"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a2bbc76aeadab8a326a08ddab7aae3412"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">b</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity1bEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Computed by fstats_diplo() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity2b1Ev">
<span id="egglib::SiteDiversity::b1"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1aa957fa6217a0e60aed096794d9ef531d"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">b1</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity2b1Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Computed by fstats_hier() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity2b2Ev">
<span id="egglib::SiteDiversity::b2"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a5c7ca38c01ab1565e93150420ac85123"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">b2</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity2b2Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Computed by fstats_hier() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity1cEv">
<span id="egglib::SiteDiversity::c"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1adef0fed0cb1426f67a076769f9867786"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">c</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity1cEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Computed by fstats_diplo() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity2c0Ev">
<span id="egglib::SiteDiversity::c0"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1acfa525aa03415b40f313c06bf0063a4a"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">c0</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity2c0Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Computed by fstats_hier() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity1dEv">
<span id="egglib::SiteDiversity::d"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1adfd619b6c8ac6b847feee99d688c85ad"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">d</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity1dEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Computed by fstats_haplo() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity1DEv">
<span id="egglib::SiteDiversity::D"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a01f1ec34f0907b90a3d974577bf07bc8"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">D</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity1DEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Computed by hstats() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity7derivedEj">
<span id="egglib::SiteDiversity::derived__unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a2543bb2ed8e123c7d8994f2659b366e4"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">derived</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity7derivedEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>&lt; Number of derived alleles (stats+outgroup) </p>
<p>Derived allele frequency (stats+outgroup) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity4flagEv">
<span id="egglib::SiteDiversity::flag"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a19a7adec0d916aad8bd27e898ae113d4"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">flag</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity4flagEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get flag value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity13global_alleleEj">
<span id="egglib::SiteDiversity::global_allele__unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a41bad8a0d1fe6f03320042a378a35b5d"></span>int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">global_allele</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity13global_alleleEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get one of the global alleles. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity3GstEv">
<span id="egglib::SiteDiversity::Gst"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1af08d323a0b4ef4e08fd4e7be206adde7"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">Gst</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity3GstEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Computed by hstats() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity4GsteEv">
<span id="egglib::SiteDiversity::Gste"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a00d4d03c59b94b1a8244180c8fad300f"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">Gste</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity4GsteEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Computed by hstats() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity2HeEv">
<span id="egglib::SiteDiversity::He"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1acb41bd0c0cf0dd272f4d8369d230815d"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">He</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity2HeEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Unbiased heterozygosity (averaged if relevant) stats() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity6He_popEj">
<span id="egglib::SiteDiversity::He_pop__unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1abd82730480408f2d11c04696bbf05c3e"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">He_pop</code><span class="sig-paren">(</span>unsigned int <em>pop</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity6He_popEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Unbiased heterozygosity for a population stats() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity2HiEv">
<span id="egglib::SiteDiversity::Hi"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a9691f2f67b1c78dd728e82be84a31e5f"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">Hi</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity2HiEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Avg number of differents between individuals (stats) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity2HoEv">
<span id="egglib::SiteDiversity::Ho"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a1619f01bd91d0273239c971161b851f7"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">Ho</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity2HoEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Frequency of heterozygotes (stats) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity3HstEv">
<span id="egglib::SiteDiversity::Hst"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a2d3ae26c8a048ea6655650d768b5b384"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">Hst</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity3HstEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Computed by hstats() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity1kEv">
<span id="egglib::SiteDiversity::k"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a8055bbb029e54fb9d21d92d29fabdf23"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">k</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity1kEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of populations (stats) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity3MAFEv">
<span id="egglib::SiteDiversity::MAF"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1af8594460ba38b8c2d2b45753590f5bb8"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">MAF</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity3MAFEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Frequency of second most frequent allele. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity7MAF_popEj">
<span id="egglib::SiteDiversity::MAF_pop__unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a5bdbb2cc5037bb9002d66c14c0bd4039"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">MAF_pop</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity7MAF_popEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>MAF for a population. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity1nEv">
<span id="egglib::SiteDiversity::n"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1ac18cd1c417c27917b70f31dbd5b5685d"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">n</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity1nEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Computed by fstats_haplo() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity8nclu_effEv">
<span id="egglib::SiteDiversity::nclu_eff"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a70a32a05243f6727a668996dedf23a98"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">nclu_eff</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity8nclu_effEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of clusters with &gt;= 1 pop with &gt;= 1 indiv (fstats_hier) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity9npop_eff1Ev">
<span id="egglib::SiteDiversity::npop_eff1"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a1e94ec57f0ec46003634b055c72ca1b6"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">npop_eff1</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity9npop_eff1Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of populations with &gt;= 2 samples (stats) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity9npop_eff2Ev">
<span id="egglib::SiteDiversity::npop_eff2"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1aa40862e941b65d98933f54292bc55b16"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">npop_eff2</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity9npop_eff2Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of populations with &gt;= 1 indiv (fstats_diplo + fstats_hier) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity9npop_eff3Ev">
<span id="egglib::SiteDiversity::npop_eff3"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a45e97a0e55f4e86495954cab4cc74f2b"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">npop_eff3</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity9npop_eff3Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of populations with &gt;= 1 sample (fstats_haplo) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity2nsEv">
<span id="egglib::SiteDiversity::ns"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1aea10f173fa7030afdc6ea33546eff761"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">ns</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity2nsEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of analyzed samples (stats) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity7nsites1Ev">
<span id="egglib::SiteDiversity::nsites1"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a14b40096277b6704be03d9328d635bf1"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">nsites1</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity7nsites1Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>For average ns. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity8nsites10Ev">
<span id="egglib::SiteDiversity::nsites10"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a1591a668a89e5aeb3f0f848855110bfe"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">nsites10</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity8nsites10Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>For average Hst. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity8nsites11Ev">
<span id="egglib::SiteDiversity::nsites11"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a746b0542624a411030d91e4d0266a031"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">nsites11</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity8nsites11Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>For average Gst. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity8nsites12Ev">
<span id="egglib::SiteDiversity::nsites12"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a1e3fbd672b9f3356bfe597416b28a2a8"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">nsites12</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity8nsites12Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>For average Gste. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity7nsites2Ev">
<span id="egglib::SiteDiversity::nsites2"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a2968358320914dc3e4dce2adf3a131f6"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">nsites2</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity7nsites2Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>For average Aglob, Aing, Atot, Stot, pairdiff, He, thetaIAM, and thetaSMM. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity7nsites3Ev">
<span id="egglib::SiteDiversity::nsites3"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a804982435796d2682683c69abb9586c7"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">nsites3</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity7nsites3Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>For average thetaIAM and thetaSMM. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity7nsites4Ev">
<span id="egglib::SiteDiversity::nsites4"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a3ffae04d9c21ba51422c6484f29c7e33"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">nsites4</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity7nsites4Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>For average Ho and Hi. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity7nsites5Ev">
<span id="egglib::SiteDiversity::nsites5"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a9cbf9ed88f8dcdbfb4c8b5a15a83878c"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">nsites5</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity7nsites5Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>For average derived and Sd. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity7nsites6Ev">
<span id="egglib::SiteDiversity::nsites6"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a9b96d5e8c55a85752f185714ddcc9690"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">nsites6</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity7nsites6Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>For average n and d. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity7nsites7Ev">
<span id="egglib::SiteDiversity::nsites7"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a7ff6ca0209909560bac0dc81ed96efa8"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">nsites7</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity7nsites7Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>For average a, b, and c. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity7nsites8Ev">
<span id="egglib::SiteDiversity::nsites8"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a037d01a7d58c57d8a3e4d37a0e3f731b"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">nsites8</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity7nsites8Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>For average c0, b1, b2, a0. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity7nsites9Ev">
<span id="egglib::SiteDiversity::nsites9"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a7644ea8cdef2ed1d2a8123bbd51a4b16"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">nsites9</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity7nsites9Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>For average D. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity10orientableEv">
<span id="egglib::SiteDiversity::orientable"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a875d621eb9d88c2010d8206beab92c58"></span>bool <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">orientable</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity10orientableEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>True if the site is orientable (stats) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity8pairdiffEv">
<span id="egglib::SiteDiversity::pairdiff"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a5aef7c0055ea751277671b472054acb0"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">pairdiff</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity8pairdiffEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Average number of pairwise differences (stats) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity14pairdiff_interEjj">
<span id="egglib::SiteDiversity::pairdiff_inter__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a7adb2eb432dfa5faa611b1d498d9d184"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">pairdiff_inter</code><span class="sig-paren">(</span>unsigned int <em>pop1</em>, unsigned int <em>pop2</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity14pairdiff_interEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Average number of differences between a pair of population (stats) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity7processERK10FreqHolder">
<span id="egglib::SiteDiversity::process__FreqHolderCR"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a60ce3598ff895b75117337fc20cdca00"></span>unsigned int <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">process</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10FreqHolderE" title="egglib::FreqHolder">FreqHolder</a> &amp;<em>frq</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity7processERK10FreqHolder" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Compute toggled statistics. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity1REv">
<span id="egglib::SiteDiversity::R"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a0e6a9ffae40092ee339aaf67c0296200"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">R</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity1REv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Allelic richness (stats) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity5resetEv">
<span id="egglib::SiteDiversity::reset"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1ad20897c5c8bd47f5d4005989bead0e55"></span>void <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">reset</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity5resetEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Reset stats sums to 0 (keep toggled flags) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity1SEv">
<span id="egglib::SiteDiversity::S"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1acf06c620c0e6893dc161a6d732c44d08"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">S</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity1SEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of alleles at frequency one (singletons) (stats) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity2SdEv">
<span id="egglib::SiteDiversity::Sd"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a2c12e1da881a3f1696f819e56a1a0328"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">Sd</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity2SdEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of derived singletons (stats) (requires outgroup) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity8thetaIAMEv">
<span id="egglib::SiteDiversity::thetaIAM"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a96561360a960f6a87e318b57848381f6"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">thetaIAM</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity8thetaIAMEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Requires stats() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity8thetaSMMEv">
<span id="egglib::SiteDiversity::thetaSMM"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a2825591c01a01b3ef293f5a5cb954d93"></span>double <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">thetaSMM</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity8thetaSMMEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Requires stats() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity19toggle_fstats_diploEv">
<span id="egglib::SiteDiversity::toggle_fstats_diplo"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a0fc962fe59b92fd653e002b2a31fd7a3"></span>void <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">toggle_fstats_diplo</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity19toggle_fstats_diploEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Toggle F-statistics. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity19toggle_fstats_haploEv">
<span id="egglib::SiteDiversity::toggle_fstats_haplo"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1ac01bdd23d635ff827e0163276630a5b6"></span>void <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">toggle_fstats_haplo</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity19toggle_fstats_haploEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Toggle F-statistics. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity18toggle_fstats_hierEv">
<span id="egglib::SiteDiversity::toggle_fstats_hier"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a62bbab14b9dddaedfc210b4814f0390f"></span>void <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">toggle_fstats_hier</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity18toggle_fstats_hierEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Toggle F-statistics. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity13toggle_hstatsEv">
<span id="egglib::SiteDiversity::toggle_hstats"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1ab7d2c8efbd8be6eb51b9b83df878d690"></span>void <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">toggle_hstats</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity13toggle_hstatsEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Toggle H-statistics. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib13SiteDiversity10toggle_offEv">
<span id="egglib::SiteDiversity::toggle_off"></span><span class="target" id="project0classegglib_1_1SiteDiversity_1a898807895f3c1114c7e0e654b237ab28"></span>void <code class="descclassname">egglib::SiteDiversity::</code><code class="descname">toggle_off</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib13SiteDiversity10toggle_offEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set all flags to off. </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="codingsite">
<h3>CodingSite<a class="headerlink" href="#codingsite" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib10CodingSiteE">
<span id="egglib::CodingSite"></span><span class="target" id="project0classegglib_1_1CodingSite"></span><em class="property">class </em></dt>
<dd><p>Holds data for a coding site (as a triplet of sites) </p>
<p>This class assists the detection of polymorphism at codon sites, although diversity analyses themselves have to be performed using <a class="reference internal" href="#project0classegglib_1_1SiteDiversity"><span class="std std-ref">SiteDiversity</span></a> itself. The class perform analyses through the <a class="reference internal" href="#project0classegglib_1_1CodingSite_1a122f15b7a86d71aaf69ff3fffbb598d3"><span class="std std-ref">process()</span></a> method, which automatically resets all previously stored data.</p>
<p>Header: &lt;egglib-cpp/CodongSite.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib10CodingSite10CodingSiteEv">
<span id="egglib::CodingSite::CodingSite"></span><span class="target" id="project0classegglib_1_1CodingSite_1a95afb5aa9a2fabfd3bf794f24c17395c"></span><code class="descclassname">egglib::CodingSite::</code><code class="descname">CodingSite</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10CodingSite10CodingSiteEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10CodingSiteD0Ev">
<span id="egglib::CodingSite::~CodingSite"></span><span class="target" id="project0classegglib_1_1CodingSite_1a5a46b3e042f9edeb51c9aad1cc5d54d3"></span><code class="descclassname">egglib::CodingSite::</code><code class="descname">~CodingSite</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10CodingSiteD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10CodingSite10aminoacidsEv">
<span id="egglib::CodingSite::aminoacids"></span><span class="target" id="project0classegglib_1_1CodingSite_1a0ffab1a84a86f99747f69979d467743a"></span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<code class="descclassname">egglib::CodingSite::</code><code class="descname">aminoacids</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10CodingSite10aminoacidsEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get access to amino acid data. </p>
<p>Requires that a codon site has been analyzed using <a class="reference internal" href="#project0classegglib_1_1CodingSite_1a122f15b7a86d71aaf69ff3fffbb598d3"><span class="std std-ref">process()</span></a>.</p>
<p>This instance contains amino acids. It contains integer data representing amino acids (‘*’ for stop codons). The <a class="reference internal" href="core.html#project0classegglib_1_1GeneticCode"><span class="std std-ref">GeneticCode</span></a> instance passed to <a class="reference internal" href="#project0classegglib_1_1CodingSite_1a122f15b7a86d71aaf69ff3fffbb598d3"><span class="std std-ref">process()</span></a> determines the translation. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10CodingSite6codonsEv">
<span id="egglib::CodingSite::codons"></span><span class="target" id="project0classegglib_1_1CodingSite_1af4106aa8695aab4e4f5aaa174599bcc1"></span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<code class="descclassname">egglib::CodingSite::</code><code class="descname">codons</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10CodingSite6codonsEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get access to merged codon data. </p>
<p>Requires that a codon site has been analyzed using <a class="reference internal" href="#project0classegglib_1_1CodingSite_1a122f15b7a86d71aaf69ff3fffbb598d3"><span class="std std-ref">process()</span></a>.</p>
<p>This instance contains codon alleles. See <a class="reference internal" href="core.html#project0classegglib_1_1GeneticCode"><span class="std std-ref">GeneticCode</span></a> for information about encoding of codons. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10CodingSite7mutatedEjjj">
<span id="egglib::CodingSite::mutated__unsigned-i.unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1CodingSite_1a02b579a65a5c72c62156c68886022044"></span>bool <code class="descclassname">egglib::CodingSite::</code><code class="descname">mutated</code><span class="sig-paren">(</span>unsigned int <em>codon_allele1</em>, unsigned int <em>codon_allele2</em>, unsigned int <em>pos</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10CodingSite7mutatedEjjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Check if two given codon alleles differ at a given position. </p>
<p>Tells, for two given codon alleles (see alleles()), if they differ (at least) at the specified position. Possible values are 0, 1 or 2.</p>
<p>Requires that a codon site has been analyzed using <a class="reference internal" href="#project0classegglib_1_1CodingSite_1a122f15b7a86d71aaf69ff3fffbb598d3"><span class="std std-ref">process()</span></a>.</p>
<p>The indexes must both be &lt; nall() but they may be passed in any order. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10CodingSite5ndiffEjj">
<span id="egglib::CodingSite::ndiff__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1CodingSite_1ab0ca285667e442f3243ac95c0c2ca80c"></span>unsigned int <code class="descclassname">egglib::CodingSite::</code><code class="descname">ndiff</code><span class="sig-paren">(</span>unsigned int <em>codon_allele1</em>, unsigned int <em>codon_allele2</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10CodingSite5ndiffEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of nucleotides differences between two given codon alleles. </p>
<p>Give, for two given codon alleles (see alleles()), the number of nucleotide differences between them. Possible values are 1, 2 or 3.</p>
<p>Requires that a codon site has been analyzed using <a class="reference internal" href="#project0classegglib_1_1CodingSite_1a122f15b7a86d71aaf69ff3fffbb598d3"><span class="std std-ref">process()</span></a>.</p>
<p>The indexes must both be &lt; nall() but they may be passed in any order. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10CodingSite2niEv">
<span id="egglib::CodingSite::ni"></span><span class="target" id="project0classegglib_1_1CodingSite_1a19fbf970025319e08ae311ace094bc72"></span>unsigned int <code class="descclassname">egglib::CodingSite::</code><code class="descname">ni</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10CodingSite2niEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of ingroup indiv. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10CodingSite2noEv">
<span id="egglib::CodingSite::no"></span><span class="target" id="project0classegglib_1_1CodingSite_1a38acab5bfb24c869bae9f34d4ad0512f"></span>unsigned int <code class="descclassname">egglib::CodingSite::</code><code class="descname">no</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10CodingSite2noEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of outgroup indiv. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10CodingSite2NSEjj">
<span id="egglib::CodingSite::NS__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1CodingSite_1aaea46e325952a83d9cc2d8acbb6e4798"></span>bool <code class="descclassname">egglib::CodingSite::</code><code class="descname">NS</code><span class="sig-paren">(</span>unsigned int <em>codon_allele1</em>, unsigned int <em>codon_allele2</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10CodingSite2NSEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>True if the two given codon alleles encode different aminoacids. </p>
<p>Requires that a codon site has been analyzed using <a class="reference internal" href="#project0classegglib_1_1CodingSite_1a122f15b7a86d71aaf69ff3fffbb598d3"><span class="std std-ref">process()</span></a>.</p>
<p>The indexes must both be &lt; nall() but they may be passed in any order. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10CodingSite5nseffEv">
<span id="egglib::CodingSite::nseff"></span><span class="target" id="project0classegglib_1_1CodingSite_1a291e81cff28e32c3a25ff41996260784"></span>unsigned int <code class="descclassname">egglib::CodingSite::</code><code class="descname">nseff</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10CodingSite5nseffEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of analyzed samples. </p>
<p>Requires that a codon site has been analyzed using <a class="reference internal" href="#project0classegglib_1_1CodingSite_1a122f15b7a86d71aaf69ff3fffbb598d3"><span class="std std-ref">process()</span></a>.</p>
<p>Value bound by 0 and ns(). Depends on the number of missing data (and stop codons, if skipstop was set to true). </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10CodingSite6nseffoEv">
<span id="egglib::CodingSite::nseffo"></span><span class="target" id="project0classegglib_1_1CodingSite_1a243bfb0daf8ebc75a976b87bfee93d8b"></span>unsigned int <code class="descclassname">egglib::CodingSite::</code><code class="descname">nseffo</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10CodingSite6nseffoEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Analyzed samples for outgroup. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10CodingSite7NSsitesEv">
<span id="egglib::CodingSite::NSsites"></span><span class="target" id="project0classegglib_1_1CodingSite_1ae8df034ba1f6652916b471d5b8dfa47d"></span>double <code class="descclassname">egglib::CodingSite::</code><code class="descname">NSsites</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10CodingSite7NSsitesEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Estimated number of nonsynonymous sites. </p>
<p>Requires that a codon site has been analyzed using <a class="reference internal" href="#project0classegglib_1_1CodingSite_1a122f15b7a86d71aaf69ff3fffbb598d3"><span class="std std-ref">process()</span></a>. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10CodingSite5nstopEv">
<span id="egglib::CodingSite::nstop"></span><span class="target" id="project0classegglib_1_1CodingSite_1a7763e212d8c162d49349abcecfc56c9f"></span>unsigned int <code class="descclassname">egglib::CodingSite::</code><code class="descname">nstop</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10CodingSite5nstopEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of stop codons met during processing of codon site. </p>
<p>Requires that a codon site has been analyzed using <a class="reference internal" href="#project0classegglib_1_1CodingSite_1a122f15b7a86d71aaf69ff3fffbb598d3"><span class="std std-ref">process()</span></a>.</p>
<p>Not affected by the skipstop option. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10CodingSite2plEv">
<span id="egglib::CodingSite::pl"></span><span class="target" id="project0classegglib_1_1CodingSite_1ad90f262222861290f0671fe901c420dd"></span>unsigned int <code class="descclassname">egglib::CodingSite::</code><code class="descname">pl</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10CodingSite2plEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Ploidy. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10CodingSite7processERK10SiteHolderRK10SiteHolderRK10SiteHolderRK11GeneticCodebj">
<span id="egglib::CodingSite::process__SiteHolderCR.SiteHolderCR.SiteHolderCR.GeneticCodeCR.b.unsigned-i"></span><span class="target" id="project0classegglib_1_1CodingSite_1a122f15b7a86d71aaf69ff3fffbb598d3"></span>bool <code class="descclassname">egglib::CodingSite::</code><code class="descname">process</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site1</em>, <em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site2</em>, <em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site3</em>, <em class="property">const</em> <a class="reference internal" href="core.html#_CPPv2N6egglib11GeneticCodeE" title="egglib::GeneticCode">GeneticCode</a> &amp;<em>code</em>, bool <em>skipstop</em>, unsigned int <em>max_missing</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10CodingSite7processERK10SiteHolderRK10SiteHolderRK10SiteHolderRK11GeneticCodebj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Analyzes a codon site. </p>
<p>The three codon positions must be loaded as Site instances containing nucleotides encoded as integer values. All values except values equal to A, C, G and T (case-independent) are treated as missing data. Obviously, the three sites must have the same number of samples and also the same number of populations (and matching affectation of samples to populations). Upon processing, the class generates and holds a Site instance (available as <a class="reference internal" href="#project0classegglib_1_1CodingSite_1af4106aa8695aab4e4f5aaa174599bcc1"><span class="std std-ref">codons()</span></a>) containing data from the three sites merged, and another (available as <a class="reference internal" href="#project0classegglib_1_1CodingSite_1a0ffab1a84a86f99747f69979d467743a"><span class="std std-ref">aminoacids()</span></a>) with the same data translated).</p>
<p><dl class="docutils">
<dt><strong>Return</strong></dt>
<dd>A boolean indicating whether analysis was completed. If False, data contained in the object should not be used since stored objects will not have been filled. Even <a class="reference internal" href="#project0classegglib_1_1CodingSite_1a19fbf970025319e08ae311ace094bc72"><span class="std std-ref">ni()</span></a> and nieff() will be invalid.</dd>
<dt><strong>Note</strong></dt>
<dd>It is not allowed to use <a class="reference internal" href="core.html#project0namespaceegglib_1a6b564311a87de05e515ec65480ebc236"><span class="std std-ref">egglib::UNKNOWN</span></a> for any of A, C, G and T argument. </dd>
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">site1</span></code> - <p>first nucleotide position of the codon. </p>
</li>
<li><code class="first docutils literal"><span class="pre">site2</span></code> - <p>second nucleotide position of the codon. </p>
</li>
<li><code class="first docutils literal"><span class="pre">site3</span></code> - <p>third nucleotide position of the codon. </p>
</li>
<li><code class="first docutils literal"><span class="pre">code</span></code> - <p><a class="reference internal" href="core.html#project0classegglib_1_1GeneticCode"><span class="std std-ref">GeneticCode</span></a> instance representing the code to be used for treating this codon. </p>
</li>
<li><code class="first docutils literal"><span class="pre">skipstop</span></code> - <p>if true, stop codons are treated as missing data and skipped. If set to true, potential mutations to stop codons are not taken into account when estimating the number of non-synonymous sites. Warning (this may be counter-intuitive). It actually assumes that stop codons are not biologically plausible and considers them as missing data. On the other hand, if skipstop is false, it takes stop codons as if they were valid amino acids. </p>
</li>
<li><code class="first docutils literal"><span class="pre">max_missing</span></code> - <p>maximum number of missing data to allow (including stop codons if skipstop if true). </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10CodingSite6SsitesEv">
<span id="egglib::CodingSite::Ssites"></span><span class="target" id="project0classegglib_1_1CodingSite_1aff5dce13b2be73aad546c1866fcea2f7"></span>double <code class="descclassname">egglib::CodingSite::</code><code class="descname">Ssites</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10CodingSite6SsitesEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Estimated number of synonymous sites. </p>
<p>Requires that a codon site has been analyzed using <a class="reference internal" href="#project0classegglib_1_1CodingSite_1a122f15b7a86d71aaf69ff3fffbb598d3"><span class="std std-ref">process()</span></a>. </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="allelestatus">
<h3>AlleleStatus<a class="headerlink" href="#allelestatus" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib12AlleleStatusE">
<span id="egglib::AlleleStatus"></span><span class="target" id="project0classegglib_1_1AlleleStatus"></span><em class="property">class </em></dt>
<dd><p>Classify alleles and site for frequencies with several populations. </p>
<p>Statistics: Sp  population-specific alleles Spd  population-specific derived alleles ShP  number of alleles segregating in at least one pair of populations ShA  number of alleles in non-null frequencies in at least one pair of populations FxA  number of alleles fixed in at least one population FxD  number of fixed differences (two different alleles fixed in a pair of populations)</p>
<p>The user must ensure that all passed sites are polymorphic. The user should also probably exclude populations with low sample sizes if they are interested in the number of fixed alleles (populations with no samples are automatically skipped).</p>
<p>The statistics are computed for each site. Sums for multi-sites are available as Sp_T and Sp_T1 (and similarly for other statistics). T1 is such as each site is counted only once for any statistic.</p>
<p>Header: &lt;egglib-cpp/AlleleStatus.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatus12AlleleStatusEv">
<span id="egglib::AlleleStatus::AlleleStatus"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1a53607b8e44855cf3d3d80795d385c837"></span><code class="descclassname">egglib::AlleleStatus::</code><code class="descname">AlleleStatus</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatus12AlleleStatusEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatusD0Ev">
<span id="egglib::AlleleStatus::~AlleleStatus"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1a91c5d0b826e9cd28032cd3aa8471a13b"></span><code class="descclassname">egglib::AlleleStatus::</code><code class="descname">~AlleleStatus</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatusD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatus3FxAEv">
<span id="egglib::AlleleStatus::FxA"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1a0a54620fcf9f2ef051d1f80839f2db4e"></span>unsigned int <code class="descclassname">egglib::AlleleStatus::</code><code class="descname">FxA</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatus3FxAEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Fixed alleles. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatus6FxA_T1Ev">
<span id="egglib::AlleleStatus::FxA_T1"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1adc31ef892bdb379c14de06d9306539ec"></span>unsigned int <code class="descclassname">egglib::AlleleStatus::</code><code class="descname">FxA_T1</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatus6FxA_T1Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Fixed alleles. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatus3FxDEv">
<span id="egglib::AlleleStatus::FxD"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1a911ab7e0fc014a9239964951179b3fa4"></span>unsigned int <code class="descclassname">egglib::AlleleStatus::</code><code class="descname">FxD</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatus3FxDEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Fixed differences. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatus6FxD_T1Ev">
<span id="egglib::AlleleStatus::FxD_T1"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1a62b391fbedc01066199b9b500646c3c7"></span>unsigned int <code class="descclassname">egglib::AlleleStatus::</code><code class="descname">FxD_T1</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatus6FxD_T1Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Fixed differences. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatus6nsitesEv">
<span id="egglib::AlleleStatus::nsites"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1a2998d731abffa878d9f98b1367e87c17"></span>unsigned int <code class="descclassname">egglib::AlleleStatus::</code><code class="descname">nsites</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatus6nsitesEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of sites with valid data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatus8nsites_oEv">
<span id="egglib::AlleleStatus::nsites_o"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1a4bcda7bfd6bef9c96bcaed9dfdc0e6c3"></span>unsigned int <code class="descclassname">egglib::AlleleStatus::</code><code class="descname">nsites_o</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatus8nsites_oEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of orientable sites with valid data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatus7processERK10FreqHolder">
<span id="egglib::AlleleStatus::process__FreqHolderCR"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1a001e66f9c079b6b5ecdbbb9d516207aa"></span>void <code class="descclassname">egglib::AlleleStatus::</code><code class="descname">process</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10FreqHolderE" title="egglib::FreqHolder">FreqHolder</a> &amp;<em>freqs</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatus7processERK10FreqHolder" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Analyze a site. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatus5resetEv">
<span id="egglib::AlleleStatus::reset"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1ad20897c5c8bd47f5d4005989bead0e55"></span>void <code class="descclassname">egglib::AlleleStatus::</code><code class="descname">reset</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatus5resetEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Reset sums (but keep toggle flag) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatus3ShAEv">
<span id="egglib::AlleleStatus::ShA"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1a2c14d850aba3231b420f861a056c0d96"></span>unsigned int <code class="descclassname">egglib::AlleleStatus::</code><code class="descname">ShA</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatus3ShAEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Shared alleles. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatus6ShA_T1Ev">
<span id="egglib::AlleleStatus::ShA_T1"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1a21161875b3d69832925267297821e713"></span>unsigned int <code class="descclassname">egglib::AlleleStatus::</code><code class="descname">ShA_T1</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatus6ShA_T1Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Shared alleles. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatus3ShPEv">
<span id="egglib::AlleleStatus::ShP"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1a99fb877318ae0d89d9b5a34297e4f5d2"></span>unsigned int <code class="descclassname">egglib::AlleleStatus::</code><code class="descname">ShP</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatus3ShPEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Shared polymorphisms. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatus6ShP_T1Ev">
<span id="egglib::AlleleStatus::ShP_T1"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1ac23d9181bcaffc2510d427b66ba499d4"></span>unsigned int <code class="descclassname">egglib::AlleleStatus::</code><code class="descname">ShP_T1</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatus6ShP_T1Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Shared polymorphisms. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatus2SpEv">
<span id="egglib::AlleleStatus::Sp"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1ae624f1359f3ae5ad1ad57b15681d726e"></span>unsigned int <code class="descclassname">egglib::AlleleStatus::</code><code class="descname">Sp</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatus2SpEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Pop-specific alleles. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatus5Sp_T1Ev">
<span id="egglib::AlleleStatus::Sp_T1"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1aad8086feaa8ffb0dff3f332f18d14333"></span>unsigned int <code class="descclassname">egglib::AlleleStatus::</code><code class="descname">Sp_T1</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatus5Sp_T1Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Pop-specific alleles. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatus3SpdEv">
<span id="egglib::AlleleStatus::Spd"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1a3ffb0d8587601bcfb4dc69f57f2b1b43"></span>unsigned int <code class="descclassname">egglib::AlleleStatus::</code><code class="descname">Spd</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatus3SpdEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Pop-specific derived alleles. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatus6Spd_T1Ev">
<span id="egglib::AlleleStatus::Spd_T1"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1a5f1869800d273ee4e498c75493336155"></span>unsigned int <code class="descclassname">egglib::AlleleStatus::</code><code class="descname">Spd_T1</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatus6Spd_T1Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Pop-specific derived alleles. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib12AlleleStatus5totalEv">
<span id="egglib::AlleleStatus::total"></span><span class="target" id="project0classegglib_1_1AlleleStatus_1abdeb2ce85125c72ddd44ceeaac74ee6f"></span>void <code class="descclassname">egglib::AlleleStatus::</code><code class="descname">total</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib12AlleleStatus5totalEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Copy all sums to director accessors. </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="computev">
<h3>ComputeV<a class="headerlink" href="#computev" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib8ComputeVE">
<span id="egglib::ComputeV"></span><span class="target" id="project0classegglib_1_1ComputeV"></span><em class="property">class </em></dt>
<dd><p>Compute allele size variance. </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib8ComputeV8ComputeVEv">
<span id="egglib::ComputeV::ComputeV"></span><span class="target" id="project0classegglib_1_1ComputeV_1a7c37ff1e03565d2c03c6577b91cfdebe"></span><code class="descclassname">egglib::ComputeV::</code><code class="descname">ComputeV</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8ComputeV8ComputeVEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8ComputeVD0Ev">
<span id="egglib::ComputeV::~ComputeV"></span><span class="target" id="project0classegglib_1_1ComputeV_1a77fb4f3251efa229be3af45217d18648"></span><code class="descclassname">egglib::ComputeV::</code><code class="descname">~ComputeV</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8ComputeVD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8ComputeV7averageEv">
<span id="egglib::ComputeV::average"></span><span class="target" id="project0classegglib_1_1ComputeV_1a03896a49e2e6908588ffb81e4f2629e0"></span>double <code class="descclassname">egglib::ComputeV::</code><code class="descname">average</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8ComputeV7averageEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get average V (UNDEF if no computed values) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8ComputeV7computeERK7FreqSet">
<span id="egglib::ComputeV::compute__FreqSetCR"></span><span class="target" id="project0classegglib_1_1ComputeV_1a5e30b793807aa5fb428e6baea6d3e21e"></span>double <code class="descclassname">egglib::ComputeV::</code><code class="descname">compute</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib7FreqSetE" title="egglib::FreqSet">FreqSet</a> &amp;<em>frq</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8ComputeV7computeERK7FreqSet" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Compute V (UNDEF if not computable) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8ComputeV9num_sitesEv">
<span id="egglib::ComputeV::num_sites"></span><span class="target" id="project0classegglib_1_1ComputeV_1aae69a5b25d3e7d7ae4027a93b0893d48"></span>unsigned int <code class="descclassname">egglib::ComputeV::</code><code class="descname">num_sites</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8ComputeV9num_sitesEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of sites with computed V. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8ComputeV5resetEv">
<span id="egglib::ComputeV::reset"></span><span class="target" id="project0classegglib_1_1ComputeV_1ad20897c5c8bd47f5d4005989bead0e55"></span>void <code class="descclassname">egglib::ComputeV::</code><code class="descname">reset</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8ComputeV5resetEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Reset. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8ComputeV10set_alleleEji">
<span id="egglib::ComputeV::set_allele__unsigned-i.i"></span><span class="target" id="project0classegglib_1_1ComputeV_1abdc5f036d7df321f818a86ae9e265ea4"></span>void <code class="descclassname">egglib::ComputeV::</code><code class="descname">set_allele</code><span class="sig-paren">(</span>unsigned int, int <em>a</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8ComputeV10set_alleleEji" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set an allele value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8ComputeV13setup_allelesEj">
<span id="egglib::ComputeV::setup_alleles__unsigned-i"></span><span class="target" id="project0classegglib_1_1ComputeV_1a2c55b1b0c960f42c7808c4352832c638"></span>void <code class="descclassname">egglib::ComputeV::</code><code class="descname">setup_alleles</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8ComputeV13setup_allelesEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Specify number of alleles. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8ComputeV23setup_alleles_from_siteERK10SiteHolder">
<span id="egglib::ComputeV::setup_alleles_from_site__SiteHolderCR"></span><span class="target" id="project0classegglib_1_1ComputeV_1ad6e34735c55189172fe7dcc597feef15"></span>void <code class="descclassname">egglib::ComputeV::</code><code class="descname">setup_alleles_from_site</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8ComputeV23setup_alleles_from_siteERK10SiteHolder" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Get alleles directly from site. </p>
</dd></dl>

</div>
</dd></dl>

</div>
</div>
<div class="section" id="generic-classes">
<h2>Generic classes<a class="headerlink" href="#generic-classes" title="Permalink to this headline">¶</a></h2>
<div class="section" id="diversity1">
<h3>Diversity1<a class="headerlink" href="#diversity1" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib10Diversity1E">
<span id="egglib::Diversity1"></span><span class="target" id="project0classegglib_1_1Diversity1"></span><em class="property">class </em></dt>
<dd><p>Compute population summary statistics from allele frequencies at several sites. </p>
<p><a class="reference internal" href="#project0classegglib_1_1Diversity1"><span class="std std-ref">Diversity1</span></a> instances cannot be copied. This class is designed to allow reuse of objects without unnecessary memory reallocation.</p>
<p>This class computes statistics that does not require access to a full Site instance and for which only frequencies are needed. The frequency for all sites that must be analyzed should be loaded.</p>
<p>Statistics:</p>
<p>code | requirement | flag | toggle flag =========|=============================|=======|============ lt | - | - | - ls | - | - | - nsmax | ls&gt;0 | 1 | - S | ls&gt;0 | 1 | - Ss | ls&gt;0 | 1 | - eta | ls&gt;0 | 1 | - Pi | ls&gt;0 | 1 | - lso | - | 4 | ori_site nsmaxo | lso&gt;0 | 8 | ori_site So | lso&gt;0 | 8 | ori_site Sso | lso&gt;0 | 8 | ori_site etao | lso&gt;0 | 8 | ori_site lM | lso&gt;0 | 8 | ori_site pM | lM&gt;0 | 16 | ori_site nseffo | lso&gt;0 | 32 | ori_div thetaH | lso&gt;0 | 32 | ori_div thetaL | lso&gt;0 | 32 | ori_div Hns | lso&gt;0 | 32 | ori_div Hsd | So&gt;0 &amp; nseffo&gt;=3 &amp; varZ&gt;0 | 1024 | ori_div E | So&gt;0 &amp; nseffo&gt;=3 &amp; varE&gt;0 | 2048 | ori_div Dfl | So&gt;0 &amp; nseffo&gt;=3 &amp; varDfl&gt;0 | 4096 | ori_div F | So&gt;0 &amp; nseffo&gt;=3 &amp; varF&gt;0 | 8192 | ori_div nseff | ls&gt;0 | 128 | basic thetaW | ls&gt;0 | 128 | basic Dxy | ls&gt;0 npop=2 | 16384 | basic Da | ls&gt;0 npop=2 | 16384 | basic Fstar | S&gt;0 &amp; ns&gt;2 | 256 | basic D | S&gt;0 &amp; ns&gt;3 | 512 | basic Deta | S&gt;0 &amp; ns&gt;3 | 512 | basic Dstar | S&gt;0 &amp; ns&gt;3 | 512 | basic sites | i&lt;S | - | site_lists sites_o | i&lt;So | - | site_lists singl | i&lt;Ss | - | site_lists singl_o | i&lt;Sso | - | site_lists</p>
<p>Header: &lt;egglib-cpp/Diversity.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib10Diversity110Diversity1Ev">
<span id="egglib::Diversity1::Diversity1"></span><span class="target" id="project0classegglib_1_1Diversity1_1a3c03c00b547418a04845eeed2e8aac46"></span><code class="descclassname">egglib::Diversity1::</code><code class="descname">Diversity1</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity110Diversity1Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity1D0Ev">
<span id="egglib::Diversity1::~Diversity1"></span><span class="target" id="project0classegglib_1_1Diversity1_1aa6ce5afd426ee174ad4276ad3fab6d49"></span><code class="descclassname">egglib::Diversity1::</code><code class="descname">~Diversity1</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity1D0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity17computeEv">
<span id="egglib::Diversity1::compute"></span><span class="target" id="project0classegglib_1_1Diversity1_1afaa6006cadbdf4267a15efba65396d95"></span>unsigned int <code class="descclassname">egglib::Diversity1::</code><code class="descname">compute</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity17computeEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Compute statistics, return flag but does not reset. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity11DEv">
<span id="egglib::Diversity1::D"></span><span class="target" id="project0classegglib_1_1Diversity1_1a01f1ec34f0907b90a3d974577bf07bc8"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">D</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity11DEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Tajima’s D. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity12DaEv">
<span id="egglib::Diversity1::Da"></span><span class="target" id="project0classegglib_1_1Diversity1_1a67abe29db90220275eb73cd423a65c37"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">Da</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity12DaEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Net pairwise distance for 1st pair. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity14DetaEv">
<span id="egglib::Diversity1::Deta"></span><span class="target" id="project0classegglib_1_1Diversity1_1a80e6b0b902ec102a1ff06a8d4eda64d5"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">Deta</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity14DetaEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Tajima’s D using eta instead of S. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity13DflEv">
<span id="egglib::Diversity1::Dfl"></span><span class="target" id="project0classegglib_1_1Diversity1_1a785d5b9383c79cf881deb32c9d1fb672"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">Dfl</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity13DflEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Fu and Li’s D. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity15DstarEv">
<span id="egglib::Diversity1::Dstar"></span><span class="target" id="project0classegglib_1_1Diversity1_1a0c395dd3049f134872f9e8a4d771150e"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">Dstar</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity15DstarEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Fu and Li’s D*. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity13DxyEv">
<span id="egglib::Diversity1::Dxy"></span><span class="target" id="project0classegglib_1_1Diversity1_1ab6b762eb2497b833711027f082b7b5ad"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">Dxy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity13DxyEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Pairwise distance for 1st pair. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity11EEv">
<span id="egglib::Diversity1::E"></span><span class="target" id="project0classegglib_1_1Diversity1_1a7e490593a4bc559143f79dfefad6a75e"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">E</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity11EEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Zeng et al.’s E. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity13etaEv">
<span id="egglib::Diversity1::eta"></span><span class="target" id="project0classegglib_1_1Diversity1_1a0fe62c2df6e0febcc982193c7ecdf200"></span>unsigned int <code class="descclassname">egglib::Diversity1::</code><code class="descname">eta</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity13etaEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>eta </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity14etaoEv">
<span id="egglib::Diversity1::etao"></span><span class="target" id="project0classegglib_1_1Diversity1_1aab2f5253f514ccb21fe09b2d6070536d"></span>unsigned int <code class="descclassname">egglib::Diversity1::</code><code class="descname">etao</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity14etaoEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>eta for orientable sites </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity11FEv">
<span id="egglib::Diversity1::F"></span><span class="target" id="project0classegglib_1_1Diversity1_1a0ce017971c484684ce5c407d7637c368"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">F</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity11FEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Fu and Li’s F. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity15FstarEv">
<span id="egglib::Diversity1::Fstar"></span><span class="target" id="project0classegglib_1_1Diversity1_1a0e71706ec3bc08037a0a05f111315749"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">Fstar</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity15FstarEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Fu and Li’s F*. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity13HnsEv">
<span id="egglib::Diversity1::Hns"></span><span class="target" id="project0classegglib_1_1Diversity1_1a0fe9f4b1a8ec827a99ca6dbf6f9e1fb1"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">Hns</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity13HnsEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Unstandardized Fay and Wu’s H. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity13HsdEv">
<span id="egglib::Diversity1::Hsd"></span><span class="target" id="project0classegglib_1_1Diversity1_1a492ffcea7cc7a37c04d7c9c9ddc92bd6"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">Hsd</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity13HsdEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Fay and Wu’s H standardized by Zeng et al. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity14loadERK10FreqHolderRK13SiteDiversityj">
<span id="egglib::Diversity1::load__FreqHolderCR.SiteDiversityCR.unsigned-i"></span><span class="target" id="project0classegglib_1_1Diversity1_1aa939788294a62e578587cc417d0a3032"></span>void <code class="descclassname">egglib::Diversity1::</code><code class="descname">load</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10FreqHolderE" title="egglib::FreqHolder">FreqHolder</a> &amp;<em>freqs</em>, <em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib13SiteDiversityE" title="egglib::SiteDiversity">SiteDiversity</a> &amp;<em>div</em>, unsigned int <em>position</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity14loadERK10FreqHolderRK13SiteDiversityj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Analyze a site. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity12lsEv">
<span id="egglib::Diversity1::ls"></span><span class="target" id="project0classegglib_1_1Diversity1_1a53e0e2eecc00ae2a9ef1f78163d114ed"></span>unsigned int <code class="descclassname">egglib::Diversity1::</code><code class="descname">ls</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity12lsEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of loaded sites (with &gt;=2 valid data) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity13lsoEv">
<span id="egglib::Diversity1::lso"></span><span class="target" id="project0classegglib_1_1Diversity1_1a5051b7f9b62345456f2447199e4304c3"></span>unsigned int <code class="descclassname">egglib::Diversity1::</code><code class="descname">lso</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity13lsoEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of loaded orientable sites (with valid data) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity12ltEv">
<span id="egglib::Diversity1::lt"></span><span class="target" id="project0classegglib_1_1Diversity1_1af689d011fa6ff6213bc0640cb79ed1b4"></span>unsigned int <code class="descclassname">egglib::Diversity1::</code><code class="descname">lt</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity12ltEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of loaded sites (total) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity12nMEv">
<span id="egglib::Diversity1::nM"></span><span class="target" id="project0classegglib_1_1Diversity1_1a44b7ba5cd6f521ed5846a5fb7ae72bba"></span>unsigned int <code class="descclassname">egglib::Diversity1::</code><code class="descname">nM</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity12nMEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Sites available for MFDM test. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity15nseffEv">
<span id="egglib::Diversity1::nseff"></span><span class="target" id="project0classegglib_1_1Diversity1_1a85659a1a93f5a3e4093e7e3ff90c4dc3"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">nseff</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity15nseffEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Average number of used samples. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity16nseffoEv">
<span id="egglib::Diversity1::nseffo"></span><span class="target" id="project0classegglib_1_1Diversity1_1afe637d873f529ac3a304b198e197e95d"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">nseffo</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity16nseffoEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Average number of used samples for orientable sites. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity17nsingldEv">
<span id="egglib::Diversity1::nsingld"></span><span class="target" id="project0classegglib_1_1Diversity1_1a63fb437516a21b96581b2c2f9cf66a29"></span>unsigned int <code class="descclassname">egglib::Diversity1::</code><code class="descname">nsingld</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity17nsingldEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number derived singletons. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity15nsmaxEv">
<span id="egglib::Diversity1::nsmax"></span><span class="target" id="project0classegglib_1_1Diversity1_1a9f75fc8936323d38b6925f9b42ab4784"></span>unsigned int <code class="descclassname">egglib::Diversity1::</code><code class="descname">nsmax</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity15nsmaxEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Largest number of used samples. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity16nsmaxoEv">
<span id="egglib::Diversity1::nsmaxo"></span><span class="target" id="project0classegglib_1_1Diversity1_1a130f8a87265928130814e12072099f68"></span>unsigned int <code class="descclassname">egglib::Diversity1::</code><code class="descname">nsmaxo</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity16nsmaxoEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Largest number of used samples for orientable sites. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity12PiEv">
<span id="egglib::Diversity1::Pi"></span><span class="target" id="project0classegglib_1_1Diversity1_1a27c9a389e504b319807eb7c6b17c4b2f"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">Pi</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity12PiEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Sum of He. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity12pMEv">
<span id="egglib::Diversity1::pM"></span><span class="target" id="project0classegglib_1_1Diversity1_1a9da34c27caa011f4de3f722fb3a2d238"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">pM</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity12pMEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Li’s MFDM test p value (large positive value by default) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity111reset_statsEv">
<span id="egglib::Diversity1::reset_stats"></span><span class="target" id="project0classegglib_1_1Diversity1_1a3a0f2873df1d7d2c10b2f98e90ebe58e"></span>void <code class="descclassname">egglib::Diversity1::</code><code class="descname">reset_stats</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity111reset_statsEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Reset counters to 0. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity11SEv">
<span id="egglib::Diversity1::S"></span><span class="target" id="project0classegglib_1_1Diversity1_1a764e5ea2b5151d53b602d846bc365997"></span>unsigned int <code class="descclassname">egglib::Diversity1::</code><code class="descname">S</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity11SEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of polymorphic sites. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity119set_option_multipleEb">
<span id="egglib::Diversity1::set_option_multiple__b"></span><span class="target" id="project0classegglib_1_1Diversity1_1ad0a6f2620831b8b21ebdf1dc5528e8da"></span>void <code class="descclassname">egglib::Diversity1::</code><code class="descname">set_option_multiple</code><span class="sig-paren">(</span>bool <em>b</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity119set_option_multipleEb" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set multiple option (default: False) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity117set_option_ns_setEj">
<span id="egglib::Diversity1::set_option_ns_set__unsigned-i"></span><span class="target" id="project0classegglib_1_1Diversity1_1a0f7f8cf9c5f0bffccdcdd737515ecaed"></span>void <code class="descclassname">egglib::Diversity1::</code><code class="descname">set_option_ns_set</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity117set_option_ns_setEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Set maximum number of samples, for H and co. (default: UNKNOWN) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity15singlEj">
<span id="egglib::Diversity1::singl__unsigned-i"></span><span class="target" id="project0classegglib_1_1Diversity1_1a20129589c9e607342e8e131b5db7772c"></span>unsigned int <code class="descclassname">egglib::Diversity1::</code><code class="descname">singl</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity15singlEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get position of site with a singleton. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity17singl_oEj">
<span id="egglib::Diversity1::singl_o__unsigned-i"></span><span class="target" id="project0classegglib_1_1Diversity1_1a142d19cf5c8b759240651552c5e8ff81"></span>unsigned int <code class="descclassname">egglib::Diversity1::</code><code class="descname">singl_o</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity17singl_oEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get position of site with an orientable singleton. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity14siteEj">
<span id="egglib::Diversity1::site__unsigned-i"></span><span class="target" id="project0classegglib_1_1Diversity1_1a2802b569a64910d32114c04cee876bfe"></span>unsigned int <code class="descclassname">egglib::Diversity1::</code><code class="descname">site</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity14siteEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get position of polymorphic site. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity16site_oEj">
<span id="egglib::Diversity1::site_o__unsigned-i"></span><span class="target" id="project0classegglib_1_1Diversity1_1a6ef6a192ceefe266f5c27b72da25deb2"></span>unsigned int <code class="descclassname">egglib::Diversity1::</code><code class="descname">site_o</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity16site_oEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get position of polymorphic orientable site. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity12SoEv">
<span id="egglib::Diversity1::So"></span><span class="target" id="project0classegglib_1_1Diversity1_1ac7bb91400ac00922ab04ae49eababe72"></span>unsigned int <code class="descclassname">egglib::Diversity1::</code><code class="descname">So</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity12SoEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of polymorphic orientable sites. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity12SsEv">
<span id="egglib::Diversity1::Ss"></span><span class="target" id="project0classegglib_1_1Diversity1_1a5c54223dd33d8e0c7b769a522c633d0e"></span>unsigned int <code class="descclassname">egglib::Diversity1::</code><code class="descname">Ss</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity12SsEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of polymorphic sites with =1 singleton. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity13SsoEv">
<span id="egglib::Diversity1::Sso"></span><span class="target" id="project0classegglib_1_1Diversity1_1a8d6ac8fb18ae5d73767496d614e3219e"></span>unsigned int <code class="descclassname">egglib::Diversity1::</code><code class="descname">Sso</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity13SsoEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of polymorphic orientable sites with =1 singleton. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity16thetaHEv">
<span id="egglib::Diversity1::thetaH"></span><span class="target" id="project0classegglib_1_1Diversity1_1ad8908842b19e1a3e062490d930a100c9"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">thetaH</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity16thetaHEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>ThetaH estimator. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity16thetaLEv">
<span id="egglib::Diversity1::thetaL"></span><span class="target" id="project0classegglib_1_1Diversity1_1a47107a46c041e6f2cab58b6549e238ed"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">thetaL</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity16thetaLEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>ThetaL estimator. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity17thetaPiEv">
<span id="egglib::Diversity1::thetaPi"></span><span class="target" id="project0classegglib_1_1Diversity1_1ae1dcda9b8649d14ba094b65e435c6a17"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">thetaPi</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity17thetaPiEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>thetaPi estimator (using orientable sites) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity16thetaWEv">
<span id="egglib::Diversity1::thetaW"></span><span class="target" id="project0classegglib_1_1Diversity1_1afcce5e71483c9cc3fd49ff4fccbf7865"></span>double <code class="descclassname">egglib::Diversity1::</code><code class="descname">thetaW</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity16thetaWEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Theta estimator based on S. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity112toggle_basicEv">
<span id="egglib::Diversity1::toggle_basic"></span><span class="target" id="project0classegglib_1_1Diversity1_1aded41978ccfe6d58cfadfc4ee69aceaf"></span>void <code class="descclassname">egglib::Diversity1::</code><code class="descname">toggle_basic</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity112toggle_basicEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Activate basic per-gene. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity110toggle_offEv">
<span id="egglib::Diversity1::toggle_off"></span><span class="target" id="project0classegglib_1_1Diversity1_1a898807895f3c1114c7e0e654b237ab28"></span>void <code class="descclassname">egglib::Diversity1::</code><code class="descname">toggle_off</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity110toggle_offEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Cancel all flags. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity114toggle_ori_divEv">
<span id="egglib::Diversity1::toggle_ori_div"></span><span class="target" id="project0classegglib_1_1Diversity1_1a3197766efb280c79fc02fca728f000eb"></span>void <code class="descclassname">egglib::Diversity1::</code><code class="descname">toggle_ori_div</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity114toggle_ori_divEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Activate per-gene oriented. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity115toggle_ori_siteEv">
<span id="egglib::Diversity1::toggle_ori_site"></span><span class="target" id="project0classegglib_1_1Diversity1_1a6a8136879c9f074de178f497670492c1"></span>void <code class="descclassname">egglib::Diversity1::</code><code class="descname">toggle_ori_site</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity115toggle_ori_siteEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Activate per-site oriented. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity117toggle_site_listsEv">
<span id="egglib::Diversity1::toggle_site_lists"></span><span class="target" id="project0classegglib_1_1Diversity1_1a7afad1561da95ca5a2193bc528441cfb"></span>void <code class="descclassname">egglib::Diversity1::</code><code class="descname">toggle_site_lists</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity117toggle_site_listsEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Activate lists of site positions. </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="diversity2">
<h3>Diversity2<a class="headerlink" href="#diversity2" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib10Diversity2E">
<span id="egglib::Diversity2"></span><span class="target" id="project0classegglib_1_1Diversity2"></span><em class="property">class </em></dt>
<dd><p>Compute population summary statistics from an array of sites. </p>
<p>Diversity instances cannot be copied. This class is designed to allow reuse of objects without unnecessary memory reallocation.</p>
<p>This class computes statistics that require access to the full Site instance (and therefore to the individual allele of each individual). Sites with missing data are ignored when computing Wall’s statistics.</p>
<p>Meaning of flag:<ul class="simple">
<li>flag&amp;1 an error occurred (+ one of 2, 4, 8, 16, 32)</li>
<li>flag&amp;2 error: less than 2 samples (including missing)</li>
<li>flag&amp;4 error: inconsistent number of samples</li>
<li>flag&amp;8 error: inconsistent ploidy</li>
<li>flag&amp;16 error: inconsistent frequency holder (sample size not checked)</li>
<li>flag&amp;32 error: provided <a class="reference internal" href="#project0classegglib_1_1SiteDiversity"><span class="std std-ref">SiteDiversity</span></a> does not have proper data</li>
<li>flag&amp;64 at least 1 polymorphic site with at least 2 non-missing samples</li>
<li>flag&amp;128 at least 1 polymorphic, orientable site with at least 2 non-missing samples</li>
<li>flag&amp;256 computed R2, R3, R4, and Ch</li>
<li>flag&amp;512 computed R2E, R3E, R4E, and ChE</li>
<li>flag&amp;1024 computed B and Q (at least 2 sites with no missing data)</li>
</ul>
</p>
<p>Header: &lt;egglib-cpp/Diversity.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib10Diversity210Diversity2Ev">
<span id="egglib::Diversity2::Diversity2"></span><span class="target" id="project0classegglib_1_1Diversity2_1a210e0d2e0996a22c4b6e54aca0ae6c9d"></span><code class="descclassname">egglib::Diversity2::</code><code class="descname">Diversity2</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity210Diversity2Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity2D0Ev">
<span id="egglib::Diversity2::~Diversity2"></span><span class="target" id="project0classegglib_1_1Diversity2_1a2bda457ac90bdc4c08fbbf66d5e62785"></span><code class="descclassname">egglib::Diversity2::</code><code class="descname">~Diversity2</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity2D0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity21BEv">
<span id="egglib::Diversity2::B"></span><span class="target" id="project0classegglib_1_1Diversity2_1ab6ee6f0f970c90d4a148a64c074f09f5"></span>double <code class="descclassname">egglib::Diversity2::</code><code class="descname">B</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity21BEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Wall’s statistic. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity22ChEv">
<span id="egglib::Diversity2::Ch"></span><span class="target" id="project0classegglib_1_1Diversity2_1ac1693c7d159172d48ca13e3379a35173"></span>double <code class="descclassname">egglib::Diversity2::</code><code class="descname">Ch</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity22ChEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Ramos-Onsins and Rozas’s statistic. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity23ChEEv">
<span id="egglib::Diversity2::ChE"></span><span class="target" id="project0classegglib_1_1Diversity2_1a20ea8eaa0d55ca6bdaed1eaea5367c82"></span>double <code class="descclassname">egglib::Diversity2::</code><code class="descname">ChE</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity23ChEEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Ramos-Onsins and Rozas’s statistic. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity27computeEv">
<span id="egglib::Diversity2::compute"></span><span class="target" id="project0classegglib_1_1Diversity2_1afaa6006cadbdf4267a15efba65396d95"></span>unsigned int <code class="descclassname">egglib::Diversity2::</code><code class="descname">compute</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity27computeEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Compute singletons and/or partitions stats, return flag. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity21kEv">
<span id="egglib::Diversity2::k"></span><span class="target" id="project0classegglib_1_1Diversity2_1af613e2b0ce080dc98b634661eca4e6ae"></span>double <code class="descclassname">egglib::Diversity2::</code><code class="descname">k</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity21kEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Average number of differences. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity22koEv">
<span id="egglib::Diversity2::ko"></span><span class="target" id="project0classegglib_1_1Diversity2_1a8e645219636cb7c7ac5b6a410e8b4c54"></span>double <code class="descclassname">egglib::Diversity2::</code><code class="descname">ko</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity22koEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Average number of differences at orientable sites. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity24loadERK10SiteHolderRK13SiteDiversityRK10FreqHolder">
<span id="egglib::Diversity2::load__SiteHolderCR.SiteDiversityCR.FreqHolderCR"></span><span class="target" id="project0classegglib_1_1Diversity2_1a50cecc9c8aa3a1c2e8fa4cb50e9f7d0b"></span>void <code class="descclassname">egglib::Diversity2::</code><code class="descname">load</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site</em>, <em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib13SiteDiversityE" title="egglib::SiteDiversity">SiteDiversity</a> &amp;<em>div</em>, <em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10FreqHolderE" title="egglib::FreqHolder">FreqHolder</a> &amp;<em>frq</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity24loadERK10SiteHolderRK13SiteDiversityRK10FreqHolder" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Load site with its matching structure (requires basic stats) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity29num_clearEv">
<span id="egglib::Diversity2::num_clear"></span><span class="target" id="project0classegglib_1_1Diversity2_1ae2463dcded7ec31ce66270ada0b3d669"></span>unsigned int <code class="descclassname">egglib::Diversity2::</code><code class="descname">num_clear</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity29num_clearEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of sites with 0 missing data (Wall stats) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity214num_orientableEv">
<span id="egglib::Diversity2::num_orientable"></span><span class="target" id="project0classegglib_1_1Diversity2_1a5894e64ddc11a1c23d0902dac33bfa3c"></span>unsigned int <code class="descclassname">egglib::Diversity2::</code><code class="descname">num_orientable</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity214num_orientableEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of orientable sites. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity213num_sequencesEv">
<span id="egglib::Diversity2::num_sequences"></span><span class="target" id="project0classegglib_1_1Diversity2_1a0a529b8aa92fb89fee9ee4ba5c76b41c"></span>unsigned int <code class="descclassname">egglib::Diversity2::</code><code class="descname">num_sequences</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity213num_sequencesEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of sequences. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity29num_sitesEv">
<span id="egglib::Diversity2::num_sites"></span><span class="target" id="project0classegglib_1_1Diversity2_1aae69a5b25d3e7d7ae4027a93b0893d48"></span>unsigned int <code class="descclassname">egglib::Diversity2::</code><code class="descname">num_sites</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity29num_sitesEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of loaded sites (only polymorphic) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity21QEv">
<span id="egglib::Diversity2::Q"></span><span class="target" id="project0classegglib_1_1Diversity2_1a37900dd97254d9a3c38b621049b1d142"></span>double <code class="descclassname">egglib::Diversity2::</code><code class="descname">Q</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity21QEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Wall’s statistic. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity22R2Ev">
<span id="egglib::Diversity2::R2"></span><span class="target" id="project0classegglib_1_1Diversity2_1ade3b82f8ae9d910a8205be169185977d"></span>double <code class="descclassname">egglib::Diversity2::</code><code class="descname">R2</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity22R2Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Ramos-Onsins and Rozas’s statistic. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity23R2EEv">
<span id="egglib::Diversity2::R2E"></span><span class="target" id="project0classegglib_1_1Diversity2_1aad6463d9b8078376c8bf85873704a7a8"></span>double <code class="descclassname">egglib::Diversity2::</code><code class="descname">R2E</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity23R2EEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Ramos-Onsins and Rozas’s statistic. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity22R3Ev">
<span id="egglib::Diversity2::R3"></span><span class="target" id="project0classegglib_1_1Diversity2_1ac6fad44a9908844aa8da9bffaafacc42"></span>double <code class="descclassname">egglib::Diversity2::</code><code class="descname">R3</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity22R3Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Ramos-Onsins and Rozas’s statistic. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity23R3EEv">
<span id="egglib::Diversity2::R3E"></span><span class="target" id="project0classegglib_1_1Diversity2_1aeeff77fb3bfbb7c9c180a852d18c002e"></span>double <code class="descclassname">egglib::Diversity2::</code><code class="descname">R3E</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity23R3EEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Ramos-Onsins and Rozas’s statistic. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity22R4Ev">
<span id="egglib::Diversity2::R4"></span><span class="target" id="project0classegglib_1_1Diversity2_1a75f3375571eb03358d3de7afe7584788"></span>double <code class="descclassname">egglib::Diversity2::</code><code class="descname">R4</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity22R4Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Ramos-Onsins and Rozas’s statistic. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity23R4EEv">
<span id="egglib::Diversity2::R4E"></span><span class="target" id="project0classegglib_1_1Diversity2_1a0715c27a99978e999b547f550d367a7f"></span>double <code class="descclassname">egglib::Diversity2::</code><code class="descname">R4E</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity23R4EEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Ramos-Onsins and Rozas’s statistic. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity25resetEv">
<span id="egglib::Diversity2::reset"></span><span class="target" id="project0classegglib_1_1Diversity2_1ad20897c5c8bd47f5d4005989bead0e55"></span>void <code class="descclassname">egglib::Diversity2::</code><code class="descname">reset</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity25resetEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Restore all variables to the default state (except toggled flags) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity219set_option_multipleEb">
<span id="egglib::Diversity2::set_option_multiple__b"></span><span class="target" id="project0classegglib_1_1Diversity2_1ad0a6f2620831b8b21ebdf1dc5528e8da"></span>void <code class="descclassname">egglib::Diversity2::</code><code class="descname">set_option_multiple</code><span class="sig-paren">(</span>bool <em>b</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity219set_option_multipleEb" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Toggle option for multiple alleles. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity210toggle_offEv">
<span id="egglib::Diversity2::toggle_off"></span><span class="target" id="project0classegglib_1_1Diversity2_1a898807895f3c1114c7e0e654b237ab28"></span>void <code class="descclassname">egglib::Diversity2::</code><code class="descname">toggle_off</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity210toggle_offEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Cancel flags. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity217toggle_partitionsEv">
<span id="egglib::Diversity2::toggle_partitions"></span><span class="target" id="project0classegglib_1_1Diversity2_1a6aac8276dbc01f56d6dd8fffe03a488a"></span>void <code class="descclassname">egglib::Diversity2::</code><code class="descname">toggle_partitions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity217toggle_partitionsEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Activate computation of B and Q stats (must be set before <a class="reference internal" href="#project0classegglib_1_1Diversity2_1a50cecc9c8aa3a1c2e8fa4cb50e9f7d0b"><span class="std std-ref">load()</span></a> </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Diversity217toggle_singletonsEv">
<span id="egglib::Diversity2::toggle_singletons"></span><span class="target" id="project0classegglib_1_1Diversity2_1afdcfaa09ebb2c896abd6fa29c623bbf5"></span>void <code class="descclassname">egglib::Diversity2::</code><code class="descname">toggle_singletons</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Diversity217toggle_singletonsEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Activate computation of Rx/Ch RxE/ChE stats. </p>
</dd></dl>

</div>
</dd></dl>

</div>
</div>
<div class="section" id="other-statistics">
<h2>Other statistics<a class="headerlink" href="#other-statistics" title="Permalink to this headline">¶</a></h2>
<div class="section" id="haplotypes">
<h3>Haplotypes<a class="headerlink" href="#haplotypes" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib10HaplotypesE">
<span id="egglib::Haplotypes"></span><span class="target" id="project0classegglib_1_1Haplotypes"></span><em class="property">class </em></dt>
<dd><p>Identifies haplotypes from a set of sites. </p>
<p>How to use this class:</p>
<p><ul class="simple">
<li>1) Setup (or reset) and an optional structure.</li>
<li>2) Load all sites with load(site). <a class="reference internal" href="#project0classegglib_1_1Haplotypes"><span class="std std-ref">Haplotypes</span></a> are computed and all samples with at least one missing data are marked as missing. While loading sites, you can monitor the values of:<ul>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a1f5d58e381c22fa999f9a650950b4b48"><span class="std std-ref">n_ing()</span></a></li>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a0e4225a0f50ef8d7234c9d0fd06e7371"><span class="std std-ref">n_otg()</span></a></li>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1ac8d838bef4e4cbe6ba664785e5ee53dc"><span class="std std-ref">n_sam()</span></a></li>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a164dcdf4c10a0af155ee6f7d6b0a1a47"><span class="std std-ref">nt_hapl()</span></a></li>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1acfc39ab76b97dc59f0fef2879c3e00cf"><span class="std std-ref">hapl()</span></a></li>
<li>map()</li>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a15862bbf68404f0d9a6ff325b5c3f4ec"><span class="std std-ref">n_sites()</span></a></li>
</ul>
</li>
<li>3) Call <a class="reference internal" href="#project0classegglib_1_1Haplotypes_1acb1df49f79cb1bdf986db0202581a5b6"><span class="std std-ref">cp_haplotypes()</span></a> to finalize haplotype processing. After that you may use:<ul>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a9f67bf51aaeaaef0b9457f82b740f62b"><span class="std std-ref">ne_ing()</span></a></li>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a5a83ef47d4663aa8d0b6e634b1e2abd7"><span class="std std-ref">ne_otg()</span></a></li>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a340c82a6185a600cb592c6bd8d18e702"><span class="std std-ref">ni_hapl()</span></a></li>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a25cf1e431bd4fd8234eaf66bb3906ef5"><span class="std std-ref">ng_hapl()</span></a></li>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a360703cbcbc75021d2469a8a7130cbc1"><span class="std std-ref">freq_i()</span></a></li>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1af254b1db3f2564798b1ccc049faaf7cc"><span class="std std-ref">freq_o()</span></a></li>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a2e67064f3bb28786504aea9f18d05e4b"><span class="std std-ref">nstot()</span></a></li>
</ul>
</li>
<li>4) If you wish (but you don’t have to), you can try and guess the haplotype of samples with missing data. For this you need first to call <a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a785550d56f3efd22112bb0906f41babc"><span class="std std-ref">prepare_impute()</span></a>. After that, you may use:<ul>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1ac444ae80906dc8051c0bb50de6837802"><span class="std std-ref">n_mis()</span></a></li>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a03f6b3c76c03889d0336211dd322d3ff"><span class="std std-ref">mis_idx()</span></a></li>
</ul>
</li>
<li>5) If you impute, load again all sites with solve(). This may change the value of:<ul>
<li>map()</li>
</ul>
</li>
<li>6) If you impute, you are required to call <a class="reference internal" href="#project0classegglib_1_1Haplotypes_1acb8eb330d631a4e3534fb34ce92ceaa9"><span class="std std-ref">impute()</span></a> (and otherwise you can’t). If you do, the following values will be updated [note: maybe impute and cp_haplotypes do exactly the same]<ul>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a340c82a6185a600cb592c6bd8d18e702"><span class="std std-ref">ni_hapl()</span></a></li>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a25cf1e431bd4fd8234eaf66bb3906ef5"><span class="std std-ref">ng_hapl()</span></a></li>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a360703cbcbc75021d2469a8a7130cbc1"><span class="std std-ref">freq_i()</span></a></li>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1af254b1db3f2564798b1ccc049faaf7cc"><span class="std std-ref">freq_o()</span></a> and these values will be set:</li>
<li>num_potential()</li>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1abb582169b5a066fbd7e035c1bf6dd46c"><span class="std std-ref">potential()</span></a></li>
</ul>
</li>
<li>7) After whether or not you performed 4-6, you can now call this to make a site:<ul>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a0d188eb99519eb87fa7bb2b9894a1056"><span class="std std-ref">get_site()</span></a></li>
</ul>
</li>
<li>8) Whether or not you performed 4-6 and/or 7, you can now call <a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a94f2525148a81a019d4e89684b86416d"><span class="std std-ref">cp_dist()</span></a> that will let you access to the distance matrix:<ul>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a929b96f5485c0aa02f6908a28b38b616"><span class="std std-ref">dist()</span></a></li>
</ul>
</li>
<li>9) To compute stats, call <a class="reference internal" href="#project0classegglib_1_1Haplotypes_1af01cc54fc86ceebdf31825f9b37bee86"><span class="std std-ref">cp_stats()</span></a> (you must have set a structure and called 8). Then you may use:<ul>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1aa403f047685ed28e3d8ad4684ed961d6"><span class="std std-ref">Fst()</span></a></li>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a6755605e91423a2999d8f095767db9ee"><span class="std std-ref">Kst()</span></a> The function <a class="reference internal" href="#project0classegglib_1_1Haplotypes_1af01cc54fc86ceebdf31825f9b37bee86"><span class="std std-ref">cp_stats()</span></a> returns a flag: 0 (no stats computed), 1 (Fst computed), or 2 (both Fst and Kst computed).</li>
</ul>
</li>
<li>10) You can also call this method (requires a structure and 8 but not 9, and valid if nstot is &gt;1):<ul>
<li><a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a54586f0e680323c8b744ed80e314a43a"><span class="std std-ref">Snn()</span></a></li>
</ul>
</li>
</ul>
</p>
<p>Header: &lt;egglib-cpp/Haplotypes.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes10HaplotypesEv">
<span id="egglib::Haplotypes::Haplotypes"></span><span class="target" id="project0classegglib_1_1Haplotypes_1ae306d6da2ea2355ca06366b8d0319bc7"></span><code class="descclassname">egglib::Haplotypes::</code><code class="descname">Haplotypes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes10HaplotypesEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10HaplotypesD0Ev">
<span id="egglib::Haplotypes::~Haplotypes"></span><span class="target" id="project0classegglib_1_1Haplotypes_1afa6a98385fc6fd681362537926f0198d"></span><code class="descclassname">egglib::Haplotypes::</code><code class="descname">~Haplotypes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10HaplotypesD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes7cp_distEv">
<span id="egglib::Haplotypes::cp_dist"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a94f2525148a81a019d4e89684b86416d"></span>void <code class="descclassname">egglib::Haplotypes::</code><code class="descname">cp_dist</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes7cp_distEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Compute distance matrix. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes13cp_haplotypesEv">
<span id="egglib::Haplotypes::cp_haplotypes"></span><span class="target" id="project0classegglib_1_1Haplotypes_1acb1df49f79cb1bdf986db0202581a5b6"></span>void <code class="descclassname">egglib::Haplotypes::</code><code class="descname">cp_haplotypes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes13cp_haplotypesEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Finalize haplotype estimation. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes8cp_statsEv">
<span id="egglib::Haplotypes::cp_stats"></span><span class="target" id="project0classegglib_1_1Haplotypes_1af01cc54fc86ceebdf31825f9b37bee86"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">cp_stats</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes8cp_statsEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Compute differentiation stats. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes4distEjj">
<span id="egglib::Haplotypes::dist__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a929b96f5485c0aa02f6908a28b38b616"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">dist</code><span class="sig-paren">(</span>unsigned int, unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes4distEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Distance matrix entry (0&lt;=j&lt;i&lt;_nt_hapl) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes6freq_iEj">
<span id="egglib::Haplotypes::freq_i__unsigned-i"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a360703cbcbc75021d2469a8a7130cbc1"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">freq_i</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes6freq_iEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Frequency of haplotype in intgroup. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes6freq_oEj">
<span id="egglib::Haplotypes::freq_o__unsigned-i"></span><span class="target" id="project0classegglib_1_1Haplotypes_1af254b1db3f2564798b1ccc049faaf7cc"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">freq_o</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes6freq_oEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Frequency of haplotype in outgroup. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes3FstEv">
<span id="egglib::Haplotypes::Fst"></span><span class="target" id="project0classegglib_1_1Haplotypes_1aa403f047685ed28e3d8ad4684ed961d6"></span>double <code class="descclassname">egglib::Haplotypes::</code><code class="descname">Fst</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes3FstEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Fst value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes8get_siteER10SiteHolder">
<span id="egglib::Haplotypes::get_site__SiteHolderR"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a0d188eb99519eb87fa7bb2b9894a1056"></span>void <code class="descclassname">egglib::Haplotypes::</code><code class="descname">get_site</code><span class="sig-paren">(</span><a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes8get_siteER10SiteHolder" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Get haplotypic data as a site (recycle passed object) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes4haplEjj">
<span id="egglib::Haplotypes::hapl__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1Haplotypes_1acfc39ab76b97dc59f0fef2879c3e00cf"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">hapl</code><span class="sig-paren">(</span>unsigned int, unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes4haplEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get site j of haplotype i. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes6imputeEv">
<span id="egglib::Haplotypes::impute"></span><span class="target" id="project0classegglib_1_1Haplotypes_1acb8eb330d631a4e3534fb34ce92ceaa9"></span>void <code class="descclassname">egglib::Haplotypes::</code><code class="descname">impute</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes6imputeEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Try to guess haplotype of samples with missing data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes3KstEv">
<span id="egglib::Haplotypes::Kst"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a6755605e91423a2999d8f095767db9ee"></span>double <code class="descclassname">egglib::Haplotypes::</code><code class="descname">Kst</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes3KstEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Gst value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes4loadERK10SiteHolder">
<span id="egglib::Haplotypes::load__SiteHolderCR"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a0832d6352cb2fcc704f56bc1cd16395a"></span>void <code class="descclassname">egglib::Haplotypes::</code><code class="descname">load</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes4loadERK10SiteHolder" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Process a site (implies call to <a class="reference internal" href="#project0classegglib_1_1Haplotypes_1a659bf68c31d60f89dc6e301e3b2245c1"><span class="std std-ref">setup()</span></a>) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes10map_sampleEj">
<span id="egglib::Haplotypes::map_sample__unsigned-i"></span><span class="target" id="project0classegglib_1_1Haplotypes_1af49396cd753f43d168c06b340b2b701c"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">map_sample</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes10map_sampleEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get haplotype index of a sample. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes7mis_idxEj">
<span id="egglib::Haplotypes::mis_idx__unsigned-i"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a03f6b3c76c03889d0336211dd322d3ff"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">mis_idx</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes7mis_idxEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Index of one of the samples with missing data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes5n_ingEv">
<span id="egglib::Haplotypes::n_ing"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a1f5d58e381c22fa999f9a650950b4b48"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">n_ing</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes5n_ingEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Total number of ingroup samples. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes5n_misEv">
<span id="egglib::Haplotypes::n_mis"></span><span class="target" id="project0classegglib_1_1Haplotypes_1ac444ae80906dc8051c0bb50de6837802"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">n_mis</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes5n_misEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of samples with missing data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes9n_missingEj">
<span id="egglib::Haplotypes::n_missing__unsigned-i"></span><span class="target" id="project0classegglib_1_1Haplotypes_1aa8407fbf7c56ea4a62dfab040de81ef7"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">n_missing</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes9n_missingEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of missing data per sample. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes5n_otgEv">
<span id="egglib::Haplotypes::n_otg"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a0e4225a0f50ef8d7234c9d0fd06e7371"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">n_otg</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes5n_otgEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Total number of outgroup samples. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes5n_popEv">
<span id="egglib::Haplotypes::n_pop"></span><span class="target" id="project0classegglib_1_1Haplotypes_1ab26b98116db68726cae2bf4f016029b9"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">n_pop</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes5n_popEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of populations. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes11n_potentialEj">
<span id="egglib::Haplotypes::n_potential__unsigned-i"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a3cf15d36170f20b803393f4c08e31d74"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">n_potential</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes11n_potentialEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of compatible haplotypes for a sample with missing data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes5n_samEv">
<span id="egglib::Haplotypes::n_sam"></span><span class="target" id="project0classegglib_1_1Haplotypes_1ac8d838bef4e4cbe6ba664785e5ee53dc"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">n_sam</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes5n_samEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Total number of samples. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes7n_sitesEv">
<span id="egglib::Haplotypes::n_sites"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a15862bbf68404f0d9a6ff325b5c3f4ec"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">n_sites</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes7n_sitesEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of sites. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes6ne_ingEv">
<span id="egglib::Haplotypes::ne_ing"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a9f67bf51aaeaaef0b9457f82b740f62b"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">ne_ing</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes6ne_ingEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Non-missing number of ingroup samples. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes6ne_otgEv">
<span id="egglib::Haplotypes::ne_otg"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a5a83ef47d4663aa8d0b6e634b1e2abd7"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">ne_otg</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes6ne_otgEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Non-missing number of outgroup samples. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes6ne_popEv">
<span id="egglib::Haplotypes::ne_pop"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a85f3347bf91dfdae57e1c1fce57195f6"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">ne_pop</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes6ne_popEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of populations with. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes7ng_haplEv">
<span id="egglib::Haplotypes::ng_hapl"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a25cf1e431bd4fd8234eaf66bb3906ef5"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">ng_hapl</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes7ng_haplEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of haplotypes at non-zero frequency overall. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes7ni_haplEv">
<span id="egglib::Haplotypes::ni_hapl"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a340c82a6185a600cb592c6bd8d18e702"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">ni_hapl</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes7ni_haplEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of haplotypes at non-zero ingroup frequency. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes5nstotEv">
<span id="egglib::Haplotypes::nstot"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a2e67064f3bb28786504aea9f18d05e4b"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">nstot</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes5nstotEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of samples (non-missing and not ignored by structure) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes7nt_haplEv">
<span id="egglib::Haplotypes::nt_hapl"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a164dcdf4c10a0af155ee6f7d6b0a1a47"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">nt_hapl</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes7nt_haplEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Total number of haplotypes (including truncated ones) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes9potentialEjj">
<span id="egglib::Haplotypes::potential__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1Haplotypes_1abb582169b5a066fbd7e035c1bf6dd46c"></span>unsigned int <code class="descclassname">egglib::Haplotypes::</code><code class="descname">potential</code><span class="sig-paren">(</span>unsigned int, unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes9potentialEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Index of a potential haplotype. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes14prepare_imputeEj">
<span id="egglib::Haplotypes::prepare_impute__unsigned-i"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a785550d56f3efd22112bb0906f41babc"></span>void <code class="descclassname">egglib::Haplotypes::</code><code class="descname">prepare_impute</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes14prepare_imputeEj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Initialize required tables for imputing. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes7resolveERK10SiteHolder">
<span id="egglib::Haplotypes::resolve__SiteHolderCR"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a98400d2ad48542385ab4adc94e6d9091"></span>void <code class="descclassname">egglib::Haplotypes::</code><code class="descname">resolve</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes7resolveERK10SiteHolder" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Second pass: try to resolve missing data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes5setupEPK15StructureHolder">
<span id="egglib::Haplotypes::setup__StructureHolderCP"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a659bf68c31d60f89dc6e301e3b2245c1"></span>void <code class="descclassname">egglib::Haplotypes::</code><code class="descname">setup</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib15StructureHolderE" title="egglib::StructureHolder">StructureHolder</a> *<em>struc</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes5setupEPK15StructureHolder" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Setup/reset instance (with optional structure object) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10Haplotypes3SnnEv">
<span id="egglib::Haplotypes::Snn"></span><span class="target" id="project0classegglib_1_1Haplotypes_1a54586f0e680323c8b744ed80e314a43a"></span>double <code class="descclassname">egglib::Haplotypes::</code><code class="descname">Snn</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10Haplotypes3SnnEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Snn value computed on the fly. </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="paralogpi">
<h3>ParalogPi<a class="headerlink" href="#paralogpi" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib9ParalogPiE">
<span id="egglib::ParalogPi"></span><span class="target" id="project0classegglib_1_1ParalogPi"></span><em class="property">class </em></dt>
<dd><p>Specifically designed to compute Innan 2003 statistics for a gene family. </p>
<p>This class computes the within- and between-paralog Pi of Innan (2003). The within-paralog Pi is the same as the standard Pi, except that it is not unbiased. The between-paralog Pi is the same as Dxy, taking the paralogs as populations, except that one pair of genes (paralogs from the same sample) is not considered.</p>
<p>Setup provides the two structure objects describing respectively the structure in paralogs and the structure in samples (two different structure objects are required because they are necessarily non-nested). It is required that the structures of interest are loaded as population levels. Cluster levels are ignored. It is required that ploidy is 1 (if you have genotypic data, skip the individual level). Required tests: ploidy of both structure and of all sites is equal to 1, the maximum index of the paralog structure is represented in all sites (other disagrements are treated as missing data). </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib9ParalogPi9ParalogPiEv">
<span id="egglib::ParalogPi::ParalogPi"></span><span class="target" id="project0classegglib_1_1ParalogPi_1a70c53376374c7dbadd8b0d036f62a0bb"></span><code class="descclassname">egglib::ParalogPi::</code><code class="descname">ParalogPi</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9ParalogPi9ParalogPiEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor (default: 0 pop) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib9ParalogPiD0Ev">
<span id="egglib::ParalogPi::~ParalogPi"></span><span class="target" id="project0classegglib_1_1ParalogPi_1ad05952c918ab47824648b661dd290f75"></span><code class="descclassname">egglib::ParalogPi::</code><code class="descname">~ParalogPi</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9ParalogPiD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib9ParalogPi4loadERK10SiteHolder">
<span id="egglib::ParalogPi::load__SiteHolderCR"></span><span class="target" id="project0classegglib_1_1ParalogPi_1a0832d6352cb2fcc704f56bc1cd16395a"></span>void <code class="descclassname">egglib::ParalogPi::</code><code class="descname">load</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9ParalogPi4loadERK10SiteHolder" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Load a site. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib9ParalogPi12num_paralogsEv">
<span id="egglib::ParalogPi::num_paralogs"></span><span class="target" id="project0classegglib_1_1ParalogPi_1ad6fcb91b70c246620186abe85ad5f742"></span>unsigned int <code class="descclassname">egglib::ParalogPi::</code><code class="descname">num_paralogs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9ParalogPi12num_paralogsEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of copies (or: number of pops) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib9ParalogPi11num_samplesEv">
<span id="egglib::ParalogPi::num_samples"></span><span class="target" id="project0classegglib_1_1ParalogPi_1aba32ed97991ec044c35d2692148f02d4"></span>unsigned int <code class="descclassname">egglib::ParalogPi::</code><code class="descname">num_samples</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9ParalogPi11num_samplesEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of samples (or: size of each pop) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib9ParalogPi14num_sites_pairEjj">
<span id="egglib::ParalogPi::num_sites_pair__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1ParalogPi_1a1bcca3f23be39ffbe1b0b15087bef548"></span>unsigned int <code class="descclassname">egglib::ParalogPi::</code><code class="descname">num_sites_pair</code><span class="sig-paren">(</span>unsigned int, unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9ParalogPi14num_sites_pairEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of analyzed sites for a pair of paralogs. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib9ParalogPi17num_sites_paralogEj">
<span id="egglib::ParalogPi::num_sites_paralog__unsigned-i"></span><span class="target" id="project0classegglib_1_1ParalogPi_1aee20e016287016ca10f492dd0aec0b9b"></span>unsigned int <code class="descclassname">egglib::ParalogPi::</code><code class="descname">num_sites_paralog</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9ParalogPi17num_sites_paralogEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of analyzed sites for a paralog. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib9ParalogPi13num_sites_totEv">
<span id="egglib::ParalogPi::num_sites_tot"></span><span class="target" id="project0classegglib_1_1ParalogPi_1a12dc65822b0566f6045b2f865b608f70"></span>unsigned int <code class="descclassname">egglib::ParalogPi::</code><code class="descname">num_sites_tot</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9ParalogPi13num_sites_totEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Total number of analyzed sites. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib9ParalogPi3PibEjj">
<span id="egglib::ParalogPi::Pib__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1ParalogPi_1abf0a16f04065addb683c683b2a6c5715"></span>double <code class="descclassname">egglib::ParalogPi::</code><code class="descname">Pib</code><span class="sig-paren">(</span>unsigned int, unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9ParalogPi3PibEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Between-paralog Pi. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib9ParalogPi3PiwEj">
<span id="egglib::ParalogPi::Piw__unsigned-i"></span><span class="target" id="project0classegglib_1_1ParalogPi_1a53766bcf147606700a5bd3b537a9ca95"></span>double <code class="descclassname">egglib::ParalogPi::</code><code class="descname">Piw</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9ParalogPi3PiwEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Within-paralog Pi. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib9ParalogPi5resetERK15StructureHolderRK15StructureHolder">
<span id="egglib::ParalogPi::reset__StructureHolderCR.StructureHolderCR"></span><span class="target" id="project0classegglib_1_1ParalogPi_1aa5feb7bf649e3e0510dc12ea68e24e42"></span>void <code class="descclassname">egglib::ParalogPi::</code><code class="descname">reset</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib15StructureHolderE" title="egglib::StructureHolder">StructureHolder</a> &amp;<em>str_prl</em>, <em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib15StructureHolderE" title="egglib::StructureHolder">StructureHolder</a> &amp;<em>str_idv</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib9ParalogPi5resetERK15StructureHolderRK15StructureHolder" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Reset and setup structure. </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="rd">
<h3>Rd<a class="headerlink" href="#rd" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib2RdE">
<span id="egglib::Rd"></span><span class="target" id="project0classegglib_1_1Rd"></span><em class="property">class </em></dt>
<dd><p>Compute the bar{r_d} (or rD) statistic. </p>
<p><a class="reference internal" href="#project0classegglib_1_1Rd"><span class="std std-ref">Rd</span></a> instances cannot be copied. The procedure is:<ul class="simple">
<li>Load as many sites as needed. They are analysed on the fly. The number of samples and ploidy are expected to match (as well as the order of individuals and alleles). If there is a mismatch in ploidy and/or number of samples, the <a class="reference internal" href="#project0classegglib_1_1Rd"><span class="std std-ref">Rd</span></a> value will be UNDEF.</li>
<li>Compute the <a class="reference internal" href="#project0classegglib_1_1Rd"><span class="std std-ref">Rd</span></a> value (resets the instance). </li>
</ul>
</p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib2Rd2RdEv">
<span id="egglib::Rd::Rd"></span><span class="target" id="project0classegglib_1_1Rd_1a56bf83f45126170f3f749a27bbe332e4"></span><code class="descclassname">egglib::Rd::</code><code class="descname">Rd</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib2Rd2RdEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib2RdD0Ev">
<span id="egglib::Rd::~Rd"></span><span class="target" id="project0classegglib_1_1Rd_1a2ac73716dd5c8214fa6d882065701335"></span><code class="descclassname">egglib::Rd::</code><code class="descname">~Rd</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib2RdD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib2Rd7computeEv">
<span id="egglib::Rd::compute"></span><span class="target" id="project0classegglib_1_1Rd_1ad13887b58165a5223e1aa89764b876d3"></span>double <code class="descclassname">egglib::Rd::</code><code class="descname">compute</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib2Rd7computeEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Compute rD (UNDEF if not computable) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib2Rd4loadERK10SiteHolder">
<span id="egglib::Rd::load__SiteHolderCR"></span><span class="target" id="project0classegglib_1_1Rd_1a0832d6352cb2fcc704f56bc1cd16395a"></span>void <code class="descclassname">egglib::Rd::</code><code class="descname">load</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib2Rd4loadERK10SiteHolder" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Load a site. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib2Rd9num_indivEv">
<span id="egglib::Rd::num_indiv"></span><span class="target" id="project0classegglib_1_1Rd_1a69ffd0fc203827dc6a8ab3a2190be461"></span>unsigned int <code class="descclassname">egglib::Rd::</code><code class="descname">num_indiv</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib2Rd9num_indivEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get number of individuals as provided to configure() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib2Rd8num_lociEv">
<span id="egglib::Rd::num_loci"></span><span class="target" id="project0classegglib_1_1Rd_1a80ace218af23abfa10b73bd240e26294"></span>unsigned int <code class="descclassname">egglib::Rd::</code><code class="descname">num_loci</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib2Rd8num_lociEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get number of processed loci (some loci may be skipped) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib2Rd6ploidyEv">
<span id="egglib::Rd::ploidy"></span><span class="target" id="project0classegglib_1_1Rd_1a526257f81406097ffa795e33ec58938b"></span>unsigned int <code class="descclassname">egglib::Rd::</code><code class="descname">ploidy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib2Rd6ploidyEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get ploidy as provided to configure() </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib2Rd5resetEv">
<span id="egglib::Rd::reset"></span><span class="target" id="project0classegglib_1_1Rd_1ad20897c5c8bd47f5d4005989bead0e55"></span>void <code class="descclassname">egglib::Rd::</code><code class="descname">reset</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib2Rd5resetEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Reset to initial state. </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="fu-s-f">
<h3>Fu’s F<a class="headerlink" href="#fu-s-f" title="Permalink to this headline">¶</a></h3>
<dl class="function">
<dt id="_CPPv2N6egglib2FsEjjd">
<span id="egglib::Fs__unsigned-i.unsigned-i.double"></span><span class="target" id="project0namespaceegglib_1af6ec9bc92bb8f419132ce4a5c012c4fe"></span>double <code class="descclassname">egglib::</code><code class="descname">Fs</code><span class="sig-paren">(</span>unsigned int <em>n</em>, unsigned int <em>K</em>, double <em>pi</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib2FsEjjd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Compute Fu’s Fs. </p>
<p>This function computes Fu’s Fs statistic using haplotype statistics (that should have been computed using the <a class="reference internal" href="#project0classegglib_1_1Haplotypes"><span class="std std-ref">Haplotypes</span></a> class) and, as a theta estimator, pi provided by the <a class="reference internal" href="#project0classegglib_1_1Diversity1"><span class="std std-ref">Diversity1</span></a> class. The values must have been computed using the same data set.</p>
<p>Warning: this function is not available for values of n (number of samples) larger than MAX_N_STIRLING. k must be &gt;= 1 and &lt;= n.</p>
<p>
The behaviour of the function is not defined if K &lt; 0. The function returns UNDEF if the value cannot be computed, which can happen:<dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">n</span></code> - <p>number of exploited samples for determining the number of haplotypes.</p>
</li>
<li><code class="first docutils literal"><span class="pre">K</span></code> - <p>number of haplotypes obtained with the same data.</p>
</li>
<li><code class="first docutils literal"><span class="pre">pi</span></code> - <p>average number of pairwise differences (as theta estimator, per gene).</p>
</li>
</ul>
</dd>
</dl>
</p>
<p><ul class="simple">
<li>if n is larger than MAX_N_STIRLING; </li>
<li>if the sum of probabilities of k values &gt;= K is too close of 0 or 1 (based on the computer’s precision); </li>
<li>if pi is 0 (no polymorphism); </li>
<li>if K &gt; n (which is an error).</li>
</ul>

Header: &lt;egglib-cpp/Fs.hpp&gt; </p>
</dd></dl>

</div>
</div>
<div class="section" id="linkage-analysis">
<h2>Linkage analysis<a class="headerlink" href="#linkage-analysis" title="Permalink to this headline">¶</a></h2>
<div class="section" id="pairwiseld">
<h3>PairwiseLD<a class="headerlink" href="#pairwiseld" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib10PairwiseLDE">
<span id="egglib::PairwiseLD"></span><span class="target" id="project0classegglib_1_1PairwiseLD"></span><em class="property">class </em></dt>
<dd><p>Analyzes linkage disequilibrium for a pair of polymorphic sites. </p>
<p>This class considers a single pair of polymorphic sites at a time. The first method, <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">process()</span></a>, detects alleles at both sites under consideration and determines whether the pairwise comparison is fit for analysis (based on the presence of polymorphism, and allele frequencies). Statistics are computed by <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1abc6a557bfa8eab268bbf219565791de6"><span class="std std-ref">compute()</span></a> for a given pair of alleles. Letting the user filter out sites for which that are more than two alleles and, if necessary, process multiple pairs of alleles.</p>
<p>One should first process a pair of sites with <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">process()</span></a>. If the return value is false, one should not process data further. Otherwise, one can access data with <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1a2f3346fb852910c23a2d1b8a0314493c"><span class="std std-ref">num_alleles1()</span></a>, <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1acd44bd61acae12ab615b2f4712124113"><span class="std std-ref">num_alleles2()</span></a>, <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1a7645b8768fc52c14161cdea92eee9250"><span class="std std-ref">index1()</span></a>, <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1af38073a5cbe4148dc39361cce6f6ce09"><span class="std std-ref">index2()</span></a>, <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ae75babdd45d3f963989ade004ee8dbe8"><span class="std std-ref">freq1()</span></a>, <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1aa1ae8966bba6c6556eea13cdce60bc77"><span class="std std-ref">freq2()</span></a>, <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1a20a46530be7dee37b54796aa892e955e"><span class="std std-ref">freq()</span></a>, and <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1a82553806ee88c1eab2710d2ffdddea42"><span class="std std-ref">nsam()</span></a>, and can also compute LD with <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1abc6a557bfa8eab268bbf219565791de6"><span class="std std-ref">compute()</span></a> (for a given pair of alleles) and then access to LD estimates. </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib10PairwiseLD10PairwiseLDEv">
<span id="egglib::PairwiseLD::PairwiseLD"></span><span class="target" id="project0classegglib_1_1PairwiseLD_1a5e955b7315707996b6d4af4df9c30ee5"></span><code class="descclassname">egglib::PairwiseLD::</code><code class="descname">PairwiseLD</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10PairwiseLD10PairwiseLDEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Default constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10PairwiseLDD0Ev">
<span id="egglib::PairwiseLD::~PairwiseLD"></span><span class="target" id="project0classegglib_1_1PairwiseLD_1a3de3d66d5e7dd843acd805d28a98cab9"></span><code class="descclassname">egglib::PairwiseLD::</code><code class="descname">~PairwiseLD</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10PairwiseLDD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10PairwiseLD7computeEjj">
<span id="egglib::PairwiseLD::compute__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1PairwiseLD_1abc6a557bfa8eab268bbf219565791de6"></span>void <code class="descclassname">egglib::PairwiseLD::</code><code class="descname">compute</code><span class="sig-paren">(</span>unsigned int <em>allele1</em>, unsigned int <em>allele2</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10PairwiseLD7computeEjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Compute D, D’, r and r^2 statistics for a given pair of alleles. </p>
<p>The method <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">process()</span></a> must have been executed and must have returned true.</p>
<p>Statistics are computed only for a given pair of alleles. If there are only two alleles, all allele pairs result in consistent results. Otherwise, some multi-allele summarizing methodology has to be applied.</p>
<p>allele1 and allele2 are the allele indexes at the first and second site, respectively. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10PairwiseLD1DEv">
<span id="egglib::PairwiseLD::D"></span><span class="target" id="project0classegglib_1_1PairwiseLD_1a01f1ec34f0907b90a3d974577bf07bc8"></span>double <code class="descclassname">egglib::PairwiseLD::</code><code class="descname">D</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10PairwiseLD1DEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the D statistic. </p>
<p>This value is reset to 0 upon call to <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">process()</span></a>.</p>
<p>Requires <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1abc6a557bfa8eab268bbf219565791de6"><span class="std std-ref">compute()</span></a>. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10PairwiseLD2DpEv">
<span id="egglib::PairwiseLD::Dp"></span><span class="target" id="project0classegglib_1_1PairwiseLD_1aebd281570393c81625e270c48259e032"></span>double <code class="descclassname">egglib::PairwiseLD::</code><code class="descname">Dp</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10PairwiseLD2DpEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the D’ statistic. </p>
<p>This value is reset to 0 upon call to <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">process()</span></a>.</p>
<p>Requires <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1abc6a557bfa8eab268bbf219565791de6"><span class="std std-ref">compute()</span></a>; </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10PairwiseLD4freqEjj">
<span id="egglib::PairwiseLD::freq__unsigned-i.unsigned-i"></span><span class="target" id="project0classegglib_1_1PairwiseLD_1a20a46530be7dee37b54796aa892e955e"></span>unsigned int <code class="descclassname">egglib::PairwiseLD::</code><code class="descname">freq</code><span class="sig-paren">(</span>unsigned int <em>allele1</em>, unsigned int <em>allele2</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10PairwiseLD4freqEjj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the frequency of a genotype. </p>
<p>The method <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">process()</span></a> must have been executed and must have returned true.</p>
<p>The indexes must be smaller than the value returned by <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1a2f3346fb852910c23a2d1b8a0314493c"><span class="std std-ref">num_alleles1()</span></a> and <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1acd44bd61acae12ab615b2f4712124113"><span class="std std-ref">num_alleles2()</span></a> respectively. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10PairwiseLD5freq1Ej">
<span id="egglib::PairwiseLD::freq1__unsigned-i"></span><span class="target" id="project0classegglib_1_1PairwiseLD_1ae75babdd45d3f963989ade004ee8dbe8"></span>unsigned int <code class="descclassname">egglib::PairwiseLD::</code><code class="descname">freq1</code><span class="sig-paren">(</span>unsigned int <em>allele</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10PairwiseLD5freq1Ej" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the frequency of an allele for the first site. </p>
<p>The method <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">process()</span></a> must have been executed and must have returned true.</p>
<p>The index must be smaller than the value returned by <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1a2f3346fb852910c23a2d1b8a0314493c"><span class="std std-ref">num_alleles1()</span></a>. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10PairwiseLD5freq2Ej">
<span id="egglib::PairwiseLD::freq2__unsigned-i"></span><span class="target" id="project0classegglib_1_1PairwiseLD_1aa1ae8966bba6c6556eea13cdce60bc77"></span>unsigned int <code class="descclassname">egglib::PairwiseLD::</code><code class="descname">freq2</code><span class="sig-paren">(</span>unsigned int <em>allele</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10PairwiseLD5freq2Ej" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the frequency of an allele for the second site. </p>
<p>The method <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">process()</span></a> must have been executed and must have returned true.</p>
<p>The index must be smaller than the value returned by <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1acd44bd61acae12ab615b2f4712124113"><span class="std std-ref">num_alleles2()</span></a>. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10PairwiseLD6index1Ej">
<span id="egglib::PairwiseLD::index1__unsigned-i"></span><span class="target" id="project0classegglib_1_1PairwiseLD_1a7645b8768fc52c14161cdea92eee9250"></span>unsigned int <code class="descclassname">egglib::PairwiseLD::</code><code class="descname">index1</code><span class="sig-paren">(</span>unsigned int <em>allele</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10PairwiseLD6index1Ej" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the index of an allele for the first site. </p>
<p>The method <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">process()</span></a> must have been executed and must have returned true.</p>
<p>For a given allele, get its index within the original <a class="reference internal" href="#project0classegglib_1_1SiteHolder"><span class="std std-ref">SiteHolder</span></a> instance. The indexes can be shifted by <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">process()</span></a> due to missing data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10PairwiseLD6index2Ej">
<span id="egglib::PairwiseLD::index2__unsigned-i"></span><span class="target" id="project0classegglib_1_1PairwiseLD_1af38073a5cbe4148dc39361cce6f6ce09"></span>unsigned int <code class="descclassname">egglib::PairwiseLD::</code><code class="descname">index2</code><span class="sig-paren">(</span>unsigned int <em>allele</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10PairwiseLD6index2Ej" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the index of an allele for the second site. </p>
<p>The method <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">process()</span></a> must have been executed and must have returned true.</p>
<p>For a given allele, get its index within the original <a class="reference internal" href="#project0classegglib_1_1SiteHolder"><span class="std std-ref">SiteHolder</span></a> instance. The indexes can be shifted by <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">process()</span></a> due to missing data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10PairwiseLD4nsamEv">
<span id="egglib::PairwiseLD::nsam"></span><span class="target" id="project0classegglib_1_1PairwiseLD_1a82553806ee88c1eab2710d2ffdddea42"></span>unsigned int <code class="descclassname">egglib::PairwiseLD::</code><code class="descname">nsam</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10PairwiseLD4nsamEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the number of analyzed samples. </p>
<p>The method <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">process()</span></a> must have been executed and must have returned true.</p>
<p>The returned value might be smaller than the initial number of samples due to missing data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10PairwiseLD12num_alleles1Ev">
<span id="egglib::PairwiseLD::num_alleles1"></span><span class="target" id="project0classegglib_1_1PairwiseLD_1a2f3346fb852910c23a2d1b8a0314493c"></span>unsigned int <code class="descclassname">egglib::PairwiseLD::</code><code class="descname">num_alleles1</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10PairwiseLD12num_alleles1Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the actual number of alleles at the first site. </p>
<p>The method <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">process()</span></a> must have been executed and must have returned true.</p>
<p>Gives the number of different alleles at the first site, considering only samples for which both sites have exploitable data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10PairwiseLD12num_alleles2Ev">
<span id="egglib::PairwiseLD::num_alleles2"></span><span class="target" id="project0classegglib_1_1PairwiseLD_1acd44bd61acae12ab615b2f4712124113"></span>unsigned int <code class="descclassname">egglib::PairwiseLD::</code><code class="descname">num_alleles2</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10PairwiseLD12num_alleles2Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the actual number of alleles at the second site. </p>
<p>The method <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">process()</span></a> must have been executed and must have returned true.</p>
<p>Gives the number of different afirst lleles at the second site, considering only samples for which both sites have exploitable data. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10PairwiseLD7processERK10SiteHolderRK10SiteHolderjdPK15StructureHolder">
<span id="egglib::PairwiseLD::process__SiteHolderCR.SiteHolderCR.unsigned.double.StructureHolderCP"></span><span class="target" id="project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"></span>bool <code class="descclassname">egglib::PairwiseLD::</code><code class="descname">process</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site1</em>, <em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site2</em>, unsigned <em>min_n</em>, double <em>max_maj</em>, <em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib15StructureHolderE" title="egglib::StructureHolder">StructureHolder</a> *<em>stru</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10PairwiseLD7processERK10SiteHolderRK10SiteHolderjdPK15StructureHolder" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Analyze a pair of sites. </p>
<p>The method takes two sites as argument. The two sites must be taken from the same data set. In particular, the sample sizes must be identical. Structure and outgroup are ignored. The indexes of samples must be matching over the two sites. Samples which are missing in either of the samples are skipped. If the remaining samples are less than the argument min_n, the whole computation is dropped. Genotypes are ignored (only alleles are considered).</p>
<p>
If this method returns true, statistics might be computed for a given pair of alleles using the <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1abc6a557bfa8eab268bbf219565791de6"><span class="std std-ref">compute()</span></a> method. The number of alleles available for analysis is available at either site using <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1a2f3346fb852910c23a2d1b8a0314493c"><span class="std std-ref">num_alleles1()</span></a> and <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1acd44bd61acae12ab615b2f4712124113"><span class="std std-ref">num_alleles2()</span></a>. When returning false, this method stops as early as possible, and the state of the object might be inconsistent. In this case, no accessor must be used and <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1abc6a557bfa8eab268bbf219565791de6"><span class="std std-ref">compute()</span></a> must not be called. <dl class="docutils">
<dt><strong>Return</strong></dt>
<dd>true if computations have been performed, false if the sites fall in one the following cases: not enough samples (based on the min_n argument); either site is fixed; the allele frequencies are too unbalanced with at least one allele at a frequency larger than max_maj.</dd>
<dt><strong>Note</strong></dt>
<dd>Due to missing data, a site that is initially polymorphic might appear to be fixed when considering only samples that are not missing for the other site, causing this method to drop the pairwise comparison. Conversely, a site that has more than two alleles might have only two when considering only samples that are not missing for the other site. For this reason, it is not trivial to filter out sites before calling this method, and sites might not be consistently included or rejected.</dd>
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">site1</span></code> - <p>first site.</p>
</li>
<li><code class="first docutils literal"><span class="pre">site2</span></code> - <p>second site.</p>
</li>
<li><code class="first docutils literal"><span class="pre">min_n</span></code> - <p>minimum number of samples used (this value must always be larger than 1).</p>
</li>
<li><code class="first docutils literal"><span class="pre">max_maj</span></code> - <p>maximum relative frequency of the majority allele (if any allele at either site has a frequency larger than this value, the pairwise comparison is dropped).</p>
</li>
<li><code class="first docutils literal"><span class="pre">stru</span></code> - <p>a Structure object (used to process a subset of samples). By default, use all samples.</p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10PairwiseLD1rEv">
<span id="egglib::PairwiseLD::r"></span><span class="target" id="project0classegglib_1_1PairwiseLD_1a2158ad8213f28ce9ad6c74b562deecca"></span>double <code class="descclassname">egglib::PairwiseLD::</code><code class="descname">r</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10PairwiseLD1rEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the r statistic. </p>
<p>This value is reset to 0 upon call to <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">process()</span></a>.</p>
<p>Requires <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1abc6a557bfa8eab268bbf219565791de6"><span class="std std-ref">compute()</span></a>. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10PairwiseLD5resetEv">
<span id="egglib::PairwiseLD::reset"></span><span class="target" id="project0classegglib_1_1PairwiseLD_1ad20897c5c8bd47f5d4005989bead0e55"></span>void <code class="descclassname">egglib::PairwiseLD::</code><code class="descname">reset</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10PairwiseLD5resetEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Reset all values to default. </p>
<p>Call to this method is usually not necessary since <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">process()</span></a> automatically resets the instance. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib10PairwiseLD3rsqEv">
<span id="egglib::PairwiseLD::rsq"></span><span class="target" id="project0classegglib_1_1PairwiseLD_1a8a33d63dcbe92e4673c4b3263b6845c2"></span>double <code class="descclassname">egglib::PairwiseLD::</code><code class="descname">rsq</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib10PairwiseLD3rsqEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the r^2 statistic. </p>
<p>Same as <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1a2158ad8213f28ce9ad6c74b562deecca"><span class="std std-ref">r()</span></a>*r(). This value is reset to 0 upon call to <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">process()</span></a>.</p>
<p>Requires <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1abc6a557bfa8eab268bbf219565791de6"><span class="std std-ref">compute()</span></a>. </p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="matrixld">
<h3>MatrixLD<a class="headerlink" href="#matrixld" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="_CPPv2N6egglib8MatrixLDE">
<span id="egglib::MatrixLD"></span><span class="target" id="project0classegglib_1_1MatrixLD"></span><em class="property">class </em></dt>
<dd><p>Analyzes linkage disequilibrium between pairs of sites. </p>
<p>This class processes a set of <a class="reference internal" href="#project0classegglib_1_1SiteHolder"><span class="std std-ref">SiteHolder</span></a> instances and computes linkage disequilibrium for all pairs of sites. A <a class="reference internal" href="#project0classegglib_1_1PairwiseLD"><span class="std std-ref">PairwiseLD</span></a> instance is provided for all comparison, skipping all pairs for which LD cannot be computed (there are several criteria). The approach consists in first calling <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1afcc2a02315827722db61e3f246726a9a"><span class="std std-ref">load()</span></a> by providing a set of <a class="reference internal" href="#project0classegglib_1_1SiteHolder"><span class="std std-ref">SiteHolder</span></a> instances. The method <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a7b44adead6e7227a294b5f0885c9631c"><span class="std std-ref">computeLD()</span></a> computes the LD for each pair and <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a9b57fd358be92f302b6f16c9a913a14d"><span class="std std-ref">computeStats()</span></a> computes the statistics of Kelly (1997) and Rozas et al. (2001). These statistics are based on the average of pairwise linkage disequilibrium statistics. In addition, <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1ad9a01d568ee6fd89c0f356715335318b"><span class="std std-ref">computeRmin()</span></a> computes Rm of Hudson and Kaplan (1985) and does not generate nor use <a class="reference internal" href="#project0classegglib_1_1PairwiseLD"><span class="std std-ref">PairwiseLD</span></a> instances, and it can be used independently. </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Types</p>
<dl class="type">
<dt id="_CPPv2N6egglib8MatrixLD12MultiAllelicE">
<span id="egglib::MatrixLD::MultiAllelic"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a38f217ceb453a34c786802494fa55cf6"></span>enum <code class="descclassname">egglib::MatrixLD::</code><code class="descname">MultiAllelic</code><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD12MultiAllelicE" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Flags for processing multiallelic sites. </p>
<p>This enum is used to specify what should be done with pairs of sites for which at least one site has more than two alleles. </p>
<p><em>Values:</em></p>
<dl class="member">
<dt id="_CPPv2N6egglib8MatrixLD6ignoreE">
<span id="egglib::MatrixLD::ignore"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a38f217ceb453a34c786802494fa55cf6a1ab39bc750d3bf0ee5d0214b3b267636"></span></dt>
<dd><p>Only process pairs of sites with exactly two alleles. </p>
</dd></dl>

<dl class="member">
<dt id="_CPPv2N6egglib8MatrixLD8use_mainE">
<span id="egglib::MatrixLD::use_main"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a38f217ceb453a34c786802494fa55cf6a6ea15be5edd1266a99dac500677fa5ff"></span></dt>
<dd><p>Use the allele with highest frequency. </p>
</dd></dl>

<dl class="member">
<dt id="_CPPv2N6egglib8MatrixLD7use_allE">
<span id="egglib::MatrixLD::use_all"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a38f217ceb453a34c786802494fa55cf6a6a3ccb3c8eb1dcd430699b1ec9b73d27"></span></dt>
<dd><p>Use all alleles. </p>
</dd></dl>

</dd></dl>

</div>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD8MatrixLDEv">
<span id="egglib::MatrixLD::MatrixLD"></span><span class="target" id="project0classegglib_1_1MatrixLD_1ae9f1f611e531ed1ccd75b3849dd6b4fd"></span><code class="descclassname">egglib::MatrixLD::</code><code class="descname">MatrixLD</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD8MatrixLDEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLDD0Ev">
<span id="egglib::MatrixLD::~MatrixLD"></span><span class="target" id="project0classegglib_1_1MatrixLD_1acf24bc88ba31080c97d53b91819c33e9"></span><code class="descclassname">egglib::MatrixLD::</code><code class="descname">~MatrixLD</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLDD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD9computeLDEjd">
<span id="egglib::MatrixLD::computeLD__unsigned.double"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a7b44adead6e7227a294b5f0885c9631c"></span>void <code class="descclassname">egglib::MatrixLD::</code><code class="descname">computeLD</code><span class="sig-paren">(</span>unsigned <em>min_n</em>, double <em>max_maj</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD9computeLDEjd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Compute LD between all pairs of sites. </p>
<p>Use sites loaded using <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1afcc2a02315827722db61e3f246726a9a"><span class="std std-ref">load()</span></a> and process all possible pairs. Each pairwise comparison is retained only if all filters are passed (see arguments of this method). After call of this method, the number of pairs can be accessed using <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1af826b088804ec352f0f7ffc35c22c274"><span class="std std-ref">num_tot()</span></a> (it is equal to n(n-1)/2 where n is the number of loaded sites); the number of analyzed pairs can be accessed using <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1aeb8b37a69ea87a93cf34885ef375fe53"><span class="std std-ref">num_pairs()</span></a>; the total number of analyzed allele pairs can be accessed using <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a5434000611eea3568aec6698fec6fec7"><span class="std std-ref">num_alleles()</span></a>. Then the method compute() can be called to compute Kelly’s statistics.</p>
<p><dl class="docutils">
<dt><strong>Note</strong></dt>
<dd>Due to missing data, it is not trivial to predict whether a pairwise comparison will be dropped. See the documentation of <a class="reference internal" href="#project0classegglib_1_1PairwiseLD_1ac27eb4d1d8fc21d1de8aa340624454ec"><span class="std std-ref">PairwiseLD::process()</span></a>. </dd>
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">min_n</span></code> - <p>minimum number of samples used (this value must always be larger than 1).</p>
</li>
<li><code class="first docutils literal"><span class="pre">max_maj</span></code> - <p>maximum relative frequency of the majority allele (if any allele at either site has a frequency larger than this value, the pairwise comparison is dropped).</p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD11computeRminEb">
<span id="egglib::MatrixLD::computeRmin__b"></span><span class="target" id="project0classegglib_1_1MatrixLD_1ad9a01d568ee6fd89c0f356715335318b"></span>void <code class="descclassname">egglib::MatrixLD::</code><code class="descname">computeRmin</code><span class="sig-paren">(</span>bool <em>oriented</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD11computeRminEb" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Computes Hudson and Kaplan’s Rmin. </p>
<p>To be used, this method requires that sites have been loaded in increasing position order. Only sites with exactly two alleles and no missing data at all are used. In addition, if the oriented argument is set to true, only orientable sites are considered. Sites with more than two alleles and sites with any missing data, and sites not orientable if oriented is set to true, are ignored.</p>
<p>When the method has finished, a few methods provide access to the results. <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1ae13cc85a02cff1a0cc1e13c6b7499ddc"><span class="std std-ref">Rmin_num_sites()</span></a> give the number of sites considered for the analysis. If the value is less than two, the statistic itself bears no signification. <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1ae81a73dd91aa5bd4965b5dbd00ec87e3"><span class="std std-ref">Rmin()</span></a> gives the minimum number of recombination events. Finally, all of the non-reductible intervals containing a recombination even can be accessed using the two methods Rmin_left(unsigned int) and Rmin_right(unsigned int). The number of intervals is always <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1ae81a73dd91aa5bd4965b5dbd00ec87e3"><span class="std std-ref">Rmin()</span></a>.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">oriented</span></code> - <p>if true, consider only orientable sites and apply the three- (instead four-) gametes rule. If false, ignore all outgroup data and include orientable and non-orientable sites. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD12computeStatsE12MultiAllelicj">
<span id="egglib::MatrixLD::computeStats__MultiAllelic.unsigned-i"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a9b57fd358be92f302b6f16c9a913a14d"></span>void <code class="descclassname">egglib::MatrixLD::</code><code class="descname">computeStats</code><span class="sig-paren">(</span><a class="reference internal" href="#_CPPv2N6egglib8MatrixLD12MultiAllelicE" title="egglib::MatrixLD::MultiAllelic">MultiAllelic</a> <em>multiallelic</em>, unsigned int <em>min_freq</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD12computeStatsE12MultiAllelicj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Computes Kelly’s and Rozas et al.’s statistics. </p>
<p>Computes ZnS, Z*nS and Z*nS* (Kelly 1997), and Za and ZZ (Rozas et al. 2001) on the basis of analyzed site pairs (requires <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a7b44adead6e7227a294b5f0885c9631c"><span class="std std-ref">computeLD()</span></a>). The number of alleles pairs used for computing ZnS, Z*nS and Z*nS* is available as <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a1ace76ef737bce92213656a52cc532dc"><span class="std std-ref">num_allele_pairs()</span></a>, and the number of allele pairs used for computing Za and ZZ is available as <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1ac496f56c25fd027ba0604caa7fbd5351"><span class="std std-ref">num_allele_pairs_adj()</span></a>. The statistics must not be used if the corresponding number of allele pairs is 0.</p>
<p>If multiallelic equals to <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a38f217ceb453a34c786802494fa55cf6a1ab39bc750d3bf0ee5d0214b3b267636"><span class="std std-ref">MatrixLD::ignore</span></a>, only pairs of sites for which both sites have exactly two alleles are processed. In this case, the first allele of each site is considered. If multiallelic is <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a38f217ceb453a34c786802494fa55cf6a6ea15be5edd1266a99dac500677fa5ff"><span class="std std-ref">MatrixLD::use_main</span></a>, the alleles with highest frequency are considered (even if one or both sites have only two alleles). In case of equality, the first allele is considered. If multiallelic is <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a38f217ceb453a34c786802494fa55cf6a6a3ccb3c8eb1dcd430699b1ec9b73d27"><span class="std std-ref">MatrixLD::use_all</span></a>, then all alleles of all sites are used, and the final statistics are averaged over <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a5434000611eea3568aec6698fec6fec7"><span class="std std-ref">num_alleles()</span></a> (rather than num_pairs).</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">multiallelic</span></code> - <p>modifies the behaviour of the method (see above).</p>
</li>
<li><code class="first docutils literal"><span class="pre">min_freq</span></code> - <p>this flag has an effect only if used in conjunction with <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a38f217ceb453a34c786802494fa55cf6a6a3ccb3c8eb1dcd430699b1ec9b73d27"><span class="std std-ref">MatrixLD::use_all</span></a> (it is ignored otherwise); if larger than 0, rather than using all alleles, use only those that have a frequency equal to or larger than the given value. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD8distanceEj">
<span id="egglib::MatrixLD::distance__unsigned-i"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a52938c9537fae21603a2036b6ed9dc2b"></span>unsigned int <code class="descclassname">egglib::MatrixLD::</code><code class="descname">distance</code><span class="sig-paren">(</span>unsigned int <em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD8distanceEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the distance between sites for a given pair. </p>
<p>Requires that enough pairs have been loaded using <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1afcc2a02315827722db61e3f246726a9a"><span class="std std-ref">load()</span></a> and that requested index must be smaller than <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1aeb8b37a69ea87a93cf34885ef375fe53"><span class="std std-ref">num_pairs()</span></a>. The distance is returned as an absolute value. It is not possible to determine to which sites the pair index corresponds. If you need it, you might want to use <a class="reference internal" href="#project0classegglib_1_1PairwiseLD"><span class="std std-ref">PairwiseLD</span></a> directly. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD6index1Ej">
<span id="egglib::MatrixLD::index1__unsigned-i"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a7645b8768fc52c14161cdea92eee9250"></span>unsigned int <code class="descclassname">egglib::MatrixLD::</code><code class="descname">index1</code><span class="sig-paren">(</span>unsigned int <em>allele</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD6index1Ej" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Index of first site for a given pair. </p>
<p>See <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a2d2b15f3d81d9652144172f9369c8df6"><span class="std std-ref">pairLD()</span></a>. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD6index2Ej">
<span id="egglib::MatrixLD::index2__unsigned-i"></span><span class="target" id="project0classegglib_1_1MatrixLD_1af38073a5cbe4148dc39361cce6f6ce09"></span>unsigned int <code class="descclassname">egglib::MatrixLD::</code><code class="descname">index2</code><span class="sig-paren">(</span>unsigned int <em>allele</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD6index2Ej" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Index of second site for a given pair. </p>
<p>See <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a2d2b15f3d81d9652144172f9369c8df6"><span class="std std-ref">pairLD()</span></a>. Note that index2 is always &gt; index1. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD4loadERK10SiteHolderd">
<span id="egglib::MatrixLD::load__SiteHolderCR.double"></span><span class="target" id="project0classegglib_1_1MatrixLD_1afcc2a02315827722db61e3f246726a9a"></span>void <code class="descclassname">egglib::MatrixLD::</code><code class="descname">load</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> &amp;<em>site</em>, double <em>position</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD4loadERK10SiteHolderd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Load a site. </p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">site</span></code> - <p>One of the sites.</p>
</li>
<li><code class="first docutils literal"><span class="pre">position</span></code> - <p>All sites must have a valid position. Positions are required to be increasing. For computing Rmin, positions are ignored (they only are fed back if interval limits are required). </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD16num_allele_pairsEv">
<span id="egglib::MatrixLD::num_allele_pairs"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a1ace76ef737bce92213656a52cc532dc"></span>unsigned int <code class="descclassname">egglib::MatrixLD::</code><code class="descname">num_allele_pairs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD16num_allele_pairsEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of allele pairs used to compute Kelly’s statistics. </p>
<p>Return the number of allele pairs used by <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a9b57fd358be92f302b6f16c9a913a14d"><span class="std std-ref">computeStats()</span></a> to compute Kelly’s ZnS, Z*nS and Z*nS* statistics. If multiallelic equals ignore, this value equals the number of pairs of sites with exactly two alleles each (at most, <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1aeb8b37a69ea87a93cf34885ef375fe53"><span class="std std-ref">num_pairs()</span></a>); if multiallelic was use_main, this value equals <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1aeb8b37a69ea87a93cf34885ef375fe53"><span class="std std-ref">num_pairs()</span></a>; if multiallelic was use_all, this value equals <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a5434000611eea3568aec6698fec6fec7"><span class="std std-ref">num_alleles()</span></a>. If the returned value is 0 (no loaded pairs of sites, or no pairs of diallelic sites, if multiallelic was set to <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a38f217ceb453a34c786802494fa55cf6a1ab39bc750d3bf0ee5d0214b3b267636"><span class="std std-ref">MatrixLD::ignore</span></a>), Kelly’s statistics have been reset to 0 but should then be considered as not computable. If <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a1ace76ef737bce92213656a52cc532dc"><span class="std std-ref">num_allele_pairs()</span></a> is null, none of the Kelly’s and Rozas et al.’s statistics can be computed. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD20num_allele_pairs_adjEv">
<span id="egglib::MatrixLD::num_allele_pairs_adj"></span><span class="target" id="project0classegglib_1_1MatrixLD_1ac496f56c25fd027ba0604caa7fbd5351"></span>unsigned int <code class="descclassname">egglib::MatrixLD::</code><code class="descname">num_allele_pairs_adj</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD20num_allele_pairs_adjEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of allele pairs used to compute Rozas et al.’s statistics. </p>
<p>Return the number of allele pairs used by <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a9b57fd358be92f302b6f16c9a913a14d"><span class="std std-ref">computeStats()</span></a> to compute Rozas et al.’s Za and ZZ statistics. See the documentation of <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a1ace76ef737bce92213656a52cc532dc"><span class="std std-ref">num_allele_pairs()</span></a> for reference. The meaning of this value is similar, except that it applies only to adjacent polymorphic sites (the value hence can only be smaller, or equal is limit cases). If this vallue is 0, Rozas et al.’s statistics have been reset to 0 but should then be considered as not computable. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD11num_allelesEv">
<span id="egglib::MatrixLD::num_alleles"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a5434000611eea3568aec6698fec6fec7"></span>unsigned int <code class="descclassname">egglib::MatrixLD::</code><code class="descname">num_alleles</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD11num_allelesEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the total number of allele pairs. </p>
<p>Requires <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a7b44adead6e7227a294b5f0885c9631c"><span class="std std-ref">computeLD()</span></a>. Returns the sum of allele pairs over all analyzed sites (see <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1aeb8b37a69ea87a93cf34885ef375fe53"><span class="std std-ref">num_pairs()</span></a>). The minimum value is twice <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1aeb8b37a69ea87a93cf34885ef375fe53"><span class="std std-ref">num_pairs()</span></a> (since, by definition, there must be at least two alleles at each retained site). </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD9num_pairsEv">
<span id="egglib::MatrixLD::num_pairs"></span><span class="target" id="project0classegglib_1_1MatrixLD_1aeb8b37a69ea87a93cf34885ef375fe53"></span>unsigned int <code class="descclassname">egglib::MatrixLD::</code><code class="descname">num_pairs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD9num_pairsEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the number of analyzed pairs of sites. </p>
<p>Requires <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a7b44adead6e7227a294b5f0885c9631c"><span class="std std-ref">computeLD()</span></a>. The returned value excludes all pairwise comparisons with no polymorphism failing any other criterion (see the <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a7b44adead6e7227a294b5f0885c9631c"><span class="std std-ref">computeLD()</span></a> method). </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD7num_totEv">
<span id="egglib::MatrixLD::num_tot"></span><span class="target" id="project0classegglib_1_1MatrixLD_1af826b088804ec352f0f7ffc35c22c274"></span>unsigned int <code class="descclassname">egglib::MatrixLD::</code><code class="descname">num_tot</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD7num_totEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the total number of pairs of sites. </p>
<p>Requires that site pairs have been processed using <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a7b44adead6e7227a294b5f0885c9631c"><span class="std std-ref">computeLD()</span></a>. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD6pairLDEj">
<span id="egglib::MatrixLD::pairLD__unsigned-i"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a2d2b15f3d81d9652144172f9369c8df6"></span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10PairwiseLDE" title="egglib::PairwiseLD">PairwiseLD</a> &amp;<code class="descclassname">egglib::MatrixLD::</code><code class="descname">pairLD</code><span class="sig-paren">(</span>unsigned int <em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD6pairLDEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get linkage disequilibrium for a given pair of sites. </p>
<p>Requires that enough pairs have been loaded using <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1afcc2a02315827722db61e3f246726a9a"><span class="std std-ref">load()</span></a> and that requested index must be smaller than <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1aeb8b37a69ea87a93cf34885ef375fe53"><span class="std std-ref">num_pairs()</span></a>. Use methods <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a7645b8768fc52c14161cdea92eee9250"><span class="std std-ref">index1()</span></a> and <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1af38073a5cbe4148dc39361cce6f6ce09"><span class="std std-ref">index2()</span></a> to obtain the corresponding site indexes. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD7processEjd12MultiAllelicjb">
<span id="egglib::MatrixLD::process__unsigned.double.MultiAllelic.unsigned-i.b"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a58e3d09a8ed4e7ef6e6dd9c26a90d2a7"></span>unsigned int <code class="descclassname">egglib::MatrixLD::</code><code class="descname">process</code><span class="sig-paren">(</span>unsigned <em>min_n</em>, double <em>max_maj</em>, <a class="reference internal" href="#_CPPv2N6egglib8MatrixLD12MultiAllelicE" title="egglib::MatrixLD::MultiAllelic">MultiAllelic</a> <em>multiallelic</em>, unsigned int <em>min_freq</em>, bool <em>oriented</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD7processEjd12MultiAllelicjb" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Call {<a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a7b44adead6e7227a294b5f0885c9631c"><span class="std std-ref">computeLD()</span></a> and <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a9b57fd358be92f302b6f16c9a913a14d"><span class="std std-ref">computeStats()</span></a>} and/or <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1ad9a01d568ee6fd89c0f356715335318b"><span class="std std-ref">computeRmin()</span></a> based on toggled flags. </p>
<p>Arguments are like for the three methods. Enter anything if they are not used.</p>
<p>Return value is a flag with the following bits:<ul class="simple">
<li>0: ZnS, Z*nS, and Z*nS* are computed.</li>
<li>1: Za and ZZ are computed.</li>
<li>2: Rmin was computed. </li>
</ul>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD5resetEv">
<span id="egglib::MatrixLD::reset"></span><span class="target" id="project0classegglib_1_1MatrixLD_1ad20897c5c8bd47f5d4005989bead0e55"></span>void <code class="descclassname">egglib::MatrixLD::</code><code class="descname">reset</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD5resetEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Reset to defaults. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD4RminEv">
<span id="egglib::MatrixLD::Rmin"></span><span class="target" id="project0classegglib_1_1MatrixLD_1ae81a73dd91aa5bd4965b5dbd00ec87e3"></span>unsigned int <code class="descclassname">egglib::MatrixLD::</code><code class="descname">Rmin</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD4RminEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Minimal number of recombination events. </p>
<p>Requires <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1ad9a01d568ee6fd89c0f356715335318b"><span class="std std-ref">computeRmin()</span></a>. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD9Rmin_leftEj">
<span id="egglib::MatrixLD::Rmin_left__unsigned-i"></span><span class="target" id="project0classegglib_1_1MatrixLD_1aa69867c70e35d19d8b44b84c961faf00"></span>unsigned int <code class="descclassname">egglib::MatrixLD::</code><code class="descname">Rmin_left</code><span class="sig-paren">(</span>unsigned int <em>i</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD9Rmin_leftEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Left bound of a recombination interval. </p>
<p>Requires <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1ad9a01d568ee6fd89c0f356715335318b"><span class="std std-ref">computeRmin()</span></a> and that <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1ae13cc85a02cff1a0cc1e13c6b7499ddc"><span class="std std-ref">Rmin_num_sites()</span></a> is at least 2.</p>
<p><dl class="docutils">
<dt><strong>Return</strong></dt>
<dd>Position of the site at the 5’ end of the given interval (as provided to <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1afcc2a02315827722db61e3f246726a9a"><span class="std std-ref">load()</span></a>). </dd>
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">i</span></code> - <p>interval index (there are <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1ae81a73dd91aa5bd4965b5dbd00ec87e3"><span class="std std-ref">Rmin()</span></a> intervals).</p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD14Rmin_num_sitesEv">
<span id="egglib::MatrixLD::Rmin_num_sites"></span><span class="target" id="project0classegglib_1_1MatrixLD_1ae13cc85a02cff1a0cc1e13c6b7499ddc"></span>unsigned int <code class="descclassname">egglib::MatrixLD::</code><code class="descname">Rmin_num_sites</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD14Rmin_num_sitesEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of sites used for computing Rmin. </p>
<p>Requires <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1ad9a01d568ee6fd89c0f356715335318b"><span class="std std-ref">computeRmin()</span></a>.</p>
<p>Fixed to 0 if there are less than two sites in total (no computation is performed by <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1ad9a01d568ee6fd89c0f356715335318b"><span class="std std-ref">computeRmin()</span></a> in that case). If <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1ae13cc85a02cff1a0cc1e13c6b7499ddc"><span class="std std-ref">Rmin_num_sites()</span></a> is less than 2, <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1ae81a73dd91aa5bd4965b5dbd00ec87e3"><span class="std std-ref">Rmin()</span></a> is not defined and fixed to 0. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD10Rmin_rightEj">
<span id="egglib::MatrixLD::Rmin_right__unsigned-i"></span><span class="target" id="project0classegglib_1_1MatrixLD_1affe2d5ee6978cc205d82a986364577a1"></span>unsigned int <code class="descclassname">egglib::MatrixLD::</code><code class="descname">Rmin_right</code><span class="sig-paren">(</span>unsigned int <em>i</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD10Rmin_rightEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Right bound of a recombination interval. </p>
<p>Requires <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1ad9a01d568ee6fd89c0f356715335318b"><span class="std std-ref">computeRmin()</span></a> and that <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1ae13cc85a02cff1a0cc1e13c6b7499ddc"><span class="std std-ref">Rmin_num_sites()</span></a> is at least 2.</p>
<p><dl class="docutils">
<dt><strong>Return</strong></dt>
<dd>Position of the site at the 5’ end of the given interval (as provided to <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1afcc2a02315827722db61e3f246726a9a"><span class="std std-ref">load()</span></a>). </dd>
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">i</span></code> - <p>interval index (there are <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1ae81a73dd91aa5bd4965b5dbd00ec87e3"><span class="std std-ref">Rmin()</span></a> intervals).</p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD10toggle_offEv">
<span id="egglib::MatrixLD::toggle_off"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a898807895f3c1114c7e0e654b237ab28"></span>void <code class="descclassname">egglib::MatrixLD::</code><code class="descname">toggle_off</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD10toggle_offEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Toggle all off. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD11toggle_RminEv">
<span id="egglib::MatrixLD::toggle_Rmin"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a67e1bd9634cfca25d4bce9de73772c5b"></span>void <code class="descclassname">egglib::MatrixLD::</code><code class="descname">toggle_Rmin</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD11toggle_RminEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Toggle Rmin. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD12toggle_statsEv">
<span id="egglib::MatrixLD::toggle_stats"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a1638a69161abd6083f12ffc54060a93e"></span>void <code class="descclassname">egglib::MatrixLD::</code><code class="descname">toggle_stats</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD12toggle_statsEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Toggle summary statistics. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD2ZaEv">
<span id="egglib::MatrixLD::Za"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a1860253eb2f3d758b72396a9634069b7"></span>double <code class="descclassname">egglib::MatrixLD::</code><code class="descname">Za</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD2ZaEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get Rozas et al.s Za statistic. </p>
<p>Requires <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a9b57fd358be92f302b6f16c9a913a14d"><span class="std std-ref">computeStats()</span></a>. See documation of this method to known when this value is defined. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD3ZnSEv">
<span id="egglib::MatrixLD::ZnS"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a231b6cdb0b197407fbd0a5d9959328fd"></span>double <code class="descclassname">egglib::MatrixLD::</code><code class="descname">ZnS</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD3ZnSEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get Kelly’s ZnS statistic. </p>
<p>Requires <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a9b57fd358be92f302b6f16c9a913a14d"><span class="std std-ref">computeStats()</span></a>. See documation of this method to known when this value is defined. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD9ZnS_star1Ev">
<span id="egglib::MatrixLD::ZnS_star1"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a1750a2390fe921af2538ccbec0b6e18d"></span>double <code class="descclassname">egglib::MatrixLD::</code><code class="descname">ZnS_star1</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD9ZnS_star1Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get Kelly’s Z*nS statistic. </p>
<p>Requires <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a9b57fd358be92f302b6f16c9a913a14d"><span class="std std-ref">computeStats()</span></a>. See documation of this method to known when this value is defined. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD9ZnS_star2Ev">
<span id="egglib::MatrixLD::ZnS_star2"></span><span class="target" id="project0classegglib_1_1MatrixLD_1a6491b3c2260a3fb60a4335cc900db4b3"></span>double <code class="descclassname">egglib::MatrixLD::</code><code class="descname">ZnS_star2</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD9ZnS_star2Ev" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get Kelly’s Z*nS* statistic. </p>
<p>Requires <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a9b57fd358be92f302b6f16c9a913a14d"><span class="std std-ref">computeStats()</span></a>. See documation of this method to known when this value is defined. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib8MatrixLD2ZZEv">
<span id="egglib::MatrixLD::ZZ"></span><span class="target" id="project0classegglib_1_1MatrixLD_1ac7bf1480b0007c2b261649321123efce"></span>double <code class="descclassname">egglib::MatrixLD::</code><code class="descname">ZZ</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib8MatrixLD2ZZEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get Rozas et al.s ZZ statistic. </p>
<p>Requires <a class="reference internal" href="#project0classegglib_1_1MatrixLD_1a9b57fd358be92f302b6f16c9a913a14d"><span class="std std-ref">computeStats()</span></a>. See documation of this method to known when this value is defined. </p>
</dd></dl>

</div>
</dd></dl>

<p>See also the <a class="reference internal" href="#_CPPv2N6egglib2RdE" title="egglib::Rd"><code class="xref cpp cpp-class docutils literal"><span class="pre">Rd</span></code></a> class.</p>
</div>
</div>
<div class="section" id="extended-haplotype-homozygosity">
<h2>Extended haplotype homozygosity<a class="headerlink" href="#extended-haplotype-homozygosity" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="_CPPv2N6egglib3EHHE">
<span id="egglib::EHH"></span><span class="target" id="project0classegglib_1_1EHH"></span><em class="property">class </em></dt>
<dd><p>Compute Extended Haplotype Homozygosity statistics. </p>
<p>Compute statistics described in Sabeti et al. (Nature 2002), Voight et al. (PLoS Biology 2006), Ramirez-Soriano et al. (Genetics 2008) and Tang et al. (PLoS Biology 2007).</p>
<p>The user must first load the core haplotype or site using the <a class="reference internal" href="#project0classegglib_1_1EHH_1acc4ac27b1c06a72fecee74fefb36ed31"><span class="std std-ref">set_core()</span></a> method which also allows to specify option values, and then all needed distant sites using <a class="reference internal" href="#project0classegglib_1_1EHH_1a96aaf2772f366419358f19d0f54ba5b9"><span class="std std-ref">load_distant()</span></a>. Distant sites must be loaded for one side only and with always increasing distance relatively to the core. To load sites of the other side, the user needs to call <a class="reference internal" href="#project0classegglib_1_1EHH_1acc4ac27b1c06a72fecee74fefb36ed31"><span class="std std-ref">set_core()</span></a> again with the same core site in order to reset statistics. Statistics are automatically computed and updated at each loaded distant site. It is required to load at least one valid core site before using accessors.</p>
<p>Header: &lt;egglib-cpp/EHH.hpp&gt; </p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric">Public Functions</p>
<dl class="function">
<dt id="_CPPv2N6egglib3EHH3EHHEv">
<span id="egglib::EHH::EHH"></span><span class="target" id="project0classegglib_1_1EHH_1a5d5c8842d65af78dbd2d48f89d337eaf"></span><code class="descclassname">egglib::EHH::</code><code class="descname">EHH</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH3EHHEv" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Constructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHHD0Ev">
<span id="egglib::EHH::~EHH"></span>virtual <span class="target" id="project0classegglib_1_1EHH_1a136f8e7e9f05df2a64be02990e09c32b"></span><code class="descclassname">egglib::EHH::</code><code class="descname">~EHH</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHHD0Ev" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Destructor. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH4dEHHEj">
<span id="egglib::EHH::dEHH__unsigned-i"></span><span class="target" id="project0classegglib_1_1EHH_1ae34630bb656d9fed9216413ce55837df"></span>double <code class="descclassname">egglib::EHH::</code><code class="descname">dEHH</code><span class="sig-paren">(</span>unsigned int <em>haplotype</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH4dEHHEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an <a class="reference internal" href="#project0classegglib_1_1EHH"><span class="std std-ref">EHH</span></a> decay value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH8dEHH_maxEv">
<span id="egglib::EHH::dEHH_max"></span><span class="target" id="project0classegglib_1_1EHH_1a696842786dbff3f6eae252588db3d984"></span>double <code class="descclassname">egglib::EHH::</code><code class="descname">dEHH_max</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH8dEHH_maxEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the maximum <a class="reference internal" href="#project0classegglib_1_1EHH"><span class="std std-ref">EHH</span></a> decay value (computed on the fly) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH9dEHH_meanEv">
<span id="egglib::EHH::dEHH_mean"></span><span class="target" id="project0classegglib_1_1EHH_1a1ad8e14004491ea5dcbc36fb00894ccb"></span>double <code class="descclassname">egglib::EHH::</code><code class="descname">dEHH_mean</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH9dEHH_meanEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get the average <a class="reference internal" href="#project0classegglib_1_1EHH"><span class="std std-ref">EHH</span></a> decay value (on the fly) </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH5dEHHcEj">
<span id="egglib::EHH::dEHHc__unsigned-i"></span><span class="target" id="project0classegglib_1_1EHH_1a71d76b0077ec662d0d82e1453343b2d8"></span>double <code class="descclassname">egglib::EHH::</code><code class="descname">dEHHc</code><span class="sig-paren">(</span>unsigned int <em>haplotype</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH5dEHHcEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an EHHc decay value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH5dEHHGEv">
<span id="egglib::EHH::dEHHG"></span><span class="target" id="project0classegglib_1_1EHH_1a307f713fc3a96cecd5349c35df32724b"></span>double <code class="descclassname">egglib::EHH::</code><code class="descname">dEHHG</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH5dEHHGEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an EHHS (genotypes) decay value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH5dEHHSEv">
<span id="egglib::EHH::dEHHS"></span><span class="target" id="project0classegglib_1_1EHH_1ae13af3e08b6de7a353fdf4b60f1c5534"></span>double <code class="descclassname">egglib::EHH::</code><code class="descname">dEHHS</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH5dEHHSEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an EHHS decay value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH4EHHcEj">
<span id="egglib::EHH::EHHc__unsigned-i"></span><span class="target" id="project0classegglib_1_1EHH_1a2c4c18b785b05749eecb3ab29fe1599c"></span>double <code class="descclassname">egglib::EHH::</code><code class="descname">EHHc</code><span class="sig-paren">(</span>unsigned int <em>haplotype</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH4EHHcEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an EHHc value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH4EHHGEv">
<span id="egglib::EHH::EHHG"></span><span class="target" id="project0classegglib_1_1EHH_1a373d56baf51f140864f5032851af7a9b"></span>double <code class="descclassname">egglib::EHH::</code><code class="descname">EHHG</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH4EHHGEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an EHHG value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH4EHHiEj">
<span id="egglib::EHH::EHHi__unsigned-i"></span><span class="target" id="project0classegglib_1_1EHH_1ab6bd50ecf02967ba98c5dede71e2c97e"></span>double <code class="descclassname">egglib::EHH::</code><code class="descname">EHHi</code><span class="sig-paren">(</span>unsigned int <em>haplotype</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH4EHHiEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an <a class="reference internal" href="#project0classegglib_1_1EHH"><span class="std std-ref">EHH</span></a> value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH4EHHSEv">
<span id="egglib::EHH::EHHS"></span><span class="target" id="project0classegglib_1_1EHH_1a5ee5b44b81b757296379a2e95e7ec33b"></span>double <code class="descclassname">egglib::EHH::</code><code class="descname">EHHS</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH4EHHSEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an EHHS value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH14flag_EHHG_doneEv">
<span id="egglib::EHH::flag_EHHG_done"></span><span class="target" id="project0classegglib_1_1EHH_1a5f6e90d464919dffcc3c9f3fb531cbb3"></span>bool <code class="descclassname">egglib::EHH::</code><code class="descname">flag_EHHG_done</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH14flag_EHHG_doneEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Tell if decay has been reached for EHHG. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH14flag_EHHS_doneEv">
<span id="egglib::EHH::flag_EHHS_done"></span><span class="target" id="project0classegglib_1_1EHH_1ab03923ddb96a10e1f63219bf66ac9732"></span>bool <code class="descclassname">egglib::EHH::</code><code class="descname">flag_EHHS_done</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH14flag_EHHS_doneEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Tell if decay has been reached for EHHS. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH3iEGEv">
<span id="egglib::EHH::iEG"></span><span class="target" id="project0classegglib_1_1EHH_1a9ceecec8824169d030ab4efb15c58ea5"></span>double <code class="descclassname">egglib::EHH::</code><code class="descname">iEG</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH3iEGEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an iEG value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH3iESEv">
<span id="egglib::EHH::iES"></span><span class="target" id="project0classegglib_1_1EHH_1add0924a7a189eea068fe4968ab93047b"></span>double <code class="descclassname">egglib::EHH::</code><code class="descname">iES</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH3iESEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an iES value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH3IHHEj">
<span id="egglib::EHH::IHH__unsigned-i"></span><span class="target" id="project0classegglib_1_1EHH_1ab1ac45de855f47ed3e4c19d92c4564bc"></span>double <code class="descclassname">egglib::EHH::</code><code class="descname">IHH</code><span class="sig-paren">(</span>unsigned int <em>haplotype</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH3IHHEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an IHH value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH4IHHcEj">
<span id="egglib::EHH::IHHc__unsigned-i"></span><span class="target" id="project0classegglib_1_1EHH_1aa4012e502f6a8c58c019445a74780ff5"></span>double <code class="descclassname">egglib::EHH::</code><code class="descname">IHHc</code><span class="sig-paren">(</span>unsigned int <em>haplotype</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH4IHHcEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an IHHc value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH3iHSEj">
<span id="egglib::EHH::iHS__unsigned-i"></span><span class="target" id="project0classegglib_1_1EHH_1a3f41d37c4e1de48f7fa19550d510b124"></span>double <code class="descclassname">egglib::EHH::</code><code class="descname">iHS</code><span class="sig-paren">(</span>unsigned int <em>haplotype</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH3iHSEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an iHS value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH6K_coreEv">
<span id="egglib::EHH::K_core"></span><span class="target" id="project0classegglib_1_1EHH_1af347f96ade3050d0262dcb2dbc4753d4"></span>unsigned int <code class="descclassname">egglib::EHH::</code><code class="descname">K_core</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH6K_coreEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of used haplotypes of the core. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH5K_curEv">
<span id="egglib::EHH::K_cur"></span><span class="target" id="project0classegglib_1_1EHH_1ace323e11da8cb1877f3c9293fa31333d"></span>unsigned int <code class="descclassname">egglib::EHH::</code><code class="descname">K_cur</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH5K_curEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Current number of haplotypes. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH12load_distantEPK10SiteHolderd">
<span id="egglib::EHH::load_distant__SiteHolderCP.double"></span><span class="target" id="project0classegglib_1_1EHH_1a96aaf2772f366419358f19d0f54ba5b9"></span>void <code class="descclassname">egglib::EHH::</code><code class="descname">load_distant</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> *<em>site</em>, double <em>distance</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH12load_distantEPK10SiteHolderd" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Load a distant site. </p>
<p>For each core haplotype, compute or update all statistics.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">site</span></code> - <p>the distant site to be loaded. The method will only throw an exception if the number of samples differ. </p>
</li>
<li><code class="first docutils literal"><span class="pre">distance</span></code> - <p>between the core and the distant site. The nature of the distance metrics is up to the user but must be consistent over sites. Distances must be &gt;=0 this must correspond to the positions of distant sites relatively to the core region or site. Site must be loaded such as distances must always be increasing. </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH14num_avail_coreEj">
<span id="egglib::EHH::num_avail_core__unsigned-i"></span><span class="target" id="project0classegglib_1_1EHH_1aad036e31673152a4fbab06cf1c7db2c2"></span>unsigned int <code class="descclassname">egglib::EHH::</code><code class="descname">num_avail_core</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH14num_avail_coreEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Current number of non-missing samples for a core haplotype. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH13num_avail_curEj">
<span id="egglib::EHH::num_avail_cur__unsigned-i"></span><span class="target" id="project0classegglib_1_1EHH_1a16ee98ce9693ac289931172f019afb05"></span>unsigned int <code class="descclassname">egglib::EHH::</code><code class="descname">num_avail_cur</code><span class="sig-paren">(</span>unsigned int<span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH13num_avail_curEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Current number of non-missing samples for a current haplotype. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH13num_avail_totEv">
<span id="egglib::EHH::num_avail_tot"></span><span class="target" id="project0classegglib_1_1EHH_1a512558acf8a72bfbe0914c5aeb1d94e8"></span>unsigned int <code class="descclassname">egglib::EHH::</code><code class="descname">num_avail_tot</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH13num_avail_totEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Current number of non-missing samples. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH12num_EHH_doneEv">
<span id="egglib::EHH::num_EHH_done"></span><span class="target" id="project0classegglib_1_1EHH_1a2112920f93141849fe4630725be62e6e"></span>unsigned int <code class="descclassname">egglib::EHH::</code><code class="descname">num_EHH_done</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH12num_EHH_doneEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of haplotypes for which computation of dEHH and IHH has been completed. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH13num_EHHc_doneEv">
<span id="egglib::EHH::num_EHHc_done"></span><span class="target" id="project0classegglib_1_1EHH_1a306e4cd338a06646d1047c5a79bb2fc3"></span>unsigned int <code class="descclassname">egglib::EHH::</code><code class="descname">num_EHHc_done</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH13num_EHHc_doneEv" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Number of haplotypes for which computation of dEHHc and IHHc has been completed. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH4rEHHEj">
<span id="egglib::EHH::rEHH__unsigned-i"></span><span class="target" id="project0classegglib_1_1EHH_1acc1780501e5e9fc14f36e8e0d7558512"></span>double <code class="descclassname">egglib::EHH::</code><code class="descname">rEHH</code><span class="sig-paren">(</span>unsigned int <em>haplotype</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH4rEHHEj" title="Permalink to this definition">¶</a><br /> const</dt>
<dd><p>Get an rEHH value. </p>
</dd></dl>

<dl class="function">
<dt id="_CPPv2N6egglib3EHH8set_coreEPK10SiteHolderbddddjjb">
<span id="egglib::EHH::set_core__SiteHolderCP.b.double.double.double.double.unsigned-i.unsigned-i.b"></span><span class="target" id="project0classegglib_1_1EHH_1acc4ac27b1c06a72fecee74fefb36ed31"></span>void <code class="descclassname">egglib::EHH::</code><code class="descname">set_core</code><span class="sig-paren">(</span><em class="property">const</em> <a class="reference internal" href="#_CPPv2N6egglib10SiteHolderE" title="egglib::SiteHolder">SiteHolder</a> *<em>site</em>, bool <em>genotypes</em>, double <em>EHH_thr</em>, double <em>EHHc_thr</em>, double <em>EHHS_thr</em>, double <em>EHHG_thr</em>, unsigned int <em>min_freq</em>, unsigned int <em>min_sam</em>, bool <em>crop</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib3EHH8set_coreEPK10SiteHolderbddddjjb" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Load the core site or region. </p>
<p>This method automatically resets the instance (clear all previously computed data and reallocate arrays to proper sizes). The Site instance passed as core is only used by this method. All counters will be incremented, until the next call to <a class="reference internal" href="#project0classegglib_1_1EHH_1acc4ac27b1c06a72fecee74fefb36ed31"><span class="std std-ref">set_core()</span></a>, or eventual destruction of object. All thresholds are understood as either <a class="reference internal" href="#project0classegglib_1_1EHH"><span class="std std-ref">EHH</span></a> or EHHS values and therefore must lie between 0.0 and 1.0.</p>
<p><dl class="docutils">
<dt><strong>Parameters</strong></dt>
<dd><ul class="breatheparameterlist first last">
<li><code class="first docutils literal"><span class="pre">site</span></code> - <p>core site or region. If a region, haplotypes within the core region must have been identified previously and should be loaded as a Site instance. The site may contain missing data. The samples containing missing data at the core site will be ignored for all subsequently loaded distant site. </p>
</li>
<li><code class="first docutils literal"><span class="pre">genotypes</span></code> - <p>if true, consider that data are entered as unphased genotypes (the Site instance must have consistent data). </p>
</li>
<li><code class="first docutils literal"><span class="pre">EHH_thr</span></code> - <p>threshold <a class="reference internal" href="#project0classegglib_1_1EHH"><span class="std std-ref">EHH</span></a> value. </p>
</li>
<li><code class="first docutils literal"><span class="pre">EHHc_thr</span></code> - <p>threshold EHHc value. </p>
</li>
<li><code class="first docutils literal"><span class="pre">EHHS_thr</span></code> - <p>threshold EHHS value. </p>
</li>
<li><code class="first docutils literal"><span class="pre">EHHG_thr</span></code> - <p>threshold EHHS (genotypes) value. </p>
</li>
<li><code class="first docutils literal"><span class="pre">min_freq</span></code> - <p>minimal absolute frequency for haplotypes (haplotypes with lower frequencies are ignored). Required to be strictly larger than zero. </p>
</li>
<li><code class="first docutils literal"><span class="pre">min_sam</span></code> - <p>minimal number of samples to continue computing (applied both within core haplotypes and for the total). </p>
</li>
<li><code class="first docutils literal"><span class="pre">crop</span></code> - <p>if True, set values of EHHS that are below the threshold to 0 to emulate the behaviour of the R package rehh (also affects iES). </p>
</li>
</ul>
</dd>
</dl>
</p>
</dd></dl>

</div>
</dd></dl>

</div>
<div class="section" id="helpers">
<h2>Helpers<a class="headerlink" href="#helpers" title="Permalink to this headline">¶</a></h2>
<div class="section" id="constants">
<h3>Constants<a class="headerlink" href="#constants" title="Permalink to this headline">¶</a></h3>
<dl class="member">
<dt id="_CPPv214MAX_N_STIRLING">
<span id="MAX_N_STIRLING__unsigned-iC"></span><span class="target" id="project0group__diversity_1ga74fe58b1123ad75f7e6bf4314b048295"></span><em class="property">const</em> unsigned int <code class="descclassname"></code><code class="descname">MAX_N_STIRLING</code><a class="headerlink" href="#_CPPv214MAX_N_STIRLING" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Maximal n values for pre-computed Stirling numbers. </p>
<p>Header: &lt;egglib-cpp/stirling.hpp&gt; </p>
</dd></dl>

<dl class="member">
<dt id="_CPPv212NUM_STIRLING">
<span id="NUM_STIRLING__unsigned-iC"></span><span class="target" id="project0group__diversity_1gaa785012d514d465a939dbc3418a9e91c"></span><em class="property">const</em> unsigned int <code class="descclassname"></code><code class="descname">NUM_STIRLING</code><a class="headerlink" href="#_CPPv212NUM_STIRLING" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Size of the Stirling numbers table. </p>
<p>Header: &lt;egglib-cpp/stirling.hpp&gt; </p>
</dd></dl>

<dl class="member">
<dt id="_CPPv214STIRLING_TABLE">
<span id="STIRLING_TABLE__doubleCA"></span><span class="target" id="project0group__diversity_1ga496bfb0f1ec8c0751f9247e7c7be2d70"></span><em class="property">const</em> double <code class="descclassname"></code><code class="descname">STIRLING_TABLE</code>[500499]<a class="headerlink" href="#_CPPv214STIRLING_TABLE" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Array of log(|S(n,k)|) (Stirling numbers of the 1st kind) </p>
<p>The values must be accessed using the <a class="reference internal" href="#project0group__diversity_1ga16615001364d52080bae649792438b47"><span class="std std-ref">stirling_table()</span></a> function.</p>
<p>Header: &lt;egglib-cpp/stirling.hpp&gt; </p>
</dd></dl>

</div>
<div class="section" id="functions">
<h3>Functions<a class="headerlink" href="#functions" title="Permalink to this headline">¶</a></h3>
<dl class="function">
<dt id="_CPPv2N6egglib14stirling_tableEjj">
<span id="egglib::stirling_table__unsigned-i.unsigned-i"></span><span class="target" id="project0group__diversity_1ga16615001364d52080bae649792438b47"></span>double <code class="descclassname">egglib::</code><code class="descname">stirling_table</code><span class="sig-paren">(</span>unsigned int <em>n</em>, unsigned int <em>k</em><span class="sig-paren">)</span><a class="headerlink" href="#_CPPv2N6egglib14stirling_tableEjj" title="Permalink to this definition">¶</a><br /></dt>
<dd><p>Get a pre-computed Stirling number of the first kind. </p>
<p>The n parameter must be &lt;= MAX_N_STIRLING and k must be &gt; 0 and &lt;= n.</p>
<p>Header: &lt;egglib-cpp/stirling.hpp&gt; </p>
</dd></dl>

</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="coalesce.html" title="Coalescent simulator"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../contents.html">EggLib&#39;s documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../toc.html" >Contents</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="index.html" >C++ Library</a> &#187;</li> 
      </ul>
    </div>
<!--
        <address style="text-align: right;"><small>Hosted by&nbsp;
        <a href="http://sourceforge.net/projects/egglib"><img src="http://sflogo.sourceforge.net/sflogo.php?group_id=232965&amp;type=8" width="80" height="15" alt="Get EggLib at SourceForge.net. Fast, secure and Free Open Source software downloads" / ALIGN=RIGHT BORDER=0></a>
        </address>
-->

  </body>
</html>