

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Genetic data holder types &#8212; EggLib&#39;s documentation</title>
    <link rel="stylesheet" href="../_static/egglibdoc.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '3.0.0b21',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/icon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Generic tools" href="tools.html" />
    <link rel="prev" title="Python Module" href="index.html" />
    <script type="text/javascript">
      // intelligent scrolling of the sidebar content
      $(window).scroll(function() {
        var sb = $('.sphinxsidebarwrapper');
        var win = $(window);
        var sbh = sb.height();
        var offset = $('.sphinxsidebar').position()['top'];
        var wintop = win.scrollTop();
        var winbot = wintop + win.innerHeight();
        var curtop = sb.position()['top'];
        var curbot = curtop + sbh;
        // does sidebar fit in window?
        if (sbh < win.innerHeight()) {
          // yes: easy case -- always keep at the top
          sb.css('top', $u.min([$u.max([0, wintop - offset - 10]),
                                $(document).height() - sbh - 200]));
        } else {
          // no: only scroll if top/bottom edge of sidebar is at
          // top/bottom edge of window
          if (curtop > wintop && curbot > winbot) {
            sb.css('top', $u.max([wintop - offset - 10, 0]));
          } else if (curtop < wintop && curbot < winbot) {
            sb.css('top', $u.min([winbot - sbh - offset - 20,
                                  $(document).height() - sbh - 200]));
          }
        }
      });
    </script>
  </head>
  <body>
<div style="background-color: white; text-align: left; padding: 0px 0px 0px 0px">
<a href="../index.html"><img src="../_static/banner.png" border="0" alt="EggLib"/></a>
</div>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="tools.html" title="Generic tools"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Python Module"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../contents.html">EggLib&#39;s documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../toc.html" >Contents</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="index.html" accesskey="U">Python Module</a> &#187;</li> 
      </ul>
    </div>

      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../contents.html">
              <img class="logo" src="../_static/egglib_small.bmp" alt="Logo"/>
            </a></p>
  <h3><a href="../contents.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Genetic data holder types</a><ul>
<li><a class="reference internal" href="#align">Align</a></li>
<li><a class="reference internal" href="#container">Container</a></li>
<li><a class="reference internal" href="#tree">Tree</a></li>
<li><a class="reference internal" href="#helpers">Helpers</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Python Module</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tools.html"
                        title="next chapter">Generic tools</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/py/toplevel.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="genetic-data-holder-types">
<span id="generic"></span><h1>Genetic data holder types<a class="headerlink" href="#genetic-data-holder-types" title="Permalink to this headline">¶</a></h1>
<p>The classes <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> and <a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> allow store and
manipulates genetic data without restriction on data encoding.
<a class="reference internal" href="#egglib.Tree" title="egglib.Tree"><code class="xref py py-class docutils literal"><span class="pre">Tree</span></code></a> allows to store and manipulate trees.</p>
<div class="section" id="align">
<h2>Align<a class="headerlink" href="#align" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="egglib.Align">
<em class="property">class </em><code class="descclassname">egglib.</code><code class="descname">Align</code><span class="sig-paren">(</span><em>nsam=0</em>, <em>nout=0</em>, <em>nsit=0</em>, <em>init=0</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#egglib._interface.DataBase" title="egglib._interface.DataBase"><code class="xref py py-class docutils literal"><span class="pre">egglib._interface.DataBase</span></code></a></p>
<p>Holds a data set with associated sample names and group information.
The data consists of given numbers of ingroup and outgroup samples,
each with the same number of sites. There can be any number of group
levels (but this number must be the same for all samples), meaning
that samples can be described by several group labels in addition to
their name. Group labels are not group indices (they do not need to
be consecutive). There is a separate data set for samples belonging
to the outgroup. There can be any number of outgroup samples.
Outgroup samples always have one level of group labels that should
be used to specify individuals (when appropriate). All data are
represented by signed integers</p>
<p>By default, the constructor generates an empty instance (0
samples and  0 sites).</p>
<p>By default, samples have empty names and no group levels are
defined.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>num_sam</strong> – number of ingroup samples.</li>
<li><strong>num_out</strong> – number of outgroup samples.</li>
<li><strong>num_sit</strong> – number of sites.</li>
<li><strong>init</strong> – initial values for all data entries (may be a
signed integer or a single-character string; ignored if
<em>num_sam</em> or <em>num_sit</em> is 0).</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 3.0.0: </span>Reimplementation of the Align class.</p>
</div>
<dl class="method">
<dt id="egglib.Align.add_outgroup">
<code class="descname">add_outgroup</code><span class="sig-paren">(</span><em>name</em>, <em>data</em>, <em>group=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.add_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an outgroup sample to the instance.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> – name of the new sample.</li>
<li><strong>data</strong> – an iterable (may be a string or a list) containing
the data to set for the new sample. For an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>
instance and if the sample is not the first, the number of
data must fit any previous ones (ingroup or outgroup).</li>
<li><strong>group</strong> – if not None, must be an unsigned integer (the
default value is 0).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.add_sample">
<code class="descname">add_sample</code><span class="sig-paren">(</span><em>name</em>, <em>data</em>, <em>groups=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.add_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a sample to the instance.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> – name of the new sample.</li>
<li><strong>data</strong> – an iterable (may be a string or a list) containing
the data to set for the new sample. For an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>
instance and if the sample is not the first, the number of
data must fit any previous ones (including outgroup samples,
if any).</li>
<li><strong>groups</strong> – if not None, must be an iterable with an
unsigned integer for each group levels of the instance (if
None, all group labels are set to 0), or a single integer
value if only one level needs to be specified.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.add_samples">
<code class="descname">add_samples</code><span class="sig-paren">(</span><em>items</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.add_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Add several samples at the end of the instance.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>items</strong> – items must be an iterable that have a length (for
example a list, an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> instance or a
<a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> instance. Each item of <em>items</em> must be
of length 2 or 3. The first item must be the sample name
string, the second item is the data values (as a list of
signed integers or a single string) and the third item, if
provided, is the list of group labels (one unsigned integer
for each level). See the method <a class="reference internal" href="#egglib.Align.add_sample" title="egglib.Align.add_sample"><code class="xref py py-meth docutils literal"><span class="pre">add_sample()</span></code></a>
for more details about each added item. If the current
instance is an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>, all added items must have
the same length which must be the same as any items
currently present in the instance. For a
<a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a>, the items may have different lengths.
The number of group levels is set by the sample with the
larger number of group labels, and all samples with less
labels are completed by e default value (0).</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 2.0.1: </span>Original name is <code class="xref py py-meth docutils literal"><span class="pre">addSequences()</span></code>.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Renamed as <code class="docutils literal"><span class="pre">add_samples</span></code>. All items are added in one shot
(rather than calling the one-sample add method iteratively).</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.clear">
<code class="descname">clear</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Clear the instance and release all memory. In most cases, it is
preferable to use the method <a class="reference internal" href="#egglib._interface.DataBase.reset" title="egglib._interface.DataBase.reset"><code class="xref py py-meth docutils literal"><span class="pre">reset()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.column">
<code class="descname">column</code><span class="sig-paren">(</span><em>index</em>, <em>ingroup=True</em>, <em>outgroup=True</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.column" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract the allele values of a site at a given position.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – the index of a site within the alignment.</li>
<li><strong>ingroup</strong> – A boolean indicating whether ingroup samples
should be extracted.</li>
<li><strong>outgroup</strong> – A boolean indicating whether outgroup samples
should be extracted.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>Returns one or two lists of integers providing the allele for
all samples. If both <em>ingroup</em> and <em>outgroup</em> are <code class="docutils literal"><span class="pre">True</span></code>, two
lists are returned, respectively for the ingroup and for the
outgroup. If only one of <em>ingroup</em> and <em>outgroup</em> is <code class="docutils literal"><span class="pre">True</span></code>,
a single list is returned with data for the selected group. If
neither is <code class="docutils literal"><span class="pre">True</span></code>, the returned value is <code class="docutils literal"><span class="pre">None</span></code> (no error).</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.consensus">
<code class="descname">consensus</code><span class="sig-paren">(</span><em>ingroup=True</em>, <em>outgroup=False</em>, <em>ignore=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.consensus" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates the consensus of the object, assuming nucleotide
sequences. The consensus is generated based on standard
ambiguity (IUPAC) codes. The consensus is returned as a string
object, of length matching the alignment length. The input
alignment can contain nucleotide bases (<code class="docutils literal"><span class="pre">A</span></code>, <code class="docutils literal"><span class="pre">C</span></code>, <code class="docutils literal"><span class="pre">G</span></code> and
<code class="docutils literal"><span class="pre">T</span></code>), and all ambiguity codes (<code class="docutils literal"><span class="pre">V</span></code>, <code class="docutils literal"><span class="pre">H</span></code>, <code class="docutils literal"><span class="pre">M</span></code>, <code class="docutils literal"><span class="pre">D</span></code>,
<code class="docutils literal"><span class="pre">R</span></code>, <code class="docutils literal"><span class="pre">W</span></code>, <code class="docutils literal"><span class="pre">B</span></code>, <code class="docutils literal"><span class="pre">S</span></code>, <code class="docutils literal"><span class="pre">Y</span></code>, <code class="docutils literal"><span class="pre">K</span></code> and <code class="docutils literal"><span class="pre">N</span></code>). <code class="docutils literal"><span class="pre">N</span></code>
stands for any of <code class="docutils literal"><span class="pre">A</span></code>, <code class="docutils literal"><span class="pre">C</span></code>, <code class="docutils literal"><span class="pre">G</span></code> and <code class="docutils literal"><span class="pre">T</span></code>. <code class="docutils literal"><span class="pre">B</span></code> stands
for any of <code class="docutils literal"><span class="pre">C</span></code>, <code class="docutils literal"><span class="pre">G</span></code> and <code class="docutils literal"><span class="pre">T</span></code> and so on. Case is ignored.
<code class="docutils literal"><span class="pre">U</span></code> is treated exactly as <code class="docutils literal"><span class="pre">T</span></code>. Gaps are represented by <code class="docutils literal"><span class="pre">-</span></code>
and missing data by <code class="docutils literal"><span class="pre">?</span></code>. Any other value will result in a
ValueError. If a site is not variable, the fixed value is
incorporated in the consensus in all cases.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>ingroup</strong> – A boolean indicating whether the ingroup must be
exported.</li>
<li><strong>outgroup</strong> – A boolean indicating whether the outgroup must
be exported.</li>
<li><strong>ignore</strong> – a boolean indicating whether missing data should
be ignored. If missing data are not ignored (by default),
any missing data (including gaps) cause the whole site to
have ? as consensus. If <em>ignore</em> is <code class="docutils literal"><span class="pre">True</span></code>, missing data
are skipped and the consensus is based on the other
(non-missing) data. If there is no non-missing data, the
consensus is <code class="docutils literal"><span class="pre">N</span></code>.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Options are added, and implementation is modified.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.create">
<code class="descname">create</code><span class="sig-paren">(</span><em>obj</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.create" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new instance by copying date from the data container
passed as <em>obj</em>. The object <em>obj</em> can be:</p>
<ul class="simple">
<li>an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>,</li>
<li>a <a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> (but all sequences are required to have
the same length if the target type is <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>),</li>
<li>any iterable type yielding a acceptable items (see the
documentation for <a class="reference internal" href="#egglib._interface.DataBase.add_samples" title="egglib._interface.DataBase.add_samples"><code class="xref py py-meth docutils literal"><span class="pre">add_samples()</span></code></a> for more details).</li>
</ul>
<div class="versionadded">
<p><span class="versionmodified">New in version 2.0.1.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.del_columns">
<code class="descname">del_columns</code><span class="sig-paren">(</span><em>site</em>, <em>num=1</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.del_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete full columns for all ingroup and outgroup samples.</p>
<p>By default (if <code class="docutils literal"><span class="pre">num=1</span></code>), remove a single site. If <em>num</em> is
larger than 1, remove a range of sites.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>site</strong> – index of the (first) site to remove. This site must
be a valid index.</li>
<li><strong>num</strong> – maximal number of sites to remove. The value cannot
be negative.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.del_outgroup">
<code class="descname">del_outgroup</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.del_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete an item of the instance.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – index of the sample to delete.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.del_sample">
<code class="descname">del_sample</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.del_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete an item of the instance.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – index of the sample to delete.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.encode">
<code class="descname">encode</code><span class="sig-paren">(</span><em>nbits=10</em>, <em>random=None</em>, <em>include_outgroup=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.encode" title="Permalink to this definition">¶</a></dt>
<dd><p>Renames all sequences using a random mapping of names to unique
keys of length <em>nbits</em>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>nbits</strong> – length of the keys (encoded names). This value
must be &gt;= 4 and &lt;= 63.</li>
<li><strong>random</strong> – a <a class="reference internal" href="tools.html#egglib.tools.Random" title="egglib.tools.Random"><code class="xref py py-class docutils literal"><span class="pre">Random</span></code></a> instance to be used a random
generator. By default, use a default instance.</li>
<li><strong>include_outgroup</strong> – a <a class="reference external" href="https://docs.python.org/2.7/library/functions.html#bool" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> indicating whether the
outgroup samples should also be considered (in this case, a
single mapping is returned including both ingroup and
outgroup samples).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">A dictionary mapping all the generated keys to the
actual sequence names. The keys are case-dependent and
guaranteed not to start with a number.</p>
</td>
</tr>
</tbody>
</table>
<p>The returned mapping can be used to restore the original names
using <a class="reference internal" href="#egglib._interface.DataBase.rename" title="egglib._interface.DataBase.rename"><code class="xref py py-meth docutils literal"><span class="pre">rename()</span></code></a>. This method is not affected by
the presence of sequences with identical names in the original
instance (and <a class="reference internal" href="#egglib._interface.DataBase.rename" title="egglib._interface.DataBase.rename"><code class="xref py py-meth docutils literal"><span class="pre">rename()</span></code></a> will also work properly
in that case).</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 2.0.1.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Keys are forced to start with a capital letter. Now take
a <a class="reference internal" href="tools.html#egglib.tools.Random" title="egglib.tools.Random"><code class="xref py py-class docutils literal"><span class="pre">Random</span></code></a> instance. Use library’s own random
number generator. Added an option for the outgroup.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.extract">
<code class="descname">extract</code><span class="sig-paren">(</span><em>*args</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.extract" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract given positions (or columns) of the alignment and
returns a new alignment.</p>
<p>The two possible ways to call this method are:</p>
<blockquote>
<div><code class="docutils literal"><span class="pre">extract(start,</span> <span class="pre">stop)</span></code> to extract a continuous range of
sites,</div></blockquote>
<p>and</p>
<blockquote>
<div><code class="docutils literal"><span class="pre">extract(indexes)</span></code> to extract a random list of positions
(in any order).</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>start</strong> – first position to extract. This position must be
a valid index for this alignment.</li>
<li><strong>stop</strong> – stop position for the range to extract. <strong>This
position is not extracted.</strong> If this position is equal to or
smaller than <em>start</em>, empty sequences are extracted. If this
position is equal to or larger than the length of the
alignment, or if it is equal to <code class="docutils literal"><span class="pre">None</span></code>, all positions
until the end of the alignment are extracted.</li>
<li><strong>indexes</strong> – a list (or other iterable type with a length) of
alignment positions (or column indexes). This list may
contain repetitions and does not need to be sorted. The
positions will be extracted in the specified order.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>Keyword arguments are not supported.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 2.0.1.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.filter">
<code class="descname">filter</code><span class="sig-paren">(</span><em>ratio</em>, <em>valid='ACGTacgt'</em>, <em>ingroup=True</em>, <em>outgroup=True</em>, <em>relative=True</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes the sequences with too few valid sites. This method
modifies the current instance and returns <code class="docutils literal"><span class="pre">None</span></code>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>ratio</strong> – limit threshold, expressed as a proportion of
either the maximum number of valid data over all processed
samples (if the <em>relative</em> argument is <code class="docutils literal"><span class="pre">True</span></code>) or the
alignment length (otherwise).</li>
<li><strong>valid</strong> – a string or an interable of one-character strings
or integers giving the allelic values considered to be valid
(note that the comparisons are case-dependent).</li>
<li><strong>ingroup</strong> – A boolean indicating whether ingroup samples
must be processed.</li>
<li><strong>outgroup</strong> – A boolean indicating whether outgroup samples
must be processed.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>If the length of the alignment is 0, or if both <em>ingroup</em> and
<em>outgroup</em> are <code class="docutils literal"><span class="pre">False</span></code>, nothing is done.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.find">
<code class="descname">find</code><span class="sig-paren">(</span><em>name</em>, <em>include_outgroup=False</em>, <em>regex=False</em>, <em>multi=False</em>, <em>flags=None</em>, <em>index=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.find" title="Permalink to this definition">¶</a></dt>
<dd><p>Find a sample by its name.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>name</strong> – name of sample to identify.</li>
<li><strong>include_outgroup</strong> – a boolean indicating whether the
outgroup should be considered.</li>
<li><strong>regex</strong> – a boolean indicating whether the value passed a
<em>name</em> is a regular expression. If so, the string is passed
as is to the re module (using function
<a class="reference external" href="https://docs.python.org/2.7/library/re.html#re.search" title="(in Python v2.7)"><code class="xref py py-func docutils literal"><span class="pre">re.search()</span></code></a>).
Otherwise, only exact matches will be considered.</li>
<li><strong>multi</strong> – a boolean indicating whether all hits should be
returned. If so, a list of <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instances
is always returned (the list will be empty in case of no
hits). Otherwise, a single <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instance
(or its index)  will be returned for the first hit, or
<code class="docutils literal"><span class="pre">None</span></code> in case of no hits.</li>
<li><strong>flags</strong> – list of flags to be passed to <a class="reference external" href="https://docs.python.org/2.7/library/re.html#re.search" title="(in Python v2.7)"><code class="xref py py-func docutils literal"><span class="pre">re.search()</span></code></a>
(ignored if <em>regex</em> is <code class="docutils literal"><span class="pre">False</span></code>). For example, when looking
for samples containing the term “sample”, for being case
insensitive, use the following syntax:
<code class="docutils literal"><span class="pre">align.find(&quot;sample&quot;,</span> <span class="pre">regex=True,</span> <span class="pre">flags=[re.I])</span></code>. By
default (<code class="docutils literal"><span class="pre">None</span></code>) no further argument is passed.</li>
<li><strong>index</strong> – boolean indicating whether the index of the sample
should be returned. In that case return values for hits are
<a class="reference external" href="https://docs.python.org/2.7/library/functions.html#int" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> (by default, <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a>) instances.
Warning: it is not allowed to set both <em>include_outgroup</em>
and <em>index</em> to <code class="docutils literal"><span class="pre">True</span></code> as there would be no way to
distinguish between ingroup and outgroup indexes.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><code class="docutils literal"><span class="pre">None</span></code> if no hits were found, a (potentially empty)
list of <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instances or class:<cite>int</cite> if
<em>multi</em> is <code class="docutils literal"><span class="pre">True</span></code>, or a single <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> or
<a class="reference external" href="https://docs.python.org/2.7/library/functions.html#int" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> otherwise.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.find_motif">
<code class="descname">find_motif</code><span class="sig-paren">(</span><em>sample</em>, <em>motif</em>, <em>start=0</em>, <em>stop=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.find_motif" title="Permalink to this definition">¶</a></dt>
<dd><p>Locate the first instance of a motif for an ingroup sample</p>
<p>Returns the index of the first exact hit to a given substring.
The returned value is the position of the first base of the hit.
Only exact matches are implemented. To use regular expression
(for example to find degenerated motifs), one should extract the
string for the sequence and use a tool such as the regular
expression module (<a class="reference external" href="https://docs.python.org/2.7/library/re.html#module-re" title="(in Python v2.7)"><code class="xref py py-mod docutils literal"><span class="pre">re</span></code></a>).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>motif</strong> – a list of integer, or one-character strings (or
mixing both) or a string constituting the motif to search.</li>
<li><strong>start</strong> – position at which to start searching. The method
will never return a value smaller than <em>start</em>. By default,
search from the start of the sequence.</li>
<li><strong>stop</strong> – position at which to stop search (the motif
cannot overlap this position). No returned value will be
larger than stop-<em>n</em>. By default, or if <em>stop</em> is equal to
or larger than the length of the sequence, search until the
end of the sequence.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.find_motif_o">
<code class="descname">find_motif_o</code><span class="sig-paren">(</span><em>sample</em>, <em>motif</em>, <em>start=0</em>, <em>stop=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.find_motif_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Locate the first instance of a motif for an outgroup sample</p>
<p>Returns the index of the first exact hit to a given substring.
The returned value is the position of the first base of the hit.
Only exact matches are implemented. To use regular expression
(for example to find degenerated motifs), one should extract the
string for the sequence and use a tool such as the regular
expression module (<a class="reference external" href="https://docs.python.org/2.7/library/re.html#module-re" title="(in Python v2.7)"><code class="xref py py-mod docutils literal"><span class="pre">re</span></code></a>).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – outgroup sample index.</li>
<li><strong>motif</strong> – a list of integer, or one-character strings (or
mixing both) or a string constituting the motif to search.</li>
<li><strong>start</strong> – position at which to start searching. The method
will never return a value smaller than <em>start</em>. By default,
search from the start of the sequence.</li>
<li><strong>stop</strong> – position at which to stop search (the motif
cannot overlap this position). No returned value will be
larger than <em>stop</em> minus the length of the motif. By
default, or if <em>stop</em> is equal to or larger than the length
of the sequence, search until the end of the sequence.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.fix_ends">
<code class="descname">fix_ends</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.fix_ends" title="Permalink to this definition">¶</a></dt>
<dd><p>Designed for nucleotide sequence alignments. Replaces all
leading and trailing occurrence of the alignment gap symbol
(the numeric equivalent of <code class="docutils literal"><span class="pre">-</span></code>) by missing data symbols
(<code class="docutils literal"><span class="pre">?</span></code>). Internal alignment gaps (those having at least one
character other than <code class="docutils literal"><span class="pre">-</span></code> and <code class="docutils literal"><span class="pre">?</span></code> at each side) are left
unchanged.</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Renamed from <code class="xref py py-meth docutils literal"><span class="pre">fix_gap_ends()</span></code> to <a class="reference internal" href="#egglib.Align.fix_ends" title="egglib.Align.fix_ends"><code class="xref py py-meth docutils literal"><span class="pre">fix_ends()</span></code></a>.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.get_i">
<code class="descname">get_i</code><span class="sig-paren">(</span><em>sample</em>, <em>site</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.get_i" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a data entry.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>site</strong> – site index.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.get_label">
<code class="descname">get_label</code><span class="sig-paren">(</span><em>sample</em>, <em>level</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.get_label" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a group label.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>level</strong> – level index.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.get_label_o">
<code class="descname">get_label_o</code><span class="sig-paren">(</span><em>sample</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.get_label_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Get an outgroup label.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>sample</strong> – sample index.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.get_name">
<code class="descname">get_name</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.get_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the name of a sample.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – sample index.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.get_name_o">
<code class="descname">get_name_o</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.get_name_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the name of an outgroup sample.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – outgroup sample index.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.get_o">
<code class="descname">get_o</code><span class="sig-paren">(</span><em>sample</em>, <em>site</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.get_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Get an outgroup data entry.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – outgroup sample index.</li>
<li><strong>site</strong> – site index.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.get_outgroup">
<code class="descname">get_outgroup</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.get_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instance corresponding to the
requested index for the outgroup. The returned object allows to
modify the underlying data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – index of the outgroup sample to access.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.get_sample">
<code class="descname">get_sample</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.get_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instance corresponding to the
requested index. The returned object allows to modify the
underlying data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – index of the sample to access.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.get_sequence">
<code class="descname">get_sequence</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.get_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Access to the data entries of a given ingroup sample. Returns a
<a class="reference internal" href="#egglib.SequenceView" title="egglib.SequenceView"><code class="xref py py-class docutils literal"><span class="pre">SequenceView</span></code></a> instance which allow modifying the
underlying sequence container object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – ingroup sample index.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.get_sequence_o">
<code class="descname">get_sequence_o</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.get_sequence_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Access to the data entries of a given outgroup sample. Returns a
<a class="reference internal" href="#egglib.SequenceView" title="egglib.SequenceView"><code class="xref py py-class docutils literal"><span class="pre">SequenceView</span></code></a> instance which allow modifying the
underlying sequence container object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – outgroup sample index.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.group_mapping">
<code class="descname">group_mapping</code><span class="sig-paren">(</span><em>level=0</em>, <em>indices=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.group_mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a dictionary mapping group labels to either
<a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instances (by default) or their indexes
representing all samples of this instance. It can process and
ingroup label level or the outgroup.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>level</strong> – index of the level to consider. If <code class="docutils literal"><span class="pre">None</span></code>,
processes the outgroup.</li>
<li><strong>indices</strong> – if <code class="docutils literal"><span class="pre">True</span></code>, represent samples by their index
(within the ingroup or the outgroup depending on <em>level</em>
being non-<code class="docutils literal"><span class="pre">None</span></code> or <code class="docutils literal"><span class="pre">None</span></code>) instead of
<a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instances.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.insert_columns">
<code class="descname">insert_columns</code><span class="sig-paren">(</span><em>position</em>, <em>values</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.insert_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Insert sites at a given position to an alignment.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>position</strong> – the position at which to insert sites. Sites
are inserted <em>before</em> the specified position, so the user
can use 0 to insert sites at the beginning of the sequence.
To insert sites at the end of the sequence, pass the current
length of the alignment, or <code class="docutils literal"><span class="pre">None</span></code>. If <em>position</em> is
larger than the length of the sequence or <code class="docutils literal"><span class="pre">None</span></code>, new
sites are inserted at the end of the alignment. The position
might be negative to count from the end. Warning: the
position -1 means <em>before</em> the last position.</li>
<li><strong>values</strong> – a list of signed integers, or a string, providing
data to insert into the instance. The same sequence will be
inserted for all ingroup and outgroup samples.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.intersperse">
<code class="descname">intersperse</code><span class="sig-paren">(</span><em>length</em>, <em>positions=None</em>, <em>alleles='A'</em>, <em>random=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.intersperse" title="Permalink to this definition">¶</a></dt>
<dd><p>Insert non-varying sites within the alignment. The current
object is permanently modified.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>length</strong> – Desired length of the final alignment. If the
value is smaller than the original (current) alignment
length, nothing is done and the alignment is unchanged.</li>
<li><strong>positions</strong> – List of positions of sites of this alignment.
The number of positions must be equal to the number of sites
of the alignment (before interspersing). The argument value
must be either a sequence of positive integers or a sequence
of real numbers comprised between 0 and 1. In either case,
values must be in increasing order. In the former case, the
last (maximal) value must be smaller than the desired length
of the final alignment. In the latter case, values are
expressed relatively, and they will be converted to integer
indexes by the method. In that case, if site positioning is
non-trivial (typically, if conversion of positions to
integer yield identical position for different conscutive
sites), it will be resolved randomly. By default (if
<code class="docutils literal"><span class="pre">None</span></code>), sites are placed regularly along the final
alignment length. If <a class="reference external" href="https://docs.python.org/2.7/library/functions.html#int" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> and <a class="reference external" href="https://docs.python.org/2.7/library/functions.html#float" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">float</span></code></a> types
are mixed, the first occurring type will condition what will
happen.</li>
<li><strong>alleles</strong> – String (or any sequence of one-character
strings) providing the alleles to be used to fill
non-varying positions of the resulting alignment. If there
is more than one allele, the allele will be picked randomly
for each site, independently for each interted site.</li>
<li><strong>random</strong> – A <a class="reference internal" href="tools.html#egglib.tools.Random" title="egglib.tools.Random"><code class="xref py py-class docutils literal"><span class="pre">Random</span></code></a> instance to be used a random
generator. By default, use a default instance.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="egglib.Align.is_matrix">
<code class="descname">is_matrix</code><a class="headerlink" href="#egglib.Align.is_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p><code class="docutils literal"><span class="pre">True</span></code> if the instance is an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>, and <code class="docutils literal"><span class="pre">False</span></code> if
it is a <a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.iter_outgroup">
<code class="descname">iter_outgroup</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.iter_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterator over the outgroup.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.iter_samples">
<code class="descname">iter_samples</code><span class="sig-paren">(</span><em>ingrp</em>, <em>outgrp</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.iter_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterator over either ingroup samples or outgroup samples, or
both ingroup and outgroup (in that order) samples. Provided for
flexibility.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>ingrp</strong> – include ingroup samples.</li>
<li><strong>outgroup</strong> – include outgroup samples.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="egglib.Align.ls">
<code class="descname">ls</code><a class="headerlink" href="#egglib.Align.ls" title="Permalink to this definition">¶</a></dt>
<dd><p>Alignment length. This value cannot be set or modified directly.
It is not possible to get the number of samples of a single
ingroup or ingroup sample as they are all equal to this value.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.name_mapping">
<code class="descname">name_mapping</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.name_mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a dictionary mapping names to <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a>
instances representing all samples of this instance. This method
is most useful when several sequences have the same name. It may
be used to detect and process duplicates. It processes the
ingroup only.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.names">
<code class="descname">names</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.names" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate the list of ingroup sample names.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.names_outgroup">
<code class="descname">names_outgroup</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.names_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate the list of outgroup sample names.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.nexus">
<code class="descname">nexus</code><span class="sig-paren">(</span><em>prot=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.nexus" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a simple nexus-formatted string. If <em>prot</em> is
<code class="docutils literal"><span class="pre">True</span></code>, adds <code class="docutils literal"><span class="pre">datatype=protein</span></code> in the file, allowing it to
be imported as proteins (but doesn’t perform further checking).</p>
<p>Returns a nexus-formatted string. Note: any spaces and tabs in
sequence names are replaced by underscores. This nexus
implementation is minimal but will normally suffice to export
sequences to programs expecting nexus.</p>
<p>Note: only the ingroup is exported. The data must be exportable
as strings.</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib.Align.ng">
<code class="descname">ng</code><a class="headerlink" href="#egglib.Align.ng" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of group levels for ingroup (for outgroup, this number is
always one). It possible to change the value directly.</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib.Align.no">
<code class="descname">no</code><a class="headerlink" href="#egglib.Align.no" title="Permalink to this definition">¶</a></dt>
<dd><p>Current number of outgroup samples (cannot be modified
directly).</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib.Align.ns">
<code class="descname">ns</code><a class="headerlink" href="#egglib.Align.ns" title="Permalink to this definition">¶</a></dt>
<dd><p>Current number of samples, not considering the outgroup (cannot
be modified directly).</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.phylip">
<code class="descname">phylip</code><span class="sig-paren">(</span><em>format='I'</em>, <em>ingroup=True</em>, <em>outgroup=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.phylip" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a phyml-formatted string representing the content of the
instance. The phyml format is suitable as input data for PhyML
and PAML software. Raises a <a class="reference external" href="https://docs.python.org/2.7/library/exceptions.html#exceptions.ValueError" title="(in Python v2.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> if any
name of the instance contains at least one character of the
following list: <code class="docutils literal"><span class="pre">()[]{},;</span></code> or a space, tab, newline or
linefeed. Group labels are never exported. Sequence names cannot
be longer than 10 characters. A <a class="reference external" href="https://docs.python.org/2.7/library/exceptions.html#exceptions.ValueError" title="(in Python v2.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a>
will be raised if a longer name is met. <em>format</em> must be ‘I’ or
‘S’ (case-independent), indicating whether the data should be
formatted in the sequential (S) or interleaved (I) format (see
PHYLIP’s documentation for definitions). The user is responsible
of ensuring that all names are unique. If not, the exported file
my cause subsequent programs to fail.</p>
<p>The sequences must all be convertible into characters.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>ingroup</strong> – A boolean indicating whether the ingroup must be
exported.</li>
<li><strong>outgroup</strong> – A boolean indicating whether the outgroup must
be exported.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Added the <em>ingroup</em> and <em>outgroup</em> options.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.phyml">
<code class="descname">phyml</code><span class="sig-paren">(</span><em>ingroup=True</em>, <em>outgroup=False</em>, <em>strict=True</em>, <em>dtype=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.phyml" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a phyml-formatted string representing the content of the
instance. The phyml format is suitable as input data for the
PhyML and PAML programmes. Raises a
<a class="reference external" href="https://docs.python.org/2.7/library/exceptions.html#exceptions.ValueError" title="(in Python v2.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> if any name of the instance
contains at least one character in the following list:
<code class="docutils literal"><span class="pre">()[]{},;</span></code> or a space, tab, newline or linefeed.
Group information is never exported.</p>
<p>The sequences must all be convertible into characters.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>ingroup</strong> – A boolean indicating whether the ingroup must be
exported.</li>
<li><strong>outgroup</strong> – A boolean indicating whether the outgroup must
be exported.</li>
<li><strong>strict</strong> – enforce that all characters within the instances
are valid (assuming either nucleotide or amino acid sequences,
see <em>dtype</em>). If <code class="docutils literal"><span class="pre">False</span></code>, characters are not checked. Also,
enforce that their is no blank character, round bracket,
colon, or comma in sequence names. When checking, both ingroup
and outgroup are always checked.</li>
<li><strong>dtype</strong> – one of <code class="docutils literal"><span class="pre">None</span></code> (default), <code class="docutils literal"><span class="pre">nt</span></code> (nucleotides), or <code class="docutils literal"><span class="pre">aa</span></code>
(amino acids). Type of data assumed (only if <em>strict</em> is set to
<code class="docutils literal"><span class="pre">True</span></code>). By default, allow for either nucleotides or amino acids
(but not a combination of both).</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Added the <em>ingroup</em>, <em>outgroup</em>, <em>strict</em>, and <em>dtype</em> options.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.random_missing">
<code class="descname">random_missing</code><span class="sig-paren">(</span><em>rate</em>, <em>ch='N'</em>, <em>valid='ACGTacgt'</em>, <em>ingroup=True</em>, <em>outgroup=True</em>, <em>random=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.random_missing" title="Permalink to this definition">¶</a></dt>
<dd><p>Randomly introduces missing data in the current instance. Random
positions of the alignment are changed to missing data. Only
data that are currently non-missing data are considered.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>rate</strong> – probability that a non-mssing (as defined after the
<em>valid</em> argument) data is turned into missing data.</li>
<li><strong>ch</strong> – missing data character, to be used for all
replacements (as a single-character string or an integer).</li>
<li><strong>valid</strong> – a string or list of integers (or any iterable,
possibly missing integers and one-character strings) given
the allele values that may be turned into missing data.</li>
<li><strong>ingroup</strong> – A boolean indicating whether the ingroup must be
processed.</li>
<li><strong>outgroup</strong> – A boolean indicating whether the outgroup must
be processed.</li>
<li><strong>random</strong> – a <a class="reference internal" href="tools.html#egglib.tools.Random" title="egglib.tools.Random"><code class="xref py py-class docutils literal"><span class="pre">Random</span></code></a> instance. If <code class="docutils literal"><span class="pre">None</span></code>,
the method will use a default instance.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 2.1.0: </span>Restricted to <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> instances.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Takes the <em>ch</em>, <em>valid</em>, <em>ingroup</em>, <em>outgroup</em> and <em>random</em>
arguments. Reimplementation dropping dependency on the
non-default <code class="docutils literal"><span class="pre">numpy</span></code> package.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.remove_duplicates">
<code class="descname">remove_duplicates</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.remove_duplicates" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove all duplicates, based on name exact matching. For all
pairs of samples with identical name, only the one occurring
first is conserved. The current instance is modified and this
method returns <code class="docutils literal"><span class="pre">None</span></code>.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.rename">
<code class="descname">rename</code><span class="sig-paren">(</span><em>mapping</em>, <em>liberal=False</em>, <em>include_outgroup=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.rename" title="Permalink to this definition">¶</a></dt>
<dd><p>Rename sequences of the instance using the provided mapping.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>mapping</strong> – a <a class="reference external" href="https://docs.python.org/2.7/library/stdtypes.html#dict" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">dict</span></code></a> providing the mapping of old
names (as keys) to new names (which may, if needed, contain
duplicated).</li>
<li><strong>liberal</strong> – if this argument is <cite>False</cite> and a name does not
appear in <em>mapping</em>, a <a class="reference external" href="https://docs.python.org/2.7/library/exceptions.html#exceptions.ValueError" title="(in Python v2.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> is
raised. If <em>liberal</em> is <code class="docutils literal"><span class="pre">True</span></code>, names that don’t appear in
<em>mapping</em> are left unchanged.</li>
<li><strong>include_outgroup</strong> – if <code class="docutils literal"><span class="pre">True</span></code>, consider both ingroup and
outgroup samples together (they should be provided together
in the same mapping).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">The number of samples that have been actually renamed,
overall.</p>
</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 2.0.1.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Added an option for the outgroup. Added return value.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.reserve">
<code class="descname">reserve</code><span class="sig-paren">(</span><em>nsam=0</em>, <em>nout=0</em>, <em>lnames=0</em>, <em>ngrp=0</em>, <em>nsit=0</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.reserve" title="Permalink to this definition">¶</a></dt>
<dd><p>Pre-allocate memory. This method can be used when the size of
arrays is known a priori, in order to speed up memory
allocations. It is not necessary to set all values. Values less
than 0 are ignored.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>nsam</strong> – number of samples in the ingroup.</li>
<li><strong>nout</strong> – number of samples in the outgroup.</li>
<li><strong>lnames</strong> – length of sample names.</li>
<li><strong>ngrp</strong> – number of group labels.</li>
<li><strong>nsit</strong> – number of sites.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.reset">
<code class="descname">reset</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.reset" title="Permalink to this definition">¶</a></dt>
<dd><p>Reset the instance.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.set_i">
<code class="descname">set_i</code><span class="sig-paren">(</span><em>sample</em>, <em>site</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.set_i" title="Permalink to this definition">¶</a></dt>
<dd><p>Set an ingroup data entry. The value must be a signed integer or a
one-character string. In the latter case, <a class="reference external" href="https://docs.python.org/2.7/library/functions.html#ord" title="(in Python v2.7)"><code class="xref py py-func docutils literal"><span class="pre">ord()</span></code></a> is
called on the behalf of the user.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>site</strong> – site index.</li>
<li><strong>value</strong> – allele value.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.set_label">
<code class="descname">set_label</code><span class="sig-paren">(</span><em>sample</em>, <em>level</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.set_label" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a group label.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>level</strong> – level index.</li>
<li><strong>value</strong> – new group value (unsigned integer).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.set_label_o">
<code class="descname">set_label_o</code><span class="sig-paren">(</span><em>sample</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.set_label_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Get an outgroup label.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>value</strong> – new group value (unsigned integer).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.set_name">
<code class="descname">set_name</code><span class="sig-paren">(</span><em>index</em>, <em>name</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.set_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the name of a sample.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – index of the sample</li>
<li><strong>name</strong> – new name value.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.set_name_o">
<code class="descname">set_name_o</code><span class="sig-paren">(</span><em>index</em>, <em>name</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.set_name_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the name of an outgroup sample.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – index of the outgroup sample</li>
<li><strong>name</strong> – new name value.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.set_o">
<code class="descname">set_o</code><span class="sig-paren">(</span><em>sample</em>, <em>site</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.set_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Set an outgroup data entry. The value must be a signed integer
or a one-character string. In the latter case, <a class="reference external" href="https://docs.python.org/2.7/library/functions.html#ord" title="(in Python v2.7)"><code class="xref py py-func docutils literal"><span class="pre">ord()</span></code></a> 
is called on the behalf of the user.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>site</strong> – site index.</li>
<li><strong>value</strong> – allele value.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.set_outgroup">
<code class="descname">set_outgroup</code><span class="sig-paren">(</span><em>index</em>, <em>name</em>, <em>data</em>, <em>group=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.set_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the values for a sample of the outgroup</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – index of the sample to access (slices are not
permitted).</li>
<li><strong>name</strong> – new name of the sample.</li>
<li><strong>data</strong> – string or list of integers given the new values to
set. In case of an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>, it is required to pass a
sequence with length matching the number of sites of the
instance.</li>
<li><strong>groups</strong> – a positive integer to serve as group label (the
default value is 0.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.set_sample">
<code class="descname">set_sample</code><span class="sig-paren">(</span><em>index</em>, <em>name</em>, <em>data</em>, <em>groups=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.set_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the values for the sample corresponding to the requested
index.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – index of the sample to access (slices are not
permitted).</li>
<li><strong>name</strong> – new name of the sample.</li>
<li><strong>data</strong> – string or list of integers given the new values to
set. In case of an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>, it is required to pass a
sequence with length matching the number of sites of the
instance.</li>
<li><strong>groups</strong> – a list of integer label values, or a single
integer value, to set as group labels. The class will ensure
that all samples have the same number of group labels,
padding with 0 as necessary. The default corresponds to an
empty list.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.set_sequence">
<code class="descname">set_sequence</code><span class="sig-paren">(</span><em>index</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.set_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Replace all data entries for a given ingroup sample by new
values.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – ingroup sample index.</li>
<li><strong>value</strong> – can be a <a class="reference internal" href="#egglib.SequenceView" title="egglib.SequenceView"><code class="xref py py-class docutils literal"><span class="pre">SequenceView</span></code></a> instance, a
string or a list of integers (or any iterable with a length
which may mix single-character strings and integers). When
modifying an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> instance, all sequences must
have the same length as the current alignment length.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.set_sequence_o">
<code class="descname">set_sequence_o</code><span class="sig-paren">(</span><em>index</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.set_sequence_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Replace all data entries for a given outgroup sample by new
values.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – outgroup sample index.</li>
<li><strong>value</strong> – can be a <a class="reference internal" href="#egglib.SequenceView" title="egglib.SequenceView"><code class="xref py py-class docutils literal"><span class="pre">SequenceView</span></code></a> instance, a
string or a list of integers (or any iterable with a length
which may mix single-character strings and integers). When
modifying an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> instance, all sequences must
have the same length as the current alignment length.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.shuffle">
<code class="descname">shuffle</code><span class="sig-paren">(</span><em>level=0</em>, <em>random=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.shuffle" title="Permalink to this definition">¶</a></dt>
<dd><p>Shuffle group labels.</p>
<p>Randomly reassigns group labels. Modifies the current instance
and returns <code class="docutils literal"><span class="pre">None</span></code>. Only the specified level is affected, and
only the group labels are modified (the order of samples is not
changed.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>level</strong> – index of the group level to shuffle. To shuffle
the outgroup’s labels, set <em>level</em> to <code class="docutils literal"><span class="pre">None</span></code>.</li>
<li><strong>random</strong> – A <a class="reference internal" href="tools.html#egglib.tools.Random" title="egglib.tools.Random"><code class="xref py py-class docutils literal"><span class="pre">Random</span></code></a> instance to be used a random
generator. By default, use a default instance.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>The outgroup is necessarily processed separately from the
rest of the instance.  Allow pass a <a class="reference internal" href="tools.html#egglib.tools.Random" title="egglib.tools.Random"><code class="xref py py-class docutils literal"><span class="pre">Random</span></code></a>
instance. Use library’s own random number generator.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.slider">
<code class="descname">slider</code><span class="sig-paren">(</span><em>wwidth</em>, <em>wstep</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.slider" title="Permalink to this definition">¶</a></dt>
<dd><p>Provides a means to perform sliding-windows analysis over the
alignment. This method returns a generator that can be used as
in <code class="docutils literal"><span class="pre">for</span> <span class="pre">window</span> <span class="pre">in</span> <span class="pre">align.slider(wwidth,</span> <span class="pre">wstep)</span></code>, where each
step <em>window</em> of the iteration will be the reference to a
<a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> instance of length <em>wwidth</em> (or less if not
enough sequence is available near the end of the alignment).
Each step moves forward following the value of <em>wstep</em>.</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>The returned <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> is actually a reference to the
same object, which is stored within the instance and
repeatively returned.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.subgroup">
<code class="descname">subgroup</code><span class="sig-paren">(</span><em>groups</em>, <em>outgroup_all=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.subgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate and return a copy of the instance with only samples
from the specified groups (identified by their group labels,
including outgroup).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>groups</strong> – an integer or a dictionary providing the labels
of the groups that are selected. If an integer, it is
understood as an ingroup label corresponding to the first
grouping level (if more than one). If a dictionary, the keys
of this dictionary must be integer corresponding to grouping
levels (or <code class="docutils literal"><span class="pre">None</span></code> for the outgroup) and the values must be
lists containing the requested labels. It is not required to
include all levels and the outgroup in the dictionary. It is
allowed to specify labels that are actually not represented
in the data.</li>
<li><strong>outgroup_all</strong> – if <code class="docutils literal"><span class="pre">True</span></code>, always include all outgroup
samples in the returned instance, regardless of whether
their group labels are specified in the first argument.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 3.0.0.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.subset">
<code class="descname">subset</code><span class="sig-paren">(</span><em>samples</em>, <em>outgroup=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate and return a copy of the instance with only a specified
list of samples. It is possible to select ingroup and/or
outgroup samples and the sample indexes are not required to be
consecutive.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>samples</strong> – a list (or other iterable type with a length) of
sample indexes giving the list of ingroup samples that must
be exported to the return value object. If <code class="docutils literal"><span class="pre">None</span></code>, do not
export any ingroup samples.</li>
<li><strong>outgroup</strong> – a list (or other iterable type with a length)
of sample indexes giving the list of outgroup samples that
must be exported to the return value object. If <code class="docutils literal"><span class="pre">None</span></code>, do
not export any outgroup samples.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 3.0.0: </span>Established as a method for <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> and
<a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a>.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.to_fasta">
<code class="descname">to_fasta</code><span class="sig-paren">(</span><em>fname=None</em>, <em>first=0</em>, <em>last=4294967295</em>, <em>mapping=None</em>, <em>groups=False</em>, <em>shift_labels=False</em>, <em>include_outgroup=True</em>, <em>linelength=50</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.to_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Export alignment in the fasta format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>fname</strong> – Name of the file to export data to. By default, the
file is created (or overwritten if it already exists). If the
option <em>append</em> is <code class="docutils literal"><span class="pre">True</span></code>, data is appended at the end of the
file (and it must exist). If <em>fname</em> is <code class="docutils literal"><span class="pre">None</span></code> (default), no
file is created and the formatted data is returned as a
<a class="reference external" href="https://docs.python.org/2.7/library/functions.html#str" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a>. In the alternative case, nothing is returned.</li>
<li><strong>first</strong> – If only part of the sequences should be exported:
index of the first sequence to export.</li>
<li><strong>last</strong> – If only part of the sequences should be exported: index
of the last sequence to export. If the value is larger than the
index of the last sequence, all sequences are exported until the
last (this is the default). If <em>last*&lt;*first</em>, no sequences are
exported.</li>
<li><strong>mapping</strong> – A string providing the character mapping. Use the
specified list of characters to map integer allelic values. If
a non-empty string is provided, the length of the string must be
larger than the largest possible allele values. In that case,
the allele values will be used as indexes in order to determine
which character from this string must be used for outputting. In
the case that this method is used with an empty string, the
mapping will not be used and the allele values will be casted
directly to characters.</li>
<li><strong>groups</strong> – A boolean indicating whether group labels
should be exported, or ignored.</li>
<li><strong>shift_labels</strong> – A boolean indicating whether group labels
should be incremented of one unit when exporting.</li>
<li><strong>include_outgroup</strong> – A boolean indicating whether the outgroup
should be exported. It will be exported at the end of the
ingroup, and without discriminating label if <em>include_labels</em> is
<code class="docutils literal"><span class="pre">False</span></code>.</li>
<li><strong>linelength</strong> – The length of lines for internal breaking of
sequences.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Align.to_outgroup">
<code class="descname">to_outgroup</code><span class="sig-paren">(</span><em>index</em>, <em>label=0</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Align.to_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Transfer a sample to the outgroup.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – index of the sample to move to the outgroup.</li>
<li><strong>label</strong> – label to assign to the sample (usually, this label
assigned the sample to an individual).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="container">
<h2>Container<a class="headerlink" href="#container" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="egglib.Container">
<em class="property">class </em><code class="descclassname">egglib.</code><code class="descname">Container</code><a class="headerlink" href="#egglib.Container" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#egglib._interface.DataBase" title="egglib._interface.DataBase"><code class="xref py py-class docutils literal"><span class="pre">egglib._interface.DataBase</span></code></a></p>
<p>Holds a data set with associated sample names and group information.
The data consists of given numbers of ingroup and outgroup samples,
and each of those may have a different number of sites. There can be
any number of group levels (but this number must be the same for all
samples), meaning that samples can be described by several group
labels in addition to their name. Group labels are not group indices
(they do not need to be consecutive). There is a separate data set
for samples belonging to the outgroup. There can be any number of
outgroup samples. Outgroup samples always have one level of group
labels that should be used to specify individuals (when
appropriate). All data are represented by signed integers</p>
<p>Default instance is empty (0 samples and 0 sites).</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 3.0.0: </span>Reimplementation of the Container class.</p>
</div>
<dl class="method">
<dt id="egglib.Container.add_outgroup">
<code class="descname">add_outgroup</code><span class="sig-paren">(</span><em>name</em>, <em>data</em>, <em>group=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.add_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an outgroup sample to the instance.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> – name of the new sample.</li>
<li><strong>data</strong> – an iterable (may be a string or a list) containing
the data to set for the new sample. For an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>
instance and if the sample is not the first, the number of
data must fit any previous ones (ingroup or outgroup).</li>
<li><strong>group</strong> – if not None, must be an unsigned integer (the
default value is 0).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.add_sample">
<code class="descname">add_sample</code><span class="sig-paren">(</span><em>name</em>, <em>data</em>, <em>groups=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.add_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a sample to the instance.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> – name of the new sample.</li>
<li><strong>data</strong> – an iterable (may be a string or a list) containing
the data to set for the new sample. For an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>
instance and if the sample is not the first, the number of
data must fit any previous ones (including outgroup samples,
if any).</li>
<li><strong>groups</strong> – if not None, must be an iterable with an
unsigned integer for each group levels of the instance (if
None, all group labels are set to 0), or a single integer
value if only one level needs to be specified.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.add_samples">
<code class="descname">add_samples</code><span class="sig-paren">(</span><em>items</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.add_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Add several samples at the end of the instance.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>items</strong> – items must be an iterable that have a length (for
example a list, an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> instance or a
<a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> instance. Each item of <em>items</em> must be
of length 2 or 3. The first item must be the sample name
string, the second item is the data values (as a list of
signed integers or a single string) and the third item, if
provided, is the list of group labels (one unsigned integer
for each level). See the method <a class="reference internal" href="#egglib.Align.add_sample" title="egglib.Align.add_sample"><code class="xref py py-meth docutils literal"><span class="pre">add_sample()</span></code></a>
for more details about each added item. If the current
instance is an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>, all added items must have
the same length which must be the same as any items
currently present in the instance. For a
<a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a>, the items may have different lengths.
The number of group levels is set by the sample with the
larger number of group labels, and all samples with less
labels are completed by e default value (0).</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 2.0.1: </span>Original name is <code class="xref py py-meth docutils literal"><span class="pre">addSequences()</span></code>.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Renamed as <code class="docutils literal"><span class="pre">add_samples</span></code>. All items are added in one shot
(rather than calling the one-sample add method iteratively).</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.clear">
<code class="descname">clear</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Clear the instance and release all memory. In most cases, it is
preferable to use the method <a class="reference internal" href="#egglib._interface.DataBase.reset" title="egglib._interface.DataBase.reset"><code class="xref py py-meth docutils literal"><span class="pre">reset()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.create">
<code class="descname">create</code><span class="sig-paren">(</span><em>obj</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.create" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new instance by copying date from the data container
passed as <em>obj</em>. The object <em>obj</em> can be:</p>
<ul class="simple">
<li>an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>,</li>
<li>a <a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> (but all sequences are required to have
the same length if the target type is <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>),</li>
<li>any iterable type yielding a acceptable items (see the
documentation for <a class="reference internal" href="#egglib._interface.DataBase.add_samples" title="egglib._interface.DataBase.add_samples"><code class="xref py py-meth docutils literal"><span class="pre">add_samples()</span></code></a> for more details).</li>
</ul>
<div class="versionadded">
<p><span class="versionmodified">New in version 2.0.1.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.del_outgroup">
<code class="descname">del_outgroup</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.del_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete an item of the instance.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – index of the sample to delete.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.del_sample">
<code class="descname">del_sample</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.del_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete an item of the instance.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – index of the sample to delete.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.del_sites">
<code class="descname">del_sites</code><span class="sig-paren">(</span><em>sample</em>, <em>site</em>, <em>num=1</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.del_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete data entries from an ingroup sample.</p>
<p>By default (if <code class="docutils literal"><span class="pre">num=1</span></code>), remove a single site. If <em>num</em> is
larger than 1, remove a range of sites.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – ingroup sample index.</li>
<li><strong>site</strong> – index of the (first) site to remove. This site must
be a valid index.</li>
<li><strong>num</strong> – maximal number of sites to remove. The value cannot
be negative.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.del_sites_o">
<code class="descname">del_sites_o</code><span class="sig-paren">(</span><em>sample</em>, <em>site</em>, <em>num=1</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.del_sites_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete data entries from an outgroup sample.</p>
<p>By default (if just <code class="docutils literal"><span class="pre">num=1</span></code>), remove a single site. If <em>num</em>
is larger than 1, remove a range of sites.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – outgroup sample index.</li>
<li><strong>site</strong> – index of the (first) site to remove. This site must
be a valid index.</li>
<li><strong>num</strong> – maximal number of sites to remove. The value cannot
be negative.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.encode">
<code class="descname">encode</code><span class="sig-paren">(</span><em>nbits=10</em>, <em>random=None</em>, <em>include_outgroup=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.encode" title="Permalink to this definition">¶</a></dt>
<dd><p>Renames all sequences using a random mapping of names to unique
keys of length <em>nbits</em>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>nbits</strong> – length of the keys (encoded names). This value
must be &gt;= 4 and &lt;= 63.</li>
<li><strong>random</strong> – a <a class="reference internal" href="tools.html#egglib.tools.Random" title="egglib.tools.Random"><code class="xref py py-class docutils literal"><span class="pre">Random</span></code></a> instance to be used a random
generator. By default, use a default instance.</li>
<li><strong>include_outgroup</strong> – a <a class="reference external" href="https://docs.python.org/2.7/library/functions.html#bool" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> indicating whether the
outgroup samples should also be considered (in this case, a
single mapping is returned including both ingroup and
outgroup samples).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">A dictionary mapping all the generated keys to the
actual sequence names. The keys are case-dependent and
guaranteed not to start with a number.</p>
</td>
</tr>
</tbody>
</table>
<p>The returned mapping can be used to restore the original names
using <a class="reference internal" href="#egglib._interface.DataBase.rename" title="egglib._interface.DataBase.rename"><code class="xref py py-meth docutils literal"><span class="pre">rename()</span></code></a>. This method is not affected by
the presence of sequences with identical names in the original
instance (and <a class="reference internal" href="#egglib._interface.DataBase.rename" title="egglib._interface.DataBase.rename"><code class="xref py py-meth docutils literal"><span class="pre">rename()</span></code></a> will also work properly
in that case).</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 2.0.1.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Keys are forced to start with a capital letter. Now take
a <a class="reference internal" href="tools.html#egglib.tools.Random" title="egglib.tools.Random"><code class="xref py py-class docutils literal"><span class="pre">Random</span></code></a> instance. Use library’s own random
number generator. Added an option for the outgroup.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.equalize">
<code class="descname">equalize</code><span class="sig-paren">(</span><em>value='?'</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.equalize" title="Permalink to this definition">¶</a></dt>
<dd><p>Extend sequences such as they all have the length of the longest
sequence (over both ingroup and outgroup).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>value</strong> – the value to use to extend sequences, as an
integer or a single-character string.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.find">
<code class="descname">find</code><span class="sig-paren">(</span><em>name</em>, <em>include_outgroup=False</em>, <em>regex=False</em>, <em>multi=False</em>, <em>flags=None</em>, <em>index=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.find" title="Permalink to this definition">¶</a></dt>
<dd><p>Find a sample by its name.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>name</strong> – name of sample to identify.</li>
<li><strong>include_outgroup</strong> – a boolean indicating whether the
outgroup should be considered.</li>
<li><strong>regex</strong> – a boolean indicating whether the value passed a
<em>name</em> is a regular expression. If so, the string is passed
as is to the re module (using function
<a class="reference external" href="https://docs.python.org/2.7/library/re.html#re.search" title="(in Python v2.7)"><code class="xref py py-func docutils literal"><span class="pre">re.search()</span></code></a>).
Otherwise, only exact matches will be considered.</li>
<li><strong>multi</strong> – a boolean indicating whether all hits should be
returned. If so, a list of <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instances
is always returned (the list will be empty in case of no
hits). Otherwise, a single <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instance
(or its index)  will be returned for the first hit, or
<code class="docutils literal"><span class="pre">None</span></code> in case of no hits.</li>
<li><strong>flags</strong> – list of flags to be passed to <a class="reference external" href="https://docs.python.org/2.7/library/re.html#re.search" title="(in Python v2.7)"><code class="xref py py-func docutils literal"><span class="pre">re.search()</span></code></a>
(ignored if <em>regex</em> is <code class="docutils literal"><span class="pre">False</span></code>). For example, when looking
for samples containing the term “sample”, for being case
insensitive, use the following syntax:
<code class="docutils literal"><span class="pre">align.find(&quot;sample&quot;,</span> <span class="pre">regex=True,</span> <span class="pre">flags=[re.I])</span></code>. By
default (<code class="docutils literal"><span class="pre">None</span></code>) no further argument is passed.</li>
<li><strong>index</strong> – boolean indicating whether the index of the sample
should be returned. In that case return values for hits are
<a class="reference external" href="https://docs.python.org/2.7/library/functions.html#int" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> (by default, <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a>) instances.
Warning: it is not allowed to set both <em>include_outgroup</em>
and <em>index</em> to <code class="docutils literal"><span class="pre">True</span></code> as there would be no way to
distinguish between ingroup and outgroup indexes.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><code class="docutils literal"><span class="pre">None</span></code> if no hits were found, a (potentially empty)
list of <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instances or class:<cite>int</cite> if
<em>multi</em> is <code class="docutils literal"><span class="pre">True</span></code>, or a single <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> or
<a class="reference external" href="https://docs.python.org/2.7/library/functions.html#int" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> otherwise.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.find_motif">
<code class="descname">find_motif</code><span class="sig-paren">(</span><em>sample</em>, <em>motif</em>, <em>start=0</em>, <em>stop=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.find_motif" title="Permalink to this definition">¶</a></dt>
<dd><p>Locate the first instance of a motif for an ingroup sample</p>
<p>Returns the index of the first exact hit to a given substring.
The returned value is the position of the first base of the hit.
Only exact matches are implemented. To use regular expression
(for example to find degenerated motifs), one should extract the
string for the sequence and use a tool such as the regular
expression module (<a class="reference external" href="https://docs.python.org/2.7/library/re.html#module-re" title="(in Python v2.7)"><code class="xref py py-mod docutils literal"><span class="pre">re</span></code></a>).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>motif</strong> – a list of integer, or one-character strings (or
mixing both) or a string constituting the motif to search.</li>
<li><strong>start</strong> – position at which to start searching. The method
will never return a value smaller than <em>start</em>. By default,
search from the start of the sequence.</li>
<li><strong>stop</strong> – position at which to stop search (the motif
cannot overlap this position). No returned value will be
larger than stop-<em>n</em>. By default, or if <em>stop</em> is equal to
or larger than the length of the sequence, search until the
end of the sequence.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.find_motif_o">
<code class="descname">find_motif_o</code><span class="sig-paren">(</span><em>sample</em>, <em>motif</em>, <em>start=0</em>, <em>stop=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.find_motif_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Locate the first instance of a motif for an outgroup sample</p>
<p>Returns the index of the first exact hit to a given substring.
The returned value is the position of the first base of the hit.
Only exact matches are implemented. To use regular expression
(for example to find degenerated motifs), one should extract the
string for the sequence and use a tool such as the regular
expression module (<a class="reference external" href="https://docs.python.org/2.7/library/re.html#module-re" title="(in Python v2.7)"><code class="xref py py-mod docutils literal"><span class="pre">re</span></code></a>).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – outgroup sample index.</li>
<li><strong>motif</strong> – a list of integer, or one-character strings (or
mixing both) or a string constituting the motif to search.</li>
<li><strong>start</strong> – position at which to start searching. The method
will never return a value smaller than <em>start</em>. By default,
search from the start of the sequence.</li>
<li><strong>stop</strong> – position at which to stop search (the motif
cannot overlap this position). No returned value will be
larger than <em>stop</em> minus the length of the motif. By
default, or if <em>stop</em> is equal to or larger than the length
of the sequence, search until the end of the sequence.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.get_i">
<code class="descname">get_i</code><span class="sig-paren">(</span><em>sample</em>, <em>site</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.get_i" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a data entry.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>site</strong> – site index.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.get_label">
<code class="descname">get_label</code><span class="sig-paren">(</span><em>sample</em>, <em>level</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.get_label" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a group label.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>level</strong> – level index.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.get_label_o">
<code class="descname">get_label_o</code><span class="sig-paren">(</span><em>sample</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.get_label_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Get an outgroup label.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>sample</strong> – sample index.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.get_name">
<code class="descname">get_name</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.get_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the name of a sample.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – sample index.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.get_name_o">
<code class="descname">get_name_o</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.get_name_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the name of an outgroup sample.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – outgroup sample index.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.get_o">
<code class="descname">get_o</code><span class="sig-paren">(</span><em>sample</em>, <em>site</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.get_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Get an outgroup data entry.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – outgroup sample index.</li>
<li><strong>site</strong> – site index.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.get_outgroup">
<code class="descname">get_outgroup</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.get_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instance corresponding to the
requested index for the outgroup. The returned object allows to
modify the underlying data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – index of the outgroup sample to access.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.get_sample">
<code class="descname">get_sample</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.get_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instance corresponding to the
requested index. The returned object allows to modify the
underlying data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – index of the sample to access.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.get_sequence">
<code class="descname">get_sequence</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.get_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Access to the data entries of a given ingroup sample. Returns a
<a class="reference internal" href="#egglib.SequenceView" title="egglib.SequenceView"><code class="xref py py-class docutils literal"><span class="pre">SequenceView</span></code></a> instance which allow modifying the
underlying sequence container object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – ingroup sample index.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.get_sequence_o">
<code class="descname">get_sequence_o</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.get_sequence_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Access to the data entries of a given outgroup sample. Returns a
<a class="reference internal" href="#egglib.SequenceView" title="egglib.SequenceView"><code class="xref py py-class docutils literal"><span class="pre">SequenceView</span></code></a> instance which allow modifying the
underlying sequence container object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – outgroup sample index.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.group_mapping">
<code class="descname">group_mapping</code><span class="sig-paren">(</span><em>level=0</em>, <em>indices=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.group_mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a dictionary mapping group labels to either
<a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instances (by default) or their indexes
representing all samples of this instance. It can process and
ingroup label level or the outgroup.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>level</strong> – index of the level to consider. If <code class="docutils literal"><span class="pre">None</span></code>,
processes the outgroup.</li>
<li><strong>indices</strong> – if <code class="docutils literal"><span class="pre">True</span></code>, represent samples by their index
(within the ingroup or the outgroup depending on <em>level</em>
being non-<code class="docutils literal"><span class="pre">None</span></code> or <code class="docutils literal"><span class="pre">None</span></code>) instead of
<a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instances.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.insert_sites">
<code class="descname">insert_sites</code><span class="sig-paren">(</span><em>sample</em>, <em>position</em>, <em>values</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.insert_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Insert sites at a given position for a given ingroup sample</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – index of the ingroup sample to which insert
sites.</li>
<li><strong>position</strong> – the position at which to insert sites. Sites
are inserted <em>before</em> the specified position, so the user
can use 0 to insert sites at the beginning of the sequence.
To insert sites at the end of the sequence, pass the current
length of the sequence. If <em>position</em> is larger than the
length of the sequence or <code class="docutils literal"><span class="pre">None</span></code>, new sites are inserted
at the end of the sequence. The position might be negative
Warning: the position -1 means <em>before</em> the last position.</li>
<li><strong>values</strong> – a list of signed integers, or a string, providing
data to insert into the instance.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.insert_sites_o">
<code class="descname">insert_sites_o</code><span class="sig-paren">(</span><em>sample</em>, <em>position</em>, <em>values</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.insert_sites_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Insert sites at a given position for a given outgroup samples</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – index of the outgroup sample to which insert
sites.</li>
<li><strong>position</strong> – the position at which to insert sites. Sites
are inserted <em>before</em> the specified position, so the user
can use 0 to insert sites at the beginning of the sequence.
To insert sites at the end of the sequence, pass the current
length of the sequence. If <em>position</em> is larger than the
length of the sequence or <code class="docutils literal"><span class="pre">None</span></code>, new sites are inserted
at the end of the sequence. The position might be negative
Warning: the position -1 means <em>before</em> the last position.</li>
<li><strong>values</strong> – a list of signed integers, or a string, providing
data to insert into the instance.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="egglib.Container.is_matrix">
<code class="descname">is_matrix</code><a class="headerlink" href="#egglib.Container.is_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p><code class="docutils literal"><span class="pre">True</span></code> if the instance is an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>, and <code class="docutils literal"><span class="pre">False</span></code> if
it is a <a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.iter_outgroup">
<code class="descname">iter_outgroup</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.iter_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterator over the outgroup.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.iter_samples">
<code class="descname">iter_samples</code><span class="sig-paren">(</span><em>ingrp</em>, <em>outgrp</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.iter_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterator over either ingroup samples or outgroup samples, or
both ingroup and outgroup (in that order) samples. Provided for
flexibility.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>ingrp</strong> – include ingroup samples.</li>
<li><strong>outgroup</strong> – include outgroup samples.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.lo">
<code class="descname">lo</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.lo" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the number of sites of an outgroup sample.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – sample index.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.ls">
<code class="descname">ls</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.ls" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the number of sites of an ingroup sample.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – sample index.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.name_mapping">
<code class="descname">name_mapping</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.name_mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a dictionary mapping names to <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a>
instances representing all samples of this instance. This method
is most useful when several sequences have the same name. It may
be used to detect and process duplicates. It processes the
ingroup only.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.names">
<code class="descname">names</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.names" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate the list of ingroup sample names.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.names_outgroup">
<code class="descname">names_outgroup</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.names_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate the list of outgroup sample names.</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib.Container.ng">
<code class="descname">ng</code><a class="headerlink" href="#egglib.Container.ng" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of group levels for ingroup (for outgroup, this number is
always one). It possible to change the value directly.</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib.Container.no">
<code class="descname">no</code><a class="headerlink" href="#egglib.Container.no" title="Permalink to this definition">¶</a></dt>
<dd><p>Current number of outgroup samples (cannot be modified
directly).</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib.Container.ns">
<code class="descname">ns</code><a class="headerlink" href="#egglib.Container.ns" title="Permalink to this definition">¶</a></dt>
<dd><p>Current number of samples, not considering the outgroup (cannot
be modified directly).</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.remove_duplicates">
<code class="descname">remove_duplicates</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.remove_duplicates" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove all duplicates, based on name exact matching. For all
pairs of samples with identical name, only the one occurring
first is conserved. The current instance is modified and this
method returns <code class="docutils literal"><span class="pre">None</span></code>.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.rename">
<code class="descname">rename</code><span class="sig-paren">(</span><em>mapping</em>, <em>liberal=False</em>, <em>include_outgroup=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.rename" title="Permalink to this definition">¶</a></dt>
<dd><p>Rename sequences of the instance using the provided mapping.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>mapping</strong> – a <a class="reference external" href="https://docs.python.org/2.7/library/stdtypes.html#dict" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">dict</span></code></a> providing the mapping of old
names (as keys) to new names (which may, if needed, contain
duplicated).</li>
<li><strong>liberal</strong> – if this argument is <cite>False</cite> and a name does not
appear in <em>mapping</em>, a <a class="reference external" href="https://docs.python.org/2.7/library/exceptions.html#exceptions.ValueError" title="(in Python v2.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> is
raised. If <em>liberal</em> is <code class="docutils literal"><span class="pre">True</span></code>, names that don’t appear in
<em>mapping</em> are left unchanged.</li>
<li><strong>include_outgroup</strong> – if <code class="docutils literal"><span class="pre">True</span></code>, consider both ingroup and
outgroup samples together (they should be provided together
in the same mapping).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">The number of samples that have been actually renamed,
overall.</p>
</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 2.0.1.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Added an option for the outgroup. Added return value.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.reserve">
<code class="descname">reserve</code><span class="sig-paren">(</span><em>nsam=0</em>, <em>nout=0</em>, <em>lnames=0</em>, <em>ngrp=0</em>, <em>nsit=0</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.reserve" title="Permalink to this definition">¶</a></dt>
<dd><p>Pre-allocate memory. This method can be used when the size of
arrays is known a priori, in order to speed up memory
allocations. It is not necessary to set all values. Values less
than 0 are ignored.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>nsam</strong> – number of samples in the ingroup.</li>
<li><strong>nout</strong> – number of samples in the outgroup.</li>
<li><strong>lnames</strong> – length of sample names.</li>
<li><strong>ngrp</strong> – number of group labels.</li>
<li><strong>nsit</strong> – number of sites.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.reset">
<code class="descname">reset</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.reset" title="Permalink to this definition">¶</a></dt>
<dd><p>Reset the instance.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.set_i">
<code class="descname">set_i</code><span class="sig-paren">(</span><em>sample</em>, <em>site</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.set_i" title="Permalink to this definition">¶</a></dt>
<dd><p>Set an ingroup data entry. The value must be a signed integer or a
one-character string. In the latter case, <a class="reference external" href="https://docs.python.org/2.7/library/functions.html#ord" title="(in Python v2.7)"><code class="xref py py-func docutils literal"><span class="pre">ord()</span></code></a> is
called on the behalf of the user.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>site</strong> – site index.</li>
<li><strong>value</strong> – allele value.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.set_label">
<code class="descname">set_label</code><span class="sig-paren">(</span><em>sample</em>, <em>level</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.set_label" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a group label.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>level</strong> – level index.</li>
<li><strong>value</strong> – new group value (unsigned integer).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.set_label_o">
<code class="descname">set_label_o</code><span class="sig-paren">(</span><em>sample</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.set_label_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Get an outgroup label.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>value</strong> – new group value (unsigned integer).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.set_name">
<code class="descname">set_name</code><span class="sig-paren">(</span><em>index</em>, <em>name</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.set_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the name of a sample.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – index of the sample</li>
<li><strong>name</strong> – new name value.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.set_name_o">
<code class="descname">set_name_o</code><span class="sig-paren">(</span><em>index</em>, <em>name</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.set_name_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the name of an outgroup sample.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – index of the outgroup sample</li>
<li><strong>name</strong> – new name value.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.set_o">
<code class="descname">set_o</code><span class="sig-paren">(</span><em>sample</em>, <em>site</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.set_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Set an outgroup data entry. The value must be a signed integer
or a one-character string. In the latter case, <a class="reference external" href="https://docs.python.org/2.7/library/functions.html#ord" title="(in Python v2.7)"><code class="xref py py-func docutils literal"><span class="pre">ord()</span></code></a> 
is called on the behalf of the user.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>site</strong> – site index.</li>
<li><strong>value</strong> – allele value.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.set_outgroup">
<code class="descname">set_outgroup</code><span class="sig-paren">(</span><em>index</em>, <em>name</em>, <em>data</em>, <em>group=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.set_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the values for a sample of the outgroup</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – index of the sample to access (slices are not
permitted).</li>
<li><strong>name</strong> – new name of the sample.</li>
<li><strong>data</strong> – string or list of integers given the new values to
set. In case of an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>, it is required to pass a
sequence with length matching the number of sites of the
instance.</li>
<li><strong>groups</strong> – a positive integer to serve as group label (the
default value is 0.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.set_sample">
<code class="descname">set_sample</code><span class="sig-paren">(</span><em>index</em>, <em>name</em>, <em>data</em>, <em>groups=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.set_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the values for the sample corresponding to the requested
index.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – index of the sample to access (slices are not
permitted).</li>
<li><strong>name</strong> – new name of the sample.</li>
<li><strong>data</strong> – string or list of integers given the new values to
set. In case of an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>, it is required to pass a
sequence with length matching the number of sites of the
instance.</li>
<li><strong>groups</strong> – a list of integer label values, or a single
integer value, to set as group labels. The class will ensure
that all samples have the same number of group labels,
padding with 0 as necessary. The default corresponds to an
empty list.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.set_sequence">
<code class="descname">set_sequence</code><span class="sig-paren">(</span><em>index</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.set_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Replace all data entries for a given ingroup sample by new
values.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – ingroup sample index.</li>
<li><strong>value</strong> – can be a <a class="reference internal" href="#egglib.SequenceView" title="egglib.SequenceView"><code class="xref py py-class docutils literal"><span class="pre">SequenceView</span></code></a> instance, a
string or a list of integers (or any iterable with a length
which may mix single-character strings and integers). When
modifying an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> instance, all sequences must
have the same length as the current alignment length.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.set_sequence_o">
<code class="descname">set_sequence_o</code><span class="sig-paren">(</span><em>index</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.set_sequence_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Replace all data entries for a given outgroup sample by new
values.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – outgroup sample index.</li>
<li><strong>value</strong> – can be a <a class="reference internal" href="#egglib.SequenceView" title="egglib.SequenceView"><code class="xref py py-class docutils literal"><span class="pre">SequenceView</span></code></a> instance, a
string or a list of integers (or any iterable with a length
which may mix single-character strings and integers). When
modifying an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> instance, all sequences must
have the same length as the current alignment length.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.shuffle">
<code class="descname">shuffle</code><span class="sig-paren">(</span><em>level=0</em>, <em>random=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.shuffle" title="Permalink to this definition">¶</a></dt>
<dd><p>Shuffle group labels.</p>
<p>Randomly reassigns group labels. Modifies the current instance
and returns <code class="docutils literal"><span class="pre">None</span></code>. Only the specified level is affected, and
only the group labels are modified (the order of samples is not
changed.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>level</strong> – index of the group level to shuffle. To shuffle
the outgroup’s labels, set <em>level</em> to <code class="docutils literal"><span class="pre">None</span></code>.</li>
<li><strong>random</strong> – A <a class="reference internal" href="tools.html#egglib.tools.Random" title="egglib.tools.Random"><code class="xref py py-class docutils literal"><span class="pre">Random</span></code></a> instance to be used a random
generator. By default, use a default instance.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>The outgroup is necessarily processed separately from the
rest of the instance.  Allow pass a <a class="reference internal" href="tools.html#egglib.tools.Random" title="egglib.tools.Random"><code class="xref py py-class docutils literal"><span class="pre">Random</span></code></a>
instance. Use library’s own random number generator.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.subgroup">
<code class="descname">subgroup</code><span class="sig-paren">(</span><em>groups</em>, <em>outgroup_all=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.subgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate and return a copy of the instance with only samples
from the specified groups (identified by their group labels,
including outgroup).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>groups</strong> – an integer or a dictionary providing the labels
of the groups that are selected. If an integer, it is
understood as an ingroup label corresponding to the first
grouping level (if more than one). If a dictionary, the keys
of this dictionary must be integer corresponding to grouping
levels (or <code class="docutils literal"><span class="pre">None</span></code> for the outgroup) and the values must be
lists containing the requested labels. It is not required to
include all levels and the outgroup in the dictionary. It is
allowed to specify labels that are actually not represented
in the data.</li>
<li><strong>outgroup_all</strong> – if <code class="docutils literal"><span class="pre">True</span></code>, always include all outgroup
samples in the returned instance, regardless of whether
their group labels are specified in the first argument.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 3.0.0.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.subset">
<code class="descname">subset</code><span class="sig-paren">(</span><em>samples</em>, <em>outgroup=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate and return a copy of the instance with only a specified
list of samples. It is possible to select ingroup and/or
outgroup samples and the sample indexes are not required to be
consecutive.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>samples</strong> – a list (or other iterable type with a length) of
sample indexes giving the list of ingroup samples that must
be exported to the return value object. If <code class="docutils literal"><span class="pre">None</span></code>, do not
export any ingroup samples.</li>
<li><strong>outgroup</strong> – a list (or other iterable type with a length)
of sample indexes giving the list of outgroup samples that
must be exported to the return value object. If <code class="docutils literal"><span class="pre">None</span></code>, do
not export any outgroup samples.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 3.0.0: </span>Established as a method for <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> and
<a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a>.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.to_fasta">
<code class="descname">to_fasta</code><span class="sig-paren">(</span><em>fname=None</em>, <em>first=0</em>, <em>last=4294967295</em>, <em>mapping=None</em>, <em>groups=False</em>, <em>shift_labels=False</em>, <em>include_outgroup=True</em>, <em>linelength=50</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.to_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Export alignment in the fasta format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>fname</strong> – Name of the file to export data to. By default, the
file is created (or overwritten if it already exists). If the
option <em>append</em> is <code class="docutils literal"><span class="pre">True</span></code>, data is appended at the end of the
file (and it must exist). If <em>fname</em> is <code class="docutils literal"><span class="pre">None</span></code> (default), no
file is created and the formatted data is returned as a
<a class="reference external" href="https://docs.python.org/2.7/library/functions.html#str" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a>. In the alternative case, nothing is returned.</li>
<li><strong>first</strong> – If only part of the sequences should be exported:
index of the first sequence to export.</li>
<li><strong>last</strong> – If only part of the sequences should be exported: index
of the last sequence to export. If the value is larger than the
index of the last sequence, all sequences are exported until the
last (this is the default). If <em>last*&lt;*first</em>, no sequences are
exported.</li>
<li><strong>mapping</strong> – A string providing the character mapping. Use the
specified list of characters to map integer allelic values. If
a non-empty string is provided, the length of the string must be
larger than the largest possible allele values. In that case,
the allele values will be used as indexes in order to determine
which character from this string must be used for outputting. In
the case that this method is used with an empty string, the
mapping will not be used and the allele values will be casted
directly to characters.</li>
<li><strong>groups</strong> – A boolean indicating whether group labels
should be exported, or ignored.</li>
<li><strong>shift_labels</strong> – A boolean indicating whether group labels
should be incremented of one unit when exporting.</li>
<li><strong>include_outgroup</strong> – A boolean indicating whether the outgroup
should be exported. It will be exported at the end of the
ingroup, and without discriminating label if <em>include_labels</em> is
<code class="docutils literal"><span class="pre">False</span></code>.</li>
<li><strong>linelength</strong> – The length of lines for internal breaking of
sequences.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Container.to_outgroup">
<code class="descname">to_outgroup</code><span class="sig-paren">(</span><em>index</em>, <em>label=0</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Container.to_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Transfer a sample to the outgroup.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – index of the sample to move to the outgroup.</li>
<li><strong>label</strong> – label to assign to the sample (usually, this label
assigned the sample to an individual).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="tree">
<h2>Tree<a class="headerlink" href="#tree" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="egglib.Tree">
<em class="property">class </em><code class="descclassname">egglib.</code><code class="descname">Tree</code><span class="sig-paren">(</span><em>fname=None</em>, <em>string=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree" title="Permalink to this definition">¶</a></dt>
<dd><p>This class Handle trees. A tree is a linked collection of nodes
which all have one parent (except the ultimate base node) and any
number of children.</p>
<p>Nodes are implemented as <a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a> instances. A node without
children is a <em>leaf</em> (others are <em>internal</em>). A node with exactly
one child is generally meaningless, but is allowed. All nodes
(internal nodes as well as leaves) have a <em>label</em> which in the case
of leaves can be used as leaf name. It is not possible to apply a
name <em>and</em> a label to leaf node, in agreement with the newick
format. All connections between nodes (<em>branches</em>) are oriented and
can have a length (although the lengths can be omitted) but note
that labels are applied to nodes, not branches. All <a class="reference internal" href="#egglib.Tree" title="egglib.Tree"><code class="xref py py-class docutils literal"><span class="pre">Tree</span></code></a>
instances have at least one base node which is the only one allowed
not to have a parent. Network-like structures are not allowed
(because nodes must have exactly one parent).</p>
<p>Import and export to/from strings and files are in the bracket-based
<a class="reference external" href="http://evolution.genetics.washington.edu/phylip/newicktree.html">newick</a>
format (the parser treats terminal node labels as strings, and
internal node labels as integers or, by default, floats).
<a class="reference internal" href="#egglib.Tree" title="egglib.Tree"><code class="xref py py-class docutils literal"><span class="pre">Tree</span></code></a> instances can be exported using the built-in
<a class="reference external" href="https://docs.python.org/2.7/library/functions.html#str" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> function, and the method <a class="reference internal" href="#egglib.Tree.newick" title="egglib.Tree.newick"><code class="xref py py-meth docutils literal"><span class="pre">newick()</span></code></a>. Nodes
also have a <a class="reference internal" href="#egglib.Node.newick" title="egglib.Node.newick"><code class="xref py py-meth docutils literal"><span class="pre">newick()</span></code></a> method.</p>
<p><a class="reference internal" href="#egglib.Tree" title="egglib.Tree"><code class="xref py py-class docutils literal"><span class="pre">Tree</span></code></a> instances are iterable. Three iterators are
provided: one is depth-first (<a class="reference internal" href="#egglib.Tree.depth_iter" title="egglib.Tree.depth_iter"><code class="xref py py-meth docutils literal"><span class="pre">depth_iter()</span></code></a>), another is
breath-first (<a class="reference internal" href="#egglib.Tree.breadth_iter" title="egglib.Tree.breadth_iter"><code class="xref py py-meth docutils literal"><span class="pre">breadth_iter()</span></code></a>) and one iterates on terminal
nodes (leaves) only (<a class="reference internal" href="#egglib.Tree.iter_leaves" title="egglib.Tree.iter_leaves"><code class="xref py py-meth docutils literal"><span class="pre">iter_leaves()</span></code></a>).</p>
<p>The instance can be initialized as an empty tree (with only a root
node), or from a newick-formatted string. By default, the string is
read from the file name passed as the <em>fname</em> argument to the
constructor, but it can be passed directly through the constructor
argument <em>string</em>. It is not allowed to set both <em>fname</em> and
<em>string</em> at the same time. The newick parser expects a well-formed
newick string (including the trailing semicolon).</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 2.0.1: </span>(This change concerns the constructor.) Imports directly from a
file. If a string is passed, it is interpreted as a file name by
default.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Several interface changes. Trees are not allowed to be network at
all anymore.</p>
</div>
<dl class="method">
<dt id="egglib.Tree.add_node">
<code class="descname">add_node</code><span class="sig-paren">(</span><em>parent</em>, <em>label=None</em>, <em>brlen=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.add_node" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a node to the tree.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>parent</strong> – one of the nodes of this instance, as a
<a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a> reference.</li>
<li><strong>label</strong> – node label which will be internal node label or
leaf name according to the final structure of the tree. The
new node has initially no children and is therefore a leaf
until it is itself connected to a child (if ever).</li>
<li><strong>brlen</strong> – length of the branche connecting <em>parent</em> to the
new node.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">The new node as a <a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a> reference.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="egglib.Tree.base">
<code class="descname">base</code><a class="headerlink" href="#egglib.Tree.base" title="Permalink to this definition">¶</a></dt>
<dd><p>Basal node of the tree (if the tree is unrooted, it is a
trifurcation whose location should be considered as arbitrary;
if the tree is rooted, it is the root). This attribute is a
<a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a> instance which can be modified, but it cannot be
replaced.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.breadth_iter">
<code class="descname">breadth_iter</code><span class="sig-paren">(</span><em>start=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.breadth_iter" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a breadth-first iterator. Iterate over the
<a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a> instances of the trees, starting from the base
but, then, following a breadth-first order.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>start</strong> – start point of the iteration, as a <a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a>
instance of this tree. By default, start from the base of
the tree.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.clean_branch_lengths">
<code class="descname">clean_branch_lengths</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.clean_branch_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove all branch lengths. In practice, they are set to
<code class="docutils literal"><span class="pre">None</span></code>.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.clean_internal_labels">
<code class="descname">clean_internal_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.clean_internal_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove all internal node labels, including the base of the tree.
In practice, they are set to <code class="docutils literal"><span class="pre">None</span></code>.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.collapse">
<code class="descname">collapse</code><span class="sig-paren">(</span><em>node</em>, <em>ignore_len=False</em>, <em>ignore_label=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.collapse" title="Permalink to this definition">¶</a></dt>
<dd><p>Collapse a branch of the tree.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>node</strong> – <a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a> representing the branch to remove.</li>
<li><strong>ignore_len</strong> – don’t try to transfer branch lengths to
children.</li>
<li><strong>ignore_label</strong> – don’t transfer label of the destroyed node
to its parent.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p><em>node</em> represents the branch that must be removed from the tree
(this node is destroyed in the process). It must be one of the
nodes contained in the tree (as a <a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a> instance), but
not the base of the tree. It cannot be an terminal node (leaf).</p>
<p>If <em>ignore_label</em> is not set to <code class="docutils literal"><span class="pre">True</span></code>, the label of the
destroyed node is transferred to the parent based on the
following procedure: (1) if the destroyed node’s label is
<code class="docutils literal"><span class="pre">None</span></code>, nothing is done; (2) if the destroyed node’s parent’s
label is <code class="docutils literal"><span class="pre">None</span></code>, the destroyed node’s label is copied to its
parent as is; (3) otherwise both labels are converted to strings
(if they are not yet) and concatenated as in the string <code class="docutils literal"><span class="pre">a;b</span></code>
(where <code class="docutils literal"><span class="pre">a</span></code> is the parent’s label and <code class="docutils literal"><span class="pre">b</span></code> is the destroyed
node’s label), even if the two labels are identical.</p>
<p>If <em>ignore_len</em> is not set to <code class="docutils literal"><span class="pre">True</span></code>, and if the length of the
removed branch (branch from the specified node to its parent) is
specified, it will be spread equally among the branches to its
children (see example below). This requires that the branch
length to all children are specified. If the removed branch has
no specified length, nothing is done.</p>
<p>Collapsing node <code class="docutils literal"><span class="pre">[4]</span></code> on the following tree:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span> <span class="o">/-------------------------------------------&gt;</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
 <span class="o">|</span>
 <span class="o">|</span>             <span class="o">/-----------------------------&gt;</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
 <span class="o">|</span>             <span class="o">|</span>
 <span class="o">|-----------&gt;</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>             <span class="o">/--------------&gt;</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
 <span class="o">|</span>             <span class="o">|</span>              <span class="o">|</span>
 <span class="o">|</span>             \<span class="o">------------&gt;</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
<span class="p">[</span><span class="mi">0</span><span class="p">]</span>                           <span class="o">|</span>
 <span class="o">|</span>                            \<span class="o">--------------&gt;</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>
 <span class="o">|</span>
 <span class="o">|</span>              <span class="o">/----------------------------&gt;</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
 <span class="o">|</span>              <span class="o">|</span> 
 \<span class="o">------------&gt;</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>            <span class="o">/-------------&gt;</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
                <span class="o">|</span>             <span class="o">|</span>
                \<span class="o">-----------&gt;</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span>
                              <span class="o">|</span>
                              \<span class="o">-------------&gt;</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span>
</pre></div>
</div>
<p>will generate the following tree, with the correction of edge
lengths as  depicted:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span> <span class="o">/-------------------------------------------&gt;</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
 <span class="o">|</span>
 <span class="o">|</span>             <span class="o">/-----------------------------&gt;</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
 <span class="o">|</span>             <span class="o">|</span>
 <span class="o">|-----------&gt;</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
 <span class="o">|</span>             <span class="o">|</span>
 <span class="o">|</span>             <span class="o">|--------------------&gt;</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>        <span class="n">L5</span> <span class="o">=</span> <span class="n">L5</span><span class="o">+</span><span class="n">L4</span><span class="o">/</span><span class="mi">2</span>
<span class="p">[</span><span class="mi">0</span><span class="p">]</span>            <span class="o">|</span>
 <span class="o">|</span>             \<span class="o">--------------------&gt;</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>        <span class="n">L6</span> <span class="o">=</span> <span class="n">L6</span><span class="o">+</span><span class="n">L4</span><span class="o">/</span><span class="mi">2</span>
 <span class="o">|</span>
 <span class="o">|</span>              <span class="o">/----------------------------&gt;</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
 <span class="o">|</span>              <span class="o">|</span> 
 \<span class="o">------------&gt;</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>            <span class="o">/-------------&gt;</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
                <span class="o">|</span>             <span class="o">|</span>
                \<span class="o">-----------&gt;</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span>
                              <span class="o">|</span>
                              \<span class="o">-------------&gt;</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span>
</pre></div>
</div>
<p>Although the total edge length of the tree is not modified, the
relationships will be altered: the distance between the
descendants of the collapsed node (nodes 5 and 6 in the example
above) will be artificially increased.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.copy">
<code class="descname">copy</code><span class="sig-paren">(</span><em>node=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new instance of <a class="reference internal" href="#egglib.Tree" title="egglib.Tree"><code class="xref py py-class docutils literal"><span class="pre">Tree</span></code></a> that is a deep copy of
a subtree of the the current tree.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>node</strong> – a <a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a> instance (one of the nodes of
the current tree) at the base of the subtree that should be
copied. By default, or if <code class="docutils literal"><span class="pre">None</span></code> or the base of the tree
is passed, the whole tree is copied. It is not allowed to
pass a leaf.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.depth_iter">
<code class="descname">depth_iter</code><span class="sig-paren">(</span><em>start=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.depth_iter" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a depth-first iterator. Iterate over the <a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a>
instances of the trees, starting from the base but, then,
following a depth-first order.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>start</strong> – start point of the iteration, as a <a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a>
instance of this tree. By default, start from the base of
the tree.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.extract">
<code class="descname">extract</code><span class="sig-paren">(</span><em>node</em>, <em>label=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.extract" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove a subtree of the current tree and return it as a new
instance of <a class="reference internal" href="#egglib.Tree" title="egglib.Tree"><code class="xref py py-class docutils literal"><span class="pre">Tree</span></code></a>. All nodes of the subtree descending
from the requested node will now belong to the new tree. The
label of the node at the base of the selected clade is deleted.
In the original tree, the extracted clade is replaced by a
terminal node which has, by default, the label of the node at
the base of the extracted clade, or the label passed as
argument.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>node</strong> – a <a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a> instance (one of the nodes of
the current tree) at the base of the subtree that should be
extracted. It is not allowed to pass a leaf or the base of
the tree.</li>
<li><strong>label</strong> – label to affect to the terminal node that is
introduced to replace the extracted clade in the orginal
tree. By default (or if <code class="docutils literal"><span class="pre">None</span></code>), use the label of the node
passed as first argument (note that this label should be in
principle a number).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.find_clade">
<code class="descname">find_clade</code><span class="sig-paren">(</span><em>names</em>, <em>ancestral=False</em>, <em>both_sides=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.find_clade" title="Permalink to this definition">¶</a></dt>
<dd><p>Check whether a group is one of the clades defined by the tree.</p>
<p>The leaf names must be provided as a list, or any other iterable
(ideally, a <a class="reference external" href="https://docs.python.org/2.7/library/stdtypes.html#set" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">set</span></code></a>). Leaf names are normally <a class="reference external" href="https://docs.python.org/2.7/library/functions.html#str" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a>
instances. All leaves must be present in the tree.</p>
<p>With the option <code class="docutils literal"><span class="pre">ancestral=False</span></code>, search for the clade that
contains the provided list of names as descendant. There must
not be any other name amongst descendant. If the tree is
unrooted and the representation uses a base within the clade, it
will not be detected. It possible (with the option
<code class="docutils literal"><span class="pre">both_sides=True</span></code> to allow searching for the complement of the
clade, thereby detecting the right clade even if it is at the
root. By default, if this situation occurs, the clade will not
be detected. Return <code class="docutils literal"><span class="pre">None</span></code> if the clade is not found.</p>
<p>With the option <code class="docutils literal"><span class="pre">ancestral=True</span></code>, search for the most recent
common ancestor of all leaves specified in <em>names</em>. Use of this
method necessarily supposes that the tree is rooted (however,
there is no requirements regarding its shape such as bifurcation
at the base) and it is not allowed to specify
<code class="docutils literal"><span class="pre">both_sides=True</span></code>. With this option, it is not possible to
have a <code class="docutils literal"><span class="pre">None</span></code> return value (since it is required that all
leaves are present in the tree, in the worse case the base of
the tree is returned).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>name</strong> – a <a class="reference external" href="https://docs.python.org/2.7/library/stdtypes.html#set" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">set</span></code></a> (or compatible) specifying the
requested leaves (as node labels, normally <a class="reference external" href="https://docs.python.org/2.7/library/functions.html#str" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a>
instances).</li>
<li><strong>ancestral</strong> – whether to look for the most recent common
ancestral clade containing requested leaves (by default,
looking for the clade containing the exact same list of
leaves).</li>
<li><strong>both_sides</strong> – only allowed with <code class="docutils literal"><span class="pre">ancestral=False</span></code>. Look
for both the requested list of leaves and its complement,
allowing to detect a clade even if it is spanning the base
of the tree.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">The <a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a> instance, if it exists, which has
the exact same list of descendants than <em>taxa</em>. If no such
clade is found, returns <code class="docutils literal"><span class="pre">None</span></code>.</p>
</td>
</tr>
</tbody>
</table>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">This method assumes that all leaf names of the tree are
unique, as well as the list of names provided as argument.
If this condition is not fulfilled, the right clade might
not be found even if it exists.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Replaces previous methods <code class="xref py py-meth docutils literal"><span class="pre">findGroup()</span></code>,
<code class="xref py py-meth docutils literal"><span class="pre">findMonophyleticGroup()</span></code>, <code class="xref py py-meth docutils literal"><span class="pre">smallest_group()</span></code> and
<code class="xref py py-meth docutils literal"><span class="pre">smallest_monophyleticGroup()</span></code> with a modification of
the underlying algorithm.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.frequency_nodes">
<code class="descname">frequency_nodes</code><span class="sig-paren">(</span><em>trees</em>, <em>relative=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.frequency_nodes" title="Permalink to this definition">¶</a></dt>
<dd><p>Labels all nodes of the current instances by integers counting
the number of trees where the same node exists among the trees
in the iterable <em>trees</em>. It is required that all leaf labels are
unique.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>trees</strong> – an iterable containing <a class="reference internal" href="#egglib.Tree" title="egglib.Tree"><code class="xref py py-class docutils literal"><span class="pre">Tree</span></code></a> instances
with exactly the same set of leaf labels.</li>
<li><strong>relative</strong> – node frequencies are expressed as fractions.
The use of this option requires that at least one tree is
provided.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">Nothing (result is available as node labels).</p>
</td>
</tr>
</tbody>
</table>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">With the exception of the base of the tree (which is ignored
by this function) and leaf labels, all previously set labels
are erased.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Labels are not converted to strings anymore.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.get_leaf">
<code class="descname">get_leaf</code><span class="sig-paren">(</span><em>label</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.get_leaf" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the terminal node (as a <a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a> instance) that
has the requested leaf label. If several nodes have this label,
returns the first one. If no nodes have this label, returns
<code class="docutils literal"><span class="pre">None</span></code>.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.iter_leaves">
<code class="descname">iter_leaves</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.iter_leaves" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an iterator over the leaf nodes.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 3.0.0: </span>Replaces the method <code class="xref py py-meth docutils literal"><span class="pre">get_terminal_nodes()</span></code> (the user
needs to get labels of those to replace the method
<code class="xref py py-meth docutils literal"><span class="pre">all_leaves()</span></code>).</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.lateralize">
<code class="descname">lateralize</code><span class="sig-paren">(</span><em>reverse=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.lateralize" title="Permalink to this definition">¶</a></dt>
<dd><p>Modify the order of children of all node of the trees in such a
way such that the children are sorted from the smallest to the
largest number of descending leaves.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>reverse</strong> – sort from in the more-descendants to
less-descendants order instead.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.map_descendants">
<code class="descname">map_descendants</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.map_descendants" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate and return a <a class="reference external" href="https://docs.python.org/2.7/library/stdtypes.html#dict" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">dict</span></code></a> where keys are all nodes (as
<a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a> instances) of the tree excepting its base and
all terminal node, and values are the <code class="xref py py-class docutils literal"><span class="pre">tuple</span></code> of leaf
labels that descend from this node.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 3.0.0.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.midroot">
<code class="descname">midroot</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.midroot" title="Permalink to this definition">¶</a></dt>
<dd><p>Automatic midpoint rooting of the tree. The tree must be
initially unrooted (trifurcation at the root). This method
identifies the most distant pair of terminal nodes (in case of
a draw, one is picked randomly) and the root of the tree (as a
new node) placed at the middle point of this path.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.newick">
<code class="descname">newick</code><span class="sig-paren">(</span><em>labels=True</em>, <em>brlens=True</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.newick" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the newick-formatted string representing the instance.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>labels</strong> – if <code class="docutils literal"><span class="pre">False</span></code>, omit the internal branch labels.</li>
<li><strong>brlens</strong> – if <code class="docutils literal"><span class="pre">False</span></code>, omit the branch lengths.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="egglib.Tree.num_leaves">
<code class="descname">num_leaves</code><a class="headerlink" href="#egglib.Tree.num_leaves" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of terminal nodes in the tree. If the tree is empty (only
the default base node), the number of leaves is 0.</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib.Tree.num_nodes">
<code class="descname">num_nodes</code><a class="headerlink" href="#egglib.Tree.num_nodes" title="Permalink to this definition">¶</a></dt>
<dd><p>Total number of nodes in the tree (the number is never smaller
than 1, even for empty trees).</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.remove_node">
<code class="descname">remove_node</code><span class="sig-paren">(</span><em>node</em>, <em>drop_parent=True</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.remove_node" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove a node from the tree, as well as all its descendants.
Since this operation may create a node with a single child, this
method may remove the parent or the brother of the removed node
depending on the structure of the tree (see <em>drop_parent</em>),
unless specified otherwise.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>node</strong> – the node to remove, as a <a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a> instance
belowing to the current tree. Terminal nodes can be removed,
but not the base of the tree.</li>
<li><strong>drop_parent</strong> – if <code class="docutils literal"><span class="pre">True</span></code>, remove the parent of the
removed node if it is left with only one child. If the
parent is the base of the tree, remove the other descendant
if it is not terminal (see below).</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>Assume we remove node <code class="docutils literal"><span class="pre">[3]</span></code> from the tree with this
structure:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>               <span class="o">/----------------------------&gt;</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
               <span class="o">|</span>
 <span class="o">/-----------&gt;</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>           <span class="o">/---------------&gt;</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
 <span class="o">|</span>             <span class="o">|</span>            <span class="o">|</span>
 <span class="o">|</span>             \<span class="o">----------&gt;</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
 <span class="o">|</span>                          <span class="o">|</span>
<span class="p">[</span><span class="mi">0</span><span class="p">]</span>                         \<span class="o">---------------&gt;</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
 <span class="o">|</span>
 <span class="o">|</span>              <span class="o">/---------------------------&gt;</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>
 <span class="o">|</span>              <span class="o">|</span> 
 \<span class="o">------------&gt;</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>            <span class="o">/-------------&gt;</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span>
                <span class="o">|</span>             <span class="o">|</span>
                \<span class="o">-----------&gt;</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
                              <span class="o">|</span>
                              \<span class="o">------------&gt;</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
</pre></div>
</div>
<p>Then, we would end up with the following tree:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span> <span class="o">/-----------&gt;</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">---------------------------&gt;</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
 <span class="o">|</span>
 <span class="o">|</span>
<span class="p">[</span><span class="mi">0</span><span class="p">]</span>
 <span class="o">|</span>              <span class="o">/---------------------------&gt;</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>
 <span class="o">|</span>              <span class="o">|</span> 
 \<span class="o">------------&gt;</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>            <span class="o">/-------------&gt;</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span>
                <span class="o">|</span>             <span class="o">|</span>
                \<span class="o">-----------&gt;</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
                              <span class="o">|</span>
                              \<span class="o">------------&gt;</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
</pre></div>
</div>
<p>The default behaviour is then to remove node <code class="docutils literal"><span class="pre">[1]</span></code> (and delete
its label if it exists) and to set the length of the branch
from <code class="docutils literal"><span class="pre">[0]</span></code> to <code class="docutils literal"><span class="pre">[2]</span></code> to the sum of the <code class="docutils literal"><span class="pre">[0]</span></code> to <code class="docutils literal"><span class="pre">[1]</span></code> and
<code class="docutils literal"><span class="pre">[1]</span></code> to <code class="docutils literal"><span class="pre">[2]</span></code>. But, with <code class="docutils literal"><span class="pre">drop_parent=False</span></code>, the tree is
left as is.</p>
<p>There is a special case with the base of the tree. Assume that
we remove node <code class="docutils literal"><span class="pre">[1]</span></code> from the original tree above. We then
would have a non-standard structure with a single child to the
base of the tree:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>                <span class="o">/---------------------------&gt;</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>
                <span class="o">|</span> 
<span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">-----------&gt;</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>            <span class="o">/-------------&gt;</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span>
                <span class="o">|</span>             <span class="o">|</span>
                \<span class="o">-----------&gt;</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
                              <span class="o">|</span>
                              \<span class="o">------------&gt;</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
</pre></div>
</div>
<p>In that case, the base is not removed, but node <code class="docutils literal"><span class="pre">[6]</span></code> is
removed using the <a class="reference internal" href="#egglib.Tree.collapse" title="egglib.Tree.collapse"><code class="xref py py-meth docutils literal"><span class="pre">collapse()</span></code></a> method (using options
<code class="docutils literal"><span class="pre">ignore_len=False</span></code> but <code class="docutils literal"><span class="pre">ignore_label=True</span></code> since the base is
not suppose to bear a label). We end up with the following
structure:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span> <span class="o">/---------------------------------&gt;</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>
 <span class="o">|</span> 
<span class="p">[</span><span class="mi">0</span><span class="p">]</span>                  <span class="o">/-------------&gt;</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span>
 <span class="o">|</span>                   <span class="o">|</span>
 \<span class="o">-----------------&gt;</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
                     <span class="o">|</span>
                     \<span class="o">------------&gt;</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
</pre></div>
</div>
<p>The length of the branch from <code class="docutils literal"><span class="pre">[0]</span></code> to <code class="docutils literal"><span class="pre">[6]</span></code> is spread
equally between the branch from <code class="docutils literal"><span class="pre">[0]</span></code> to <code class="docutils literal"><span class="pre">[7]</span></code> and the
branch from <code class="docutils literal"><span class="pre">[0]</span></code> to <code class="docutils literal"><span class="pre">[8]</span></code> (and so on if there are actually
more than one descendants). if <code class="docutils literal"><span class="pre">drop_parent=False</span></code> or if
<code class="docutils literal"><span class="pre">[6]</span></code> is a terminal node, it is not removed.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.root">
<code class="descname">root</code><span class="sig-paren">(</span><em>outgroup</em>, <em>branch_split=0.5</em>, <em>reoriente=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.root" title="Permalink to this definition">¶</a></dt>
<dd><p>Roots or reoriente the tree. By default, a new node is created
to represent the root and is placed on the branch leading to
the provided outgroup node (the second argument determines where
the new node is placed on this branch). Otherwise, the tree is
reoriented such as its base is placed at the location of the
provided outgroup. In the former case, its ends with a
bifurcation at the root; in the latter case, a trifurcation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>outgroup</strong> – <a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a> instance contained in this
tree. It can be a leaf or any internal node, but not the
current base of the tree (unless <em>reoriente</em> is <code class="docutils literal"><span class="pre">True</span></code>: in
that case, it <em>might</em> be the base of the tree [it will not
change anything] and it <em>cannot</em> be a leaf).</li>
<li><strong>branch_split</strong> – where to cut the branch leading to the
outgroup.</li>
<li><strong>reoriente</strong> – don’t create any root node (<em>branch_site</em> is
therefore not considered) and only place the node provided
as the <em>outgroup</em> argument at the base of the tree, thereby
merely changing the representation of the tree.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>The information below describes the case where
<code class="docutils literal"><span class="pre">reoriented=False</span></code> (proper rooting).</p>
<p>If this branch to the provided outgroup doesn’t have a branch
length, the <em>branch_split</em> argument is ignored. Otherwise,
<em>branch_split</em> must be a real number between 0 and 1 and gives
the proportion of the branch that must be allocated to the basal
branch leading to the outgroup, the complement being allocated
to the branch leading to the rest of the tree. If <em>branch_split</em>
is either 0 or 1, one of the branch will have a length of 0, but
it will exist anyway.</p>
<p>It is illegal to call this method on trees that are already
rooted (have a difurcation at the root).</p>
<p>If the original tree has this structure:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span> <span class="o">/-------------------------------------------&gt;</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
 <span class="o">|</span>
 <span class="o">|</span>             <span class="o">/-----------------------------&gt;</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
 <span class="o">|</span>             <span class="o">|</span>
 <span class="o">|-----------&gt;</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>             <span class="o">/--------------&gt;</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
 <span class="o">|</span>             <span class="o">|</span>              <span class="o">|</span>
 <span class="o">|</span>             \<span class="o">------------&gt;</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
<span class="p">[</span><span class="mi">0</span><span class="p">]</span>                           <span class="o">|</span>
 <span class="o">|</span>                            \<span class="o">--------------&gt;</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>
 <span class="o">|</span>
 <span class="o">|</span>              <span class="o">/----------------------------&gt;</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
 <span class="o">|</span>              <span class="o">|</span> 
 \<span class="o">------------&gt;</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>            <span class="o">/-------------&gt;</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
                <span class="o">|</span>             <span class="o">|</span>
                \<span class="o">---</span><span class="p">[</span><span class="n">ROOT</span><span class="p">]</span><span class="o">--&gt;</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span>
                              <span class="o">|</span>
                              \<span class="o">-------------&gt;</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span>
</pre></div>
</div>
<p>And rooting is requested at node <code class="docutils literal"><span class="pre">[9]</span></code>, the root will be
placed on the branch marked by <code class="docutils literal"><span class="pre">[ROOT]</span></code>.  The outcome will be
as depicted below, with the introduction of a new node (marked
<code class="docutils literal"><span class="pre">[ROOT]</span></code>) and the reorientation of the tree to place it at the
base:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>                     <span class="o">/-------------------------</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                     <span class="o">|</span>
              <span class="o">/-----</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>     <span class="o">/------------------</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
              <span class="o">|</span>      <span class="o">|</span>      <span class="o">|</span>
              <span class="o">|</span>      \<span class="o">-----</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>      <span class="o">/----------</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
              <span class="o">|</span>             <span class="o">|</span>       <span class="o">|</span>
  <span class="o">/---</span><span class="p">[</span><span class="n">E2</span><span class="p">]</span><span class="o">---</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>            \<span class="o">------</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
  <span class="o">|</span>           <span class="o">|</span>                     <span class="o">|</span>
  <span class="o">|</span>           <span class="o">|</span>                     \<span class="o">----------</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>
  <span class="o">|</span>           <span class="o">|</span>
<span class="p">[</span><span class="n">ROOT</span><span class="p">]</span>        \<span class="o">--------------------------------</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
  <span class="o">|</span>
  <span class="o">|</span>                     <span class="o">/---------------------</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
  <span class="o">|</span>                     <span class="o">|</span>
  \<span class="o">--------</span><span class="p">[</span><span class="n">E1</span><span class="p">]</span><span class="o">--------</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span>
                        <span class="o">|</span>
                        \<span class="o">---------------------</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span>
</pre></div>
</div>
<p>In this example, the relationship between nodes <code class="docutils literal"><span class="pre">[7]</span></code> and
<code class="docutils literal"><span class="pre">[0]</span></code> (the previous base of the tree) is reverted. The label
of node <code class="docutils literal"><span class="pre">[7]</span></code> is automatically transferred to node <code class="docutils literal"><span class="pre">[0]</span></code>.
This is consistent with the idea that internal node labels
describe a property of the branch. The original label of the
base, if it exists, is discarded. Since the branch between
<code class="docutils literal"><span class="pre">[7]</span></code> and <code class="docutils literal"><span class="pre">[9]</span></code> is cut in two, the original label of node
<code class="docutils literal"><span class="pre">[9]</span></code> is copied to node <code class="docutils literal"><span class="pre">[7]</span></code>, leaving them both with the
same label. However, if the outgroup is a terminal node, the
label is not copied and the other basal branch is left without
label.</p>
<p>Let <code class="docutils literal"><span class="pre">L</span></code> be the length of the branch from <code class="docutils literal"><span class="pre">[7]</span></code> and <code class="docutils literal"><span class="pre">[9]</span></code>
in the original tree, and <code class="docutils literal"><span class="pre">r</span></code> the value of the parameter
<code class="docutils literal"><span class="pre">branch_split</span></code>. The length of the branch <code class="docutils literal"><span class="pre">[E1]</span></code> will be set
to <code class="docutils literal"><span class="pre">rL</span></code>, and the branch <code class="docutils literal"><span class="pre">[E2]</span></code> to <code class="docutils literal"><span class="pre">(1-r)L</span></code>. Overall, the
length of the tree will not be modified.</p>
<p>In the case that <code class="docutils literal"><span class="pre">reoriente=True</span></code>, the final tree is rather:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span> <span class="o">/-------------------------------------------&gt;</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
 <span class="o">|</span>
 <span class="o">|-------------------------------------------&gt;</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span>
 <span class="o">|</span><span class="n">n</span>
<span class="p">[</span><span class="mi">9</span><span class="p">]</span>       <span class="o">/-----------------------------------&gt;</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
 <span class="o">|</span>        <span class="o">|</span>
 \<span class="o">------&gt;</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>       <span class="o">/--------------------------&gt;</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
          <span class="o">|</span>        <span class="o">|</span>
          \<span class="o">------&gt;</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>      <span class="o">/------------------&gt;</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
                   <span class="o">|</span>       <span class="o">|</span>
                   \<span class="o">-----&gt;</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>        <span class="o">/--------&gt;</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
                           <span class="o">|</span>         <span class="o">|</span>
                           \<span class="o">-------&gt;</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
                                     <span class="o">|</span>
                                     \<span class="o">--------&gt;</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>
</pre></div>
</div>
<p>The topology of the tree is the same as the initial one, except
that the base is now <code class="docutils literal"><span class="pre">[9]</span></code>. The lengths of all branches are
conserved. However, node labels between the old and the new base
are reverted: the node label of the new base (<code class="docutils literal"><span class="pre">[9]</span></code> in the
example) is affected to the next node (<code class="docutils literal"><span class="pre">[7]</span></code> in the example)
and so on until the old base (<code class="docutils literal"><span class="pre">[0]</span></code> in the example), whose
labe, if it exists, is discarded.</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Merged with <code class="xref py py-meth docutils literal"><span class="pre">reoriente()</span></code>.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.total_length">
<code class="descname">total_length</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.total_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the sum of all branch lengths across all nodes of the
tree. All branch lengths must be defined (non-<code class="docutils literal"><span class="pre">None</span></code>),
otherwise a <code class="xref py py-class docutils literal"><span class="pre">ValueError</span></code> will be raised.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Tree.unroot">
<code class="descname">unroot</code><span class="sig-paren">(</span><em>reverse=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Tree.unroot" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove the root. The tree must be initially unrooted
(bifurcation at the root). This method removes the root node and
places the base of the tree at one of the two basal nodes (the
nodes that are ancestral to the two basal groups). This method
does not change to total length of the tree. And error is raised
if only one of the two basal branches has a length. If the
initial basal node has a label, it is lost. If the node that
becomes the base has a label, it is left there (it will appear
a the base of the tree).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>reverse</strong> – if <code class="docutils literal"><span class="pre">True</span></code>, place the base of the tree at the
second basal node (by default, the first basal node is
used).</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="helpers">
<h2>Helpers<a class="headerlink" href="#helpers" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="egglib._interface.DataBase">
<em class="property">class </em><code class="descclassname">egglib._interface.</code><code class="descname">DataBase</code><a class="headerlink" href="#egglib._interface.DataBase" title="Permalink to this definition">¶</a></dt>
<dd><p>Base class for <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> and <a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a>.</p>
<p>This base class cannot be instanciated. Attempting to do so will
raise an exception.</p>
<dl class="method">
<dt id="egglib._interface.DataBase.add_outgroup">
<code class="descname">add_outgroup</code><span class="sig-paren">(</span><em>name</em>, <em>data</em>, <em>group=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.add_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an outgroup sample to the instance.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> – name of the new sample.</li>
<li><strong>data</strong> – an iterable (may be a string or a list) containing
the data to set for the new sample. For an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>
instance and if the sample is not the first, the number of
data must fit any previous ones (ingroup or outgroup).</li>
<li><strong>group</strong> – if not None, must be an unsigned integer (the
default value is 0).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.add_sample">
<code class="descname">add_sample</code><span class="sig-paren">(</span><em>name</em>, <em>data</em>, <em>groups=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.add_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a sample to the instance.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> – name of the new sample.</li>
<li><strong>data</strong> – an iterable (may be a string or a list) containing
the data to set for the new sample. For an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>
instance and if the sample is not the first, the number of
data must fit any previous ones (including outgroup samples,
if any).</li>
<li><strong>groups</strong> – if not None, must be an iterable with an
unsigned integer for each group levels of the instance (if
None, all group labels are set to 0), or a single integer
value if only one level needs to be specified.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.add_samples">
<code class="descname">add_samples</code><span class="sig-paren">(</span><em>items</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.add_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Add several samples at the end of the instance.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>items</strong> – items must be an iterable that have a length (for
example a list, an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> instance or a
<a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> instance. Each item of <em>items</em> must be
of length 2 or 3. The first item must be the sample name
string, the second item is the data values (as a list of
signed integers or a single string) and the third item, if
provided, is the list of group labels (one unsigned integer
for each level). See the method <a class="reference internal" href="#egglib.Align.add_sample" title="egglib.Align.add_sample"><code class="xref py py-meth docutils literal"><span class="pre">add_sample()</span></code></a>
for more details about each added item. If the current
instance is an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>, all added items must have
the same length which must be the same as any items
currently present in the instance. For a
<a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a>, the items may have different lengths.
The number of group levels is set by the sample with the
larger number of group labels, and all samples with less
labels are completed by e default value (0).</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 2.0.1: </span>Original name is <code class="xref py py-meth docutils literal"><span class="pre">addSequences()</span></code>.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Renamed as <code class="docutils literal"><span class="pre">add_samples</span></code>. All items are added in one shot
(rather than calling the one-sample add method iteratively).</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.clear">
<code class="descname">clear</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Clear the instance and release all memory. In most cases, it is
preferable to use the method <a class="reference internal" href="#egglib._interface.DataBase.reset" title="egglib._interface.DataBase.reset"><code class="xref py py-meth docutils literal"><span class="pre">reset()</span></code></a>.</p>
</dd></dl>

<dl class="classmethod">
<dt id="egglib._interface.DataBase.create">
<em class="property">classmethod </em><code class="descname">create</code><span class="sig-paren">(</span><em>obj</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.create" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new instance by copying date from the data container
passed as <em>obj</em>. The object <em>obj</em> can be:</p>
<ul class="simple">
<li>an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>,</li>
<li>a <a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> (but all sequences are required to have
the same length if the target type is <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>),</li>
<li>any iterable type yielding a acceptable items (see the
documentation for <a class="reference internal" href="#egglib._interface.DataBase.add_samples" title="egglib._interface.DataBase.add_samples"><code class="xref py py-meth docutils literal"><span class="pre">add_samples()</span></code></a> for more details).</li>
</ul>
<div class="versionadded">
<p><span class="versionmodified">New in version 2.0.1.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.del_outgroup">
<code class="descname">del_outgroup</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.del_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete an item of the instance.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – index of the sample to delete.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.del_sample">
<code class="descname">del_sample</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.del_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete an item of the instance.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – index of the sample to delete.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.encode">
<code class="descname">encode</code><span class="sig-paren">(</span><em>nbits=10</em>, <em>random=None</em>, <em>include_outgroup=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.encode" title="Permalink to this definition">¶</a></dt>
<dd><p>Renames all sequences using a random mapping of names to unique
keys of length <em>nbits</em>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>nbits</strong> – length of the keys (encoded names). This value
must be &gt;= 4 and &lt;= 63.</li>
<li><strong>random</strong> – a <a class="reference internal" href="tools.html#egglib.tools.Random" title="egglib.tools.Random"><code class="xref py py-class docutils literal"><span class="pre">Random</span></code></a> instance to be used a random
generator. By default, use a default instance.</li>
<li><strong>include_outgroup</strong> – a <a class="reference external" href="https://docs.python.org/2.7/library/functions.html#bool" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> indicating whether the
outgroup samples should also be considered (in this case, a
single mapping is returned including both ingroup and
outgroup samples).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">A dictionary mapping all the generated keys to the
actual sequence names. The keys are case-dependent and
guaranteed not to start with a number.</p>
</td>
</tr>
</tbody>
</table>
<p>The returned mapping can be used to restore the original names
using <a class="reference internal" href="#egglib._interface.DataBase.rename" title="egglib._interface.DataBase.rename"><code class="xref py py-meth docutils literal"><span class="pre">rename()</span></code></a>. This method is not affected by
the presence of sequences with identical names in the original
instance (and <a class="reference internal" href="#egglib._interface.DataBase.rename" title="egglib._interface.DataBase.rename"><code class="xref py py-meth docutils literal"><span class="pre">rename()</span></code></a> will also work properly
in that case).</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 2.0.1.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Keys are forced to start with a capital letter. Now take
a <a class="reference internal" href="tools.html#egglib.tools.Random" title="egglib.tools.Random"><code class="xref py py-class docutils literal"><span class="pre">Random</span></code></a> instance. Use library’s own random
number generator. Added an option for the outgroup.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.find">
<code class="descname">find</code><span class="sig-paren">(</span><em>name</em>, <em>include_outgroup=False</em>, <em>regex=False</em>, <em>multi=False</em>, <em>flags=None</em>, <em>index=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.find" title="Permalink to this definition">¶</a></dt>
<dd><p>Find a sample by its name.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>name</strong> – name of sample to identify.</li>
<li><strong>include_outgroup</strong> – a boolean indicating whether the
outgroup should be considered.</li>
<li><strong>regex</strong> – a boolean indicating whether the value passed a
<em>name</em> is a regular expression. If so, the string is passed
as is to the re module (using function
<a class="reference external" href="https://docs.python.org/2.7/library/re.html#re.search" title="(in Python v2.7)"><code class="xref py py-func docutils literal"><span class="pre">re.search()</span></code></a>).
Otherwise, only exact matches will be considered.</li>
<li><strong>multi</strong> – a boolean indicating whether all hits should be
returned. If so, a list of <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instances
is always returned (the list will be empty in case of no
hits). Otherwise, a single <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instance
(or its index)  will be returned for the first hit, or
<code class="docutils literal"><span class="pre">None</span></code> in case of no hits.</li>
<li><strong>flags</strong> – list of flags to be passed to <a class="reference external" href="https://docs.python.org/2.7/library/re.html#re.search" title="(in Python v2.7)"><code class="xref py py-func docutils literal"><span class="pre">re.search()</span></code></a>
(ignored if <em>regex</em> is <code class="docutils literal"><span class="pre">False</span></code>). For example, when looking
for samples containing the term “sample”, for being case
insensitive, use the following syntax:
<code class="docutils literal"><span class="pre">align.find(&quot;sample&quot;,</span> <span class="pre">regex=True,</span> <span class="pre">flags=[re.I])</span></code>. By
default (<code class="docutils literal"><span class="pre">None</span></code>) no further argument is passed.</li>
<li><strong>index</strong> – boolean indicating whether the index of the sample
should be returned. In that case return values for hits are
<a class="reference external" href="https://docs.python.org/2.7/library/functions.html#int" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> (by default, <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a>) instances.
Warning: it is not allowed to set both <em>include_outgroup</em>
and <em>index</em> to <code class="docutils literal"><span class="pre">True</span></code> as there would be no way to
distinguish between ingroup and outgroup indexes.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><code class="docutils literal"><span class="pre">None</span></code> if no hits were found, a (potentially empty)
list of <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instances or class:<cite>int</cite> if
<em>multi</em> is <code class="docutils literal"><span class="pre">True</span></code>, or a single <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> or
<a class="reference external" href="https://docs.python.org/2.7/library/functions.html#int" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> otherwise.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.find_motif">
<code class="descname">find_motif</code><span class="sig-paren">(</span><em>sample</em>, <em>motif</em>, <em>start=0</em>, <em>stop=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.find_motif" title="Permalink to this definition">¶</a></dt>
<dd><p>Locate the first instance of a motif for an ingroup sample</p>
<p>Returns the index of the first exact hit to a given substring.
The returned value is the position of the first base of the hit.
Only exact matches are implemented. To use regular expression
(for example to find degenerated motifs), one should extract the
string for the sequence and use a tool such as the regular
expression module (<a class="reference external" href="https://docs.python.org/2.7/library/re.html#module-re" title="(in Python v2.7)"><code class="xref py py-mod docutils literal"><span class="pre">re</span></code></a>).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>motif</strong> – a list of integer, or one-character strings (or
mixing both) or a string constituting the motif to search.</li>
<li><strong>start</strong> – position at which to start searching. The method
will never return a value smaller than <em>start</em>. By default,
search from the start of the sequence.</li>
<li><strong>stop</strong> – position at which to stop search (the motif
cannot overlap this position). No returned value will be
larger than stop-<em>n</em>. By default, or if <em>stop</em> is equal to
or larger than the length of the sequence, search until the
end of the sequence.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.find_motif_o">
<code class="descname">find_motif_o</code><span class="sig-paren">(</span><em>sample</em>, <em>motif</em>, <em>start=0</em>, <em>stop=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.find_motif_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Locate the first instance of a motif for an outgroup sample</p>
<p>Returns the index of the first exact hit to a given substring.
The returned value is the position of the first base of the hit.
Only exact matches are implemented. To use regular expression
(for example to find degenerated motifs), one should extract the
string for the sequence and use a tool such as the regular
expression module (<a class="reference external" href="https://docs.python.org/2.7/library/re.html#module-re" title="(in Python v2.7)"><code class="xref py py-mod docutils literal"><span class="pre">re</span></code></a>).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – outgroup sample index.</li>
<li><strong>motif</strong> – a list of integer, or one-character strings (or
mixing both) or a string constituting the motif to search.</li>
<li><strong>start</strong> – position at which to start searching. The method
will never return a value smaller than <em>start</em>. By default,
search from the start of the sequence.</li>
<li><strong>stop</strong> – position at which to stop search (the motif
cannot overlap this position). No returned value will be
larger than <em>stop</em> minus the length of the motif. By
default, or if <em>stop</em> is equal to or larger than the length
of the sequence, search until the end of the sequence.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.get_i">
<code class="descname">get_i</code><span class="sig-paren">(</span><em>sample</em>, <em>site</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.get_i" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a data entry.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>site</strong> – site index.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.get_label">
<code class="descname">get_label</code><span class="sig-paren">(</span><em>sample</em>, <em>level</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.get_label" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a group label.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>level</strong> – level index.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.get_label_o">
<code class="descname">get_label_o</code><span class="sig-paren">(</span><em>sample</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.get_label_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Get an outgroup label.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>sample</strong> – sample index.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.get_name">
<code class="descname">get_name</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.get_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the name of a sample.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – sample index.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.get_name_o">
<code class="descname">get_name_o</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.get_name_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the name of an outgroup sample.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – outgroup sample index.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.get_o">
<code class="descname">get_o</code><span class="sig-paren">(</span><em>sample</em>, <em>site</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.get_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Get an outgroup data entry.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – outgroup sample index.</li>
<li><strong>site</strong> – site index.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.get_outgroup">
<code class="descname">get_outgroup</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.get_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instance corresponding to the
requested index for the outgroup. The returned object allows to
modify the underlying data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – index of the outgroup sample to access.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.get_sample">
<code class="descname">get_sample</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.get_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instance corresponding to the
requested index. The returned object allows to modify the
underlying data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – index of the sample to access.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.get_sequence">
<code class="descname">get_sequence</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.get_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Access to the data entries of a given ingroup sample. Returns a
<a class="reference internal" href="#egglib.SequenceView" title="egglib.SequenceView"><code class="xref py py-class docutils literal"><span class="pre">SequenceView</span></code></a> instance which allow modifying the
underlying sequence container object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – ingroup sample index.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.get_sequence_o">
<code class="descname">get_sequence_o</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.get_sequence_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Access to the data entries of a given outgroup sample. Returns a
<a class="reference internal" href="#egglib.SequenceView" title="egglib.SequenceView"><code class="xref py py-class docutils literal"><span class="pre">SequenceView</span></code></a> instance which allow modifying the
underlying sequence container object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> – outgroup sample index.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.group_mapping">
<code class="descname">group_mapping</code><span class="sig-paren">(</span><em>level=0</em>, <em>indices=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.group_mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a dictionary mapping group labels to either
<a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instances (by default) or their indexes
representing all samples of this instance. It can process and
ingroup label level or the outgroup.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>level</strong> – index of the level to consider. If <code class="docutils literal"><span class="pre">None</span></code>,
processes the outgroup.</li>
<li><strong>indices</strong> – if <code class="docutils literal"><span class="pre">True</span></code>, represent samples by their index
(within the ingroup or the outgroup depending on <em>level</em>
being non-<code class="docutils literal"><span class="pre">None</span></code> or <code class="docutils literal"><span class="pre">None</span></code>) instead of
<code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code> instances.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="egglib._interface.DataBase.is_matrix">
<code class="descname">is_matrix</code><a class="headerlink" href="#egglib._interface.DataBase.is_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p><code class="docutils literal"><span class="pre">True</span></code> if the instance is an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>, and <code class="docutils literal"><span class="pre">False</span></code> if
it is a <a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.iter_outgroup">
<code class="descname">iter_outgroup</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.iter_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterator over the outgroup.</p>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.iter_samples">
<code class="descname">iter_samples</code><span class="sig-paren">(</span><em>ingrp</em>, <em>outgrp</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.iter_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterator over either ingroup samples or outgroup samples, or
both ingroup and outgroup (in that order) samples. Provided for
flexibility.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>ingrp</strong> – include ingroup samples.</li>
<li><strong>outgroup</strong> – include outgroup samples.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.name_mapping">
<code class="descname">name_mapping</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.name_mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a dictionary mapping names to <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a>
instances representing all samples of this instance. This method
is most useful when several sequences have the same name. It may
be used to detect and process duplicates. It processes the
ingroup only.</p>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.names">
<code class="descname">names</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.names" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate the list of ingroup sample names.</p>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.names_outgroup">
<code class="descname">names_outgroup</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.names_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate the list of outgroup sample names.</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib._interface.DataBase.ng">
<code class="descname">ng</code><a class="headerlink" href="#egglib._interface.DataBase.ng" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of group levels for ingroup (for outgroup, this number is
always one). It possible to change the value directly.</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib._interface.DataBase.no">
<code class="descname">no</code><a class="headerlink" href="#egglib._interface.DataBase.no" title="Permalink to this definition">¶</a></dt>
<dd><p>Current number of outgroup samples (cannot be modified
directly).</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib._interface.DataBase.ns">
<code class="descname">ns</code><a class="headerlink" href="#egglib._interface.DataBase.ns" title="Permalink to this definition">¶</a></dt>
<dd><p>Current number of samples, not considering the outgroup (cannot
be modified directly).</p>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.remove_duplicates">
<code class="descname">remove_duplicates</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.remove_duplicates" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove all duplicates, based on name exact matching. For all
pairs of samples with identical name, only the one occurring
first is conserved. The current instance is modified and this
method returns <code class="docutils literal"><span class="pre">None</span></code>.</p>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.rename">
<code class="descname">rename</code><span class="sig-paren">(</span><em>mapping</em>, <em>liberal=False</em>, <em>include_outgroup=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.rename" title="Permalink to this definition">¶</a></dt>
<dd><p>Rename sequences of the instance using the provided mapping.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>mapping</strong> – a <a class="reference external" href="https://docs.python.org/2.7/library/stdtypes.html#dict" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">dict</span></code></a> providing the mapping of old
names (as keys) to new names (which may, if needed, contain
duplicated).</li>
<li><strong>liberal</strong> – if this argument is <cite>False</cite> and a name does not
appear in <em>mapping</em>, a <a class="reference external" href="https://docs.python.org/2.7/library/exceptions.html#exceptions.ValueError" title="(in Python v2.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> is
raised. If <em>liberal</em> is <code class="docutils literal"><span class="pre">True</span></code>, names that don’t appear in
<em>mapping</em> are left unchanged.</li>
<li><strong>include_outgroup</strong> – if <code class="docutils literal"><span class="pre">True</span></code>, consider both ingroup and
outgroup samples together (they should be provided together
in the same mapping).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">The number of samples that have been actually renamed,
overall.</p>
</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 2.0.1.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Added an option for the outgroup. Added return value.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.reserve">
<code class="descname">reserve</code><span class="sig-paren">(</span><em>nsam=0</em>, <em>nout=0</em>, <em>lnames=0</em>, <em>ngrp=0</em>, <em>nsit=0</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.reserve" title="Permalink to this definition">¶</a></dt>
<dd><p>Pre-allocate memory. This method can be used when the size of
arrays is known a priori, in order to speed up memory
allocations. It is not necessary to set all values. Values less
than 0 are ignored.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>nsam</strong> – number of samples in the ingroup.</li>
<li><strong>nout</strong> – number of samples in the outgroup.</li>
<li><strong>lnames</strong> – length of sample names.</li>
<li><strong>ngrp</strong> – number of group labels.</li>
<li><strong>nsit</strong> – number of sites.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.reset">
<code class="descname">reset</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.reset" title="Permalink to this definition">¶</a></dt>
<dd><p>Reset the instance.</p>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.set_i">
<code class="descname">set_i</code><span class="sig-paren">(</span><em>sample</em>, <em>site</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.set_i" title="Permalink to this definition">¶</a></dt>
<dd><p>Set an ingroup data entry. The value must be a signed integer or a
one-character string. In the latter case, <a class="reference external" href="https://docs.python.org/2.7/library/functions.html#ord" title="(in Python v2.7)"><code class="xref py py-func docutils literal"><span class="pre">ord()</span></code></a> is
called on the behalf of the user.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>site</strong> – site index.</li>
<li><strong>value</strong> – allele value.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.set_label">
<code class="descname">set_label</code><span class="sig-paren">(</span><em>sample</em>, <em>level</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.set_label" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a group label.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>level</strong> – level index.</li>
<li><strong>value</strong> – new group value (unsigned integer).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.set_label_o">
<code class="descname">set_label_o</code><span class="sig-paren">(</span><em>sample</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.set_label_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Get an outgroup label.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>value</strong> – new group value (unsigned integer).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.set_name">
<code class="descname">set_name</code><span class="sig-paren">(</span><em>index</em>, <em>name</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.set_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the name of a sample.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – index of the sample</li>
<li><strong>name</strong> – new name value.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.set_name_o">
<code class="descname">set_name_o</code><span class="sig-paren">(</span><em>index</em>, <em>name</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.set_name_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the name of an outgroup sample.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – index of the outgroup sample</li>
<li><strong>name</strong> – new name value.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.set_o">
<code class="descname">set_o</code><span class="sig-paren">(</span><em>sample</em>, <em>site</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.set_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Set an outgroup data entry. The value must be a signed integer
or a one-character string. In the latter case, <a class="reference external" href="https://docs.python.org/2.7/library/functions.html#ord" title="(in Python v2.7)"><code class="xref py py-func docutils literal"><span class="pre">ord()</span></code></a> 
is called on the behalf of the user.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> – sample index.</li>
<li><strong>site</strong> – site index.</li>
<li><strong>value</strong> – allele value.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.set_outgroup">
<code class="descname">set_outgroup</code><span class="sig-paren">(</span><em>index</em>, <em>name</em>, <em>data</em>, <em>group=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.set_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the values for a sample of the outgroup</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – index of the sample to access (slices are not
permitted).</li>
<li><strong>name</strong> – new name of the sample.</li>
<li><strong>data</strong> – string or list of integers given the new values to
set. In case of an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>, it is required to pass a
sequence with length matching the number of sites of the
instance.</li>
<li><strong>groups</strong> – a positive integer to serve as group label (the
default value is 0.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.set_sample">
<code class="descname">set_sample</code><span class="sig-paren">(</span><em>index</em>, <em>name</em>, <em>data</em>, <em>groups=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.set_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the values for the sample corresponding to the requested
index.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – index of the sample to access (slices are not
permitted).</li>
<li><strong>name</strong> – new name of the sample.</li>
<li><strong>data</strong> – string or list of integers given the new values to
set. In case of an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>, it is required to pass a
sequence with length matching the number of sites of the
instance.</li>
<li><strong>groups</strong> – a list of integer label values, or a single
integer value, to set as group labels. The class will ensure
that all samples have the same number of group labels,
padding with 0 as necessary. The default corresponds to an
empty list.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.set_sequence">
<code class="descname">set_sequence</code><span class="sig-paren">(</span><em>index</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.set_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Replace all data entries for a given ingroup sample by new
values.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – ingroup sample index.</li>
<li><strong>value</strong> – can be a <a class="reference internal" href="#egglib.SequenceView" title="egglib.SequenceView"><code class="xref py py-class docutils literal"><span class="pre">SequenceView</span></code></a> instance, a
string or a list of integers (or any iterable with a length
which may mix single-character strings and integers). When
modifying an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> instance, all sequences must
have the same length as the current alignment length.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.set_sequence_o">
<code class="descname">set_sequence_o</code><span class="sig-paren">(</span><em>index</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.set_sequence_o" title="Permalink to this definition">¶</a></dt>
<dd><p>Replace all data entries for a given outgroup sample by new
values.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – outgroup sample index.</li>
<li><strong>value</strong> – can be a <a class="reference internal" href="#egglib.SequenceView" title="egglib.SequenceView"><code class="xref py py-class docutils literal"><span class="pre">SequenceView</span></code></a> instance, a
string or a list of integers (or any iterable with a length
which may mix single-character strings and integers). When
modifying an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> instance, all sequences must
have the same length as the current alignment length.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.shuffle">
<code class="descname">shuffle</code><span class="sig-paren">(</span><em>level=0</em>, <em>random=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.shuffle" title="Permalink to this definition">¶</a></dt>
<dd><p>Shuffle group labels.</p>
<p>Randomly reassigns group labels. Modifies the current instance
and returns <code class="docutils literal"><span class="pre">None</span></code>. Only the specified level is affected, and
only the group labels are modified (the order of samples is not
changed.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>level</strong> – index of the group level to shuffle. To shuffle
the outgroup’s labels, set <em>level</em> to <code class="docutils literal"><span class="pre">None</span></code>.</li>
<li><strong>random</strong> – A <a class="reference internal" href="tools.html#egglib.tools.Random" title="egglib.tools.Random"><code class="xref py py-class docutils literal"><span class="pre">Random</span></code></a> instance to be used a random
generator. By default, use a default instance.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>The outgroup is necessarily processed separately from the
rest of the instance.  Allow pass a <a class="reference internal" href="tools.html#egglib.tools.Random" title="egglib.tools.Random"><code class="xref py py-class docutils literal"><span class="pre">Random</span></code></a>
instance. Use library’s own random number generator.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.subgroup">
<code class="descname">subgroup</code><span class="sig-paren">(</span><em>groups</em>, <em>outgroup_all=False</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.subgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate and return a copy of the instance with only samples
from the specified groups (identified by their group labels,
including outgroup).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>groups</strong> – an integer or a dictionary providing the labels
of the groups that are selected. If an integer, it is
understood as an ingroup label corresponding to the first
grouping level (if more than one). If a dictionary, the keys
of this dictionary must be integer corresponding to grouping
levels (or <code class="docutils literal"><span class="pre">None</span></code> for the outgroup) and the values must be
lists containing the requested labels. It is not required to
include all levels and the outgroup in the dictionary. It is
allowed to specify labels that are actually not represented
in the data.</li>
<li><strong>outgroup_all</strong> – if <code class="docutils literal"><span class="pre">True</span></code>, always include all outgroup
samples in the returned instance, regardless of whether
their group labels are specified in the first argument.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 3.0.0.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.subset">
<code class="descname">subset</code><span class="sig-paren">(</span><em>samples</em>, <em>outgroup=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate and return a copy of the instance with only a specified
list of samples. It is possible to select ingroup and/or
outgroup samples and the sample indexes are not required to be
consecutive.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>samples</strong> – a list (or other iterable type with a length) of
sample indexes giving the list of ingroup samples that must
be exported to the return value object. If <code class="docutils literal"><span class="pre">None</span></code>, do not
export any ingroup samples.</li>
<li><strong>outgroup</strong> – a list (or other iterable type with a length)
of sample indexes giving the list of outgroup samples that
must be exported to the return value object. If <code class="docutils literal"><span class="pre">None</span></code>, do
not export any outgroup samples.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 3.0.0: </span>Established as a method for <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> and
<a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a>.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.to_fasta">
<code class="descname">to_fasta</code><span class="sig-paren">(</span><em>fname=None</em>, <em>first=0</em>, <em>last=4294967295</em>, <em>mapping=None</em>, <em>groups=False</em>, <em>shift_labels=False</em>, <em>include_outgroup=True</em>, <em>linelength=50</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.to_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Export alignment in the fasta format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>fname</strong> – Name of the file to export data to. By default, the
file is created (or overwritten if it already exists). If the
option <em>append</em> is <code class="docutils literal"><span class="pre">True</span></code>, data is appended at the end of the
file (and it must exist). If <em>fname</em> is <code class="docutils literal"><span class="pre">None</span></code> (default), no
file is created and the formatted data is returned as a
<a class="reference external" href="https://docs.python.org/2.7/library/functions.html#str" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a>. In the alternative case, nothing is returned.</li>
<li><strong>first</strong> – If only part of the sequences should be exported:
index of the first sequence to export.</li>
<li><strong>last</strong> – If only part of the sequences should be exported: index
of the last sequence to export. If the value is larger than the
index of the last sequence, all sequences are exported until the
last (this is the default). If <em>last*&lt;*first</em>, no sequences are
exported.</li>
<li><strong>mapping</strong> – A string providing the character mapping. Use the
specified list of characters to map integer allelic values. If
a non-empty string is provided, the length of the string must be
larger than the largest possible allele values. In that case,
the allele values will be used as indexes in order to determine
which character from this string must be used for outputting. In
the case that this method is used with an empty string, the
mapping will not be used and the allele values will be casted
directly to characters.</li>
<li><strong>groups</strong> – A boolean indicating whether group labels
should be exported, or ignored.</li>
<li><strong>shift_labels</strong> – A boolean indicating whether group labels
should be incremented of one unit when exporting.</li>
<li><strong>include_outgroup</strong> – A boolean indicating whether the outgroup
should be exported. It will be exported at the end of the
ingroup, and without discriminating label if <em>include_labels</em> is
<code class="docutils literal"><span class="pre">False</span></code>.</li>
<li><strong>linelength</strong> – The length of lines for internal breaking of
sequences.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib._interface.DataBase.to_outgroup">
<code class="descname">to_outgroup</code><span class="sig-paren">(</span><em>index</em>, <em>label=0</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib._interface.DataBase.to_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Transfer a sample to the outgroup.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>index</strong> – index of the sample to move to the outgroup.</li>
<li><strong>label</strong> – label to assign to the sample (usually, this label
assigned the sample to an individual).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="egglib.SampleView">
<em class="property">class </em><code class="descclassname">egglib.</code><code class="descname">SampleView</code><span class="sig-paren">(</span><em>parent</em>, <em>index</em>, <em>outgroup</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.SampleView" title="Permalink to this definition">¶</a></dt>
<dd><p>This class manages the name, sequence and groups of an item of the
<a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> and <a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> classes.
<a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> objects allow iteration and general
manipulation of (large) data sets without unnecessary extraction of
full sequences. Modifications of <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> objects are
immediately applied to the underlying data holder object.
<a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> objects are iterable and allow indexing (the
values are: name, sequence, group, in that order).</p>
<p>In principle, only <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> and <a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> instances
are supposed to build <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a> instances.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>parent</strong> – a <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> or <a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> instance.</li>
<li><strong>index</strong> – an index within the parent instance.</li>
<li><strong>outgroup</strong> – a boolean indicating whether the instance
represents an outgroup (rather than ingroup) sample.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 2.0.1.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Renamed from <code class="xref py py-class docutils literal"><span class="pre">SequenceItem</span></code>. No more index check at
construction. No more string formatting. Supports multiple
groups. Sequence managed by <a class="reference internal" href="#egglib.SequenceView" title="egglib.SequenceView"><code class="xref py py-class docutils literal"><span class="pre">SequenceView</span></code></a> and groups by
<a class="reference internal" href="#egglib.GroupView" title="egglib.GroupView"><code class="xref py py-class docutils literal"><span class="pre">GroupView</span></code></a>.</p>
</div>
<dl class="attribute">
<dt id="egglib.SampleView.group">
<code class="descname">group</code><a class="headerlink" href="#egglib.SampleView.group" title="Permalink to this definition">¶</a></dt>
<dd><p>Access to group labels. Returns a <a class="reference internal" href="#egglib.GroupView" title="egglib.GroupView"><code class="xref py py-class docutils literal"><span class="pre">GroupView</span></code></a> instance
that can be modified. The values used for assignment may be a
<a class="reference internal" href="#egglib.GroupView" title="egglib.GroupView"><code class="xref py py-class docutils literal"><span class="pre">GroupView</span></code></a>, a list of integers or a single integer. If
there are less values defined than the number of group levels,
the values after the last one are left unchanged.</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib.SampleView.index">
<code class="descname">index</code><a class="headerlink" href="#egglib.SampleView.index" title="Permalink to this definition">¶</a></dt>
<dd><p>Index of the sample in the parent  (<a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> or
<a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> instance containing this sample).</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib.SampleView.ls">
<code class="descname">ls</code><a class="headerlink" href="#egglib.SampleView.ls" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of the sequence for this sample.</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib.SampleView.name">
<code class="descname">name</code><a class="headerlink" href="#egglib.SampleView.name" title="Permalink to this definition">¶</a></dt>
<dd><p>Get or set the sample name.</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib.SampleView.outgroup">
<code class="descname">outgroup</code><a class="headerlink" href="#egglib.SampleView.outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p><code class="docutils literal"><span class="pre">True</span></code> if the sample is part of the outgroup. This value
cannot be modifed.</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib.SampleView.parent">
<code class="descname">parent</code><a class="headerlink" href="#egglib.SampleView.parent" title="Permalink to this definition">¶</a></dt>
<dd><p>Reference of the parent instance (<a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> or
<a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> instance containing this sample).</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib.SampleView.sequence">
<code class="descname">sequence</code><a class="headerlink" href="#egglib.SampleView.sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Access to data entries. This attribute is represented by a
<a class="reference internal" href="#egglib.SequenceView" title="egglib.SequenceView"><code class="xref py py-class docutils literal"><span class="pre">SequenceView</span></code></a> instance which allow modifying the
underlying sequence container object. It is possible to set this
attribute using either a string, a list of integers (it can be
any iterable with a length and it may mix single-character
strings and integers to represent individual data entries). When
modifying an <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> instance, all sequences must have
the same length as the current alignment length.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="egglib.SequenceView">
<em class="property">class </em><code class="descclassname">egglib.</code><code class="descname">SequenceView</code><span class="sig-paren">(</span><em>parent</em>, <em>index</em>, <em>outgroup</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.SequenceView" title="Permalink to this definition">¶</a></dt>
<dd><p>This class manages the sequence of an item of the <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a>
and <a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> classes. Supports iteration and random
access (including with slices) which is the preferred way if
sequences are large because it prevents extraction of the full
sequence.</p>
<p>In principle, only <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a>, <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> and
<a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> instances are supposed to build
<a class="reference internal" href="#egglib.SequenceView" title="egglib.SequenceView"><code class="xref py py-class docutils literal"><span class="pre">SequenceView</span></code></a> instances.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>parent</strong> – a <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> or <a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> instance.</li>
<li><strong>index</strong> – an index within the parent instance.</li>
<li><strong>outgroup</strong> – a boolean indicating whether the instance
represents an outgroup (rather than ingroup) sample.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 3.0.0.</span></p>
</div>
<dl class="method">
<dt id="egglib.SequenceView.find">
<code class="descname">find</code><span class="sig-paren">(</span><em>motif</em>, <em>start=0</em>, <em>stop=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.SequenceView.find" title="Permalink to this definition">¶</a></dt>
<dd><p>Locate the first instance of a motif.</p>
<p>Returns the index of the first exact hit to a given substring.
The returned value is the position of the first base of the hit.
Only exact matches are implemented. To use regular expression
(for example to find degenerated motifs), one should extract the
string for the sequence and use a tool such as the regular
expression module (<a class="reference external" href="https://docs.python.org/2.7/library/re.html#module-re" title="(in Python v2.7)"><code class="xref py py-mod docutils literal"><span class="pre">re</span></code></a>).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>motif</strong> – a list of integer, or one-character strings (or
mixing both) or a string constituting the motif to search.</li>
<li><strong>start</strong> – position at which to start searching. The method
will never return a value smaller than <em>start</em>. By default,
search from the start of the sequence.</li>
<li><strong>stop</strong> – position at which to stop search (the motif
cannot overlap this position). No returned value will be
larger than stop-<em>n</em>. By default, or if <em>stop</em> is equal to
or larger than the length of the sequence, search until the
end of the sequence.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.SequenceView.insert">
<code class="descname">insert</code><span class="sig-paren">(</span><em>position</em>, <em>values</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.SequenceView.insert" title="Permalink to this definition">¶</a></dt>
<dd><p>Insert data entries.</p>
<p>This method is only available for samples belonging to a
<a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> instance. For <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> instances,
(for which it is possible to insert data entries to all ingroup
and outgroup samples) use the method <a class="reference internal" href="#egglib.Align.insert_columns" title="egglib.Align.insert_columns"><code class="xref py py-meth docutils literal"><span class="pre">insert_columns()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>position</strong> – the position at which to insert sites. The new
sites are inserted before the specified index. Use 0 to add
sites at the beginning of the sequence, and the current
number of sites for this sample to add sites at the end. If
the value is larger than the current number of sites for
this sample, or if <code class="docutils literal"><span class="pre">None</span></code> is provided, new sites are added
at the end of the sequence.</li>
<li><strong>values</strong> – a list of integers or a string containing the
data entries to insert in the sequence. It is allowed to mix
integers and one-character strings in a list or use other
types provided that they are iterable and have a length</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.SequenceView.string">
<code class="descname">string</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.SequenceView.string" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a string from all data entries. All allele values must
be &gt;= 0. Even if this condition is met, the generated string
might be non printable.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.SequenceView.strip">
<code class="descname">strip</code><span class="sig-paren">(</span><em>values</em>, <em>left=True</em>, <em>right=True</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.SequenceView.strip" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete leading and/or trailing occurrences of any characters
given in the <em>values</em> argument. The underlying object is
modified and this method returns <code class="docutils literal"><span class="pre">None</span></code>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>values</strong> – <em>values</em> should be a list of integers but can
also be a string.</li>
<li><strong>left</strong> – A bolean indicating whether left-side characters
should be stripped.</li>
<li><strong>right</strong> – A bolean indicating whether left-side characters
should be stripped.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.SequenceView.to_lower">
<code class="descname">to_lower</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.SequenceView.to_lower" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts all allele values of this sample to lower case. More
specifically, transforms all values lying in the range <code class="docutils literal"><span class="pre">A-Z</span></code>
to their equivalent in the range <code class="docutils literal"><span class="pre">a-z</span></code>. All other allele
values are ignored. The underlying object data is modified and
this method returns <code class="docutils literal"><span class="pre">None</span></code>.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.SequenceView.to_upper">
<code class="descname">to_upper</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.SequenceView.to_upper" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts all allele values of this sample to upper case. More
specifically, transforms all values lying in the range <code class="docutils literal"><span class="pre">a-z</span></code>
to their equivalent in the range <code class="docutils literal"><span class="pre">A-Z</span></code>. All other allele
values are ignored. The underlying object data is modified and
this method returns <code class="docutils literal"><span class="pre">None</span></code>.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="egglib.GroupView">
<em class="property">class </em><code class="descclassname">egglib.</code><code class="descname">GroupView</code><span class="sig-paren">(</span><em>parent</em>, <em>index</em>, <em>outgroup</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.GroupView" title="Permalink to this definition">¶</a></dt>
<dd><p>This class manages the list of group labels of an item of the
<a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> and <a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> classes. This class can be
represented a list of unsigned integers, is iterable and allows
random access (read and write) but not slices.</p>
<p>In principle, only <a class="reference internal" href="#egglib.SampleView" title="egglib.SampleView"><code class="xref py py-class docutils literal"><span class="pre">SampleView</span></code></a>, <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> and
<a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> instances are supposed to build
<a class="reference internal" href="#egglib.GroupView" title="egglib.GroupView"><code class="xref py py-class docutils literal"><span class="pre">GroupView</span></code></a> instances.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>parent</strong> – a <a class="reference internal" href="#egglib.Align" title="egglib.Align"><code class="xref py py-class docutils literal"><span class="pre">Align</span></code></a> or <a class="reference internal" href="#egglib.Container" title="egglib.Container"><code class="xref py py-class docutils literal"><span class="pre">Container</span></code></a> instance.</li>
<li><strong>index</strong> – an index within the parent instance.</li>
<li><strong>outgroup</strong> – a boolean indicating whether the instance
represents an outgroup (rather than ingroup) sample.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 3.0.0.</span></p>
</div>
<dl class="attribute">
<dt id="egglib.GroupView.outgroup">
<code class="descname">outgroup</code><a class="headerlink" href="#egglib.GroupView.outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p><code class="docutils literal"><span class="pre">True</span></code> if the sample is part of the outgroup.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="egglib.Node">
<em class="property">class </em><code class="descclassname">egglib.</code><code class="descname">Node</code><span class="sig-paren">(</span><em>label=None</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Node" title="Permalink to this definition">¶</a></dt>
<dd><p>This class provides an interface to a <a class="reference internal" href="#egglib.Tree" title="egglib.Tree"><code class="xref py py-class docutils literal"><span class="pre">Tree</span></code></a> instance’s
nodes and allows access and modification of data attached to a given
node as well as the tree descending from that node. A node must be
understood as the point <em>below</em> a branch. Branches (connections
between nodes) have a direction: they go <em>from</em> a node <em>to</em> another
node. Nodes have therefore <em>children</em>  and <em>parents</em> (a given node
can have one, or possibly no, parent). Connecting a node to itself,
making a two-way branch (to branches connecting the same two nodes
in opposite directions) or duplicate branches (between the same two
nodes and in the same direction) are illegal.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>label</strong> – node label (in case of a terminal node, its leaf
label), if needed. Labels, if provided, are expected to be
strings for terminal nodes, and numeric values for internal
nodes, but technically, all user-supplied values are accepted
(however, some <a class="reference internal" href="#egglib.Tree" title="egglib.Tree"><code class="xref py py-class docutils literal"><span class="pre">Tree</span></code></a> methods require proper types).</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="egglib.Node.branch_to">
<code class="descname">branch_to</code><span class="sig-paren">(</span><em>child</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Node.branch_to" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the length of the branch to one of this node’s children.
Non-specified branch lengths are represented by <code class="docutils literal"><span class="pre">None</span></code>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>child</strong> – nodes whose branch should be returned. It can be
represented by a direct reference (as a :class:~.Node`), or
by its index in the children list. In that case, ensure that
the index is currently valid.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Node.child">
<code class="descname">child</code><span class="sig-paren">(</span><em>idx</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Node.child" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a given child, as a <a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a> instance.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Node.children">
<code class="descname">children</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Node.children" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an iterator over this node’s children.</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 3.0.0: </span>Returns an iterator instead of a list.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.Node.has_descendant">
<code class="descname">has_descendant</code><span class="sig-paren">(</span><em>node</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Node.has_descendant" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a boolean indicating whether the <a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a> passed
as argument can be found amongst the descendants (children,
children of children, and so on down to the leaves) of this
node.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Node.is_child">
<code class="descname">is_child</code><span class="sig-paren">(</span><em>node</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Node.is_child" title="Permalink to this definition">¶</a></dt>
<dd><p>Return <code class="docutils literal"><span class="pre">True</span></code> if the <a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a> instance passed as
argument is one of the children of the current node.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Node.is_parent">
<code class="descname">is_parent</code><span class="sig-paren">(</span><em>node</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Node.is_parent" title="Permalink to this definition">¶</a></dt>
<dd><p>Return <code class="docutils literal"><span class="pre">True</span></code> if the <a class="reference internal" href="#egglib.Node" title="egglib.Node"><code class="xref py py-class docutils literal"><span class="pre">Node</span></code></a> instance passed as
argument is the parent of the current node. As a side effect,
passing <code class="docutils literal"><span class="pre">None</span></code> to an instance that has no parent will return
<code class="docutils literal"><span class="pre">True</span></code>.</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib.Node.label">
<code class="descname">label</code><a class="headerlink" href="#egglib.Node.label" title="Permalink to this definition">¶</a></dt>
<dd><p>Node’s label (modifiable).</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Node.leaves_down">
<code class="descname">leaves_down</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Node.leaves_down" title="Permalink to this definition">¶</a></dt>
<dd><p>Recursively gets all leaf labels descending from that node. If
this is a terminal node, returns its label in a one-item list.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Node.leaves_up">
<code class="descname">leaves_up</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Node.leaves_up" title="Permalink to this definition">¶</a></dt>
<dd><p>Recursively gets all leaf labels contained on the other side of
the tree (that is, all leaves except those descending from this
node). If this is the root node, returns an empty list).</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Node.newick">
<code class="descname">newick</code><span class="sig-paren">(</span><em>labels=True</em>, <em>brlens=True</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Node.newick" title="Permalink to this definition">¶</a></dt>
<dd><p>Formats the node and the subtree descending from is as a
newick string. If <em>labels</em> is <code class="docutils literal"><span class="pre">False</span></code>, omit internal branch
labels. If <em>brlens</em> is <code class="docutils literal"><span class="pre">False</span></code>, omit branch lengths.</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib.Node.num_children">
<code class="descname">num_children</code><a class="headerlink" href="#egglib.Node.num_children" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of children connected to this node.</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib.Node.parent">
<code class="descname">parent</code><a class="headerlink" href="#egglib.Node.parent" title="Permalink to this definition">¶</a></dt>
<dd><p>Reference to the parent of the current node (<code class="docutils literal"><span class="pre">None</span></code> if the
node has no parent).</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib.Node.parent_branch">
<code class="descname">parent_branch</code><a class="headerlink" href="#egglib.Node.parent_branch" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of the branch to this node’s parent. Non-specified branch
lengths are represented by <code class="docutils literal"><span class="pre">None</span></code>. An exception is thrown if
this node has no parent. This attribute can be modified.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.Node.set_branch_to">
<code class="descname">set_branch_to</code><span class="sig-paren">(</span><em>child</em>, <em>brlen</em><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Node.set_branch_to" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the length of the branch to one of this node’s children.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>child</strong> – node whose branch should be resized. It can be
represented by a direct reference (as a :class:~.Node`), or
by its index in the child list. In that case, ensure that
the index is currently valid.</li>
<li><strong>brlen</strong> – new branch length (it is allowed to pass <code class="docutils literal"><span class="pre">None</span></code>).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="egglib.Node.siblings">
<code class="descname">siblings</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#egglib.Node.siblings" title="Permalink to this definition">¶</a></dt>
<dd><p>List of other children of this node’s parent. It is required
that this node has a parent.</p>
</dd></dl>

</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="tools.html" title="Generic tools"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Python Module"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../contents.html">EggLib&#39;s documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../toc.html" >Contents</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="index.html" >Python Module</a> &#187;</li> 
      </ul>
    </div>
<!--
        <address style="text-align: right;"><small>Hosted by&nbsp;
        <a href="http://sourceforge.net/projects/egglib"><img src="http://sflogo.sourceforge.net/sflogo.php?group_id=232965&amp;type=8" width="80" height="15" alt="Get EggLib at SourceForge.net. Fast, secure and Free Open Source software downloads" / ALIGN=RIGHT BORDER=0></a>
        </address>
-->

  </body>
</html>