import glob
import re
import sys
from os.path import join 

###############################################################################
###############################################################################
# Adapt here so you can get all read pairs and ref:
#--------------------------------------------------

# directory contained sequencing data files
datadir="/homedir/gladieux/work/magMax_project/0_Assembly/0_rawdata/souche_adel/"

# Out directory
OutDir = '/homedir/gladieux/work/magMax_project/0_Assembly/4_test_result/'


# file termination
suffixFileR1 = '_R1.fastq.gz'
suffixFileR2 = '_R2.fastq.gz'

if not os.path.exists(OutDir):
	os.makedirs(OutDir)
################################################################################
################################################################################

SAMPLES, = glob_wildcards(datadir+"{smp}"+suffixFileR1)
kmere = [20,30,40,50,60,70,80,90]	
	
rule final:
	input: 
		expand( OutDir + '2_data-final/')
		
		
		


rule select_assemblage :
	threads : 1 
	input : file = OutDir + '3_report/data_report/select.csv'
	output : OutDir + '2_data-final/'
	run : 
		directory = input.file[0:-31] + '1_ABySS/'
		f = open(input.file,'r')
		lines = f.readlines()
		f.close()
		for line in lines : 
			select = line.split('\t')[0]
			path_old = '%s/%s/%s-8.fa'%(directory + select.split('_')[0],select,select)
			path_New = '%s.fasta'%(output[0]+select)
			os.system('mv %s %s'%(path_old,path_New))
	
	
rule report:
	threads : 1 
	input : OutDir + '3_report/data_report/Quality.txt'
	output : report = OutDir + '3_report/report.html', select = OutDir + '3_report/data_report/select.csv'
	shell : """
	module load bioinfo/R/3.4.3
	Rscript -e 'rmarkdown::render("script/report.Rmd", output_file="{output.report}", quiet=TRUE, params = list(quality = "{input}"))'
	"""
		
rule recup_stat :
	threads : 1  
	input :  lien = expand(OutDir + '1_ABySS/{smp}/{smp}_{km}/',smp = SAMPLES, km = kmere), outdir = OutDir
	output : OutDir + '3_report/data_report/Quality.txt'
	shell : """
	script/Quality.py -d {input.outdir}1_ABySS/ -o {output}
	"""	

rule ABySS_Launch:
	threads : 1 
	input: R1 = datadir + '{smp}' + suffixFileR1, R2 = datadir + '{smp}' + suffixFileR2
	params : suffix = suffixFileR1
	output: OutDir + '1_ABySS/{smp}/{smp}_{km}/'
	shell :"""
	module load bioinfo/abyss/1.9.0
	cd {output}
	/usr/local/bioinfo/abyss/1.9.0/bin/abyss-pe name={wildcards.smp}_{wildcards.km} k={wildcards.km} in='{input.R1} {input.R2}' -o {output}abyss_assembly_{wildcards.smp}_{wildcards.km}.fasta
	"""
		
