---
title: "Analyse des données RNAseqs"
author: "Florian Charriat"
date: "`r format(Sys.time(), '%d %B, %Y')`"
params:
   mapping: "None"
   count: "None"
output:
  html_document:
    fig_caption: yes
    highlight: zenburn
    theme: cerulean
    toc: yes
    toc_depth: 5
    toc_float:
      collapsed: false
      smooth_scroll: false
---
```{r knitr setup, include=FALSE,  eval=TRUE, echo=FALSE, warning=FALSE}
library(knitr)
options(width=300)
knitr::opts_chunk$set(
  fig.width = 10, 
  fig.height = 5, 
  fig.align = "center", 
  size = "tiny", 
  echo = FALSE, eval=TRUE, 
  warning = FALSE, message = FALSE, 
  results = TRUE, comment = "")
```

```{r installation package}
if(!require("ggplot2"))
{
  install.packages("ggplot2", repos='http://cran.us.r-project.org')
}
if(!require("gridExtra"))
{
  install.packages("gridExtra", repos='http://cran.us.r-project.org')
}
if(!require("cowplot"))
{
  install.packages("cowplot", repos='http://cran.us.r-project.org')
}
if(!require('DT'))
{
  install.packages("DT", repos='http://cran.us.r-project.org')
}
if(!require('stringr'))
{
  install.packages("stringr", repos='http://cran.us.r-project.org')
}
library("ggplot2")
library("gridExtra")
library("cowplot")
library('DT')
library('stringr')
```

&nbsp;

##Représentation des données RNAseq

&nbsp;

```{r Représentation des Reads}
data.mapping <- read.table(params$mapping, sep = '\t')
colnames(data.mapping) <- c('Id','nb.read','type')
name.RNAseq <- unique(data.mapping$Id)


text <- c()
for (elt in name.RNAseq)
{
  text <- c(text,paste('Total read : ',sum(data.mapping[data.mapping$Id == elt,2])),'')
}



all.read.plot <- ggplot(data.mapping,aes(x = Id,y = as.numeric(as.character(nb.read)),group = type,fill = type)) + geom_bar(stat = 'identity',color = 'black') +
  coord_flip() + ylab('Nombre de read')+ ggtitle('Nombre total de reads')+scale_x_discrete(name ="")  +
  scale_fill_manual(values = c('forestgreen','firebrick')) + geom_text(aes(x = Id, y = 1,label =text),hjust = 0,color = 'white')


data.aln <- data.mapping[data.mapping$type == 'Align 1 times',]
align.read.plot <- ggplot(data.aln,aes(x = Id,y = as.numeric(as.character(nb.read)))) + geom_bar(stat = 'identity',fill = 'forestgreen',color = 'black') + ggtitle('Nombre de reads aligné')+
  coord_flip()+scale_x_discrete(name ="")  + ylab('Nombre de read') + geom_text(aes(label = nb.read),color = 'white',hjust = 1.3)

plot_grid(all.read.plot, align.read.plot, ncol = 1, nrow = 2)
```


---- 

##Représentation du nombre de gènes exprimés en fonction des données RNAseq

&nbsp;

```{r summary expression}
data <- read.table(params$count)
colnames(data) <- c('Id','Position','FPKM','couverture','RNAseq')
data$FPKM <- log2(data$FPKM)
data$FPKM[data$FPKM == '-Inf'] <- 0
min_level <- 1
No_expression <- data[data$FPKM < min_level ,]
expression <- data[data$FPKM > min_level,]
stat_expression = summary(expression$FPKM)
sd.expression <- sd(expression$FPKM)
meanExpression <- mean(expression$FPKM)

summaryExpression<-data.frame(matrix(NA,ncol=3,nrow=(length(name.RNAseq)*2)))
for (indice in seq(1,(length(name.RNAseq)*2),by=2))
{
  RNAseq <- as.character(name.RNAseq[round((indice+1)/2)])
  summaryExpression[indice,] <- c(RNAseq,sum(expression$RNAseq == RNAseq),'Expression')
  summaryExpression[indice+1,] <- c(RNAseq, sum(No_expression$RNAseq == RNAseq),'In-expression')
}

colnames(summaryExpression) <- c('Id','Count','Type')
ggplot(summaryExpression,aes(x = Id, y = as.numeric(as.character(Count)), group = Type, fill = Type)) +
  geom_bar(stat = 'identity', position=position_dodge(-0.92),color = 'white')+ coord_flip() +  
  scale_x_discrete(name ="") +ylab('Nb Gene')+
  scale_fill_manual(values = c('forestgreen','red3'))+ 
  geom_text(aes(label = Count),hjust = 1.1, color = 'white',position=position_dodge(-0.9))
```


---- 

##Représentation des niveaux d'expressions de chaque données RNAseq

&nbsp;

```{r level expression}
color <- rep('orange',length(expression$FPKM))
color[expression$FPKM > meanExpression + sd.expression] <- 'brown1'
color[expression$FPKM > meanExpression + 2*sd.expression] <- 'red3'
color[expression$FPKM < meanExpression - sd.expression] <- 'green'
color[expression$FPKM < meanExpression - 2*sd.expression] <- 'darkgreen'

ggplot(expression, aes(x = RNAseq, y = FPKM)) + geom_boxplot()+geom_point(color =color)
```

---- 

##Représentation des niveaux d'expressions des données RNAseq réunis

&nbsp;

```{r traitement data}
name <- unique(expression$Id)
liste.sd <- c()
liste.name <- c()
liste.mean <- c()
nb <- 0 
for (elt in name)
{
  if (str_sub(elt,1,5) != 'CUFF.')
  {
    sd <- sd(expression$FPKM[expression$Id == elt])
    mean <- mean(expression$FPKM[expression$Id == elt])
    if (!(is.na(sd)))
    {
      liste.name <- c(liste.name,elt)
      liste.sd <- c(liste.sd,sd)
      liste.mean <- c(liste.mean,mean)
    }
  }
  else 
  {
    nb <- nb +1
  }
}
```

Les données RNAseq ont données `r nb` nouveau transcripts trouvés par rapport au GFF données en Input.

&nbsp;


```{r disperssion}

dispersion <- data.frame(cbind(as.character(liste.name),liste.sd,liste.mean))
colnames(dispersion) <- c('Id','sd','mean')
dispersion.sort <- dispersion
dispersion.sort$mean <- as.numeric(as.character(dispersion.sort$mean))
dispersion.sort$sd <- as.numeric(as.character(dispersion.sort$sd))
dispersion.sort$Id = factor(dispersion.sort$Id, levels=dispersion.sort[order(dispersion.sort$mean), "Id"])
len.data <- length(dispersion.sort$mean)
ggplot(dispersion.sort,aes(x = Id, y = mean)) +geom_point(size = 0.05,color = 'red')+  
  annotate("rect",xmin = 0, xmax = len.data,ymin = meanExpression + 2*sd.expression, ymax = max(dispersion.sort$mean),alpha=0.5,fill = 'red3')+
  annotate("rect",xmin = 0, xmax = len.data,ymin = meanExpression + sd.expression, ymax = meanExpression + 2*sd.expression,alpha=0.5,fill = 'brown1')+
  annotate("rect",xmin = 0, xmax = len.data,ymin = meanExpression - sd.expression, ymax = meanExpression + sd.expression,alpha=0.5,fill = 'orange')+
  annotate("rect",xmin = 0, xmax = len.data,ymin = meanExpression - 2*sd.expression, ymax = meanExpression - sd.expression,alpha=0.5,fill = 'green')+
  annotate("rect",xmin = 0, xmax = len.data,ymin = 0, ymax = meanExpression - 2*sd.expression,alpha=0.5,fill = 'darkgreen')+
  geom_errorbar(aes(ymin=mean-sd,ymax = mean+sd),color='steelblue')+geom_point(size = 0.05,color = 'red')+
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +xlab('Gene')+ylab('log2(FPKM)')
```

----

##Tableau des données d'expression

&nbsp;

```{r tableau expression}
line1 <- meanExpression - 2*sd.expression
line2 <- meanExpression - sd.expression
line3 <- meanExpression + sd.expression
line4 <- meanExpression + 2*sd.expression

datatable(data,extensions = 'Buttons', options = list(dom = 'Blfrtip',buttons = c('copy', 'csv', 'excel', 'pdf', 'print'))) %>%
  formatStyle('FPKM',color = 'white',font ='bold',background = styleInterval(c(line1,line2,line3,line4), c('darkgreen','green','orange', 'red', 'brown')))

```









